// finalhandler/index.js
/*!
 * finalhandler
 * Copyright(c) 2014-2017 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var debug = require('debug')('finalhandler')
var encodeUrl = require('encodeurl')
var escapeHtml = require('escape-html')
var onFinished = require('on-finished')
var parseUrl = require('parseurl')
var statuses = require('statuses')
var unpipe = require('unpipe')

/**
 * Module variables.
 * @private
 */

var DOUBLE_SPACE_REGEXP = /\x20{2***REMOVED***/g
var NEWLINE_REGEXP = /\n/g

/* istanbul ignore next */
var defer = typeof setImmediate === 'function'
  ? setImmediate
  : function (fn) { process.nextTick(fn.bind.apply(fn, arguments)) ***REMOVED***
var isFinished = onFinished.isFinished

/**
 * Create a minimal HTML document.
 *
 * @param {string***REMOVED*** message
 * @private
 */

function createHtmlDocument (message) {
  var body = escapeHtml(message)
    .replace(NEWLINE_REGEXP, '<br>')
    .replace(DOUBLE_SPACE_REGEXP, ' &nbsp;')

  return '<!DOCTYPE html>\n' +
    '<html lang="en">\n' +
    '<head>\n' +
    '<meta charset="utf-8">\n' +
    '<title>Error</title>\n' +
    '</head>\n' +
    '<body>\n' +
    '<pre>' + body + '</pre>\n' +
    '</body>\n' +
    '</html>\n'
***REMOVED***

/**
 * Module exports.
 * @public
 */

module.exports = finalhandler

/**
 * Create a function to handle the final response.
 *
 * @param {Request***REMOVED*** req
 * @param {Response***REMOVED*** res
 * @param {Object***REMOVED*** [options]
 * @return {Function***REMOVED***
 * @public
 */

function finalhandler (req, res, options) {
  var opts = options || {***REMOVED***

  // get environment
  var env = opts.env || process.env.NODE_ENV || 'development'

  // get error callback
  var onerror = opts.onerror

  return function (err) {
    var headers
    var msg
    var status

    // ignore 404 on in-flight response
    if (!err && headersSent(res)) {
      debug('cannot 404 after headers sent')
      return
    ***REMOVED***

    // unhandled error
    if (err) {
      // respect status code from error
      status = getErrorStatusCode(err)

      if (status === undefined) {
        // fallback to status code on response
        status = getResponseStatusCode(res)
      ***REMOVED***
        // respect headers from error
        headers = getErrorHeaders(err)
      ***REMOVED***

      // get error message
      msg = getErrorMessage(err, status, env)
    ***REMOVED***
      // not found
      status = 404
      msg = 'Cannot ' + req.method + ' ' + encodeUrl(getResourceName(req))
    ***REMOVED***

    debug('default %s', status)

    // schedule onerror callback
    if (err && onerror) {
      defer(onerror, err, req, res)
    ***REMOVED***

    // cannot actually respond
    if (headersSent(res)) {
      debug('cannot %d after headers sent', status)
      req.socket.destroy()
      return
    ***REMOVED***

    // send response
    send(req, res, status, headers, msg)
  ***REMOVED***
***REMOVED***

/**
 * Get headers from Error object.
 *
 * @param {Error***REMOVED*** err
 * @return {object***REMOVED***
 * @private
 */

function getErrorHeaders (err) {
  if (!err.headers || typeof err.headers !== 'object') {
    return undefined
  ***REMOVED***

  var headers = Object.create(null)
  var keys = Object.keys(err.headers)

  for (var i = 0; i < keys.length; i++) {
    var key = keys[i]
    headers[key] = err.headers[key]
  ***REMOVED***

  return headers
***REMOVED***

/**
 * Get message from Error object, fallback to status message.
 *
 * @param {Error***REMOVED*** err
 * @param {number***REMOVED*** status
 * @param {string***REMOVED*** env
 * @return {string***REMOVED***
 * @private
 */

function getErrorMessage (err, status, env) {
  var msg

  if (env !== 'production') {
    // use err.stack, which typically includes err.message
    msg = err.stack

    // fallback to err.toString() when possible
    if (!msg && typeof err.toString === 'function') {
      msg = err.toString()
    ***REMOVED***
  ***REMOVED***

  return msg || statuses[status]
***REMOVED***

/**
 * Get status code from Error object.
 *
 * @param {Error***REMOVED*** err
 * @return {number***REMOVED***
 * @private
 */

function getErrorStatusCode (err) {
  // check err.status
  if (typeof err.status === 'number' && err.status >= 400 && err.status < 600) {
    return err.status
  ***REMOVED***

  // check err.statusCode
  if (typeof err.statusCode === 'number' && err.statusCode >= 400 && err.statusCode < 600) {
    return err.statusCode
  ***REMOVED***

  return undefined
***REMOVED***

/**
 * Get resource name for the request.
 *
 * This is typically just the original pathname of the request
 * but will fallback to "resource" is that cannot be determined.
 *
 * @param {IncomingMessage***REMOVED*** req
 * @return {string***REMOVED***
 * @private
 */

function getResourceName (req) {
  try {
    return parseUrl.original(req).pathname
  ***REMOVED*** catch (e) {
    return 'resource'
  ***REMOVED***
***REMOVED***

/**
 * Get status code from response.
 *
 * @param {OutgoingMessage***REMOVED*** res
 * @return {number***REMOVED***
 * @private
 */

function getResponseStatusCode (res) {
  var status = res.statusCode

  // default status code to 500 if outside valid range
  if (typeof status !== 'number' || status < 400 || status > 599) {
    status = 500
  ***REMOVED***

  return status
***REMOVED***

/**
 * Determine if the response headers have been sent.
 *
 * @param {object***REMOVED*** res
 * @returns {boolean***REMOVED***
 * @private
 */

function headersSent (res) {
  return typeof res.headersSent !== 'boolean'
    ? Boolean(res._header)
    : res.headersSent
***REMOVED***

/**
 * Send response.
 *
 * @param {IncomingMessage***REMOVED*** req
 * @param {OutgoingMessage***REMOVED*** res
 * @param {number***REMOVED*** status
 * @param {object***REMOVED*** headers
 * @param {string***REMOVED*** message
 * @private
 */

function send (req, res, status, headers, message) {
  function write () {
    // response body
    var body = createHtmlDocument(message)

    // response status
    res.statusCode = status
    res.statusMessage = statuses[status]

    // response headers
    setHeaders(res, headers)

    // security headers
    res.setHeader('Content-Security-Policy', "default-src 'self'")
    res.setHeader('X-Content-Type-Options', 'nosniff')

    // standard headers
    res.setHeader('Content-Type', 'text/html; charset=utf-8')
    res.setHeader('Content-Length', Buffer.byteLength(body, 'utf8'))

    if (req.method === 'HEAD') {
      res.end()
      return
    ***REMOVED***

    res.end(body, 'utf8')
  ***REMOVED***

  if (isFinished(req)) {
    write()
    return
  ***REMOVED***

  // unpipe everything from the request
  unpipe(req)

  // flush the request
  onFinished(req, write)
  req.resume()
***REMOVED***

/**
 * Set response headers from an object.
 *
 * @param {OutgoingMessage***REMOVED*** res
 * @param {object***REMOVED*** headers
 * @private
 */

function setHeaders (res, headers) {
  if (!headers) {
    return
  ***REMOVED***

  var keys = Object.keys(headers)
  for (var i = 0; i < keys.length; i++) {
    var key = keys[i]
    res.setHeader(key, headers[key])
  ***REMOVED***
***REMOVED***
