// mongoose/lib/helpers/query/castUpdate.js
'use strict';

const CastError = require('../../error/cast');
const StrictModeError = require('../../error/strict');
const ValidationError = require('../../error/validation');
const castNumber = require('../../cast/number');
const getEmbeddedDiscriminatorPath = require('./getEmbeddedDiscriminatorPath');
const utils = require('../../utils');

/*!
 * Casts an update op based on the given schema
 *
 * @param {Schema***REMOVED*** schema
 * @param {Object***REMOVED*** obj
 * @param {Object***REMOVED*** options
 * @param {Boolean***REMOVED*** [options.overwrite] defaults to false
 * @param {Boolean|String***REMOVED*** [options.strict] defaults to true
 * @param {Query***REMOVED*** context passed to setters
 * @return {Boolean***REMOVED*** true iff the update is non-empty
 */

module.exports = function castUpdate(schema, obj, options, context, filter) {
  if (!obj) {
    return undefined;
  ***REMOVED***

  const ops = Object.keys(obj);
  let i = ops.length;
  const ret = {***REMOVED***;
  let hasKeys;
  let val;
  let hasDollarKey = false;
  const overwrite = options.overwrite;

  filter = filter || {***REMOVED***;

  while (i--) {
    const op = ops[i];
    // if overwrite is set, don't do any of the special $set stuff
    if (op[0] !== '$' && !overwrite) {
      // fix up $set sugar
      if (!ret.$set) {
        if (obj.$set) {
          ret.$set = obj.$set;
        ***REMOVED***
          ret.$set = {***REMOVED***;
        ***REMOVED***
      ***REMOVED***
      ret.$set[op] = obj[op];
      ops.splice(i, 1);
      if (!~ops.indexOf('$set')) ops.push('$set');
    ***REMOVED*** else if (op === '$set') {
      if (!ret.$set) {
        ret[op] = obj[op];
      ***REMOVED***
    ***REMOVED***
      ret[op] = obj[op];
    ***REMOVED***
  ***REMOVED***

  // cast each value
  i = ops.length;

  // if we get passed {***REMOVED*** for the update, we still need to respect that when it
  // is an overwrite scenario
  if (overwrite) {
    hasKeys = true;
  ***REMOVED***

  while (i--) {
    const op = ops[i];
    val = ret[op];
    hasDollarKey = hasDollarKey || op.charAt(0) === '$';

    if (val &&
        typeof val === 'object' &&
        !Buffer.isBuffer(val) &&
        (!overwrite || hasDollarKey)) {
      hasKeys |= walkUpdatePath(schema, val, op, options, context, filter);
    ***REMOVED*** else if (overwrite && ret && typeof ret === 'object') {
      // if we are just using overwrite, cast the query and then we will
      // *always* return the value, even if it is an empty object. We need to
      // set hasKeys above because we need to account for the case where the
      // user passes {***REMOVED*** and wants to clobber the whole document
      // Also, _walkUpdatePath expects an operation, so give it $set since that
      // is basically what we're doing
      walkUpdatePath(schema, ret, '$set', options, context, filter);
    ***REMOVED***
      const msg = 'Invalid atomic update value for ' + op + '. '
          + 'Expected an object, received ' + typeof val;
      throw new Error(msg);
    ***REMOVED***
  ***REMOVED***

  return hasKeys && ret;
***REMOVED***;

/*!
 * Walk each path of obj and cast its values
 * according to its schema.
 *
 * @param {Schema***REMOVED*** schema
 * @param {Object***REMOVED*** obj - part of a query
 * @param {String***REMOVED*** op - the atomic operator ($pull, $set, etc)
 * @param {Object***REMOVED*** options
 * @param {Boolean|String***REMOVED*** [options.strict]
 * @param {Boolean***REMOVED*** [options.omitUndefined]
 * @param {Query***REMOVED*** context
 * @param {String***REMOVED*** pref - path prefix (internal only)
 * @return {Bool***REMOVED*** true if this path has keys to update
 * @api private
 */

function walkUpdatePath(schema, obj, op, options, context, filter, pref) {
  const strict = options.strict;
  const prefix = pref ? pref + '.' : '';
  const keys = Object.keys(obj);
  let i = keys.length;
  let hasKeys = false;
  let schematype;
  let key;
  let val;

  let aggregatedError = null;

  let useNestedStrict;
  if (options.useNestedStrict === undefined) {
    useNestedStrict = schema.options.useNestedStrict;
  ***REMOVED***
    useNestedStrict = options.useNestedStrict;
  ***REMOVED***

  while (i--) {
    key = keys[i];
    val = obj[key];

    if (val && val.constructor.name === 'Object') {
      // watch for embedded doc schemas
      schematype = schema._getSchema(prefix + key);
      if (schematype && schematype.caster && op in castOps) {
        // embedded doc schema
        if ('$each' in val) {
          hasKeys = true;
          try {
            obj[key] = {
              $each: castUpdateVal(schematype, val.$each, op, key, context, prefix + key)
            ***REMOVED***;
          ***REMOVED*** catch (error) {
            aggregatedError = _handleCastError(error, context, key, aggregatedError);
          ***REMOVED***

          if (val.$slice != null) {
            obj[key].$slice = val.$slice | 0;
          ***REMOVED***

          if (val.$sort) {
            obj[key].$sort = val.$sort;
          ***REMOVED***

          if (!!val.$position || val.$position === 0) {
            obj[key].$position = val.$position;
          ***REMOVED***
        ***REMOVED***
          try {
            obj[key] = castUpdateVal(schematype, val, op, key, context, prefix + key);
          ***REMOVED*** catch (error) {
            aggregatedError = _handleCastError(error, context, key, aggregatedError);
          ***REMOVED***

          if (options.omitUndefined && obj[key] === void 0) {
            delete obj[key];
            continue;
          ***REMOVED***

          hasKeys = true;
        ***REMOVED***
      ***REMOVED*** else if ((op === '$currentDate') || (op in castOps && schematype)) {
        // $currentDate can take an object
        try {
          obj[key] = castUpdateVal(schematype, val, op, key, context, prefix + key);
        ***REMOVED*** catch (error) {
          aggregatedError = _handleCastError(error, context, key, aggregatedError);
        ***REMOVED***

        if (options.omitUndefined && obj[key] === void 0) {
          delete obj[key];
          continue;
        ***REMOVED***

        hasKeys = true;
      ***REMOVED***
        const pathToCheck = (prefix + key);
        const v = schema._getPathType(pathToCheck);
        let _strict = strict;
        if (useNestedStrict &&
            v &&
            v.schema &&
            'strict' in v.schema.options) {
          _strict = v.schema.options.strict;
        ***REMOVED***

        if (v.pathType === 'undefined') {
          if (_strict === 'throw') {
            throw new StrictModeError(pathToCheck);
          ***REMOVED*** else if (_strict) {
            delete obj[key];
            continue;
          ***REMOVED***
        ***REMOVED***

        // gh-2314
        // we should be able to set a schema-less field
        // to an empty object literal
        hasKeys |= walkUpdatePath(schema, val, op, options, context, filter, prefix + key) ||
          (utils.isObject(val) && Object.keys(val).length === 0);
      ***REMOVED***
    ***REMOVED***
      const checkPath = (key === '$each' || key === '$or' || key === '$and' || key === '$in') ?
        pref : prefix + key;
      schematype = schema._getSchema(checkPath);
      let pathDetails = schema._getPathType(checkPath);

      // If no schema type, check for embedded discriminators
      if (schematype == null) {
        const _res = getEmbeddedDiscriminatorPath(schema, obj, filter, checkPath);
        if (_res.schematype != null) {
          schematype = _res.schematype;
          pathDetails = _res.type;
        ***REMOVED***
      ***REMOVED***

      let isStrict = strict;
      if (useNestedStrict &&
          pathDetails &&
          pathDetails.schema &&
          'strict' in pathDetails.schema.options) {
        isStrict = pathDetails.schema.options.strict;
      ***REMOVED***

      const skip = isStrict &&
        !schematype &&
        !/real|nested/.test(pathDetails.pathType);

      if (skip) {
        // Even if strict is `throw`, avoid throwing an error because of
        // virtuals because of #6731
        if (isStrict === 'throw' && schema.virtuals[checkPath] == null) {
          throw new StrictModeError(prefix + key);
        ***REMOVED***
          delete obj[key];
        ***REMOVED***
      ***REMOVED***
        // gh-1845 temporary fix: ignore $rename. See gh-3027 for tracking
        // improving this.
        if (op === '$rename') {
          hasKeys = true;
          continue;
        ***REMOVED***

        try {
          obj[key] = castUpdateVal(schematype, val, op, key, context, prefix + key);
        ***REMOVED*** catch (error) {
          aggregatedError = _handleCastError(error, context, key, aggregatedError);
        ***REMOVED***

        if (Array.isArray(obj[key]) && (op === '$addToSet' || op === '$push') && key !== '$each') {
          if (schematype && schematype.caster && !schematype.caster.$isMongooseArray) {
            obj[key] = { $each: obj[key] ***REMOVED***;
          ***REMOVED***
        ***REMOVED***

        if (options.omitUndefined && obj[key] === void 0) {
          delete obj[key];
          continue;
        ***REMOVED***

        hasKeys = true;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  if (aggregatedError != null) {
    throw aggregatedError;
  ***REMOVED***

  return hasKeys;
***REMOVED***

/*!
 * ignore
 */

function _handleCastError(error, query, key, aggregatedError) {
  if (typeof query !== 'object' || !query.options.multipleCastError) {
    throw error;
  ***REMOVED***
  aggregatedError = aggregatedError || new ValidationError();
  aggregatedError.addError(key, error);
  return aggregatedError;
***REMOVED***

/*!
 * These operators should be cast to numbers instead
 * of their path schema type.
 */

const numberOps = {
  $pop: 1,
  $inc: 1
***REMOVED***;

/*!
 * These ops require no casting because the RHS doesn't do anything.
 */

const noCastOps = {
  $unset: 1
***REMOVED***;

/*!
 * These operators require casting docs
 * to real Documents for Update operations.
 */

const castOps = {
  $push: 1,
  $addToSet: 1,
  $set: 1,
  $setOnInsert: 1
***REMOVED***;

/*!
 * ignore
 */

const overwriteOps = {
  $set: 1,
  $setOnInsert: 1
***REMOVED***;

/*!
 * Casts `val` according to `schema` and atomic `op`.
 *
 * @param {SchemaType***REMOVED*** schema
 * @param {Object***REMOVED*** val
 * @param {String***REMOVED*** op - the atomic operator ($pull, $set, etc)
 * @param {String***REMOVED*** $conditional
 * @param {Query***REMOVED*** context
 * @api private
 */

function castUpdateVal(schema, val, op, $conditional, context, path) {
  if (!schema) {
    // non-existing schema path
    return op in numberOps ?
      castNumber(val, path) :
      val;
  ***REMOVED***

  const cond = schema.caster && op in castOps &&
      (utils.isObject(val) || Array.isArray(val));
  if (cond && op !== '$set') {
    // Cast values for ops that add data to MongoDB.
    // Ensures embedded documents get ObjectIds etc.
    const tmp = schema.cast(val);
    if (Array.isArray(val)) {
      val = tmp;
    ***REMOVED*** else if (Array.isArray(tmp)) {
      val = tmp[0];
    ***REMOVED***
      val = tmp;
    ***REMOVED***
    return val;
  ***REMOVED*** else if (cond && op === '$set') {
    return schema.cast(val);
  ***REMOVED***

  if (op in noCastOps) {
    return val;
  ***REMOVED***
  if (op in numberOps) {
    // Null and undefined not allowed for $pop, $inc
    if (val == null) {
      throw new CastError('number', val, schema.path);
    ***REMOVED***
    if (op === '$inc') {
      // Support `$inc` with long, int32, etc. (gh-4283)
      return schema.castForQueryWrapper({
        val: val,
        context: context
      ***REMOVED***);
    ***REMOVED***
    return castNumber(val, schema.path);
  ***REMOVED***
  if (op === '$currentDate') {
    if (typeof val === 'object') {
      return {$type: val.$type***REMOVED***;
    ***REMOVED***
    return Boolean(val);
  ***REMOVED***

  if (/^\$/.test($conditional)) {
    return schema.castForQueryWrapper({
      $conditional: $conditional,
      val: val,
      context: context
    ***REMOVED***);
  ***REMOVED***

  if (overwriteOps[op]) {
    return schema.castForQueryWrapper({
      val: val,
      context: context,
      $skipQueryCastForUpdate: val != null && schema.$isMongooseArray && schema.$parentSchema
    ***REMOVED***);
  ***REMOVED***

  return schema.castForQueryWrapper({ val: val, context: context ***REMOVED***);
***REMOVED***
