// mongodb/lib/operations/cursor_ops.js
'use strict';

const buildCountCommand = require('./collection_ops').buildCountCommand;
const formattedOrderClause = require('../utils').formattedOrderClause;
const handleCallback = require('../utils').handleCallback;
const MongoError = require('mongodb-core').MongoError;
const push = Array.prototype.push;

/**
 * Get the count of documents for this cursor.
 *
 * @method
 * @param {Cursor***REMOVED*** cursor The Cursor instance on which to count.
 * @param {boolean***REMOVED*** [applySkipLimit=true] Specifies whether the count command apply limit and skip settings should be applied on the cursor or in the provided options.
 * @param {object***REMOVED*** [options] Optional settings. See Cursor.prototype.count for a list of options.
 * @param {Cursor~countResultCallback***REMOVED*** [callback] The result callback.
 */
function count(cursor, applySkipLimit, opts, callback) {
  if (applySkipLimit) {
    if (typeof cursor.cursorSkip() === 'number') opts.skip = cursor.cursorSkip();
    if (typeof cursor.cursorLimit() === 'number') opts.limit = cursor.cursorLimit();
  ***REMOVED***

  // Ensure we have the right read preference inheritance
  if (opts.readPreference) {
    cursor.setReadPreference(opts.readPreference);
  ***REMOVED***

  if (
    typeof opts.maxTimeMS !== 'number' &&
    cursor.s.cmd &&
    typeof cursor.s.cmd.maxTimeMS === 'number'
  ) {
    opts.maxTimeMS = cursor.s.cmd.maxTimeMS;
  ***REMOVED***

  let options = {***REMOVED***;
  options.skip = opts.skip;
  options.limit = opts.limit;
  options.hint = opts.hint;
  options.maxTimeMS = opts.maxTimeMS;

  // Command
  const delimiter = cursor.s.ns.indexOf('.');
  options.collectionName = cursor.s.ns.substr(delimiter + 1);

  const command = buildCountCommand(cursor, cursor.s.cmd.query, options);

  // Set cursor server to the same as the topology
  cursor.server = cursor.topology.s.coreTopology;

  // Execute the command
  cursor.s.topology.command(
    `${cursor.s.ns.substr(0, delimiter)***REMOVED***.$cmd`,
    command,
    cursor.s.options,
    (err, result) => {
      callback(err, result ? result.result.n : null);
    ***REMOVED***
  );
***REMOVED***

/**
 * Iterates over all the documents for this cursor. See Cursor.prototype.each for more information.
 *
 * @method
 * @deprecated
 * @param {Cursor***REMOVED*** cursor The Cursor instance on which to run.
 * @param {Cursor~resultCallback***REMOVED*** callback The result callback.
 */
function each(cursor, callback) {
  const Cursor = require('../cursor');

  if (!callback) throw MongoError.create({ message: 'callback is mandatory', driver: true ***REMOVED***);
  if (cursor.isNotified()) return;
  if (cursor.s.state === Cursor.CLOSED || cursor.isDead()) {
    return handleCallback(
      callback,
      MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***)
    );
  ***REMOVED***

  if (cursor.s.state === Cursor.INIT) cursor.s.state = Cursor.OPEN;

  // Define function to avoid global scope escape
  let fn = null;
  // Trampoline all the entries
  if (cursor.bufferedCount() > 0) {
    while ((fn = loop(cursor, callback))) fn(cursor, callback);
    each(cursor, callback);
  ***REMOVED***
    cursor.next((err, item) => {
      if (err) return handleCallback(callback, err);
      if (item == null) {
        return cursor.close({ skipKillCursors: true ***REMOVED***, () => handleCallback(callback, null, null));
      ***REMOVED***

      if (handleCallback(callback, null, item) === false) return;
      each(cursor, callback);
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

/**
 * Check if there is any document still available in the cursor.
 *
 * @method
 * @param {Cursor***REMOVED*** cursor The Cursor instance on which to run.
 * @param {Cursor~resultCallback***REMOVED*** [callback] The result callback.
 */
function hasNext(cursor, callback) {
  const Cursor = require('../cursor');

  if (cursor.s.currentDoc) {
    return callback(null, true);
  ***REMOVED***

  if (cursor.isNotified()) {
    return callback(null, false);
  ***REMOVED***

  nextObject(cursor, (err, doc) => {
    if (err) return callback(err, null);
    if (cursor.s.state === Cursor.CLOSED || cursor.isDead()) return callback(null, false);
    if (!doc) return callback(null, false);
    cursor.s.currentDoc = doc;
    callback(null, true);
  ***REMOVED***);
***REMOVED***

// Trampoline emptying the number of retrieved items
// without incurring a nextTick operation
function loop(cursor, callback) {
  // No more items we are done
  if (cursor.bufferedCount() === 0) return;
  // Get the next document
  cursor._next(callback);
  // Loop
  return loop;
***REMOVED***

/**
 * Get the next available document from the cursor. Returns null if no more documents are available.
 *
 * @method
 * @param {Cursor***REMOVED*** cursor The Cursor instance from which to get the next document.
 * @param {Cursor~resultCallback***REMOVED*** [callback] The result callback.
 */
function next(cursor, callback) {
  // Return the currentDoc if someone called hasNext first
  if (cursor.s.currentDoc) {
    const doc = cursor.s.currentDoc;
    cursor.s.currentDoc = null;
    return callback(null, doc);
  ***REMOVED***

  // Return the next object
  nextObject(cursor, callback);
***REMOVED***

// Get the next available document from the cursor, returns null if no more documents are available.
function nextObject(cursor, callback) {
  const Cursor = require('../cursor');

  if (cursor.s.state === Cursor.CLOSED || (cursor.isDead && cursor.isDead()))
    return handleCallback(
      callback,
      MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***)
    );
  if (cursor.s.state === Cursor.INIT && cursor.s.cmd.sort) {
    try {
      cursor.s.cmd.sort = formattedOrderClause(cursor.s.cmd.sort);
    ***REMOVED*** catch (err) {
      return handleCallback(callback, err);
    ***REMOVED***
  ***REMOVED***

  // Get the next object
  cursor._next((err, doc) => {
    cursor.s.state = Cursor.OPEN;
    if (err) return handleCallback(callback, err);
    handleCallback(callback, null, doc);
  ***REMOVED***);
***REMOVED***

/**
 * Returns an array of documents. See Cursor.prototype.toArray for more information.
 *
 * @method
 * @param {Cursor***REMOVED*** cursor The Cursor instance from which to get the next document.
 * @param {Cursor~toArrayResultCallback***REMOVED*** [callback] The result callback.
 */
function toArray(cursor, callback) {
  const Cursor = require('../cursor');

  const items = [];

  // Reset cursor
  cursor.rewind();
  cursor.s.state = Cursor.INIT;

  // Fetch all the documents
  const fetchDocs = () => {
    cursor._next((err, doc) => {
      if (err) {
        return cursor._endSession
          ? cursor._endSession(() => handleCallback(callback, err))
          : handleCallback(callback, err);
      ***REMOVED***
      if (doc == null) {
        return cursor.close({ skipKillCursors: true ***REMOVED***, () => handleCallback(callback, null, items));
      ***REMOVED***

      // Add doc to items
      items.push(doc);

      // Get all buffered objects
      if (cursor.bufferedCount() > 0) {
        let docs = cursor.readBufferedDocuments(cursor.bufferedCount());

        // Transform the doc if transform method added
        if (cursor.s.transforms && typeof cursor.s.transforms.doc === 'function') {
          docs = docs.map(cursor.s.transforms.doc);
        ***REMOVED***

        push.apply(items, docs);
      ***REMOVED***

      // Attempt a fetch
      fetchDocs();
    ***REMOVED***);
  ***REMOVED***;

  fetchDocs();
***REMOVED***

module.exports = { count, each, hasNext, next, toArray ***REMOVED***;
