// body-parser/lib/types/urlencoded.js
/*!
 * body-parser
 * Copyright(c) 2014 Jonathan Ong
 * Copyright(c) 2014-2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var bytes = require('bytes')
var contentType = require('content-type')
var createError = require('http-errors')
var debug = require('debug')('body-parser:urlencoded')
var deprecate = require('depd')('body-parser')
var read = require('../read')
var typeis = require('type-is')

/**
 * Module exports.
 */

module.exports = urlencoded

/**
 * Cache of parser modules.
 */

var parsers = Object.create(null)

/**
 * Create a middleware to parse urlencoded bodies.
 *
 * @param {object***REMOVED*** [options]
 * @return {function***REMOVED***
 * @public
 */

function urlencoded (options) {
  var opts = options || {***REMOVED***

  // notice because option default will flip in next major
  if (opts.extended === undefined) {
    deprecate('undefined extended: provide extended option')
  ***REMOVED***

  var extended = opts.extended !== false
  var inflate = opts.inflate !== false
  var limit = typeof opts.limit !== 'number'
    ? bytes.parse(opts.limit || '100kb')
    : opts.limit
  var type = opts.type || 'application/x-www-form-urlencoded'
  var verify = opts.verify || false

  if (verify !== false && typeof verify !== 'function') {
    throw new TypeError('option verify must be function')
  ***REMOVED***

  // create the appropriate query parser
  var queryparse = extended
    ? extendedparser(opts)
    : simpleparser(opts)

  // create the appropriate type checking function
  var shouldParse = typeof type !== 'function'
    ? typeChecker(type)
    : type

  function parse (body) {
    return body.length
      ? queryparse(body)
      : {***REMOVED***
  ***REMOVED***

  return function urlencodedParser (req, res, next) {
    if (req._body) {
      debug('body already parsed')
      next()
      return
    ***REMOVED***

    req.body = req.body || {***REMOVED***

    // skip requests without bodies
    if (!typeis.hasBody(req)) {
      debug('skip empty body')
      next()
      return
    ***REMOVED***

    debug('content-type %j', req.headers['content-type'])

    // determine if request should be parsed
    if (!shouldParse(req)) {
      debug('skip parsing')
      next()
      return
    ***REMOVED***

    // assert charset
    var charset = getCharset(req) || 'utf-8'
    if (charset !== 'utf-8') {
      debug('invalid charset')
      next(createError(415, 'unsupported charset "' + charset.toUpperCase() + '"', {
        charset: charset,
        type: 'charset.unsupported'
      ***REMOVED***))
      return
    ***REMOVED***

    // read
    read(req, res, next, parse, debug, {
      debug: debug,
      encoding: charset,
      inflate: inflate,
      limit: limit,
      verify: verify
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

/**
 * Get the extended query parser.
 *
 * @param {object***REMOVED*** options
 */

function extendedparser (options) {
  var parameterLimit = options.parameterLimit !== undefined
    ? options.parameterLimit
    : 1000
  var parse = parser('qs')

  if (isNaN(parameterLimit) || parameterLimit < 1) {
    throw new TypeError('option parameterLimit must be a positive number')
  ***REMOVED***

  if (isFinite(parameterLimit)) {
    parameterLimit = parameterLimit | 0
  ***REMOVED***

  return function queryparse (body) {
    var paramCount = parameterCount(body, parameterLimit)

    if (paramCount === undefined) {
      debug('too many parameters')
      throw createError(413, 'too many parameters', {
        type: 'parameters.too.many'
      ***REMOVED***)
    ***REMOVED***

    var arrayLimit = Math.max(100, paramCount)

    debug('parse extended urlencoding')
    return parse(body, {
      allowPrototypes: true,
      arrayLimit: arrayLimit,
      depth: Infinity,
      parameterLimit: parameterLimit
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

/**
 * Get the charset of a request.
 *
 * @param {object***REMOVED*** req
 * @api private
 */

function getCharset (req) {
  try {
    return (contentType.parse(req).parameters.charset || '').toLowerCase()
  ***REMOVED*** catch (e) {
    return undefined
  ***REMOVED***
***REMOVED***

/**
 * Count the number of parameters, stopping once limit reached
 *
 * @param {string***REMOVED*** body
 * @param {number***REMOVED*** limit
 * @api private
 */

function parameterCount (body, limit) {
  var count = 0
  var index = 0

  while ((index = body.indexOf('&', index)) !== -1) {
    count++
    index++

    if (count === limit) {
      return undefined
    ***REMOVED***
  ***REMOVED***

  return count
***REMOVED***

/**
 * Get parser for module name dynamically.
 *
 * @param {string***REMOVED*** name
 * @return {function***REMOVED***
 * @api private
 */

function parser (name) {
  var mod = parsers[name]

  if (mod !== undefined) {
    return mod.parse
  ***REMOVED***

  // this uses a switch for static require analysis
  switch (name) {
    case 'qs':
      mod = require('qs')
      break
    case 'querystring':
      mod = require('querystring')
      break
  ***REMOVED***

  // store to prevent invoking require()
  parsers[name] = mod

  return mod.parse
***REMOVED***

/**
 * Get the simple query parser.
 *
 * @param {object***REMOVED*** options
 */

function simpleparser (options) {
  var parameterLimit = options.parameterLimit !== undefined
    ? options.parameterLimit
    : 1000
  var parse = parser('querystring')

  if (isNaN(parameterLimit) || parameterLimit < 1) {
    throw new TypeError('option parameterLimit must be a positive number')
  ***REMOVED***

  if (isFinite(parameterLimit)) {
    parameterLimit = parameterLimit | 0
  ***REMOVED***

  return function queryparse (body) {
    var paramCount = parameterCount(body, parameterLimit)

    if (paramCount === undefined) {
      debug('too many parameters')
      throw createError(413, 'too many parameters', {
        type: 'parameters.too.many'
      ***REMOVED***)
    ***REMOVED***

    debug('parse urlencoding')
    return parse(body, undefined, undefined, {maxKeys: parameterLimit***REMOVED***)
  ***REMOVED***
***REMOVED***

/**
 * Get the simple type checker.
 *
 * @param {string***REMOVED*** type
 * @return {function***REMOVED***
 */

function typeChecker (type) {
  return function checkType (req) {
    return Boolean(typeis(req, type))
  ***REMOVED***
***REMOVED***
