// mongoose/lib/document.js
'use strict';

/*!
 * Module dependencies.
 */

const EventEmitter = require('events').EventEmitter;
const InternalCache = require('./internal');
const MongooseError = require('./error');
const MixedSchema = require('./schema/mixed');
const ObjectExpectedError = require('./error/objectExpected');
const ObjectParameterError = require('./error/objectParameter');
const StrictModeError = require('./error/strict');
const ValidatorError = require('./schematype').ValidatorError;
const VirtualType = require('./virtualtype');
const cleanModifiedSubpaths = require('./helpers/document/cleanModifiedSubpaths');
const compile = require('./helpers/document/compile').compile;
const defineKey = require('./helpers/document/compile').defineKey;
const flatten = require('./helpers/common').flatten;
const get = require('lodash.get');
const getEmbeddedDiscriminatorPath = require('./helpers/document/getEmbeddedDiscriminatorPath');
const idGetter = require('./plugins/idGetter');
const isDefiningProjection = require('./helpers/projection/isDefiningProjection');
const isExclusive = require('./helpers/projection/isExclusive');
const inspect = require('util').inspect;
const internalToObjectOptions = require('./options').internalToObjectOptions;
const mpath = require('mpath');
const utils = require('./utils');

const ValidationError = MongooseError.ValidationError;
const clone = utils.clone;
const deepEqual = utils.deepEqual;
const isMongooseObject = utils.isMongooseObject;

let DocumentArray;
let MongooseArray;
let Embedded;

const specialProperties = ['__proto__', 'constructor', 'prototype'];

/**
 * The core Mongoose document constructor. You should not call this directly,
 * the Mongoose [Model constructor](./api.html#Model) calls this for you.
 *
 * @param {Object***REMOVED*** obj the values to set
 * @param {Object***REMOVED*** [fields] optional object containing the fields which were selected in the query returning this document and any populated paths data
 * @param {Boolean***REMOVED*** [skipId] bool, should we auto create an ObjectId _id
 * @inherits NodeJS EventEmitter http://nodejs.org/api/events.html#events_class_events_eventemitter
 * @event `init`: Emitted on a document after it has was retreived from the db and fully hydrated by Mongoose.
 * @event `save`: Emitted when the document is successfully saved
 * @api private
 */

function Document(obj, fields, skipId, options) {
  if (typeof skipId === 'object' && skipId != null) {
    options = skipId;
    skipId = options.skipId;
  ***REMOVED***
  options = options || {***REMOVED***;

  this.$__ = new InternalCache;
  this.$__.emitter = new EventEmitter();
  this.isNew = 'isNew' in options ? options.isNew : true;
  this.errors = undefined;
  this.$__.$options = options || {***REMOVED***;

  if (obj != null && typeof obj !== 'object') {
    throw new ObjectParameterError(obj, 'obj', 'Document');
  ***REMOVED***

  const schema = this.schema;

  if (typeof fields === 'boolean') {
    this.$__.strictMode = fields;
    fields = undefined;
  ***REMOVED***
    this.$__.strictMode = schema.options.strict;
    this.$__.selected = fields;
  ***REMOVED***

  const required = schema.requiredPaths(true);
  for (let i = 0; i < required.length; ++i) {
    this.$__.activePaths.require(required[i]);
  ***REMOVED***

  this.$__.emitter.setMaxListeners(0);

  let exclude = null;

  // determine if this doc is a result of a query with
  // excluded fields
  if (fields && utils.getFunctionName(fields.constructor) === 'Object') {
    exclude = isExclusive(fields);
  ***REMOVED***

  const hasIncludedChildren = exclude === false && fields ?
    $__hasIncludedChildren(fields) :
    {***REMOVED***;

  this.$__buildDoc(obj, fields, skipId, exclude, hasIncludedChildren, false);

  // By default, defaults get applied **before** setting initial values
  // Re: gh-6155
  $__applyDefaults(this, fields, skipId, exclude, hasIncludedChildren, true, {
    isNew: this.isNew
  ***REMOVED***);

  if (obj) {
    if (obj instanceof Document) {
      this.isNew = obj.isNew;
    ***REMOVED***
    // Skip set hooks
    if (this.$__original_set) {
      this.$__original_set(obj, undefined, true);
    ***REMOVED***
      this.$set(obj, undefined, true);
    ***REMOVED***
  ***REMOVED***

  // Function defaults get applied **after** setting initial values so they
  // see the full doc rather than an empty one, unless they opt out.
  // Re: gh-3781, gh-6155
  if (options.willInit) {
    this.once('init', () => {
      $__applyDefaults(this, fields, skipId, exclude, hasIncludedChildren, false, options.skipDefaults, {
        isNew: this.isNew
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***
    $__applyDefaults(this, fields, skipId, exclude, hasIncludedChildren, false, options.skipDefaults, {
      isNew: this.isNew
    ***REMOVED***);
  ***REMOVED***

  this.$__._id = this._id;

  if (!schema.options.strict && obj) {
    const _this = this;
    const keys = Object.keys(this._doc);

    keys.forEach(function(key) {
      if (!(key in schema.tree)) {
        defineKey(key, null, _this);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***

  applyQueue(this);
***REMOVED***

/*!
 * Document exposes the NodeJS event emitter API, so you can use
 * `on`, `once`, etc.
 */
utils.each(
  ['on', 'once', 'emit', 'listeners', 'removeListener', 'setMaxListeners',
    'removeAllListeners', 'addListener'],
  function(emitterFn) {
    Document.prototype[emitterFn] = function() {
      return this.$__.emitter[emitterFn].apply(this.$__.emitter, arguments);
    ***REMOVED***;
  ***REMOVED***);

Document.prototype.constructor = Document;

/**
 * The documents schema.
 *
 * @api public
 * @property schema
 * @memberOf Document
 * @instance
 */

Document.prototype.schema;

/**
 * Boolean flag specifying if the document is new.
 *
 * @api public
 * @property isNew
 * @memberOf Document
 * @instance
 */

Document.prototype.isNew;

/**
 * The string version of this documents _id.
 *
 * ####Note:
 *
 * This getter exists on all documents by default. The getter can be disabled by setting the `id` [option](/docs/guide.html#id) of its `Schema` to false at construction time.
 *
 *     new Schema({ name: String ***REMOVED***, { id: false ***REMOVED***);
 *
 * @api public
 * @see Schema options /docs/guide.html#options
 * @property id
 * @memberOf Document
 * @instance
 */

Document.prototype.id;

/**
 * Hash containing current validation errors.
 *
 * @api public
 * @property errors
 * @memberOf Document
 * @instance
 */

Document.prototype.errors;

/*!
 * ignore
 */

function $__hasIncludedChildren(fields) {
  const hasIncludedChildren = {***REMOVED***;
  const keys = Object.keys(fields);
  for (let j = 0; j < keys.length; ++j) {
    const parts = keys[j].split('.');
    const c = [];
    for (let k = 0; k < parts.length; ++k) {
      c.push(parts[k]);
      hasIncludedChildren[c.join('.')] = 1;
    ***REMOVED***
  ***REMOVED***

  return hasIncludedChildren;
***REMOVED***

/*!
 * ignore
 */

function $__applyDefaults(doc, fields, skipId, exclude, hasIncludedChildren, isBeforeSetters, pathsToSkip, options) {
  const paths = Object.keys(doc.schema.paths);
  const plen = paths.length;
  const init = options && 'isNew' in options ? options.isNew : false;

  for (let i = 0; i < plen; ++i) {
    let def;
    let curPath = '';
    const p = paths[i];

    if (p === '_id' && skipId) {
      continue;
    ***REMOVED***

    const type = doc.schema.paths[p];
    const path = p.split('.');
    const len = path.length;
    let included = false;
    let doc_ = doc._doc;

    for (let j = 0; j < len; ++j) {
      if (doc_ == null) {
        break;
      ***REMOVED***

      const piece = path[j];
      curPath += (!curPath.length ? '' : '.') + piece;

      if (exclude === true) {
        if (curPath in fields) {
          break;
        ***REMOVED***
      ***REMOVED*** else if (exclude === false && fields && !included) {
        if (curPath in fields) {
          included = true;
        ***REMOVED*** else if (!hasIncludedChildren[curPath]) {
          break;
        ***REMOVED***
      ***REMOVED***

      if (j === len - 1) {
        if (doc_[piece] !== void 0) {
          break;
        ***REMOVED***

        if (typeof type.defaultValue === 'function') {
          if (!type.defaultValue.$runBeforeSetters && isBeforeSetters) {
            break;
          ***REMOVED***
          if (type.defaultValue.$runBeforeSetters && !isBeforeSetters) {
            break;
          ***REMOVED***
        ***REMOVED*** else if (!isBeforeSetters) {
          // Non-function defaults should always run **before** setters
          continue;
        ***REMOVED***

        if (pathsToSkip && pathsToSkip[curPath]) {
          break;
        ***REMOVED***

        if (fields && exclude !== null) {
          if (exclude === true) {
            // apply defaults to all non-excluded fields
            if (p in fields) {
              continue;
            ***REMOVED***

            def = type.getDefault(doc, init);
            if (typeof def !== 'undefined') {
              doc_[piece] = def;
              doc.$__.activePaths.default(p);
            ***REMOVED***
          ***REMOVED*** else if (included) {
            // selected field
            def = type.getDefault(doc, init);
            if (typeof def !== 'undefined') {
              doc_[piece] = def;
              doc.$__.activePaths.default(p);
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***
          def = type.getDefault(doc, init);
          if (typeof def !== 'undefined') {
            doc_[piece] = def;
            doc.$__.activePaths.default(p);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
        doc_ = doc_[piece];
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * Builds the default doc structure
 *
 * @param {Object***REMOVED*** obj
 * @param {Object***REMOVED*** [fields]
 * @param {Boolean***REMOVED*** [skipId]
 * @api private
 * @method $__buildDoc
 * @memberOf Document
 * @instance
 */

Document.prototype.$__buildDoc = function(obj, fields, skipId, exclude, hasIncludedChildren) {
  const doc = {***REMOVED***;

  const paths = Object.keys(this.schema.paths).
    // Don't build up any paths that are underneath a map, we don't know
    // what the keys will be
    filter(p => !p.includes('$*'));
  const plen = paths.length;
  let ii = 0;

  for (; ii < plen; ++ii) {
    const p = paths[ii];

    if (p === '_id') {
      if (skipId) {
        continue;
      ***REMOVED***
      if (obj && '_id' in obj) {
        continue;
      ***REMOVED***
    ***REMOVED***

    const path = p.split('.');
    const len = path.length;
    const last = len - 1;
    let curPath = '';
    let doc_ = doc;
    let included = false;

    for (let i = 0; i < len; ++i) {
      const piece = path[i];

      curPath += (!curPath.length ? '' : '.') + piece;

      // support excluding intermediary levels
      if (exclude === true) {
        if (curPath in fields) {
          break;
        ***REMOVED***
      ***REMOVED*** else if (exclude === false && fields && !included) {
        if (curPath in fields) {
          included = true;
        ***REMOVED*** else if (!hasIncludedChildren[curPath]) {
          break;
        ***REMOVED***
      ***REMOVED***

      if (i < last) {
        doc_ = doc_[piece] || (doc_[piece] = {***REMOVED***);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  this._doc = doc;
***REMOVED***;

/*!
 * Converts to POJO when you use the document for querying
 */

Document.prototype.toBSON = function() {
  return this.toObject(internalToObjectOptions);
***REMOVED***;

/**
 * Initializes the document without setters or marking anything modified.
 *
 * Called internally after a document is returned from mongodb. Normally,
 * you do **not** need to call this function on your own.
 *
 * This function triggers `init` [middleware](/docs/middleware.html).
 * Note that `init` hooks are [synchronous](/docs/middleware.html#synchronous).
 *
 * @param {Object***REMOVED*** doc document returned by mongo
 * @api public
 * @memberOf Document
 * @instance
 */

Document.prototype.init = function(doc, opts, fn) {
  if (typeof opts === 'function') {
    fn = opts;
    opts = null;
  ***REMOVED***

  this.$__init(doc, opts);

  if (fn) {
    fn(null, this);
  ***REMOVED***

  return this;
***REMOVED***;

/*!
 * ignore
 */

Document.prototype.$__init = function(doc, opts) {
  this.isNew = false;
  this.$init = true;

  // handle docs with populated paths
  // If doc._id is not null or undefined
  if (doc._id !== null && doc._id !== undefined &&
    opts && opts.populated && opts.populated.length) {
    const id = String(doc._id);
    for (let i = 0; i < opts.populated.length; ++i) {
      const item = opts.populated[i];
      if (item.isVirtual) {
        this.populated(item.path, utils.getValue(item.path, doc), item);
      ***REMOVED***
        this.populated(item.path, item._docs[id], item);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  init(this, doc, this._doc);

  this.emit('init', this);
  this.constructor.emit('init', this);

  this.$__._id = this._id;

  return this;
***REMOVED***;

/*!
 * Init helper.
 *
 * @param {Object***REMOVED*** self document instance
 * @param {Object***REMOVED*** obj raw mongodb doc
 * @param {Object***REMOVED*** doc object we are initializing
 * @api private
 */

function init(self, obj, doc, prefix) {
  prefix = prefix || '';

  const keys = Object.keys(obj);
  const len = keys.length;
  let schema;
  let path;
  let i;
  let index = 0;

  while (index < len) {
    _init(index++);
  ***REMOVED***

  function _init(index) {
    i = keys[index];
    path = prefix + i;
    schema = self.schema.path(path);

    // Should still work if not a model-level discriminator, but should not be
    // necessary. This is *only* to catch the case where we queried using the
    // base model and the discriminated model has a projection
    if (self.schema.$isRootDiscriminator && !self.isSelected(path)) {
      return;
    ***REMOVED***

    if (!schema && utils.isObject(obj[i]) &&
        (!obj[i].constructor || utils.getFunctionName(obj[i].constructor) === 'Object')) {
      // assume nested object
      if (!doc[i]) {
        doc[i] = {***REMOVED***;
      ***REMOVED***
      init(self, obj[i], doc[i], path + '.');
    ***REMOVED*** else if (!schema) {
      doc[i] = obj[i];
    ***REMOVED***
      if (obj[i] === null) {
        doc[i] = null;
      ***REMOVED*** else if (obj[i] !== undefined) {
        const intCache = obj[i].$__ || {***REMOVED***;
        const wasPopulated = intCache.wasPopulated || null;
        if (schema && !wasPopulated) {
          try {
            doc[i] = schema.cast(obj[i], self, true);
          ***REMOVED*** catch (e) {
            self.invalidate(e.path, new ValidatorError({
              path: e.path,
              message: e.message,
              type: 'cast',
              value: e.value
            ***REMOVED***));
          ***REMOVED***
        ***REMOVED***
          doc[i] = obj[i];
        ***REMOVED***
      ***REMOVED***
      // mark as hydrated
      if (!self.isModified(path)) {
        self.$__.activePaths.init(path);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * Sends an update command with this document `_id` as the query selector.
 *
 * ####Example:
 *
 *     weirdCar.update({$inc: {wheels:1***REMOVED******REMOVED***, { w: 1 ***REMOVED***, callback);
 *
 * ####Valid options:
 *
 *  - same as in [Model.update](#model_Model.update)
 *
 * @see Model.update #model_Model.update
 * @param {Object***REMOVED*** doc
 * @param {Object***REMOVED*** options
 * @param {Function***REMOVED*** callback
 * @return {Query***REMOVED***
 * @api public
 * @memberOf Document
 * @instance
 */

Document.prototype.update = function update() {
  const args = utils.args(arguments);
  args.unshift({_id: this._id***REMOVED***);
  return this.constructor.update.apply(this.constructor, args);
***REMOVED***;

/**
 * Sends an updateOne command with this document `_id` as the query selector.
 *
 * ####Example:
 *
 *     weirdCar.updateOne({$inc: {wheels:1***REMOVED******REMOVED***, { w: 1 ***REMOVED***, callback);
 *
 * ####Valid options:
 *
 *  - same as in [Model.updateOne](#model_Model.updateOne)
 *
 * @see Model.updateOne #model_Model.updateOne
 * @param {Object***REMOVED*** doc
 * @param {Object***REMOVED*** options
 * @param {Function***REMOVED*** callback
 * @return {Query***REMOVED***
 * @api public
 * @memberOf Document
 * @instance
 */

Document.prototype.updateOne = function updateOne() {
  const args = utils.args(arguments);
  args.unshift({_id: this._id***REMOVED***);
  return this.constructor.updateOne.apply(this.constructor, args);
***REMOVED***;

/**
 * Sends a replaceOne command with this document `_id` as the query selector.
 *
 * ####Valid options:
 *
 *  - same as in [Model.replaceOne](#model_Model.replaceOne)
 *
 * @see Model.replaceOne #model_Model.replaceOne
 * @param {Object***REMOVED*** doc
 * @param {Object***REMOVED*** options
 * @param {Function***REMOVED*** callback
 * @return {Query***REMOVED***
 * @api public
 * @memberOf Document
 * @instance
 */

Document.prototype.replaceOne = function replaceOne() {
  const args = utils.args(arguments);
  args.unshift({ _id: this._id ***REMOVED***);
  return this.constructor.replaceOne.apply(this.constructor, args);
***REMOVED***;

/**
 * Getter/setter around the session associated with this document. Used to
 * automatically set `session` if you `save()` a doc that you got from a
 * query with an associated session.
 *
 * ####Example:
 *
 *     const session = MyModel.startSession();
 *     const doc = await MyModel.findOne().session(session);
 *     doc.$session() === session; // true
 *     doc.$session(null);
 *     doc.$session() === null; // true
 *
 * If this is a top-level document, setting the session propagates to all child
 * docs.
 *
 * @param {ClientSession***REMOVED*** [session] overwrite the current session
 * @return {ClientSession***REMOVED***
 * @method $session
 * @api public
 * @memberOf Document
 */

Document.prototype.$session = function $session(session) {
  if (arguments.length === 0) {
    return this.$__.session;
  ***REMOVED***
  this.$__.session = session;

  if (!this.ownerDocument) {
    const subdocs = this.$__getAllSubdocs();
    for (const child of subdocs) {
      child.$session(session);
    ***REMOVED***
  ***REMOVED***

  return session;
***REMOVED***;

/**
 * Alias for `set()`, used internally to avoid conflicts
 *
 * @param {String|Object***REMOVED*** path path or object of key/vals to set
 * @param {Any***REMOVED*** val the value to set
 * @param {Schema|String|Number|Buffer|****REMOVED*** [type] optionally specify a type for "on-the-fly" attributes
 * @param {Object***REMOVED*** [options] optionally specify options that modify the behavior of the set
 * @method $set
 * @name $set
 * @memberOf Document
 * @instance
 * @api public
 */

Document.prototype.$set = function $set(path, val, type, options) {
  if (type && utils.getFunctionName(type.constructor) === 'Object') {
    options = type;
    type = undefined;
  ***REMOVED***

  options = options || {***REMOVED***;
  const merge = options.merge;
  const adhoc = type && type !== true;
  const constructing = type === true;
  let adhocs;
  let keys;
  let i = 0;
  let pathtype;
  let key;
  let prefix;

  const strict = 'strict' in options
    ? options.strict
    : this.$__.strictMode;

  if (adhoc) {
    adhocs = this.$__.adhocPaths || (this.$__.adhocPaths = {***REMOVED***);
    adhocs[path] = this.schema.interpretAsType(path, type, this.schema.options);
  ***REMOVED***

  if (typeof path !== 'string') {
    // new Document({ key: val ***REMOVED***)
    if (path === null || path === void 0) {
      const _ = path;
      path = val;
      val = _;
    ***REMOVED***
      prefix = val ? val + '.' : '';

      if (path instanceof Document) {
        if (path.$__isNested) {
          path = path.toObject();
        ***REMOVED***
          path = path._doc;
        ***REMOVED***
      ***REMOVED***

      keys = Object.keys(path);
      const len = keys.length;

      if (len === 0 && !this.schema.options.minimize) {
        if (val) {
          this.$set(val, {***REMOVED***);
        ***REMOVED***
        return this;
      ***REMOVED***

      while (i < len) {
        _handleIndex.call(this, i++);
      ***REMOVED***

      return this;
    ***REMOVED***
  ***REMOVED***

  function _handleIndex(i) {
    key = keys[i];
    const pathName = prefix + key;
    pathtype = this.schema.pathType(pathName);

    // On initial set, delete any nested keys if we're going to overwrite
    // them to ensure we keep the user's key order.
    if (type === true &&
        !prefix &&
        path[key] != null &&
        pathtype === 'nested' &&
        this._doc[key] != null &&
        Object.keys(this._doc[key]).length === 0) {
      delete this._doc[key];
    ***REMOVED***

    if (path[key] !== null &&
        path[key] !== void 0 &&
        // need to know if plain object - no Buffer, ObjectId, ref, etc
        utils.isObject(path[key]) &&
        (!path[key].constructor || utils.getFunctionName(path[key].constructor) === 'Object') &&
        pathtype !== 'virtual' &&
        pathtype !== 'real' &&
        !(this.$__path(pathName) instanceof MixedSchema) &&
        !(this.schema.paths[pathName] &&
        this.schema.paths[pathName].options &&
        this.schema.paths[pathName].options.ref)) {
      this.$set(path[key], prefix + key, constructing);
    ***REMOVED*** else if (strict) {
      // Don't overwrite defaults with undefined keys (gh-3981)
      if (constructing && path[key] === void 0 &&
          this.get(key) !== void 0) {
        return;
      ***REMOVED***

      if (pathtype === 'adhocOrUndefined') {
        pathtype = getEmbeddedDiscriminatorPath(this, pathName, { typeOnly: true ***REMOVED***);
      ***REMOVED***

      if (pathtype === 'real' || pathtype === 'virtual') {
        // Check for setting single embedded schema to document (gh-3535)
        let p = path[key];
        if (this.schema.paths[pathName] &&
            this.schema.paths[pathName].$isSingleNested &&
            path[key] instanceof Document) {
          p = p.toObject({ virtuals: false, transform: false ***REMOVED***);
        ***REMOVED***
        this.$set(prefix + key, p, constructing);
      ***REMOVED*** else if (pathtype === 'nested' && path[key] instanceof Document) {
        this.$set(prefix + key,
          path[key].toObject({transform: false***REMOVED***), constructing);
      ***REMOVED*** else if (strict === 'throw') {
        if (pathtype === 'nested') {
          throw new ObjectExpectedError(key, path[key]);
        ***REMOVED***
          throw new StrictModeError(key);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else if (path[key] !== void 0) {
      this.$set(prefix + key, path[key], constructing);
    ***REMOVED***
  ***REMOVED***

  const pathType = this.schema.pathType(path);
  if (pathType === 'nested' && val) {
    if (utils.isObject(val) &&
        (!val.constructor || utils.getFunctionName(val.constructor) === 'Object')) {
      if (!merge) {
        this.setValue(path, null);
        cleanModifiedSubpaths(this, path);
      ***REMOVED***
        return this.$set(val, path, constructing);
      ***REMOVED***

      const keys = Object.keys(val);
      this.setValue(path, {***REMOVED***);
      for (const key of keys) {
        this.$set(path + '.' + key, val[key], constructing);
      ***REMOVED***
      this.markModified(path);
      cleanModifiedSubpaths(this, path, { skipDocArrays: true ***REMOVED***);
      return this;
    ***REMOVED***
    this.invalidate(path, new MongooseError.CastError('Object', val, path));
    return this;
  ***REMOVED***

  let schema;
  const parts = path.split('.');

  if (pathType === 'adhocOrUndefined' && strict) {
    // check for roots that are Mixed types
    let mixed;

    for (i = 0; i < parts.length; ++i) {
      const subpath = parts.slice(0, i + 1).join('.');

      // If path is underneath a virtual, bypass everything and just set it.
      if (i + 1 < parts.length && this.schema.pathType(subpath) === 'virtual') {
        mpath.set(path, val, this);
        return this;
      ***REMOVED***

      schema = this.schema.path(subpath);
      if (schema == null) {
        continue;
      ***REMOVED***

      if (schema instanceof MixedSchema) {
        // allow changes to sub paths of mixed types
        mixed = true;
        break;
      ***REMOVED***
    ***REMOVED***

    if (schema == null) {
      // Check for embedded discriminators
      schema = getEmbeddedDiscriminatorPath(this, path);
    ***REMOVED***

    if (!mixed && !schema) {
      if (strict === 'throw') {
        throw new StrictModeError(path);
      ***REMOVED***
      return this;
    ***REMOVED***
  ***REMOVED*** else if (pathType === 'virtual') {
    schema = this.schema.virtualpath(path);
    schema.applySetters(val, this);
    return this;
  ***REMOVED***
    schema = this.$__path(path);
  ***REMOVED***

  // gh-4578, if setting a deeply nested path that doesn't exist yet, create it
  let cur = this._doc;
  let curPath = '';
  for (i = 0; i < parts.length - 1; ++i) {
    cur = cur[parts[i]];
    curPath += (curPath.length > 0 ? '.' : '') + parts[i];
    if (!cur) {
      this.$set(curPath, {***REMOVED***);
      // Hack re: gh-5800. If nested field is not selected, it probably exists
      // so `MongoError: cannot use the part (nested of nested.num) to
      // traverse the element ({nested: null***REMOVED***)` is not likely. If user gets
      // that error, its their fault for now. We should reconsider disallowing
      // modifying not selected paths for 6.x
      if (!this.isSelected(curPath)) {
        this.unmarkModified(curPath);
      ***REMOVED***
      cur = this.getValue(curPath);
    ***REMOVED***
  ***REMOVED***

  let pathToMark;

  // When using the $set operator the path to the field must already exist.
  // Else mongodb throws: "LEFT_SUBFIELD only supports Object"

  if (parts.length <= 1) {
    pathToMark = path;
  ***REMOVED***
    for (i = 0; i < parts.length; ++i) {
      const subpath = parts.slice(0, i + 1).join('.');
      if (this.isDirectModified(subpath) // earlier prefixes that are already
      // marked as dirty have precedence
          || this.get(subpath) === null) {
        pathToMark = subpath;
        break;
      ***REMOVED***
    ***REMOVED***

    if (!pathToMark) {
      pathToMark = path;
    ***REMOVED***
  ***REMOVED***

  // if this doc is being constructed we should not trigger getters
  const priorVal = (() => {
    if (this.$__.$options.priorDoc != null) {
      return this.$__.$options.priorDoc.getValue(path);
    ***REMOVED***
    if (constructing) {
      return void 0;
    ***REMOVED***
    return this.getValue(path);
  ***REMOVED***)();

  if (!schema) {
    this.$__set(pathToMark, path, constructing, parts, schema, val, priorVal);
    return this;
  ***REMOVED***

  let shouldSet = true;
  try {
    // If the user is trying to set a ref path to a document with
    // the correct model name, treat it as populated
    const refMatches = (() => {
      if (schema.options == null) {
        return false;
      ***REMOVED***
      if (!(val instanceof Document)) {
        return false;
      ***REMOVED***
      const model = val.constructor;

      // Check ref
      const ref = schema.options.ref;
      if (ref != null && (ref === model.modelName || ref === model.baseModelName)) {
        return true;
      ***REMOVED***

      // Check refPath
      const refPath = schema.options.refPath;
      if (refPath == null) {
        return false;
      ***REMOVED***
      const modelName = val.get(refPath);
      if (modelName === model.modelName || modelName === model.baseModelName) {
        return true;
      ***REMOVED***
      return false;
    ***REMOVED***)();

    let didPopulate = false;
    if (refMatches && val instanceof Document) {
      if (this.ownerDocument) {
        this.ownerDocument().populated(this.$__fullPath(path),
          val._id, {model: val.constructor***REMOVED***);
      ***REMOVED***
        this.populated(path, val._id, {model: val.constructor***REMOVED***);
      ***REMOVED***
      didPopulate = true;
    ***REMOVED***

    let popOpts;
    if (schema.options &&
        Array.isArray(schema.options[this.schema.options.typeKey]) &&
        schema.options[this.schema.options.typeKey].length &&
        schema.options[this.schema.options.typeKey][0].ref &&
        Array.isArray(val) &&
        val.length > 0 &&
        val[0] instanceof Document &&
        val[0].constructor.modelName &&
        (schema.options[this.schema.options.typeKey][0].ref === val[0].constructor.baseModelName || schema.options[this.schema.options.typeKey][0].ref === val[0].constructor.modelName)) {
      if (this.ownerDocument) {
        popOpts = { model: val[0].constructor ***REMOVED***;
        this.ownerDocument().populated(this.$__fullPath(path),
          val.map(function(v) { return v._id; ***REMOVED***), popOpts);
      ***REMOVED***
        popOpts = { model: val[0].constructor ***REMOVED***;
        this.populated(path, val.map(function(v) { return v._id; ***REMOVED***), popOpts);
      ***REMOVED***
      didPopulate = true;
    ***REMOVED***

    // If this path is underneath a single nested schema, we'll call the setter
    // later in `$__set()` because we don't take `_doc` when we iterate through
    // a single nested doc. That's to make sure we get the correct context.
    // Otherwise we would double-call the setter, see gh-7196.
    if (this.schema.singleNestedPaths[path] == null) {
      const setterContext = constructing && this.$__.$options.priorDoc ?
        this.$__.$options.priorDoc :
        this;
      val = schema.applySetters(val, setterContext, false, priorVal);
    ***REMOVED***

    if (!didPopulate && this.$__.populated) {
      delete this.$__.populated[path];
    ***REMOVED***

    this.$markValid(path);
  ***REMOVED*** catch (e) {
    this.invalidate(path,
      new MongooseError.CastError(schema.instance, val, path, e));
    shouldSet = false;
  ***REMOVED***

  if (shouldSet) {
    this.$__set(pathToMark, path, constructing, parts, schema, val, priorVal);
  ***REMOVED***

  if (schema.$isSingleNested && (this.isDirectModified(path) || val == null)) {
    cleanModifiedSubpaths(this, path);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Sets the value of a path, or many paths.
 *
 * ####Example:
 *
 *     // path, value
 *     doc.set(path, value)
 *
 *     // object
 *     doc.set({
 *         path  : value
 *       , path2 : {
 *            path  : value
 *         ***REMOVED***
 *     ***REMOVED***)
 *
 *     // on-the-fly cast to number
 *     doc.set(path, value, Number)
 *
 *     // on-the-fly cast to string
 *     doc.set(path, value, String)
 *
 *     // changing strict mode behavior
 *     doc.set(path, value, { strict: false ***REMOVED***);
 *
 * @param {String|Object***REMOVED*** path path or object of key/vals to set
 * @param {Any***REMOVED*** val the value to set
 * @param {Schema|String|Number|Buffer|****REMOVED*** [type] optionally specify a type for "on-the-fly" attributes
 * @param {Object***REMOVED*** [options] optionally specify options that modify the behavior of the set
 * @api public
 * @method set
 * @memberOf Document
 * @instance
 */

Document.prototype.set = Document.prototype.$set;

/**
 * Determine if we should mark this change as modified.
 *
 * @return {Boolean***REMOVED***
 * @api private
 * @method $__shouldModify
 * @memberOf Document
 * @instance
 */

Document.prototype.$__shouldModify = function(pathToMark, path, constructing, parts, schema, val, priorVal) {
  if (this.isNew) {
    return true;
  ***REMOVED***

  if (val === void 0 && !this.isSelected(path)) {
    // when a path is not selected in a query, its initial
    // value will be undefined.
    return true;
  ***REMOVED***

  if (val === void 0 && path in this.$__.activePaths.states.default) {
    // we're just unsetting the default value which was never saved
    return false;
  ***REMOVED***

  // gh-3992: if setting a populated field to a doc, don't mark modified
  // if they have the same _id
  if (this.populated(path) &&
      val instanceof Document &&
      deepEqual(val._id, priorVal)) {
    return false;
  ***REMOVED***

  if (!deepEqual(val, priorVal || this.get(path))) {
    return true;
  ***REMOVED***

  if (!constructing &&
      val !== null &&
      val !== undefined &&
      path in this.$__.activePaths.states.default &&
      deepEqual(val, schema.getDefault(this, constructing))) {
    // a path with a default was $unset on the server
    // and the user is setting it to the same value again
    return true;
  ***REMOVED***
  return false;
***REMOVED***;

/**
 * Handles the actual setting of the value and marking the path modified if appropriate.
 *
 * @api private
 * @method $__set
 * @memberOf Document
 * @instance
 */

Document.prototype.$__set = function(pathToMark, path, constructing, parts, schema, val, priorVal) {
  Embedded = Embedded || require('./types/embedded');

  const shouldModify = this.$__shouldModify(pathToMark, path, constructing, parts,
    schema, val, priorVal);
  const _this = this;

  if (shouldModify) {
    this.markModified(pathToMark);

    // handle directly setting arrays (gh-1126)
    MongooseArray || (MongooseArray = require('./types/array'));
    if (val && val.isMongooseArray) {
      val._registerAtomic('$set', val);

      // Update embedded document parent references (gh-5189)
      if (val.isMongooseDocumentArray) {
        val.forEach(function(item) {
          item && item.__parentArray && (item.__parentArray = val);
        ***REMOVED***);
      ***REMOVED***

      // Small hack for gh-1638: if we're overwriting the entire array, ignore
      // paths that were modified before the array overwrite
      this.$__.activePaths.forEach(function(modifiedPath) {
        if (modifiedPath.indexOf(path + '.') === 0) {
          _this.$__.activePaths.ignore(modifiedPath);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  let obj = this._doc;
  let i = 0;
  const l = parts.length;
  let cur = '';

  for (; i < l; i++) {
    const next = i + 1;
    const last = next === l;
    cur += (cur ? '.' + parts[i] : parts[i]);
    if (specialProperties.indexOf(parts[i]) !== -1) {
      return;
    ***REMOVED***

    if (last) {
      if (obj instanceof Map) {
        obj.set(parts[i], val);
      ***REMOVED***
        obj[parts[i]] = val;
      ***REMOVED***
    ***REMOVED***
      if (obj[parts[i]] && utils.getFunctionName(obj[parts[i]].constructor) === 'Object') {
        obj = obj[parts[i]];
      ***REMOVED*** else if (obj[parts[i]] && obj[parts[i]] instanceof Embedded) {
        obj = obj[parts[i]];
      ***REMOVED*** else if (obj[parts[i]] && obj[parts[i]].$isSingleNested) {
        obj = obj[parts[i]];
      ***REMOVED*** else if (obj[parts[i]] && Array.isArray(obj[parts[i]])) {
        obj = obj[parts[i]];
      ***REMOVED***
        obj[parts[i]] = obj[parts[i]] || {***REMOVED***;
        obj = obj[parts[i]];
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/**
 * Gets a raw value from a path (no getters)
 *
 * @param {String***REMOVED*** path
 * @api private
 */

Document.prototype.getValue = function(path) {
  return utils.getValue(path, this._doc);
***REMOVED***;

/**
 * Sets a raw value for a path (no casting, setters, transformations)
 *
 * @param {String***REMOVED*** path
 * @param {Object***REMOVED*** value
 * @api private
 */

Document.prototype.setValue = function(path, val) {
  utils.setValue(path, val, this._doc);
  return this;
***REMOVED***;

/**
 * Returns the value of a path.
 *
 * ####Example
 *
 *     // path
 *     doc.get('age') // 47
 *
 *     // dynamic casting to a string
 *     doc.get('age', String) // "47"
 *
 * @param {String***REMOVED*** path
 * @param {Schema|String|Number|Buffer|****REMOVED*** [type] optionally specify a type for on-the-fly attributes
 * @api public
 */

Document.prototype.get = function(path, type, options) {
  let adhoc;
  options = options || {***REMOVED***;
  if (type) {
    adhoc = this.schema.interpretAsType(path, type, this.schema.options);
  ***REMOVED***

  const schema = this.$__path(path) || this.schema.virtualpath(path);
  const pieces = path.split('.');
  let obj = this._doc;

  if (schema instanceof VirtualType) {
    if (schema.getters.length === 0) {
      return void 0;
    ***REMOVED***
    return schema.applyGetters(null, this);
  ***REMOVED***

  for (let i = 0, l = pieces.length; i < l; i++) {
    if (obj && obj._doc) {
      obj = obj._doc;
    ***REMOVED***

    if (obj == null) {
      obj = void 0;
    ***REMOVED*** else if (obj instanceof Map) {
      obj = obj.get(pieces[i]);
    ***REMOVED*** else if (i === l - 1) {
      obj = utils.getValue(pieces[i], obj);
    ***REMOVED***
      obj = obj[pieces[i]];
    ***REMOVED***
  ***REMOVED***

  if (adhoc) {
    obj = adhoc.cast(obj);
  ***REMOVED***

  if (schema) {
    obj = schema.applyGetters(obj, this);
  ***REMOVED*** else if (this.schema.nested[path] && options.virtuals) {
    // Might need to apply virtuals if this is a nested path
    return applyGetters(this, utils.clone(obj), 'virtuals', { path: path ***REMOVED***);
  ***REMOVED***

  return obj;
***REMOVED***;

/**
 * Returns the schematype for the given `path`.
 *
 * @param {String***REMOVED*** path
 * @api private
 * @method $__path
 * @memberOf Document
 * @instance
 */

Document.prototype.$__path = function(path) {
  const adhocs = this.$__.adhocPaths;
  const adhocType = adhocs && adhocs[path];

  if (adhocType) {
    return adhocType;
  ***REMOVED***
  return this.schema.path(path);
***REMOVED***;

/**
 * Marks the path as having pending changes to write to the db.
 *
 * _Very helpful when using [Mixed](./schematypes.html#mixed) types._
 *
 * ####Example:
 *
 *     doc.mixed.type = 'changed';
 *     doc.markModified('mixed.type');
 *     doc.save() // changes to mixed.type are now persisted
 *
 * @param {String***REMOVED*** path the path to mark modified
 * @param {Document***REMOVED*** [scope] the scope to run validators with
 * @api public
 */

Document.prototype.markModified = function(path, scope) {
  this.$__.activePaths.modify(path);
  if (scope != null && !this.ownerDocument) {
    this.$__.pathsToScopes[path] = scope;
  ***REMOVED***
***REMOVED***;

/**
 * Clears the modified state on the specified path.
 *
 * ####Example:
 *
 *     doc.foo = 'bar';
 *     doc.unmarkModified('foo');
 *     doc.save(); // changes to foo will not be persisted
 *
 * @param {String***REMOVED*** path the path to unmark modified
 * @api public
 */

Document.prototype.unmarkModified = function(path) {
  this.$__.activePaths.init(path);
  delete this.$__.pathsToScopes[path];
***REMOVED***;

/**
 * Don't run validation on this path or persist changes to this path.
 *
 * ####Example:
 *
 *     doc.foo = null;
 *     doc.$ignore('foo');
 *     doc.save(); // changes to foo will not be persisted and validators won't be run
 *
 * @memberOf Document
 * @instance
 * @method $ignore
 * @param {String***REMOVED*** path the path to ignore
 * @api public
 */

Document.prototype.$ignore = function(path) {
  this.$__.activePaths.ignore(path);
***REMOVED***;

/**
 * Returns the list of paths that have been modified.
 *
 * @param {Object***REMOVED*** [options]
 * @param {Boolean***REMOVED*** [options.includeChildren=false] if true, returns children of modified paths as well. For example, if false, the list of modified paths for `doc.colors = { primary: 'blue' ***REMOVED***;` will **not** contain `colors.primary`. If true, `modifiedPaths()` will return an array that contains `colors.primary`.
 * @return {Array***REMOVED***
 * @api public
 */

Document.prototype.modifiedPaths = function(options) {
  options = options || {***REMOVED***;
  const directModifiedPaths = Object.keys(this.$__.activePaths.states.modify);
  const _this = this;
  return directModifiedPaths.reduce(function(list, path) {
    const parts = path.split('.');
    list = list.concat(parts.reduce(function(chains, part, i) {
      return chains.concat(parts.slice(0, i).concat(part).join('.'));
    ***REMOVED***, []).filter(function(chain) {
      return (list.indexOf(chain) === -1);
    ***REMOVED***));

    if (!options.includeChildren) {
      return list;
    ***REMOVED***

    let cur = _this.get(path);
    if (cur != null && typeof cur === 'object') {
      if (cur._doc) {
        cur = cur._doc;
      ***REMOVED***
      if (Array.isArray(cur)) {
        const len = cur.length;
        for (let i = 0; i < len; ++i) {
          if (list.indexOf(path + '.' + i) === -1) {
            list.push(path + '.' + i);
            if (cur[i] != null && cur[i].$__) {
              const modified = cur[i].modifiedPaths();
              for (const childPath of modified) {
                list.push(path + '.' + i + '.' + childPath);
              ***REMOVED***
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
        Object.keys(cur).
          filter(function(key) {
            return list.indexOf(path + '.' + key) === -1;
          ***REMOVED***).
          forEach(function(key) {
            list.push(path + '.' + key);
          ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    return list;
  ***REMOVED***, []);
***REMOVED***;

/**
 * Returns true if this document was modified, else false.
 *
 * If `path` is given, checks if a path or any full path containing `path` as part of its path chain has been modified.
 *
 * ####Example
 *
 *     doc.set('documents.0.title', 'changed');
 *     doc.isModified()                      // true
 *     doc.isModified('documents')           // true
 *     doc.isModified('documents.0.title')   // true
 *     doc.isModified('documents otherProp') // true
 *     doc.isDirectModified('documents')     // false
 *
 * @param {String***REMOVED*** [path] optional
 * @return {Boolean***REMOVED***
 * @api public
 */

Document.prototype.isModified = function(paths, modifiedPaths) {
  if (paths) {
    if (!Array.isArray(paths)) {
      paths = paths.split(' ');
    ***REMOVED***
    const modified = modifiedPaths || this.modifiedPaths();
    const directModifiedPaths = Object.keys(this.$__.activePaths.states.modify);
    const isModifiedChild = paths.some(function(path) {
      return !!~modified.indexOf(path);
    ***REMOVED***);
    return isModifiedChild || paths.some(function(path) {
      return directModifiedPaths.some(function(mod) {
        return mod === path || path.indexOf(mod + '.') === 0;
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***
  return this.$__.activePaths.some('modify');
***REMOVED***;

/**
 * Checks if a path is set to its default.
 *
 * ####Example
 *
 *     MyModel = mongoose.model('test', { name: { type: String, default: 'Val '***REMOVED*** ***REMOVED***);
 *     var m = new MyModel();
 *     m.$isDefault('name'); // true
 *
 * @memberOf Document
 * @instance
 * @method $isDefault
 * @param {String***REMOVED*** [path]
 * @return {Boolean***REMOVED***
 * @api public
 */

Document.prototype.$isDefault = function(path) {
  return (path in this.$__.activePaths.states.default);
***REMOVED***;

/**
 * Getter/setter, determines whether the document was removed or not.
 *
 * ####Example:
 *     product.remove(function (err, product) {
 *       product.isDeleted(); // true
 *       product.remove(); // no-op, doesn't send anything to the db
 *
 *       product.isDeleted(false);
 *       product.isDeleted(); // false
 *       product.remove(); // will execute a remove against the db
 *     ***REMOVED***)
 *
 * @param {Boolean***REMOVED*** [val] optional, overrides whether mongoose thinks the doc is deleted
 * @return {Boolean***REMOVED*** whether mongoose thinks this doc is deleted.
 * @method $isDeleted
 * @memberOf Document
 * @instance
 * @api public
 */

Document.prototype.$isDeleted = function(val) {
  if (arguments.length === 0) {
    return !!this.$__.isDeleted;
  ***REMOVED***

  this.$__.isDeleted = !!val;
  return this;
***REMOVED***;

/**
 * Returns true if `path` was directly set and modified, else false.
 *
 * ####Example
 *
 *     doc.set('documents.0.title', 'changed');
 *     doc.isDirectModified('documents.0.title') // true
 *     doc.isDirectModified('documents') // false
 *
 * @param {String***REMOVED*** path
 * @return {Boolean***REMOVED***
 * @api public
 */

Document.prototype.isDirectModified = function(path) {
  return (path in this.$__.activePaths.states.modify);
***REMOVED***;

/**
 * Checks if `path` was initialized.
 *
 * @param {String***REMOVED*** path
 * @return {Boolean***REMOVED***
 * @api public
 */

Document.prototype.isInit = function(path) {
  return (path in this.$__.activePaths.states.init);
***REMOVED***;

/**
 * Checks if `path` was selected in the source query which initialized this document.
 *
 * ####Example
 *
 *     Thing.findOne().select('name').exec(function (err, doc) {
 *        doc.isSelected('name') // true
 *        doc.isSelected('age')  // false
 *     ***REMOVED***)
 *
 * @param {String***REMOVED*** path
 * @return {Boolean***REMOVED***
 * @api public
 */

Document.prototype.isSelected = function isSelected(path) {
  if (this.$__.selected) {
    if (path === '_id') {
      return this.$__.selected._id !== 0;
    ***REMOVED***

    const paths = Object.keys(this.$__.selected);
    let i = paths.length;
    let inclusive = null;
    let cur;

    if (i === 1 && paths[0] === '_id') {
      // only _id was selected.
      return this.$__.selected._id === 0;
    ***REMOVED***

    while (i--) {
      cur = paths[i];
      if (cur === '_id') {
        continue;
      ***REMOVED***
      if (!isDefiningProjection(this.$__.selected[cur])) {
        continue;
      ***REMOVED***
      inclusive = !!this.$__.selected[cur];
      break;
    ***REMOVED***

    if (inclusive === null) {
      return true;
    ***REMOVED***

    if (path in this.$__.selected) {
      return inclusive;
    ***REMOVED***

    i = paths.length;
    const pathDot = path + '.';

    while (i--) {
      cur = paths[i];
      if (cur === '_id') {
        continue;
      ***REMOVED***

      if (cur.indexOf(pathDot) === 0) {
        return inclusive || cur !== pathDot;
      ***REMOVED***

      if (pathDot.indexOf(cur + '.') === 0) {
        return inclusive;
      ***REMOVED***
    ***REMOVED***

    return !inclusive;
  ***REMOVED***

  return true;
***REMOVED***;

/**
 * Checks if `path` was explicitly selected. If no projection, always returns
 * true.
 *
 * ####Example
 *
 *     Thing.findOne().select('nested.name').exec(function (err, doc) {
 *        doc.isDirectSelected('nested.name') // true
 *        doc.isDirectSelected('nested.otherName') // false
 *        doc.isDirectSelected('nested')  // false
 *     ***REMOVED***)
 *
 * @param {String***REMOVED*** path
 * @return {Boolean***REMOVED***
 * @api public
 */

Document.prototype.isDirectSelected = function isDirectSelected(path) {
  if (this.$__.selected) {
    if (path === '_id') {
      return this.$__.selected._id !== 0;
    ***REMOVED***

    const paths = Object.keys(this.$__.selected);
    let i = paths.length;
    let inclusive = null;
    let cur;

    if (i === 1 && paths[0] === '_id') {
      // only _id was selected.
      return this.$__.selected._id === 0;
    ***REMOVED***

    while (i--) {
      cur = paths[i];
      if (cur === '_id') {
        continue;
      ***REMOVED***
      if (!isDefiningProjection(this.$__.selected[cur])) {
        continue;
      ***REMOVED***
      inclusive = !!this.$__.selected[cur];
      break;
    ***REMOVED***

    if (inclusive === null) {
      return true;
    ***REMOVED***

    if (path in this.$__.selected) {
      return inclusive;
    ***REMOVED***

    return !inclusive;
  ***REMOVED***

  return true;
***REMOVED***;

/**
 * Executes registered validation rules for this document.
 *
 * ####Note:
 *
 * This method is called `pre` save and if a validation rule is violated, [save](#model_Model-save) is aborted and the error is returned to your `callback`.
 *
 * ####Example:
 *
 *     doc.validate(function (err) {
 *       if (err) handleError(err);
 *       else // validation passed
 *     ***REMOVED***);
 *
 * @param {Object***REMOVED*** optional options internal options
 * @param {Function***REMOVED*** callback optional callback called after validation completes, passing an error if one occurred
 * @return {Promise***REMOVED*** Promise
 * @api public
 */

Document.prototype.validate = function(options, callback) {
  if (typeof options === 'function') {
    callback = options;
    options = null;
  ***REMOVED***

  return utils.promiseOrCallback(callback, cb => this.$__validate(function(error) {
    cb(error);
  ***REMOVED***));
***REMOVED***;

/*!
 * ignore
 */

function _evaluateRequiredFunctions(doc) {
  Object.keys(doc.$__.activePaths.states.require).forEach(path => {
    const p = doc.schema.path(path);

    if (p != null && typeof p.originalRequiredValue === 'function') {
      doc.$__.cachedRequired[path] = p.originalRequiredValue.call(doc);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

/*!
 * ignore
 */

function _getPathsToValidate(doc) {
  let i;
  let len;
  const skipSchemaValidators = {***REMOVED***;

  _evaluateRequiredFunctions(doc);

  // only validate required fields when necessary
  let paths = Object.keys(doc.$__.activePaths.states.require).filter(function(path) {
    if (!doc.isSelected(path) && !doc.isModified(path)) {
      return false;
    ***REMOVED***
    if (path in doc.$__.cachedRequired) {
      return doc.$__.cachedRequired[path];
    ***REMOVED***
    return true;
  ***REMOVED***);

  paths = paths.concat(Object.keys(doc.$__.activePaths.states.init));
  paths = paths.concat(Object.keys(doc.$__.activePaths.states.modify));
  paths = paths.concat(Object.keys(doc.$__.activePaths.states.default));

  if (!doc.ownerDocument) {
    const subdocs = doc.$__getAllSubdocs();
    let subdoc;
    len = subdocs.length;
    const modifiedPaths = doc.modifiedPaths();
    for (i = 0; i < len; ++i) {
      subdoc = subdocs[i];
      if (doc.isModified(subdoc.$basePath, modifiedPaths) &&
          !doc.isDirectModified(subdoc.$basePath)) {
        // Remove child paths for now, because we'll be validating the whole
        // subdoc
        paths = paths.filter(function(p) {
          return p != null && p.indexOf(subdoc.$basePath + '.') !== 0;
        ***REMOVED***);
        paths.push(subdoc.$basePath);
        skipSchemaValidators[subdoc.$basePath] = true;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  // gh-661: if a whole array is modified, make sure to run validation on all
  // the children as well
  len = paths.length;
  for (i = 0; i < len; ++i) {
    const path = paths[i];

    const _pathType = doc.schema.path(path);
    if (!_pathType ||
        !_pathType.$isMongooseArray ||
        // To avoid potential performance issues, skip doc arrays whose children
        // are not required. `getPositionalPathType()` may be slow, so avoid
        // it unless we have a case of #6364
        (_pathType.$isMongooseDocumentArray && !get(_pathType, 'schemaOptions.required'))) {
      continue;
    ***REMOVED***

    const val = doc.getValue(path);
    if (val) {
      const numElements = val.length;
      for (let j = 0; j < numElements; ++j) {
        paths.push(path + '.' + j);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  const flattenOptions = { skipArrays: true ***REMOVED***;
  len = paths.length;
  for (i = 0; i < len; ++i) {
    const pathToCheck = paths[i];
    if (doc.schema.nested[pathToCheck]) {
      let _v = doc.getValue(pathToCheck);
      if (isMongooseObject(_v)) {
        _v = _v.toObject({ transform: false ***REMOVED***);
      ***REMOVED***
      const flat = flatten(_v, '', flattenOptions);
      const _subpaths = Object.keys(flat).map(function(p) {
        return pathToCheck + '.' + p;
      ***REMOVED***);
      paths = paths.concat(_subpaths);
    ***REMOVED***
  ***REMOVED***

  len = paths.length;
  for (i = 0; i < len; ++i) {
    const path = paths[i];
    const _pathType = doc.schema.path(path);
    if (!_pathType || !_pathType.$isSchemaMap) {
      continue;
    ***REMOVED***

    const val = doc.getValue(path);
    if (val == null) {
      continue;
    ***REMOVED***
    for (const key of val.keys()) {
      paths.push(path + '.' + key);
    ***REMOVED***
  ***REMOVED***

  return [paths, skipSchemaValidators];
***REMOVED***

/*!
 * ignore
 */

Document.prototype.$__validate = function(callback) {
  const _this = this;
  const _complete = () => {
    const err = this.$__.validationError;
    this.$__.validationError = undefined;
    this.$__.cachedRequired = {***REMOVED***;
    this.emit('validate', _this);
    this.constructor.emit('validate', _this);
    if (err) {
      for (const key in err.errors) {
        // Make sure cast errors persist
        if (!this.__parent && err.errors[key] instanceof MongooseError.CastError) {
          this.invalidate(key, err.errors[key]);
        ***REMOVED***
      ***REMOVED***

      return err;
    ***REMOVED***
  ***REMOVED***;

  // only validate required fields when necessary
  const pathDetails = _getPathsToValidate(this);
  const paths = pathDetails[0];
  const skipSchemaValidators = pathDetails[1];

  if (paths.length === 0) {
    return process.nextTick(function() {
      const error = _complete();
      if (error) {
        return _this.schema.s.hooks.execPost('validate:error', _this, [ _this], { error: error ***REMOVED***, function(error) {
          callback(error);
        ***REMOVED***);
      ***REMOVED***
      callback(null, _this);
    ***REMOVED***);
  ***REMOVED***

  const validated = {***REMOVED***;
  let total = 0;

  const complete = function() {
    const error = _complete();
    if (error) {
      return _this.schema.s.hooks.execPost('validate:error', _this, [ _this], { error: error ***REMOVED***, function(error) {
        callback(error);
      ***REMOVED***);
    ***REMOVED***
    callback(null, _this);
  ***REMOVED***;

  const validatePath = function(path) {
    if (path == null || validated[path]) {
      return;
    ***REMOVED***

    validated[path] = true;
    total++;

    process.nextTick(function() {
      const p = _this.schema.path(path);

      if (!p) {
        return --total || complete();
      ***REMOVED***

      // If user marked as invalid or there was a cast error, don't validate
      if (!_this.$isValid(path)) {
        --total || complete();
        return;
      ***REMOVED***

      const val = _this.getValue(path);
      const scope = path in _this.$__.pathsToScopes ?
        _this.$__.pathsToScopes[path] :
        _this;

      p.doValidate(val, function(err) {
        if (err && (!p.$isMongooseDocumentArray || err.$isArrayValidatorError)) {
          if (p.$isSingleNested &&
              err.name === 'ValidationError' &&
              p.schema.options.storeSubdocValidationError === false) {
            return --total || complete();
          ***REMOVED***
          _this.invalidate(path, err, undefined, true);
        ***REMOVED***
        --total || complete();
      ***REMOVED***, scope, { skipSchemaValidators: skipSchemaValidators[path] ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***;

  const numPaths = paths.length;
  for (let i = 0; i < numPaths; ++i) {
    validatePath(paths[i]);
  ***REMOVED***
***REMOVED***;

/**
 * Executes registered validation rules (skipping asynchronous validators) for this document.
 *
 * ####Note:
 *
 * This method is useful if you need synchronous validation.
 *
 * ####Example:
 *
 *     var err = doc.validateSync();
 *     if ( err ){
 *       handleError( err );
 *     ***REMOVED***
 *       // validation passed
 *     ***REMOVED***
 *
 * @param {Array|string***REMOVED*** pathsToValidate only validate the given paths
 * @return {MongooseError|undefined***REMOVED*** MongooseError if there are errors during validation, or undefined if there is no error.
 * @api public
 */

Document.prototype.validateSync = function(pathsToValidate) {
  const _this = this;

  if (typeof pathsToValidate === 'string') {
    pathsToValidate = pathsToValidate.split(' ');
  ***REMOVED***

  // only validate required fields when necessary
  const pathDetails = _getPathsToValidate(this);
  let paths = pathDetails[0];
  const skipSchemaValidators = pathDetails[1];

  if (pathsToValidate && pathsToValidate.length) {
    const tmp = [];
    for (let i = 0; i < paths.length; ++i) {
      if (pathsToValidate.indexOf(paths[i]) !== -1) {
        tmp.push(paths[i]);
      ***REMOVED***
    ***REMOVED***
    paths = tmp;
  ***REMOVED***

  const validating = {***REMOVED***;

  paths.forEach(function(path) {
    if (validating[path]) {
      return;
    ***REMOVED***

    validating[path] = true;

    const p = _this.schema.path(path);
    if (!p) {
      return;
    ***REMOVED***
    if (!_this.$isValid(path)) {
      return;
    ***REMOVED***

    const val = _this.getValue(path);
    const err = p.doValidateSync(val, _this, {
      skipSchemaValidators: skipSchemaValidators[path]
    ***REMOVED***);
    if (err && (!p.$isMongooseDocumentArray || err.$isArrayValidatorError)) {
      if (p.$isSingleNested &&
          err.name === 'ValidationError' &&
          p.schema.options.storeSubdocValidationError === false) {
        return;
      ***REMOVED***
      _this.invalidate(path, err, undefined, true);
    ***REMOVED***
  ***REMOVED***);

  const err = _this.$__.validationError;
  _this.$__.validationError = undefined;
  _this.emit('validate', _this);
  _this.constructor.emit('validate', _this);

  if (err) {
    for (const key in err.errors) {
      // Make sure cast errors persist
      if (err.errors[key] instanceof MongooseError.CastError) {
        _this.invalidate(key, err.errors[key]);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  return err;
***REMOVED***;

/**
 * Marks a path as invalid, causing validation to fail.
 *
 * The `errorMsg` argument will become the message of the `ValidationError`.
 *
 * The `value` argument (if passed) will be available through the `ValidationError.value` property.
 *
 *     doc.invalidate('size', 'must be less than 20', 14);

 *     doc.validate(function (err) {
 *       console.log(err)
 *       // prints
 *       { message: 'Validation failed',
 *         name: 'ValidationError',
 *         errors:
 *          { size:
 *             { message: 'must be less than 20',
 *               name: 'ValidatorError',
 *               path: 'size',
 *               type: 'user defined',
 *               value: 14 ***REMOVED*** ***REMOVED*** ***REMOVED***
 *     ***REMOVED***)
 *
 * @param {String***REMOVED*** path the field to invalidate
 * @param {String|Error***REMOVED*** errorMsg the error which states the reason `path` was invalid
 * @param {Object|String|Number|any***REMOVED*** value optional invalid value
 * @param {String***REMOVED*** [kind] optional `kind` property for the error
 * @return {ValidationError***REMOVED*** the current ValidationError, with all currently invalidated paths
 * @api public
 */

Document.prototype.invalidate = function(path, err, val, kind) {
  if (!this.$__.validationError) {
    this.$__.validationError = new ValidationError(this);
  ***REMOVED***

  if (this.$__.validationError.errors[path]) {
    return;
  ***REMOVED***

  if (!err || typeof err === 'string') {
    err = new ValidatorError({
      path: path,
      message: err,
      type: kind || 'user defined',
      value: val
    ***REMOVED***);
  ***REMOVED***

  if (this.$__.validationError === err) {
    return this.$__.validationError;
  ***REMOVED***

  this.$__.validationError.addError(path, err);
  return this.$__.validationError;
***REMOVED***;

/**
 * Marks a path as valid, removing existing validation errors.
 *
 * @param {String***REMOVED*** path the field to mark as valid
 * @api public
 * @memberOf Document
 * @instance
 * @method $markValid
 */

Document.prototype.$markValid = function(path) {
  if (!this.$__.validationError || !this.$__.validationError.errors[path]) {
    return;
  ***REMOVED***

  delete this.$__.validationError.errors[path];
  if (Object.keys(this.$__.validationError.errors).length === 0) {
    this.$__.validationError = null;
  ***REMOVED***
***REMOVED***;

/**
 * Saves this document.
 *
 * ####Example:
 *
 *     product.sold = Date.now();
 *     product.save(function (err, product) {
 *       if (err) ..
 *     ***REMOVED***)
 *
 * The callback will receive two parameters
 *
 * 1. `err` if an error occurred
 * 2. `product` which is the saved `product`
 *
 * As an extra measure of flow control, save will return a Promise.
 * ####Example:
 *     product.save().then(function(product) {
 *        ...
 *     ***REMOVED***);
 *
 * @param {Object***REMOVED*** [options] options optional options
 * @param {Object***REMOVED*** [options.safe] (DEPRECATED) overrides [schema's safe option](http://mongoosejs.com//docs/guide.html#safe)
 * @param {Boolean***REMOVED*** [options.validateBeforeSave] set to false to save without validating.
 * @param {Function***REMOVED*** [fn] optional callback
 * @method save
 * @memberOf Document
 * @instance
 * @return {Promise|undefined***REMOVED*** Returns undefined if used with callback or a Promise otherwise.
 * @api public
 * @see middleware http://mongoosejs.com/docs/middleware.html
 */

/**
 * Checks if a path is invalid
 *
 * @param {String***REMOVED*** path the field to check
 * @method $isValid
 * @memberOf Document
 * @instance
 * @api private
 */

Document.prototype.$isValid = function(path) {
  return !this.$__.validationError || !this.$__.validationError.errors[path];
***REMOVED***;

/**
 * Resets the internal modified state of this document.
 *
 * @api private
 * @return {Document***REMOVED***
 * @method $__reset
 * @memberOf Document
 * @instance
 */

Document.prototype.$__reset = function reset() {
  let _this = this;
  DocumentArray || (DocumentArray = require('./types/documentarray'));

  this.$__.activePaths
    .map('init', 'modify', function(i) {
      return _this.getValue(i);
    ***REMOVED***)
    .filter(function(val) {
      return val && val instanceof Array && val.isMongooseDocumentArray && val.length;
    ***REMOVED***)
    .forEach(function(array) {
      let i = array.length;
      while (i--) {
        const doc = array[i];
        if (!doc) {
          continue;
        ***REMOVED***
        doc.$__reset();
      ***REMOVED***

      _this.$__.activePaths.init(array._path);

      array._atomics = {***REMOVED***;
    ***REMOVED***);

  this.$__.activePaths.
    map('init', 'modify', function(i) {
      return _this.getValue(i);
    ***REMOVED***).
    filter(function(val) {
      return val && val.$isSingleNested;
    ***REMOVED***).
    forEach(function(doc) {
      doc.$__reset();
      _this.$__.activePaths.init(doc.$basePath);
    ***REMOVED***);

  // clear atomics
  this.$__dirty().forEach(function(dirt) {
    const type = dirt.value;

    if (type && type._atomics) {
      type._atomics = {***REMOVED***;
    ***REMOVED***
  ***REMOVED***);

  // Clear 'dirty' cache
  this.$__.activePaths.clear('modify');
  this.$__.activePaths.clear('default');
  this.$__.validationError = undefined;
  this.errors = undefined;
  _this = this;
  this.schema.requiredPaths().forEach(function(path) {
    _this.$__.activePaths.require(path);
  ***REMOVED***);

  return this;
***REMOVED***;

/**
 * Returns this documents dirty paths / vals.
 *
 * @api private
 * @method $__dirty
 * @memberOf Document
 * @instance
 */

Document.prototype.$__dirty = function() {
  const _this = this;

  let all = this.$__.activePaths.map('modify', function(path) {
    return {
      path: path,
      value: _this.getValue(path),
      schema: _this.$__path(path)
    ***REMOVED***;
  ***REMOVED***);

  // gh-2558: if we had to set a default and the value is not undefined,
  // we have to save as well
  all = all.concat(this.$__.activePaths.map('default', function(path) {
    if (path === '_id' || _this.getValue(path) == null) {
      return;
    ***REMOVED***
    return {
      path: path,
      value: _this.getValue(path),
      schema: _this.$__path(path)
    ***REMOVED***;
  ***REMOVED***));

  // Sort dirty paths in a flat hierarchy.
  all.sort(function(a, b) {
    return (a.path < b.path ? -1 : (a.path > b.path ? 1 : 0));
  ***REMOVED***);

  // Ignore "foo.a" if "foo" is dirty already.
  const minimal = [];
  let lastPath;
  let top;

  all.forEach(function(item) {
    if (!item) {
      return;
    ***REMOVED***
    if (item.path.indexOf(lastPath) !== 0) {
      lastPath = item.path + '.';
      minimal.push(item);
      top = item;
    ***REMOVED***
      // special case for top level MongooseArrays
      if (top.value && top.value._atomics && top.value.hasAtomics()) {
        // the `top` array itself and a sub path of `top` are being modified.
        // the only way to honor all of both modifications is through a $set
        // of entire array.
        top.value._atomics = {***REMOVED***;
        top.value._atomics.$set = top.value;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);

  top = lastPath = null;
  return minimal;
***REMOVED***;

/**
 * Assigns/compiles `schema` into this documents prototype.
 *
 * @param {Schema***REMOVED*** schema
 * @api private
 * @method $__setSchema
 * @memberOf Document
 * @instance
 */

Document.prototype.$__setSchema = function(schema) {
  schema.plugin(idGetter, { deduplicate: true ***REMOVED***);
  compile(schema.tree, this, undefined, schema.options);

  // Apply default getters if virtual doesn't have any (gh-6262)
  for (const key of Object.keys(schema.virtuals)) {
    schema.virtuals[key]._applyDefaultGetters();
  ***REMOVED***

  this.schema = schema;
***REMOVED***;


/**
 * Get active path that were changed and are arrays
 *
 * @api private
 * @method $__getArrayPathsToValidate
 * @memberOf Document
 * @instance
 */

Document.prototype.$__getArrayPathsToValidate = function() {
  DocumentArray || (DocumentArray = require('./types/documentarray'));

  // validate all document arrays.
  return this.$__.activePaths
    .map('init', 'modify', function(i) {
      return this.getValue(i);
    ***REMOVED***.bind(this))
    .filter(function(val) {
      return val && val instanceof Array && val.isMongooseDocumentArray && val.length;
    ***REMOVED***).reduce(function(seed, array) {
      return seed.concat(array);
    ***REMOVED***, [])
    .filter(function(doc) {
      return doc;
    ***REMOVED***);
***REMOVED***;


/**
 * Get all subdocs (by bfs)
 *
 * @api private
 * @method $__getAllSubdocs
 * @memberOf Document
 * @instance
 */

Document.prototype.$__getAllSubdocs = function() {
  DocumentArray || (DocumentArray = require('./types/documentarray'));
  Embedded = Embedded || require('./types/embedded');

  function docReducer(doc, seed, path) {
    const val = path ? doc[path] : doc;
    if (val instanceof Embedded) {
      seed.push(val);
    ***REMOVED***
    else if (val instanceof Map) {
      seed = Array.from(val.keys()).reduce(function(seed, path) {
        return docReducer(val.get(path), seed, null);
      ***REMOVED***, seed);
    ***REMOVED***
    else if (val && val.$isSingleNested) {
      seed = Object.keys(val._doc).reduce(function(seed, path) {
        return docReducer(val._doc, seed, path);
      ***REMOVED***, seed);
      seed.push(val);
    ***REMOVED***
    else if (val && val.isMongooseDocumentArray) {
      val.forEach(function _docReduce(doc) {
        if (!doc || !doc._doc) {
          return;
        ***REMOVED***
        if (doc instanceof Embedded) {
          seed.push(doc);
        ***REMOVED***
        seed = Object.keys(doc._doc).reduce(function(seed, path) {
          return docReducer(doc._doc, seed, path);
        ***REMOVED***, seed);
      ***REMOVED***);
    ***REMOVED*** else if (val instanceof Document && val.$__isNested) {
      if (val) {
        seed = Object.keys(val).reduce(function(seed, path) {
          return docReducer(val, seed, path);
        ***REMOVED***, seed);
      ***REMOVED***
    ***REMOVED***
    return seed;
  ***REMOVED***

  const _this = this;
  const subDocs = Object.keys(this._doc).reduce(function(seed, path) {
    return docReducer(_this, seed, path);
  ***REMOVED***, []);

  return subDocs;
***REMOVED***;

/*!
 * Runs queued functions
 */

function applyQueue(doc) {
  const q = doc.schema && doc.schema.callQueue;
  if (!q.length) {
    return;
  ***REMOVED***
  let pair;

  for (let i = 0; i < q.length; ++i) {
    pair = q[i];
    if (pair[0] !== 'pre' && pair[0] !== 'post' && pair[0] !== 'on') {
      doc[pair[0]].apply(doc, pair[1]);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/*!
 * ignore
 */

Document.prototype.$__handleReject = function handleReject(err) {
  // emit on the Model if listening
  if (this.listeners('error').length) {
    this.emit('error', err);
  ***REMOVED*** else if (this.constructor.listeners && this.constructor.listeners('error').length) {
    this.constructor.emit('error', err);
  ***REMOVED*** else if (this.listeners && this.listeners('error').length) {
    this.emit('error', err);
  ***REMOVED***
***REMOVED***;

/**
 * Internal helper for toObject() and toJSON() that doesn't manipulate options
 *
 * @api private
 * @method $toObject
 * @memberOf Document
 * @instance
 */

Document.prototype.$toObject = function(options, json) {
  let defaultOptions = {
    transform: true,
    flattenDecimals: true
  ***REMOVED***;

  const path = json ? 'toJSON' : 'toObject';
  const baseOptions = get(this, 'constructor.base.options.' + path, {***REMOVED***);
  const schemaOptions = get(this, 'schema.options', {***REMOVED***);
  // merge base default options with Schema's set default options if available.
  // `clone` is necessary here because `utils.options` directly modifies the second input.
  defaultOptions = utils.options(defaultOptions, clone(baseOptions));
  defaultOptions = utils.options(defaultOptions, clone(schemaOptions[path] || {***REMOVED***));

  // If options do not exist or is not an object, set it to empty object
  options = options && utils.getFunctionName(options.constructor) === 'Object' ?
    clone(options) :
    {***REMOVED***;

  let _minimize;
  if (options.minimize != null) {
    _minimize = options.minimize;
  ***REMOVED*** else if (defaultOptions.minimize != null) {
    _minimize = defaultOptions.minimize;
  ***REMOVED***
    _minimize = schemaOptions.minimize;
  ***REMOVED***

  // The original options that will be passed to `clone()`. Important because
  // `clone()` will recursively call `$toObject()` on embedded docs, so we
  // need the original options the user passed in, plus `_isNested` and
  // `_parentOptions` for checking whether we need to depopulate.
  const cloneOptions = Object.assign(utils.clone(options), {
    _isNested: true,
    json: json,
    minimize: _minimize
  ***REMOVED***);

  const depopulate = options.depopulate ||
    get(options, '_parentOptions.depopulate', false);
  // _isNested will only be true if this is not the top level document, we
  // should never depopulate
  if (depopulate && options._isNested && this.$__.wasPopulated) {
    // populated paths that we set to a document
    return clone(this._id, cloneOptions);
  ***REMOVED***

  // merge default options with input options.
  options = utils.options(defaultOptions, options);
  options._isNested = true;
  options.json = json;
  options.minimize = _minimize;

  cloneOptions._parentOptions = options;

  // remember the root transform function
  // to save it from being overwritten by sub-transform functions
  const originalTransform = options.transform;

  let ret = clone(this._doc, cloneOptions) || {***REMOVED***;

  if (options.getters) {
    applyGetters(this, ret, 'paths', cloneOptions);
    // applyGetters for paths will add nested empty objects;
    // if minimize is set, we need to remove them.
    if (options.minimize) {
      ret = minimize(ret) || {***REMOVED***;
    ***REMOVED***
  ***REMOVED***

  if (options.virtuals || options.getters && options.virtuals !== false) {
    applyGetters(this, ret, 'virtuals', cloneOptions);
  ***REMOVED***

  if (options.versionKey === false && this.schema.options.versionKey) {
    delete ret[this.schema.options.versionKey];
  ***REMOVED***

  let transform = options.transform;

  // In the case where a subdocument has its own transform function, we need to
  // check and see if the parent has a transform (options.transform) and if the
  // child schema has a transform (this.schema.options.toObject) In this case,
  // we need to adjust options.transform to be the child schema's transform and
  // not the parent schema's
  if (transform === true || (schemaOptions.toObject && transform)) {
    const opts = options.json ? schemaOptions.toJSON : schemaOptions.toObject;

    if (opts) {
      transform = (typeof options.transform === 'function' ? options.transform : opts.transform);
    ***REMOVED***
  ***REMOVED***
    options.transform = originalTransform;
  ***REMOVED***

  if (typeof transform === 'function') {
    const xformed = transform(this, ret, options);
    if (typeof xformed !== 'undefined') {
      ret = xformed;
    ***REMOVED***
  ***REMOVED***

  return ret;
***REMOVED***;

/**
 * Converts this document into a plain javascript object, ready for storage in MongoDB.
 *
 * Buffers are converted to instances of [mongodb.Binary](http://mongodb.github.com/node-mongodb-native/api-bson-generated/binary.html) for proper storage.
 *
 * ####Options:
 *
 * - `getters` apply all getters (path and virtual getters), defaults to false
 * - `virtuals` apply virtual getters (can override `getters` option), defaults to false
 * - `minimize` remove empty objects (defaults to true)
 * - `transform` a transform function to apply to the resulting document before returning
 * - `depopulate` depopulate any populated paths, replacing them with their original refs (defaults to false)
 * - `versionKey` whether to include the version key (defaults to true)
 *
 * ####Getters/Virtuals
 *
 * Example of only applying path getters
 *
 *     doc.toObject({ getters: true, virtuals: false ***REMOVED***)
 *
 * Example of only applying virtual getters
 *
 *     doc.toObject({ virtuals: true ***REMOVED***)
 *
 * Example of applying both path and virtual getters
 *
 *     doc.toObject({ getters: true ***REMOVED***)
 *
 * To apply these options to every document of your schema by default, set your [schemas](#schema_Schema) `toObject` option to the same argument.
 *
 *     schema.set('toObject', { virtuals: true ***REMOVED***)
 *
 * ####Transform
 *
 * We may need to perform a transformation of the resulting object based on some criteria, say to remove some sensitive information or return a custom object. In this case we set the optional `transform` function.
 *
 * Transform functions receive three arguments
 *
 *     function (doc, ret, options) {***REMOVED***
 *
 * - `doc` The mongoose document which is being converted
 * - `ret` The plain object representation which has been converted
 * - `options` The options in use (either schema options or the options passed inline)
 *
 * ####Example
 *
 *     // specify the transform schema option
 *     if (!schema.options.toObject) schema.options.toObject = {***REMOVED***;
 *     schema.options.toObject.transform = function (doc, ret, options) {
 *       // remove the _id of every document before returning the result
 *       delete ret._id;
 *       return ret;
 *     ***REMOVED***
 *
 *     // without the transformation in the schema
 *     doc.toObject(); // { _id: 'anId', name: 'Wreck-it Ralph' ***REMOVED***
 *
 *     // with the transformation
 *     doc.toObject(); // { name: 'Wreck-it Ralph' ***REMOVED***
 *
 * With transformations we can do a lot more than remove properties. We can even return completely new customized objects:
 *
 *     if (!schema.options.toObject) schema.options.toObject = {***REMOVED***;
 *     schema.options.toObject.transform = function (doc, ret, options) {
 *       return { movie: ret.name ***REMOVED***
 *     ***REMOVED***
 *
 *     // without the transformation in the schema
 *     doc.toObject(); // { _id: 'anId', name: 'Wreck-it Ralph' ***REMOVED***
 *
 *     // with the transformation
 *     doc.toObject(); // { movie: 'Wreck-it Ralph' ***REMOVED***
 *
 * _Note: if a transform function returns `undefined`, the return value will be ignored._
 *
 * Transformations may also be applied inline, overridding any transform set in the options:
 *
 *     function xform (doc, ret, options) {
 *       return { inline: ret.name, custom: true ***REMOVED***
 *     ***REMOVED***
 *
 *     // pass the transform as an inline option
 *     doc.toObject({ transform: xform ***REMOVED***); // { inline: 'Wreck-it Ralph', custom: true ***REMOVED***
 *
 * If you want to skip transformations, use `transform: false`:
 *
 *     if (!schema.options.toObject) schema.options.toObject = {***REMOVED***;
 *     schema.options.toObject.hide = '_id';
 *     schema.options.toObject.transform = function (doc, ret, options) {
 *       if (options.hide) {
 *         options.hide.split(' ').forEach(function (prop) {
 *           delete ret[prop];
 *         ***REMOVED***);
 *       ***REMOVED***
 *       return ret;
 *     ***REMOVED***
 *
 *     var doc = new Doc({ _id: 'anId', secret: 47, name: 'Wreck-it Ralph' ***REMOVED***);
 *     doc.toObject();                                        // { secret: 47, name: 'Wreck-it Ralph' ***REMOVED***
 *     doc.toObject({ hide: 'secret _id', transform: false ***REMOVED***);// { _id: 'anId', secret: 47, name: 'Wreck-it Ralph' ***REMOVED***
 *     doc.toObject({ hide: 'secret _id', transform: true ***REMOVED***); // { name: 'Wreck-it Ralph' ***REMOVED***
 *
 * Transforms are applied _only to the document and are not applied to sub-documents_.
 *
 * Transforms, like all of these options, are also available for `toJSON`.
 *
 * See [schema options](/docs/guide.html#toObject) for some more details.
 *
 * _During save, no custom options are applied to the document before being sent to the database._
 *
 * @param {Object***REMOVED*** [options]
 * @param {Boolean***REMOVED*** [options.getters=false] if true, apply all getters, including virtuals
 * @param {Boolean***REMOVED*** [options.virtuals=false] if true, apply virtuals. Use `{ getters: true, virtuals: false ***REMOVED***` to just apply getters, not virtuals
 * @param {Boolean***REMOVED*** [options.minimize=true] if true, omit any empty objects from the output
 * @param {Function|null***REMOVED*** [options.transform=null] if set, mongoose will call this function to allow you to transform the returned object
 * @param {Boolean***REMOVED*** [options.depopulate=false] if true, replace any conventionally populated paths with the original id in the output. Has no affect on virtual populated paths.
 * @param {Boolean***REMOVED*** [options.versionKey=true] if false, exclude the version key (`__v` by default) from the output
 * @return {Object***REMOVED*** js object
 * @see mongodb.Binary http://mongodb.github.com/node-mongodb-native/api-bson-generated/binary.html
 * @api public
 * @memberOf Document
 * @instance
 */

Document.prototype.toObject = function(options) {
  return this.$toObject(options);
***REMOVED***;

/*!
 * Minimizes an object, removing undefined values and empty objects
 *
 * @param {Object***REMOVED*** object to minimize
 * @return {Object***REMOVED***
 */

function minimize(obj) {
  const keys = Object.keys(obj);
  let i = keys.length;
  let hasKeys;
  let key;
  let val;

  while (i--) {
    key = keys[i];
    val = obj[key];

    if (utils.isObject(val) && !Buffer.isBuffer(val)) {
      obj[key] = minimize(val);
    ***REMOVED***

    if (undefined === obj[key]) {
      delete obj[key];
      continue;
    ***REMOVED***

    hasKeys = true;
  ***REMOVED***

  return hasKeys
    ? obj
    : undefined;
***REMOVED***

/*!
 * Applies virtuals properties to `json`.
 *
 * @param {Document***REMOVED*** self
 * @param {Object***REMOVED*** json
 * @param {String***REMOVED*** type either `virtuals` or `paths`
 * @return {Object***REMOVED*** `json`
 */

function applyGetters(self, json, type, options) {
  const schema = self.schema;
  const paths = Object.keys(schema[type]);
  let i = paths.length;
  const numPaths = i;
  let path;
  let assignPath;
  let cur = self._doc;
  let v;

  if (!cur) {
    return json;
  ***REMOVED***

  if (type === 'virtuals') {
    options = options || {***REMOVED***;
    for (i = 0; i < numPaths; ++i) {
      path = paths[i];
      // We may be applying virtuals to a nested object, for example if calling
      // `doc.nestedProp.toJSON()`. If so, the path we assign to, `assignPath`,
      // will be a trailing substring of the `path`.
      assignPath = path;
      if (options.path != null) {
        if (!path.startsWith(options.path + '.')) {
          continue;
        ***REMOVED***
        assignPath = path.substr(options.path.length + 1);
      ***REMOVED***
      const parts = assignPath.split('.');
      v = clone(self.get(path), options);
      if (v === void 0) {
        continue;
      ***REMOVED***
      const plen = parts.length;
      cur = json;
      for (let j = 0; j < plen - 1; ++j) {
        cur[parts[j]] = cur[parts[j]] || {***REMOVED***;
        cur = cur[parts[j]];
      ***REMOVED***
      cur[parts[plen - 1]] = v;
    ***REMOVED***

    return json;
  ***REMOVED***

  while (i--) {
    path = paths[i];

    const parts = path.split('.');
    const plen = parts.length;
    const last = plen - 1;
    let branch = json;
    let part;
    cur = self._doc;

    for (let ii = 0; ii < plen; ++ii) {
      part = parts[ii];
      v = cur[part];
      if (ii === last) {
        branch[part] = clone(self.get(path), options);
      ***REMOVED*** else if (v == null) {
        if (part in cur) {
          branch[part] = v;
        ***REMOVED***
        break;
      ***REMOVED***
        branch = branch[part] || (branch[part] = {***REMOVED***);
      ***REMOVED***
      cur = v;
    ***REMOVED***
  ***REMOVED***

  return json;
***REMOVED***

/**
 * The return value of this method is used in calls to JSON.stringify(doc).
 *
 * This method accepts the same options as [Document#toObject](#document_Document-toObject). To apply the options to every document of your schema by default, set your [schemas](#schema_Schema) `toJSON` option to the same argument.
 *
 *     schema.set('toJSON', { virtuals: true ***REMOVED***)
 *
 * See [schema options](/docs/guide.html#toJSON) for details.
 *
 * @param {Object***REMOVED*** options
 * @return {Object***REMOVED***
 * @see Document#toObject #document_Document-toObject
 * @api public
 * @memberOf Document
 * @instance
 */

Document.prototype.toJSON = function(options) {
  return this.$toObject(options, true);
***REMOVED***;

/**
 * Helper for console.log
 *
 * @api public
 * @method inspect
 * @memberOf Document
 * @instance
 */

Document.prototype.inspect = function(options) {
  const isPOJO = options &&
    utils.getFunctionName(options.constructor) === 'Object';
  let opts;
  if (isPOJO) {
    opts = options;
    opts.minimize = false;
  ***REMOVED***
  return this.toObject(opts);
***REMOVED***;

if (inspect.custom) {
  /*!
  * Avoid Node deprecation warning DEP0079
  */

  Document.prototype[inspect.custom] = Document.prototype.inspect;
***REMOVED***

/**
 * Helper for console.log
 *
 * @api public
 * @method toString
 * @memberOf Document
 * @instance
 */

Document.prototype.toString = function() {
  return inspect(this.inspect());
***REMOVED***;

/**
 * Returns true if the Document stores the same data as doc.
 *
 * Documents are considered equal when they have matching `_id`s, unless neither
 * document has an `_id`, in which case this function falls back to using
 * `deepEqual()`.
 *
 * @param {Document***REMOVED*** doc a document to compare
 * @return {Boolean***REMOVED***
 * @api public
 * @memberOf Document
 * @instance
 */

Document.prototype.equals = function(doc) {
  if (!doc) {
    return false;
  ***REMOVED***

  const tid = this.get('_id');
  const docid = doc.get ? doc.get('_id') : doc;
  if (!tid && !docid) {
    return deepEqual(this, doc);
  ***REMOVED***
  return tid && tid.equals
    ? tid.equals(docid)
    : tid === docid;
***REMOVED***;

/**
 * Populates document references, executing the `callback` when complete.
 * If you want to use promises instead, use this function with
 * [`execPopulate()`](#document_Document-execPopulate)
 *
 * ####Example:
 *
 *     doc
 *     .populate('company')
 *     .populate({
 *       path: 'notes',
 *       match: /airline/,
 *       select: 'text',
 *       model: 'modelName'
 *       options: opts
 *     ***REMOVED***, function (err, user) {
 *       assert(doc._id === user._id) // the document itself is passed
 *     ***REMOVED***)
 *
 *     // summary
 *     doc.populate(path)                   // not executed
 *     doc.populate(options);               // not executed
 *     doc.populate(path, callback)         // executed
 *     doc.populate(options, callback);     // executed
 *     doc.populate(callback);              // executed
 *     doc.populate(options).execPopulate() // executed, returns promise
 *
 *
 * ####NOTE:
 *
 * Population does not occur unless a `callback` is passed *or* you explicitly
 * call `execPopulate()`.
 * Passing the same path a second time will overwrite the previous path options.
 * See [Model.populate()](#model_Model.populate) for explaination of options.
 *
 * @see Model.populate #model_Model.populate
 * @see Document.execPopulate #document_Document-execPopulate
 * @param {String|Object***REMOVED*** [path] The path to populate or an options object
 * @param {Function***REMOVED*** [callback] When passed, population is invoked
 * @api public
 * @return {Document***REMOVED*** this
 * @memberOf Document
 * @instance
 */

Document.prototype.populate = function populate() {
  if (arguments.length === 0) {
    return this;
  ***REMOVED***

  const pop = this.$__.populate || (this.$__.populate = {***REMOVED***);
  const args = utils.args(arguments);
  let fn;

  if (typeof args[args.length - 1] === 'function') {
    fn = args.pop();
  ***REMOVED***

  // allow `doc.populate(callback)`
  if (args.length) {
    // use hash to remove duplicate paths
    const res = utils.populate.apply(null, args);
    for (let i = 0; i < res.length; ++i) {
      pop[res[i].path] = res[i];
    ***REMOVED***
  ***REMOVED***

  if (fn) {
    const paths = utils.object.vals(pop);
    this.$__.populate = undefined;
    let topLevelModel = this.constructor;
    if (this.$__isNested) {
      topLevelModel = this.$__.scope.constructor;
      const nestedPath = this.$__.nestedPath;
      paths.forEach(function(populateOptions) {
        populateOptions.path = nestedPath + '.' + populateOptions.path;
      ***REMOVED***);
    ***REMOVED***

    // Use `$session()` by default if the document has an associated session
    // See gh-6754
    if (this.$session() != null) {
      const session = this.$session();
      paths.forEach(path => {
        if (path.options == null) {
          path.options = { session: session ***REMOVED***;
          return;
        ***REMOVED***
        if (!('session' in path.options)) {
          path.options.session = session;
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***

    topLevelModel.populate(this, paths, fn);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Explicitly executes population and returns a promise. Useful for ES2015
 * integration.
 *
 * ####Example:
 *
 *     var promise = doc.
 *       populate('company').
 *       populate({
 *         path: 'notes',
 *         match: /airline/,
 *         select: 'text',
 *         model: 'modelName'
 *         options: opts
 *       ***REMOVED***).
 *       execPopulate();
 *
 *     // summary
 *     doc.execPopulate().then(resolve, reject);
 *
 *
 * @see Document.populate #document_Document-populate
 * @api public
 * @param {Function***REMOVED*** [callback] optional callback. If specified, a promise will **not** be returned
 * @return {Promise***REMOVED*** promise that resolves to the document when population is done
 * @memberOf Document
 * @instance
 */

Document.prototype.execPopulate = function(callback) {
  return utils.promiseOrCallback(callback, cb => {
    this.populate(cb);
  ***REMOVED***);
***REMOVED***;

/**
 * Gets _id(s) used during population of the given `path`.
 *
 * ####Example:
 *
 *     Model.findOne().populate('author').exec(function (err, doc) {
 *       console.log(doc.author.name)         // Dr.Seuss
 *       console.log(doc.populated('author')) // '5144cf8050f071d979c118a7'
 *     ***REMOVED***)
 *
 * If the path was not populated, undefined is returned.
 *
 * @param {String***REMOVED*** path
 * @return {Array|ObjectId|Number|Buffer|String|undefined***REMOVED***
 * @memberOf Document
 * @instance
 * @api public
 */

Document.prototype.populated = function(path, val, options) {
  // val and options are internal
  if (val === null || val === void 0) {
    if (!this.$__.populated) {
      return undefined;
    ***REMOVED***
    const v = this.$__.populated[path];
    if (v) {
      return v.value;
    ***REMOVED***
    return undefined;
  ***REMOVED***

  // internal

  if (val === true) {
    if (!this.$__.populated) {
      return undefined;
    ***REMOVED***
    return this.$__.populated[path];
  ***REMOVED***

  this.$__.populated || (this.$__.populated = {***REMOVED***);
  this.$__.populated[path] = {value: val, options: options***REMOVED***;
  return val;
***REMOVED***;

/**
 * Takes a populated field and returns it to its unpopulated state.
 *
 * ####Example:
 *
 *     Model.findOne().populate('author').exec(function (err, doc) {
 *       console.log(doc.author.name); // Dr.Seuss
 *       console.log(doc.depopulate('author'));
 *       console.log(doc.author); // '5144cf8050f071d979c118a7'
 *     ***REMOVED***)
 *
 * If the path was not populated, this is a no-op.
 *
 * @param {String***REMOVED*** path
 * @return {Document***REMOVED*** this
 * @see Document.populate #document_Document-populate
 * @api public
 * @memberOf Document
 * @instance
 */

Document.prototype.depopulate = function(path) {
  if (typeof path === 'string') {
    path = path.split(' ');
  ***REMOVED***
  let i;
  let populatedIds;
  const virtualKeys = this.$$populatedVirtuals ? Object.keys(this.$$populatedVirtuals) : [];

  if (arguments.length === 0) {
    // Depopulate all
    const keys = this.$__.populated ? Object.keys(this.$__.populated) : [];

    for (i = 0; i < virtualKeys.length; i++) {
      delete this.$$populatedVirtuals[virtualKeys[i]];
      delete this._doc[virtualKeys[i]];
    ***REMOVED***

    for (i = 0; i < keys.length; i++) {
      populatedIds = this.populated(keys[i]);
      if (!populatedIds) {
        continue;
      ***REMOVED***
      delete this.$__.populated[keys[i]];
      this.$set(keys[i], populatedIds);
    ***REMOVED***
    return this;
  ***REMOVED***

  for (i = 0; i < path.length; i++) {
    populatedIds = this.populated(path[i]);
    if (!populatedIds) {
      if (virtualKeys.indexOf(path[i]) === -1) {
        continue;
      ***REMOVED***
        delete this.$$populatedVirtuals[path[i]];
        delete this._doc[path[i]];
        continue;
      ***REMOVED***
    ***REMOVED***
    delete this.$__.populated[path[i]];
    this.$set(path[i], populatedIds);
  ***REMOVED***
  return this;
***REMOVED***;


/**
 * Returns the full path to this document.
 *
 * @param {String***REMOVED*** [path]
 * @return {String***REMOVED***
 * @api private
 * @method $__fullPath
 * @memberOf Document
 * @instance
 */

Document.prototype.$__fullPath = function(path) {
  // overridden in SubDocuments
  return path || '';
***REMOVED***;

/*!
 * Module exports.
 */

Document.ValidationError = ValidationError;
module.exports = exports = Document;
