// mongoose/lib/schematype.js
'use strict';

/*!
 * Module dependencies.
 */

const MongooseError = require('./error');
const $exists = require('./schema/operators/exists');
const $type = require('./schema/operators/type');
const get = require('lodash.get');
const immediate = require('./helpers/immediate');
const utils = require('./utils');
const validatorErrorSymbol = require('./helpers/symbols').validatorErrorSymbol;

const CastError = MongooseError.CastError;
const ValidatorError = MongooseError.ValidatorError;

/**
 * SchemaType constructor. Do **not** instantiate `SchemaType` directly.
 * Mongoose converts your schema paths into SchemaTypes automatically.
 *
 * ####Example:
 *
 *     const schema = new Schema({ name: String ***REMOVED***);
 *     schema.path('name') instanceof SchemaType; // true
 *
 * @param {String***REMOVED*** path
 * @param {Object***REMOVED*** [options]
 * @param {String***REMOVED*** [instance]
 * @api public
 */

function SchemaType(path, options, instance) {
  this.path = path;
  this.instance = instance;
  this.validators = [];
  this.setters = [];
  this.getters = [];
  this.options = options;
  this._index = null;
  this.selected;

  for (const prop in options) {
    if (this[prop] && typeof this[prop] === 'function') {
      // { unique: true, index: true ***REMOVED***
      if (prop === 'index' && this._index) {
        continue;
      ***REMOVED***

      const val = options[prop];
      // Special case so we don't screw up array defaults, see gh-5780
      if (prop === 'default') {
        this.default(val);
        continue;
      ***REMOVED***

      const opts = Array.isArray(val) ? val : [val];

      this[prop].apply(this, opts);
    ***REMOVED***
  ***REMOVED***

  Object.defineProperty(this, '$$context', {
    enumerable: false,
    configurable: false,
    writable: true,
    value: null
  ***REMOVED***);
***REMOVED***

/**
 * Sets a default value for this SchemaType.
 *
 * ####Example:
 *
 *     var schema = new Schema({ n: { type: Number, default: 10 ***REMOVED***)
 *     var M = db.model('M', schema)
 *     var m = new M;
 *     console.log(m.n) // 10
 *
 * Defaults can be either `functions` which return the value to use as the default or the literal value itself. Either way, the value will be cast based on its schema type before being set during document creation.
 *
 * ####Example:
 *
 *     // values are cast:
 *     var schema = new Schema({ aNumber: { type: Number, default: 4.815162342 ***REMOVED******REMOVED***)
 *     var M = db.model('M', schema)
 *     var m = new M;
 *     console.log(m.aNumber) // 4.815162342
 *
 *     // default unique objects for Mixed types:
 *     var schema = new Schema({ mixed: Schema.Types.Mixed ***REMOVED***);
 *     schema.path('mixed').default(function () {
 *       return {***REMOVED***;
 *     ***REMOVED***);
 *
 *     // if we don't use a function to return object literals for Mixed defaults,
 *     // each document will receive a reference to the same object literal creating
 *     // a "shared" object instance:
 *     var schema = new Schema({ mixed: Schema.Types.Mixed ***REMOVED***);
 *     schema.path('mixed').default({***REMOVED***);
 *     var M = db.model('M', schema);
 *     var m1 = new M;
 *     m1.mixed.added = 1;
 *     console.log(m1.mixed); // { added: 1 ***REMOVED***
 *     var m2 = new M;
 *     console.log(m2.mixed); // { added: 1 ***REMOVED***
 *
 * @param {Function|any***REMOVED*** val the default value
 * @return {defaultValue***REMOVED***
 * @api public
 */

SchemaType.prototype.default = function(val) {
  if (arguments.length === 1) {
    if (val === void 0) {
      this.defaultValue = void 0;
      return void 0;
    ***REMOVED***
    this.defaultValue = val;
    return this.defaultValue;
  ***REMOVED*** else if (arguments.length > 1) {
    this.defaultValue = utils.args(arguments);
  ***REMOVED***
  return this.defaultValue;
***REMOVED***;

/**
 * Declares the index options for this schematype.
 *
 * ####Example:
 *
 *     var s = new Schema({ name: { type: String, index: true ***REMOVED***)
 *     var s = new Schema({ loc: { type: [Number], index: 'hashed' ***REMOVED***)
 *     var s = new Schema({ loc: { type: [Number], index: '2d', sparse: true ***REMOVED***)
 *     var s = new Schema({ loc: { type: [Number], index: { type: '2dsphere', sparse: true ***REMOVED******REMOVED***)
 *     var s = new Schema({ date: { type: Date, index: { unique: true, expires: '1d' ***REMOVED******REMOVED***)
 *     Schema.path('my.path').index(true);
 *     Schema.path('my.date').index({ expires: 60 ***REMOVED***);
 *     Schema.path('my.path').index({ unique: true, sparse: true ***REMOVED***);
 *
 * ####NOTE:
 *
 * _Indexes are created [in the background](https://docs.mongodb.com/manual/core/index-creation/#index-creation-background)
 * by default. If `background` is set to `false`, MongoDB will not execute any
 * read/write operations you send until the index build.
 * Specify `background: false` to override Mongoose's default._
 *
 * @param {Object|Boolean|String***REMOVED*** options
 * @return {SchemaType***REMOVED*** this
 * @api public
 */

SchemaType.prototype.index = function(options) {
  this._index = options;
  utils.expires(this._index);
  return this;
***REMOVED***;

/**
 * Declares an unique index.
 *
 * ####Example:
 *
 *     var s = new Schema({ name: { type: String, unique: true ***REMOVED******REMOVED***);
 *     Schema.path('name').index({ unique: true ***REMOVED***);
 *
 * _NOTE: violating the constraint returns an `E11000` error from MongoDB when saving, not a Mongoose validation error._
 *
 * @param {Boolean***REMOVED*** bool
 * @return {SchemaType***REMOVED*** this
 * @api public
 */

SchemaType.prototype.unique = function(bool) {
  if (this._index === false) {
    if (!bool) {
      return;
    ***REMOVED***
    throw new Error('Path "' + this.path + '" may not have `index` set to ' +
      'false and `unique` set to true');
  ***REMOVED***
  if (this._index == null || this._index === true) {
    this._index = {***REMOVED***;
  ***REMOVED*** else if (typeof this._index === 'string') {
    this._index = {type: this._index***REMOVED***;
  ***REMOVED***

  this._index.unique = bool;
  return this;
***REMOVED***;

/**
 * Declares a full text index.
 *
 * ###Example:
 *
 *      var s = new Schema({name : {type: String, text : true ***REMOVED***)
 *      Schema.path('name').index({text : true***REMOVED***);
 * @param {Boolean***REMOVED*** bool
 * @return {SchemaType***REMOVED*** this
 * @api public
 */

SchemaType.prototype.text = function(bool) {
  if (this._index === null || this._index === undefined ||
    typeof this._index === 'boolean') {
    this._index = {***REMOVED***;
  ***REMOVED*** else if (typeof this._index === 'string') {
    this._index = {type: this._index***REMOVED***;
  ***REMOVED***

  this._index.text = bool;
  return this;
***REMOVED***;

/**
 * Declares a sparse index.
 *
 * ####Example:
 *
 *     var s = new Schema({ name: { type: String, sparse: true ***REMOVED***)
 *     Schema.path('name').index({ sparse: true ***REMOVED***);
 *
 * @param {Boolean***REMOVED*** bool
 * @return {SchemaType***REMOVED*** this
 * @api public
 */

SchemaType.prototype.sparse = function(bool) {
  if (this._index === null || this._index === undefined ||
    typeof this._index === 'boolean') {
    this._index = {***REMOVED***;
  ***REMOVED*** else if (typeof this._index === 'string') {
    this._index = {type: this._index***REMOVED***;
  ***REMOVED***

  this._index.sparse = bool;
  return this;
***REMOVED***;

/**
 * Adds a setter to this schematype.
 *
 * ####Example:
 *
 *     function capitalize (val) {
 *       if (typeof val !== 'string') val = '';
 *       return val.charAt(0).toUpperCase() + val.substring(1);
 *     ***REMOVED***
 *
 *     // defining within the schema
 *     var s = new Schema({ name: { type: String, set: capitalize ***REMOVED******REMOVED***);
 *
 *     // or with the SchemaType
 *     var s = new Schema({ name: String ***REMOVED***)
 *     s.path('name').set(capitalize);
 *
 * Setters allow you to transform the data before it gets to the raw mongodb
 * document or query.
 *
 * Suppose you are implementing user registration for a website. Users provide
 * an email and password, which gets saved to mongodb. The email is a string
 * that you will want to normalize to lower case, in order to avoid one email
 * having more than one account -- e.g., otherwise, avenue@q.com can be registered for 2 accounts via avenue@q.com and AvEnUe@Q.CoM.
 *
 * You can set up email lower case normalization easily via a Mongoose setter.
 *
 *     function toLower(v) {
 *       return v.toLowerCase();
 *     ***REMOVED***
 *
 *     var UserSchema = new Schema({
 *       email: { type: String, set: toLower ***REMOVED***
 *     ***REMOVED***);
 *
 *     var User = db.model('User', UserSchema);
 *
 *     var user = new User({email: 'AVENUE@Q.COM'***REMOVED***);
 *     console.log(user.email); // 'avenue@q.com'
 *
 *     // or
 *     var user = new User();
 *     user.email = 'Avenue@Q.com';
 *     console.log(user.email); // 'avenue@q.com'
 *     User.updateOne({ _id: _id ***REMOVED***, { $set: { email: 'AVENUE@Q.COM' ***REMOVED*** ***REMOVED***); // update to 'avenue@q.com'
 *
 * As you can see above, setters allow you to transform the data before it
 * stored in MongoDB, or before executing a query.
 *
 * _NOTE: we could have also just used the built-in `lowercase: true` SchemaType option instead of defining our own function._
 *
 *     new Schema({ email: { type: String, lowercase: true ***REMOVED******REMOVED***)
 *
 * Setters are also passed a second argument, the schematype on which the setter was defined. This allows for tailored behavior based on options passed in the schema.
 *
 *     function inspector (val, schematype) {
 *       if (schematype.options.required) {
 *         return schematype.path + ' is required';
 *       ***REMOVED***
 *         return val;
 *       ***REMOVED***
 *     ***REMOVED***
 *
 *     var VirusSchema = new Schema({
 *       name: { type: String, required: true, set: inspector ***REMOVED***,
 *       taxonomy: { type: String, set: inspector ***REMOVED***
 *     ***REMOVED***)
 *
 *     var Virus = db.model('Virus', VirusSchema);
 *     var v = new Virus({ name: 'Parvoviridae', taxonomy: 'Parvovirinae' ***REMOVED***);
 *
 *     console.log(v.name);     // name is required
 *     console.log(v.taxonomy); // Parvovirinae
 *
 * You can also use setters to modify other properties on the document. If
 * you're setting a property `name` on a document, the setter will run with
 * `this` as the document. Be careful, in mongoose 5 setters will also run
 * when querying by `name` with `this` as the query.
 *
 * ```javascript
 * const nameSchema = new Schema({ name: String, keywords: [String] ***REMOVED***);
 * nameSchema.path('name').set(function(v) {
 *   // Need to check if `this` is a document, because in mongoose 5
 *   // setters will also run on queries, in which case `this` will be a
 *   // mongoose query object.
 *   if (this instanceof Document && v != null) {
 *     this.keywords = v.split(' ');
 *   ***REMOVED***
 *   return v;
 * ***REMOVED***);
 * ```
 *
 * @param {Function***REMOVED*** fn
 * @return {SchemaType***REMOVED*** this
 * @api public
 */

SchemaType.prototype.set = function(fn) {
  if (typeof fn !== 'function') {
    throw new TypeError('A setter must be a function.');
  ***REMOVED***
  this.setters.push(fn);
  return this;
***REMOVED***;

/**
 * Adds a getter to this schematype.
 *
 * ####Example:
 *
 *     function dob (val) {
 *       if (!val) return val;
 *       return (val.getMonth() + 1) + "/" + val.getDate() + "/" + val.getFullYear();
 *     ***REMOVED***
 *
 *     // defining within the schema
 *     var s = new Schema({ born: { type: Date, get: dob ***REMOVED***)
 *
 *     // or by retreiving its SchemaType
 *     var s = new Schema({ born: Date ***REMOVED***)
 *     s.path('born').get(dob)
 *
 * Getters allow you to transform the representation of the data as it travels from the raw mongodb document to the value that you see.
 *
 * Suppose you are storing credit card numbers and you want to hide everything except the last 4 digits to the mongoose user. You can do so by defining a getter in the following way:
 *
 *     function obfuscate (cc) {
 *       return '****-****-****-' + cc.slice(cc.length-4, cc.length);
 *     ***REMOVED***
 *
 *     var AccountSchema = new Schema({
 *       creditCardNumber: { type: String, get: obfuscate ***REMOVED***
 *     ***REMOVED***);
 *
 *     var Account = db.model('Account', AccountSchema);
 *
 *     Account.findById(id, function (err, found) {
 *       console.log(found.creditCardNumber); // '****-****-****-1234'
 *     ***REMOVED***);
 *
 * Getters are also passed a second argument, the schematype on which the getter was defined. This allows for tailored behavior based on options passed in the schema.
 *
 *     function inspector (val, schematype) {
 *       if (schematype.options.required) {
 *         return schematype.path + ' is required';
 *       ***REMOVED***
 *         return schematype.path + ' is not';
 *       ***REMOVED***
 *     ***REMOVED***
 *
 *     var VirusSchema = new Schema({
 *       name: { type: String, required: true, get: inspector ***REMOVED***,
 *       taxonomy: { type: String, get: inspector ***REMOVED***
 *     ***REMOVED***)
 *
 *     var Virus = db.model('Virus', VirusSchema);
 *
 *     Virus.findById(id, function (err, virus) {
 *       console.log(virus.name);     // name is required
 *       console.log(virus.taxonomy); // taxonomy is not
 *     ***REMOVED***)
 *
 * @param {Function***REMOVED*** fn
 * @return {SchemaType***REMOVED*** this
 * @api public
 */

SchemaType.prototype.get = function(fn) {
  if (typeof fn !== 'function') {
    throw new TypeError('A getter must be a function.');
  ***REMOVED***
  this.getters.push(fn);
  return this;
***REMOVED***;

/**
 * Adds validator(s) for this document path.
 *
 * Validators always receive the value to validate as their first argument and
 * must return `Boolean`. Returning `false` or throwing an error means
 * validation failed.
 *
 * The error message argument is optional. If not passed, the [default generic error message template](#error_messages_MongooseError-messages) will be used.
 *
 * ####Examples:
 *
 *     // make sure every value is equal to "something"
 *     function validator (val) {
 *       return val == 'something';
 *     ***REMOVED***
 *     new Schema({ name: { type: String, validate: validator ***REMOVED******REMOVED***);
 *
 *     // with a custom error message
 *
 *     var custom = [validator, 'Uh oh, {PATH***REMOVED*** does not equal "something".']
 *     new Schema({ name: { type: String, validate: custom ***REMOVED******REMOVED***);
 *
 *     // adding many validators at a time
 *
 *     var many = [
 *         { validator: validator, msg: 'uh oh' ***REMOVED***
 *       , { validator: anotherValidator, msg: 'failed' ***REMOVED***
 *     ]
 *     new Schema({ name: { type: String, validate: many ***REMOVED******REMOVED***);
 *
 *     // or utilizing SchemaType methods directly:
 *
 *     var schema = new Schema({ name: 'string' ***REMOVED***);
 *     schema.path('name').validate(validator, 'validation of `{PATH***REMOVED***` failed with value `{VALUE***REMOVED***`');
 *
 * ####Error message templates:
 *
 * From the examples above, you may have noticed that error messages support
 * basic templating. There are a few other template keywords besides `{PATH***REMOVED***`
 * and `{VALUE***REMOVED***` too. To find out more, details are available
 * [here](#error_messages_MongooseError.messages).
 *
 * If Mongoose's built-in error message templating isn't enough, Mongoose
 * supports setting the `message` property to a function.
 *
 *     schema.path('name').validate({
 *       validator: function() { return v.length > 5; ***REMOVED***,
 *       // `errors['name']` will be "name must have length 5, got 'foo'"
 *       message: function(props) {
 *         return `${props.path***REMOVED*** must have length 5, got '${props.value***REMOVED***'`;
 *       ***REMOVED***
 *     ***REMOVED***);
 *
 * To bypass Mongoose's error messages and just copy the error message that
 * the validator throws, do this:
 *
 *     schema.path('name').validate({
 *       validator: function() { throw new Error('Oops!'); ***REMOVED***,
 *       // `errors['name']` will be "Oops!"
 *       message: function(props) { return props.reason.message; ***REMOVED***
 *     ***REMOVED***);
 *
 * ####Asynchronous validation:
 *
 * Mongoose supports validators that return a promise. A validator that returns
 * a promise is called an _async validator_. Async validators run in
 * parallel, and `validate()` will wait until all async validators have settled.
 *
 *     schema.path('name').validate({
 *       validator: function (value) {
 *         return new Promise(function (resolve, reject) {
 *           resolve(false); // validation failed
 *         ***REMOVED***);
 *       ***REMOVED***
 *     ***REMOVED***);
 *
 * You might use asynchronous validators to retreive other documents from the database to validate against or to meet other I/O bound validation needs.
 *
 * Validation occurs `pre('save')` or whenever you manually execute [document#validate](#document_Document-validate).
 *
 * If validation fails during `pre('save')` and no callback was passed to receive the error, an `error` event will be emitted on your Models associated db [connection](#connection_Connection), passing the validation error object along.
 *
 *     var conn = mongoose.createConnection(..);
 *     conn.on('error', handleError);
 *
 *     var Product = conn.model('Product', yourSchema);
 *     var dvd = new Product(..);
 *     dvd.save(); // emits error on the `conn` above
 *
 * If you want to handle these errors at the Model level, add an `error`
 * listener to your Model as shown below.
 *
 *     // registering an error listener on the Model lets us handle errors more locally
 *     Product.on('error', handleError);
 *
 * @param {RegExp|Function|Object***REMOVED*** obj validator function, or hash describing options
 * @param {Function***REMOVED*** [obj.validator] validator function. If the validator function returns `undefined` or a truthy value, validation succeeds. If it returns falsy (except `undefined`) or throws an error, validation fails.
 * @param {String|Function***REMOVED*** [obj.message] optional error message. If function, should return the error message as a string
 * @param {Boolean***REMOVED*** [obj.propsParameter=false] If true, Mongoose will pass the validator properties object (with the `validator` function, `message`, etc.) as the 2nd arg to the validator function. This is disabled by default because many validators [rely on positional args](https://github.com/chriso/validator.js#validators), so turning this on may cause unpredictable behavior in external validators.
 * @param {String|Function***REMOVED*** [errorMsg] optional error message. If function, should return the error message as a string
 * @param {String***REMOVED*** [type] optional validator type
 * @return {SchemaType***REMOVED*** this
 * @api public
 */

SchemaType.prototype.validate = function(obj, message, type) {
  if (typeof obj === 'function' || obj && utils.getFunctionName(obj.constructor) === 'RegExp') {
    let properties;
    if (message instanceof Object && !type) {
      properties = utils.clone(message);
      if (!properties.message) {
        properties.message = properties.msg;
      ***REMOVED***
      properties.validator = obj;
      properties.type = properties.type || 'user defined';
    ***REMOVED***
      if (!message) {
        message = MongooseError.messages.general.default;
      ***REMOVED***
      if (!type) {
        type = 'user defined';
      ***REMOVED***
      properties = {message: message, type: type, validator: obj***REMOVED***;
    ***REMOVED***
    this.validators.push(properties);
    return this;
  ***REMOVED***

  let i;
  let length;
  let arg;

  for (i = 0, length = arguments.length; i < length; i++) {
    arg = arguments[i];
    if (!(arg && utils.getFunctionName(arg.constructor) === 'Object')) {
      const msg = 'Invalid validator. Received (' + typeof arg + ') '
          + arg
          + '. See http://mongoosejs.com/docs/api.html#schematype_SchemaType-validate';

      throw new Error(msg);
    ***REMOVED***
    this.validate(arg.validator, arg);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Adds a required validator to this SchemaType. The validator gets added
 * to the front of this SchemaType's validators array using `unshift()`.
 *
 * ####Example:
 *
 *     var s = new Schema({ born: { type: Date, required: true ***REMOVED***)
 *
 *     // or with custom error message
 *
 *     var s = new Schema({ born: { type: Date, required: '{PATH***REMOVED*** is required!' ***REMOVED***)
 *
 *     // or with a function
 *
 *     var s = new Schema({
 *       userId: ObjectId,
 *       username: {
 *         type: String,
 *         required: function() { return this.userId != null; ***REMOVED***
 *       ***REMOVED***
 *     ***REMOVED***)
 *
 *     // or with a function and a custom message
 *     var s = new Schema({
 *       userId: ObjectId,
 *       username: {
 *         type: String,
 *         required: [
 *           function() { return this.userId != null; ***REMOVED***,
 *           'username is required if id is specified'
 *         ]
 *       ***REMOVED***
 *     ***REMOVED***)
 *
 *     // or through the path API
 *
 *     Schema.path('name').required(true);
 *
 *     // with custom error messaging
 *
 *     Schema.path('name').required(true, 'grrr :( ');
 *
 *     // or make a path conditionally required based on a function
 *     var isOver18 = function() { return this.age >= 18; ***REMOVED***;
 *     Schema.path('voterRegistrationId').required(isOver18);
 *
 * The required validator uses the SchemaType's `checkRequired` function to
 * determine whether a given value satisfies the required validator. By default,
 * a value satisfies the required validator if `val != null` (that is, if
 * the value is not null nor undefined). However, most built-in mongoose schema
 * types override the default `checkRequired` function:
 *
 * @param {Boolean|Function|Object***REMOVED*** required enable/disable the validator, or function that returns required boolean, or options object
 * @param {Boolean|Function***REMOVED*** [options.isRequired] enable/disable the validator, or function that returns required boolean
 * @param {Function***REMOVED*** [options.ErrorConstructor] custom error constructor. The constructor receives 1 parameter, an object containing the validator properties.
 * @param {String***REMOVED*** [message] optional custom error message
 * @return {SchemaType***REMOVED*** this
 * @see Customized Error Messages #error_messages_MongooseError-messages
 * @see SchemaArray#checkRequired #schema_array_SchemaArray.checkRequired
 * @see SchemaBoolean#checkRequired #schema_boolean_SchemaBoolean-checkRequired
 * @see SchemaBuffer#checkRequired #schema_buffer_SchemaBuffer.schemaName
 * @see SchemaNumber#checkRequired #schema_number_SchemaNumber-min
 * @see SchemaObjectId#checkRequired #schema_objectid_ObjectId-auto
 * @see SchemaString#checkRequired #schema_string_SchemaString-checkRequired
 * @api public
 */

SchemaType.prototype.required = function(required, message) {
  let customOptions = {***REMOVED***;
  if (typeof required === 'object') {
    customOptions = required;
    message = customOptions.message || message;
    required = required.isRequired;
  ***REMOVED***

  if (required === false) {
    this.validators = this.validators.filter(function(v) {
      return v.validator !== this.requiredValidator;
    ***REMOVED***, this);

    this.isRequired = false;
    delete this.originalRequiredValue;
    return this;
  ***REMOVED***

  const _this = this;
  this.isRequired = true;

  this.requiredValidator = function(v) {
    const cachedRequired = get(this, '$__.cachedRequired');

    // no validation when this path wasn't selected in the query.
    if (cachedRequired != null && !this.isSelected(_this.path) && !this.isModified(_this.path)) {
      return true;
    ***REMOVED***

    // `$cachedRequired` gets set in `_evaluateRequiredFunctions()` so we
    // don't call required functions multiple times in one validate call
    // See gh-6801
    if (cachedRequired != null && _this.path in cachedRequired) {
      const res = cachedRequired[_this.path] ?
        _this.checkRequired(v, this) :
        true;
      delete cachedRequired[_this.path];
      return res;
    ***REMOVED*** else if (typeof required === 'function') {
      return required.apply(this) ? _this.checkRequired(v, this) : true;
    ***REMOVED***

    return _this.checkRequired(v, this);
  ***REMOVED***;
  this.originalRequiredValue = required;

  if (typeof required === 'string') {
    message = required;
    required = undefined;
  ***REMOVED***

  const msg = message || MongooseError.messages.general.required;
  this.validators.unshift(Object.assign({***REMOVED***, customOptions, {
    validator: this.requiredValidator,
    message: msg,
    type: 'required'
  ***REMOVED***));

  return this;
***REMOVED***;

/**
 * Gets the default value
 *
 * @param {Object***REMOVED*** scope the scope which callback are executed
 * @param {Boolean***REMOVED*** init
 * @api private
 */

SchemaType.prototype.getDefault = function(scope, init) {
  let ret = typeof this.defaultValue === 'function'
    ? this.defaultValue.call(scope)
    : this.defaultValue;

  if (ret !== null && ret !== undefined) {
    if (typeof ret === 'object' && (!this.options || !this.options.shared)) {
      ret = utils.clone(ret);
    ***REMOVED***

    const casted = this.cast(ret, scope, init);
    if (casted && casted.$isSingleNested) {
      casted.$parent = scope;
    ***REMOVED***
    return casted;
  ***REMOVED***
  return ret;
***REMOVED***;

/*!
 * Applies setters without casting
 *
 * @api private
 */

SchemaType.prototype._applySetters = function(value, scope, init, priorVal) {
  let v = value;
  const setters = this.setters;
  let len = setters.length;
  const caster = this.caster;

  while (len--) {
    v = setters[len].call(scope, v, this);
  ***REMOVED***

  if (Array.isArray(v) && caster && caster.setters) {
    const newVal = [];
    for (let i = 0; i < v.length; i++) {
      newVal.push(caster.applySetters(v[i], scope, init, priorVal));
    ***REMOVED***
    v = newVal;
  ***REMOVED***

  return v;
***REMOVED***;

/**
 * Applies setters
 *
 * @param {Object***REMOVED*** value
 * @param {Object***REMOVED*** scope
 * @param {Boolean***REMOVED*** init
 * @api private
 */

SchemaType.prototype.applySetters = function(value, scope, init, priorVal, options) {
  let v = this._applySetters(value, scope, init, priorVal, options);

  if (v == null) {
    return v;
  ***REMOVED***

  // do not cast until all setters are applied #665
  v = this.cast(v, scope, init, priorVal, options);

  return v;
***REMOVED***;

/**
 * Applies getters to a value
 *
 * @param {Object***REMOVED*** value
 * @param {Object***REMOVED*** scope
 * @api private
 */

SchemaType.prototype.applyGetters = function(value, scope) {
  let v = value;
  const getters = this.getters;
  const len = getters.length;

  if (len === 0) {
    return v;
  ***REMOVED***

  for (let i = 0; i < len; ++i) {
    v = getters[i].call(scope, v, this);
  ***REMOVED***

  return v;
***REMOVED***;

/**
 * Sets default `select()` behavior for this path.
 *
 * Set to `true` if this path should always be included in the results, `false` if it should be excluded by default. This setting can be overridden at the query level.
 *
 * ####Example:
 *
 *     T = db.model('T', new Schema({ x: { type: String, select: true ***REMOVED******REMOVED***));
 *     T.find(..); // field x will always be selected ..
 *     // .. unless overridden;
 *     T.find().select('-x').exec(callback);
 *
 * @param {Boolean***REMOVED*** val
 * @return {SchemaType***REMOVED*** this
 * @api public
 */

SchemaType.prototype.select = function select(val) {
  this.selected = !!val;
  return this;
***REMOVED***;

/**
 * Performs a validation of `value` using the validators declared for this SchemaType.
 *
 * @param {any***REMOVED*** value
 * @param {Function***REMOVED*** callback
 * @param {Object***REMOVED*** scope
 * @api private
 */

SchemaType.prototype.doValidate = function(value, fn, scope) {
  let err = false;
  const path = this.path;
  let count = this.validators.length;

  if (!count) {
    return fn(null);
  ***REMOVED***

  const validate = function(ok, validatorProperties) {
    if (err) {
      return;
    ***REMOVED***
    if (ok === undefined || ok) {
      if (--count <= 0) {
        immediate(function() {
          fn(null);
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***
      const ErrorConstructor = validatorProperties.ErrorConstructor || ValidatorError;
      err = new ErrorConstructor(validatorProperties);
      err[validatorErrorSymbol] = true;
      immediate(function() {
        fn(err);
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***;

  const _this = this;
  this.validators.forEach(function(v) {
    if (err) {
      return;
    ***REMOVED***

    const validator = v.validator;
    let ok;

    const validatorProperties = utils.clone(v);
    validatorProperties.path = path;
    validatorProperties.value = value;

    if (validator instanceof RegExp) {
      validate(validator.test(value), validatorProperties);
    ***REMOVED*** else if (typeof validator === 'function') {
      if (value === undefined && validator !== _this.requiredValidator) {
        validate(true, validatorProperties);
        return;
      ***REMOVED***
      if (validatorProperties.isAsync) {
        asyncValidate(validator, scope, value, validatorProperties, validate);
      ***REMOVED***
        try {
          if (validatorProperties.propsParameter) {
            ok = validator.call(scope, value, validatorProperties);
          ***REMOVED***
            ok = validator.call(scope, value);
          ***REMOVED***
        ***REMOVED*** catch (error) {
          ok = false;
          validatorProperties.reason = error;
          if (error.message) {
            validatorProperties.message = error.message;
          ***REMOVED***
        ***REMOVED***
        if (ok != null && typeof ok.then === 'function') {
          ok.then(
            function(ok) { validate(ok, validatorProperties); ***REMOVED***,
            function(error) {
              validatorProperties.reason = error;
              ok = false;
              validate(ok, validatorProperties);
            ***REMOVED***);
        ***REMOVED***
          validate(ok, validatorProperties);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);
***REMOVED***;

/*!
 * Handle async validators
 */

function asyncValidate(validator, scope, value, props, cb) {
  let called = false;
  const returnVal = validator.call(scope, value, function(ok, customMsg) {
    if (called) {
      return;
    ***REMOVED***
    called = true;
    if (customMsg) {
      props.message = customMsg;
    ***REMOVED***
    cb(ok, props);
  ***REMOVED***);
  if (typeof returnVal === 'boolean') {
    called = true;
    cb(returnVal, props);
  ***REMOVED*** else if (returnVal && typeof returnVal.then === 'function') {
    // Promise
    returnVal.then(
      function(ok) {
        if (called) {
          return;
        ***REMOVED***
        called = true;
        cb(ok, props);
      ***REMOVED***,
      function(error) {
        if (called) {
          return;
        ***REMOVED***
        called = true;

        props.reason = error;
        cb(false, props);
      ***REMOVED***);
  ***REMOVED***
***REMOVED***

/**
 * Performs a validation of `value` using the validators declared for this SchemaType.
 *
 * ####Note:
 *
 * This method ignores the asynchronous validators.
 *
 * @param {any***REMOVED*** value
 * @param {Object***REMOVED*** scope
 * @return {MongooseError|undefined***REMOVED***
 * @api private
 */

SchemaType.prototype.doValidateSync = function(value, scope) {
  let err = null;
  const path = this.path;
  const count = this.validators.length;

  if (!count) {
    return null;
  ***REMOVED***

  const validate = function(ok, validatorProperties) {
    if (err) {
      return;
    ***REMOVED***
    if (ok !== undefined && !ok) {
      const ErrorConstructor = validatorProperties.ErrorConstructor || ValidatorError;
      err = new ErrorConstructor(validatorProperties);
      err[validatorErrorSymbol] = true;
    ***REMOVED***
  ***REMOVED***;

  let validators = this.validators;
  if (value === void 0) {
    if (this.validators.length > 0 && this.validators[0].type === 'required') {
      validators = [this.validators[0]];
    ***REMOVED***
      return null;
    ***REMOVED***
  ***REMOVED***

  validators.forEach(function(v) {
    if (err) {
      return;
    ***REMOVED***

    const validator = v.validator;
    const validatorProperties = utils.clone(v);
    validatorProperties.path = path;
    validatorProperties.value = value;
    let ok;

    // Skip any explicit async validators. Validators that return a promise
    // will still run, but won't trigger any errors.
    if (validator.isAsync) {
      return;
    ***REMOVED***

    if (validator instanceof RegExp) {
      validate(validator.test(value), validatorProperties);
    ***REMOVED*** else if (typeof validator === 'function') {
      try {
        if (validatorProperties.propsParameter) {
          ok = validator.call(scope, value, validatorProperties);
        ***REMOVED***
          ok = validator.call(scope, value);
        ***REMOVED***
      ***REMOVED*** catch (error) {
        ok = false;
        validatorProperties.reason = error;
      ***REMOVED***

      // Skip any validators that return a promise, we can't handle those
      // synchronously
      if (ok != null && typeof ok.then === 'function') {
        return;
      ***REMOVED***
      validate(ok, validatorProperties);
    ***REMOVED***
  ***REMOVED***);

  return err;
***REMOVED***;

/**
 * Determines if value is a valid Reference.
 *
 * @param {SchemaType***REMOVED*** self
 * @param {Object***REMOVED*** value
 * @param {Document***REMOVED*** doc
 * @param {Boolean***REMOVED*** init
 * @return {Boolean***REMOVED***
 * @api private
 */

SchemaType._isRef = function(self, value, doc, init) {
  // fast path
  let ref = init && self.options && (self.options.ref || self.options.refPath);

  if (!ref && doc && doc.$__ != null) {
    // checks for
    // - this populated with adhoc model and no ref was set in schema OR
    // - setting / pushing values after population
    const path = doc.$__fullPath(self.path);
    const owner = doc.ownerDocument ? doc.ownerDocument() : doc;
    ref = owner.populated(path);
  ***REMOVED***

  if (ref) {
    if (value == null) {
      return true;
    ***REMOVED***
    if (!Buffer.isBuffer(value) && // buffers are objects too
        value._bsontype !== 'Binary' // raw binary value from the db
        && utils.isObject(value) // might have deselected _id in population query
    ) {
      return true;
    ***REMOVED***
  ***REMOVED***

  return false;
***REMOVED***;

/*!
 * ignore
 */

function handleSingle(val) {
  return this.castForQuery(val);
***REMOVED***

/*!
 * ignore
 */

function handleArray(val) {
  const _this = this;
  if (!Array.isArray(val)) {
    return [this.castForQuery(val)];
  ***REMOVED***
  return val.map(function(m) {
    return _this.castForQuery(m);
  ***REMOVED***);
***REMOVED***

/*!
 * Just like handleArray, except also allows `[]` because surprisingly
 * `$in: [1, []]` works fine
 */

function handle$in(val) {
  const _this = this;
  if (!Array.isArray(val)) {
    return [this.castForQuery(val)];
  ***REMOVED***
  return val.map(function(m) {
    if (Array.isArray(m) && m.length === 0) {
      return m;
    ***REMOVED***
    return _this.castForQuery(m);
  ***REMOVED***);
***REMOVED***

/*!
 * ignore
 */

SchemaType.prototype.$conditionalHandlers = {
  $all: handleArray,
  $eq: handleSingle,
  $in: handle$in,
  $ne: handleSingle,
  $nin: handleArray,
  $exists: $exists,
  $type: $type
***REMOVED***;

/*!
 * Wraps `castForQuery` to handle context
 */

SchemaType.prototype.castForQueryWrapper = function(params) {
  this.$$context = params.context;
  if ('$conditional' in params) {
    return this.castForQuery(params.$conditional, params.val);
  ***REMOVED***
  if (params.$skipQueryCastForUpdate) {
    return this._castForQuery(params.val);
  ***REMOVED***
  return this.castForQuery(params.val);
***REMOVED***;

/**
 * Cast the given value with the given optional query operator.
 *
 * @param {String***REMOVED*** [$conditional] query operator, like `$eq` or `$in`
 * @param {any***REMOVED*** val
 * @api private
 */

SchemaType.prototype.castForQuery = function($conditional, val) {
  let handler;
  if (arguments.length === 2) {
    handler = this.$conditionalHandlers[$conditional];
    if (!handler) {
      throw new Error('Can\'t use ' + $conditional);
    ***REMOVED***
    return handler.call(this, val);
  ***REMOVED***
  val = $conditional;
  return this._castForQuery(val);
***REMOVED***;

/*!
 * Internal switch for runSetters
 *
 * @api private
 */

SchemaType.prototype._castForQuery = function(val) {
  return this.applySetters(val, this.$$context);
***REMOVED***;

/**
 * Default check for if this path satisfies the `required` validator.
 *
 * @param {any***REMOVED*** val
 * @api private
 */

SchemaType.prototype.checkRequired = function(val) {
  return val != null;
***REMOVED***;

/*!
 * Module exports.
 */

module.exports = exports = SchemaType;

exports.CastError = CastError;

exports.ValidatorError = ValidatorError;
