// mongoose/lib/schema.js
'use strict';

/*!
 * Module dependencies.
 */

const EventEmitter = require('events').EventEmitter;
const Kareem = require('kareem');
const SchemaType = require('./schematype');
const VirtualType = require('./virtualtype');
const applyTimestampsToChildren = require('./helpers/update/applyTimestampsToChildren');
const applyTimestampsToUpdate = require('./helpers/update/applyTimestampsToUpdate');
const get = require('lodash.get');
const getIndexes = require('./helpers/schema/getIndexes');
const handleTimestampOption = require('./helpers/schema/handleTimestampOption');
const merge = require('./helpers/schema/merge');
const mpath = require('mpath');
const readPref = require('./driver').get().ReadPreference;
const util = require('util');
const utils = require('./utils');
const validateRef = require('./helpers/populate/validateRef');

let MongooseTypes;

const allMiddleware = require('./helpers/query/applyQueryMiddleware').
  middlewareFunctions.
  concat(require('./helpers/model/applyHooks').middlewareFunctions);

let id = 0;

/**
 * Schema constructor.
 *
 * ####Example:
 *
 *     var child = new Schema({ name: String ***REMOVED***);
 *     var schema = new Schema({ name: String, age: Number, children: [child] ***REMOVED***);
 *     var Tree = mongoose.model('Tree', schema);
 *
 *     // setting schema options
 *     new Schema({ name: String ***REMOVED***, { _id: false, autoIndex: false ***REMOVED***)
 *
 * ####Options:
 *
 * - [autoIndex](/docs/guide.html#autoIndex): bool - defaults to null (which means use the connection's autoIndex option)
 * - [autoCreate](/docs/guide.html#autoCreate): bool - defaults to null (which means use the connection's autoCreate option)
 * - [bufferCommands](/docs/guide.html#bufferCommands): bool - defaults to true
 * - [capped](/docs/guide.html#capped): bool - defaults to false
 * - [collection](/docs/guide.html#collection): string - no default
 * - [id](/docs/guide.html#id): bool - defaults to true
 * - [_id](/docs/guide.html#_id): bool - defaults to true
 * - `minimize`: bool - controls [document#toObject](#document_Document-toObject) behavior when called manually - defaults to true
 * - [read](/docs/guide.html#read): string
 * - [writeConcern](/docs/guide.html#writeConcern): object - defaults to null, use to override [the MongoDB server's default write concern settings](https://docs.mongodb.com/manual/reference/write-concern/)
 * - [shardKey](/docs/guide.html#shardKey): bool - defaults to `null`
 * - [strict](/docs/guide.html#strict): bool - defaults to true
 * - [toJSON](/docs/guide.html#toJSON) - object - no default
 * - [toObject](/docs/guide.html#toObject) - object - no default
 * - [typeKey](/docs/guide.html#typeKey) - string - defaults to 'type'
 * - [useNestedStrict](/docs/guide.html#useNestedStrict) - boolean - defaults to false
 * - [validateBeforeSave](/docs/guide.html#validateBeforeSave) - bool - defaults to `true`
 * - [versionKey](/docs/guide.html#versionKey): string - defaults to "__v"
 * - [collation](/docs/guide.html#collation): object - defaults to null (which means use no collation)
 * - [selectPopulatedPaths](/docs/guide.html#selectPopulatedPaths): boolean - defaults to `true`
 *
 * ####Note:
 *
 * _When nesting schemas, (`children` in the example above), always declare the child schema first before passing it into its parent._
 *
 * @param {Object***REMOVED*** definition
 * @param {Object***REMOVED*** [options]
 * @inherits NodeJS EventEmitter http://nodejs.org/api/events.html#events_class_events_eventemitter
 * @event `init`: Emitted after the schema is compiled into a `Model`.
 * @api public
 */

function Schema(obj, options) {
  if (!(this instanceof Schema)) {
    return new Schema(obj, options);
  ***REMOVED***

  this.obj = obj;
  this.paths = {***REMOVED***;
  this.aliases = {***REMOVED***;
  this.subpaths = {***REMOVED***;
  this.virtuals = {***REMOVED***;
  this.singleNestedPaths = {***REMOVED***;
  this.nested = {***REMOVED***;
  this.inherits = {***REMOVED***;
  this.callQueue = [];
  this._indexes = [];
  this.methods = {***REMOVED***;
  this.methodOptions = {***REMOVED***;
  this.statics = {***REMOVED***;
  this.tree = {***REMOVED***;
  this.query = {***REMOVED***;
  this.childSchemas = [];
  this.plugins = [];
  // For internal debugging. Do not use this to try to save a schema in MDB.
  this.$id = ++id;

  this.s = {
    hooks: new Kareem()
  ***REMOVED***;

  this.options = this.defaultOptions(options);

  // build paths
  if (obj) {
    this.add(obj);
  ***REMOVED***

  // check if _id's value is a subdocument (gh-2276)
  const _idSubDoc = obj && obj._id && utils.isObject(obj._id);

  // ensure the documents get an auto _id unless disabled
  const auto_id = !this.paths['_id'] &&
      (!this.options.noId && this.options._id) && !_idSubDoc;

  if (auto_id) {
    const _obj = {_id: {auto: true***REMOVED******REMOVED***;
    _obj._id[this.options.typeKey] = Schema.ObjectId;
    this.add(_obj);
  ***REMOVED***

  this.setupTimestamp(this.options.timestamps);
***REMOVED***

/*!
 * Create virtual properties with alias field
 */
function aliasFields(schema, paths) {
  paths = paths || Object.keys(schema.paths);
  for (const path of paths) {
    const options = get(schema.paths[path], 'options');
    if (options == null) {
      continue;
    ***REMOVED***

    const prop = schema.paths[path].path;
    const alias = options.alias;

    if (!alias) {
      continue;
    ***REMOVED***

    if (typeof alias !== 'string') {
      throw new Error('Invalid value for alias option on ' + prop + ', got ' + alias);
    ***REMOVED***

    schema.aliases[alias] = prop;

    schema.
      virtual(alias).
      get((function(p) {
        return function() {
          if (typeof this.get === 'function') {
            return this.get(p);
          ***REMOVED***
          return this[p];
        ***REMOVED***;
      ***REMOVED***)(prop)).
      set((function(p) {
        return function(v) {
          return this.set(p, v);
        ***REMOVED***;
      ***REMOVED***)(prop));
  ***REMOVED***
***REMOVED***

/*!
 * Inherit from EventEmitter.
 */
Schema.prototype = Object.create(EventEmitter.prototype);
Schema.prototype.constructor = Schema;
Schema.prototype.instanceOfSchema = true;

/*!
 * ignore
 */

Object.defineProperty(Schema.prototype, '$schemaType', {
  configurable: false,
  enumerable: false,
  writable: true
***REMOVED***);

/**
 * Array of child schemas (from document arrays and single nested subdocs)
 * and their corresponding compiled models. Each element of the array is
 * an object with 2 properties: `schema` and `model`.
 *
 * This property is typically only useful for plugin authors and advanced users.
 * You do not need to interact with this property at all to use mongoose.
 *
 * @api public
 * @property childSchemas
 * @memberOf Schema
 * @instance
 */

Object.defineProperty(Schema.prototype, 'childSchemas', {
  configurable: false,
  enumerable: true,
  writable: true
***REMOVED***);

/**
 * The original object passed to the schema constructor
 *
 * ####Example:
 *
 *     var schema = new Schema({ a: String ***REMOVED***).add({ b: String ***REMOVED***);
 *     schema.obj; // { a: String ***REMOVED***
 *
 * @api public
 * @property obj
 * @memberOf Schema
 * @instance
 */

Schema.prototype.obj;

/**
 * Schema as flat paths
 *
 * ####Example:
 *     {
 *         '_id'        : SchemaType,
 *       , 'nested.key' : SchemaType,
 *     ***REMOVED***
 *
 * @api private
 * @property paths
 * @memberOf Schema
 * @instance
 */

Schema.prototype.paths;

/**
 * Schema as a tree
 *
 * ####Example:
 *     {
 *         '_id'     : ObjectId
 *       , 'nested'  : {
 *             'key' : String
 *         ***REMOVED***
 *     ***REMOVED***
 *
 * @api private
 * @property tree
 * @memberOf Schema
 * @instance
 */

Schema.prototype.tree;

/**
 * Returns a deep copy of the schema
 *
 * @return {Schema***REMOVED*** the cloned schema
 * @api public
 * @memberOf Schema
 * @instance
 */

Schema.prototype.clone = function() {
  const s = new Schema({***REMOVED***, this._userProvidedOptions);
  s.obj = this.obj;
  s.options = utils.clone(this.options);
  s.callQueue = this.callQueue.map(function(f) { return f; ***REMOVED***);
  s.methods = utils.clone(this.methods);
  s.methodOptions = utils.clone(this.methodOptions);
  s.statics = utils.clone(this.statics);
  s.query = utils.clone(this.query);
  s.plugins = Array.prototype.slice.call(this.plugins);
  s._indexes = utils.clone(this._indexes);
  s.s.hooks = this.s.hooks.clone();
  s._originalSchema = this._originalSchema == null ?
    this._originalSchema :
    this._originalSchema.clone();

  s.tree = utils.clone(this.tree);
  s.paths = utils.clone(this.paths);
  s.nested = utils.clone(this.nested);
  s.subpaths = utils.clone(this.subpaths);
  s.childSchemas = this.childSchemas.slice();
  s.singleNestedPaths = utils.clone(this.singleNestedPaths);

  s.virtuals = utils.clone(this.virtuals);
  s.$globalPluginsApplied = this.$globalPluginsApplied;
  s.$isRootDiscriminator = this.$isRootDiscriminator;

  if (this.discriminatorMapping != null) {
    s.discriminatorMapping = Object.assign({***REMOVED***, this.discriminatorMapping);
  ***REMOVED***
  if (s.discriminators != null) {
    s.discriminators = Object.assign({***REMOVED***, this.discriminators);
  ***REMOVED***

  s.aliases = Object.assign({***REMOVED***, this.aliases);

  // Bubble up `init` for backwards compat
  s.on('init', v => this.emit('init', v));

  return s;
***REMOVED***;

/**
 * Returns default options for this schema, merged with `options`.
 *
 * @param {Object***REMOVED*** options
 * @return {Object***REMOVED***
 * @api private
 */

Schema.prototype.defaultOptions = function(options) {
  if (options && options.safe === false) {
    options.safe = {w: 0***REMOVED***;
  ***REMOVED***

  if (options && options.safe && options.safe.w === 0) {
    // if you turn off safe writes, then versioning goes off as well
    options.versionKey = false;
  ***REMOVED***

  this._userProvidedOptions = options == null ? {***REMOVED*** : utils.clone(options);

  const baseOptions = get(this, 'base.options', {***REMOVED***);
  options = utils.options({
    strict: 'strict' in baseOptions ? baseOptions.strict : true,
    bufferCommands: true,
    capped: false, // { size, max, autoIndexId ***REMOVED***
    versionKey: '__v',
    discriminatorKey: '__t',
    minimize: true,
    autoIndex: null,
    shardKey: null,
    read: null,
    validateBeforeSave: true,
    // the following are only applied at construction time
    noId: false, // deprecated, use { _id: false ***REMOVED***
    _id: true,
    noVirtualId: false, // deprecated, use { id: false ***REMOVED***
    id: true,
    typeKey: 'type'
  ***REMOVED***, utils.clone(options));

  if (options.read) {
    options.read = readPref(options.read);
  ***REMOVED***

  return options;
***REMOVED***;

/**
 * Adds key path / schema type pairs to this schema.
 *
 * ####Example:
 *
 *     const ToySchema = new Schema();
 *     ToySchema.add({ name: 'string', color: 'string', price: 'number' ***REMOVED***);
 *
 *     const TurboManSchema = new Schema();
 *     // You can also `add()` another schema and copy over all paths, virtuals,
 *     // getters, setters, indexes, methods, and statics.
 *     TurboManSchema.add(ToySchema).add({ year: Number ***REMOVED***);
 *
 * @param {Object|Schema***REMOVED*** obj plain object with paths to add, or another schema
 * @param {String***REMOVED*** [prefix] path to prefix the newly added paths with
 * @return {Schema***REMOVED*** the Schema instance
 * @api public
 */

Schema.prototype.add = function add(obj, prefix) {
  if (obj instanceof Schema) {
    merge(this, obj);
    return;
  ***REMOVED***

  prefix = prefix || '';
  const keys = Object.keys(obj);

  for (let i = 0; i < keys.length; ++i) {
    const key = keys[i];

    if (obj[key] == null) {
      throw new TypeError('Invalid value for schema path `' + prefix + key + '`');
    ***REMOVED***

    if (Array.isArray(obj[key]) && obj[key].length === 1 && obj[key][0] == null) {
      throw new TypeError('Invalid value for schema Array path `' + prefix + key + '`');
    ***REMOVED***

    if (utils.isObject(obj[key]) &&
        (!obj[key].constructor || utils.getFunctionName(obj[key].constructor) === 'Object') &&
        (!obj[key][this.options.typeKey] || (this.options.typeKey === 'type' && obj[key].type.type))) {
      if (Object.keys(obj[key]).length) {
        // nested object { last: { name: String ***REMOVED******REMOVED***
        this.nested[prefix + key] = true;
        this.add(obj[key], prefix + key + '.');
      ***REMOVED***
        if (prefix) {
          this.nested[prefix.substr(0, prefix.length - 1)] = true;
        ***REMOVED***
        this.path(prefix + key, obj[key]); // mixed type
      ***REMOVED***
    ***REMOVED***
      if (prefix) {
        this.nested[prefix.substr(0, prefix.length - 1)] = true;
      ***REMOVED***
      this.path(prefix + key, obj[key]);
    ***REMOVED***
  ***REMOVED***

  const addedKeys = Object.keys(obj).
    map(key => prefix ? prefix + key : key);
  aliasFields(this, addedKeys);
  return this;
***REMOVED***;

/**
 * Reserved document keys.
 *
 * Keys in this object are names that are rejected in schema declarations b/c they conflict with mongoose functionality. Using these key name will throw an error.
 *
 *      on, emit, _events, db, get, set, init, isNew, errors, schema, options, modelName, collection, _pres, _posts, toObject
 *
 * _NOTE:_ Use of these terms as method names is permitted, but play at your own risk, as they may be existing mongoose document methods you are stomping on.
 *
 *      var schema = new Schema(..);
 *      schema.methods.init = function () {***REMOVED*** // potentially breaking
 */

Schema.reserved = Object.create(null);
Schema.prototype.reserved = Schema.reserved;
const reserved = Schema.reserved;
// Core object
reserved['prototype'] =
// EventEmitter
reserved.emit =
reserved.on =
reserved.once =
reserved.listeners =
reserved.removeListener =
// document properties and functions
reserved.collection =
reserved.db =
reserved.errors =
reserved.init =
reserved.isModified =
reserved.isNew =
reserved.get =
reserved.modelName =
reserved.save =
reserved.schema =
reserved.toObject =
reserved.validate =
reserved.remove =
// hooks.js
reserved._pres = reserved._posts = 1;

/*!
 * Document keys to print warnings for
 */

const warnings = {***REMOVED***;
warnings.increment = '`increment` should not be used as a schema path name ' +
    'unless you have disabled versioning.';

/**
 * Gets/sets schema paths.
 *
 * Sets a path (if arity 2)
 * Gets a path (if arity 1)
 *
 * ####Example
 *
 *     schema.path('name') // returns a SchemaType
 *     schema.path('name', Number) // changes the schemaType of `name` to Number
 *
 * @param {String***REMOVED*** path
 * @param {Object***REMOVED*** constructor
 * @api public
 */

Schema.prototype.path = function(path, obj) {
  if (obj === undefined) {
    if (this.paths[path]) {
      return this.paths[path];
    ***REMOVED***
    if (this.subpaths[path]) {
      return this.subpaths[path];
    ***REMOVED***
    if (this.singleNestedPaths[path]) {
      return this.singleNestedPaths[path];
    ***REMOVED***

    // Look for maps
    for (const _path of Object.keys(this.paths)) {
      if (!_path.includes('.$*')) {
        continue;
      ***REMOVED***
      const re = new RegExp('^' + _path.replace(/\.\$\*/g, '.[^.]+') + '$');
      if (re.test(path)) {
        return this.paths[_path];
      ***REMOVED***
    ***REMOVED***

    // subpaths?
    return /\.\d+\.?.*$/.test(path)
      ? getPositionalPath(this, path)
      : undefined;
  ***REMOVED***

  // some path names conflict with document methods
  if (reserved[path]) {
    throw new Error('`' + path + '` may not be used as a schema pathname');
  ***REMOVED***

  if (warnings[path]) {
    console.log('WARN: ' + warnings[path]);
  ***REMOVED***

  if (typeof obj === 'object' && 'ref' in obj) {
    validateRef(obj.ref, path);
  ***REMOVED***

  // update the tree
  const subpaths = path.split(/\./);
  const last = subpaths.pop();
  let branch = this.tree;

  subpaths.forEach(function(sub, i) {
    if (!branch[sub]) {
      branch[sub] = {***REMOVED***;
    ***REMOVED***
    if (typeof branch[sub] !== 'object') {
      const msg = 'Cannot set nested path `' + path + '`. '
          + 'Parent path `'
          + subpaths.slice(0, i).concat([sub]).join('.')
          + '` already set to type ' + branch[sub].name
          + '.';
      throw new Error(msg);
    ***REMOVED***
    branch = branch[sub];
  ***REMOVED***);

  branch[last] = utils.clone(obj);

  this.paths[path] = this.interpretAsType(path, obj, this.options);
  const schemaType = this.paths[path];

  if (schemaType.$isSchemaMap) {
    // Maps can have arbitrary keys, so `$*` is internal shorthand for "any key"
    // The '$' is to imply this path should never be stored in MongoDB so we
    // can easily build a regexp out of this path, and '*' to imply "any key."
    const mapPath = path + '.$*';
    this.paths[path + '.$*'] = this.interpretAsType(mapPath,
      obj.of || { type: {***REMOVED*** ***REMOVED***, this.options);
    schemaType.$__schemaType = this.paths[path + '.$*'];
  ***REMOVED***

  if (schemaType.$isSingleNested) {
    for (const key in schemaType.schema.paths) {
      this.singleNestedPaths[path + '.' + key] = schemaType.schema.paths[key];
    ***REMOVED***
    for (const key in schemaType.schema.singleNestedPaths) {
      this.singleNestedPaths[path + '.' + key] =
        schemaType.schema.singleNestedPaths[key];
    ***REMOVED***

    this.childSchemas.push({
      schema: schemaType.schema,
      model: schemaType.caster
    ***REMOVED***);
  ***REMOVED*** else if (schemaType.$isMongooseDocumentArray) {
    this.childSchemas.push({
      schema: schemaType.schema,
      model: schemaType.casterConstructor
    ***REMOVED***);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Converts type arguments into Mongoose Types.
 *
 * @param {String***REMOVED*** path
 * @param {Object***REMOVED*** obj constructor
 * @api private
 */

Schema.prototype.interpretAsType = function(path, obj, options) {
  if (obj instanceof SchemaType) {
    return obj;
  ***REMOVED***

  // If this schema has an associated Mongoose object, use the Mongoose object's
  // copy of SchemaTypes re: gh-7158 gh-6933
  const MongooseTypes = this.base != null ? this.base.Schema.Types : Schema.Types;

  if (obj.constructor) {
    const constructorName = utils.getFunctionName(obj.constructor);
    if (constructorName !== 'Object') {
      const oldObj = obj;
      obj = {***REMOVED***;
      obj[options.typeKey] = oldObj;
    ***REMOVED***
  ***REMOVED***

  // Get the type making sure to allow keys named "type"
  // and default to mixed if not specified.
  // { type: { type: String, default: 'freshcut' ***REMOVED*** ***REMOVED***
  let type = obj[options.typeKey] && (options.typeKey !== 'type' || !obj.type.type)
    ? obj[options.typeKey]
    : {***REMOVED***;
  let name;

  if (utils.getFunctionName(type.constructor) === 'Object' || type === 'mixed') {
    return new MongooseTypes.Mixed(path, obj);
  ***REMOVED***

  if (Array.isArray(type) || Array === type || type === 'array') {
    // if it was specified through { type ***REMOVED*** look for `cast`
    let cast = (Array === type || type === 'array')
      ? obj.cast
      : type[0];

    if (cast && cast.instanceOfSchema) {
      return new MongooseTypes.DocumentArray(path, cast, obj);
    ***REMOVED***
    if (cast &&
        cast[options.typeKey] &&
        cast[options.typeKey].instanceOfSchema) {
      return new MongooseTypes.DocumentArray(path, cast[options.typeKey], obj, cast);
    ***REMOVED***

    if (Array.isArray(cast)) {
      return new MongooseTypes.Array(path, this.interpretAsType(path, cast, options), obj);
    ***REMOVED***

    if (typeof cast === 'string') {
      cast = MongooseTypes[cast.charAt(0).toUpperCase() + cast.substring(1)];
    ***REMOVED*** else if (cast && (!cast[options.typeKey] || (options.typeKey === 'type' && cast.type.type))
        && utils.getFunctionName(cast.constructor) === 'Object') {
      if (Object.keys(cast).length) {
        // The `minimize` and `typeKey` options propagate to child schemas
        // declared inline, like `{ arr: [{ val: { $type: String ***REMOVED*** ***REMOVED***] ***REMOVED***`.
        // See gh-3560
        const childSchemaOptions = {minimize: options.minimize***REMOVED***;
        if (options.typeKey) {
          childSchemaOptions.typeKey = options.typeKey;
        ***REMOVED***
        //propagate 'strict' option to child schema
        if (options.hasOwnProperty('strict')) {
          childSchemaOptions.strict = options.strict;
        ***REMOVED***
        const childSchema = new Schema(cast, childSchemaOptions);
        childSchema.$implicitlyCreated = true;
        return new MongooseTypes.DocumentArray(path, childSchema, obj);
      ***REMOVED***
        // Special case: empty object becomes mixed
        return new MongooseTypes.Array(path, MongooseTypes.Mixed, obj);
      ***REMOVED***
    ***REMOVED***

    if (cast) {
      type = cast[options.typeKey] && (options.typeKey !== 'type' || !cast.type.type)
        ? cast[options.typeKey]
        : cast;

      name = typeof type === 'string'
        ? type
        : type.schemaName || utils.getFunctionName(type);

      if (!(name in MongooseTypes)) {
        throw new TypeError('Undefined type `' + name + '` at array `' + path +
          '`');
      ***REMOVED***
    ***REMOVED***

    return new MongooseTypes.Array(path, cast || MongooseTypes.Mixed, obj, options);
  ***REMOVED***

  if (type && type.instanceOfSchema) {
    return new MongooseTypes.Embedded(type, path, obj);
  ***REMOVED***

  if (Buffer.isBuffer(type)) {
    name = 'Buffer';
  ***REMOVED***
    name = typeof type === 'string'
      ? type
      // If not string, `type` is a function. Outside of IE, function.name
      // gives you the function name. In IE, you need to compute it
      : type.schemaName || utils.getFunctionName(type);
  ***REMOVED***

  if (name) {
    name = name.charAt(0).toUpperCase() + name.substring(1);
  ***REMOVED***
  // Special case re: gh-7049 because the bson `ObjectID` class' capitalization
  // doesn't line up with Mongoose's.
  if (name === 'ObjectID') {
    name = 'ObjectId';
  ***REMOVED***

  if (undefined == MongooseTypes[name]) {
    throw new TypeError('Undefined type `' + name + '` at `' + path +
        '`\n  Did you try nesting Schemas? ' +
        'You can only nest using refs or arrays.');
  ***REMOVED***

  return new MongooseTypes[name](path, obj);
***REMOVED***;

/**
 * Iterates the schemas paths similar to Array#forEach.
 *
 * The callback is passed the pathname and schemaType as arguments on each iteration.
 *
 * @param {Function***REMOVED*** fn callback function
 * @return {Schema***REMOVED*** this
 * @api public
 */

Schema.prototype.eachPath = function(fn) {
  const keys = Object.keys(this.paths);
  const len = keys.length;

  for (let i = 0; i < len; ++i) {
    fn(keys[i], this.paths[keys[i]]);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Returns an Array of path strings that are required by this schema.
 *
 * @api public
 * @param {Boolean***REMOVED*** invalidate refresh the cache
 * @return {Array***REMOVED***
 */

Schema.prototype.requiredPaths = function requiredPaths(invalidate) {
  if (this._requiredpaths && !invalidate) {
    return this._requiredpaths;
  ***REMOVED***

  const paths = Object.keys(this.paths);
  let i = paths.length;
  const ret = [];

  while (i--) {
    const path = paths[i];
    if (this.paths[path].isRequired) {
      ret.push(path);
    ***REMOVED***
  ***REMOVED***
  this._requiredpaths = ret;
  return this._requiredpaths;
***REMOVED***;

/**
 * Returns indexes from fields and schema-level indexes (cached).
 *
 * @api private
 * @return {Array***REMOVED***
 */

Schema.prototype.indexedPaths = function indexedPaths() {
  if (this._indexedpaths) {
    return this._indexedpaths;
  ***REMOVED***
  this._indexedpaths = this.indexes();
  return this._indexedpaths;
***REMOVED***;

/**
 * Returns the pathType of `path` for this schema.
 *
 * Given a path, returns whether it is a real, virtual, nested, or ad-hoc/undefined path.
 *
 * @param {String***REMOVED*** path
 * @return {String***REMOVED***
 * @api public
 */

Schema.prototype.pathType = function(path) {
  if (path in this.paths) {
    return 'real';
  ***REMOVED***
  if (path in this.virtuals) {
    return 'virtual';
  ***REMOVED***
  if (path in this.nested) {
    return 'nested';
  ***REMOVED***
  if (path in this.subpaths) {
    return 'real';
  ***REMOVED***
  if (path in this.singleNestedPaths) {
    return 'real';
  ***REMOVED***

  // Look for maps
  for (const _path of Object.keys(this.paths)) {
    if (!_path.includes('.$*')) {
      continue;
    ***REMOVED***
    const re = new RegExp('^' + _path.replace(/\.\$\*/g, '.[^.]+') + '$');
    if (re.test(path)) {
      return this.paths[_path];
    ***REMOVED***
  ***REMOVED***

  if (/\.\d+\.|\.\d+$/.test(path)) {
    return getPositionalPathType(this, path);
  ***REMOVED***
  return 'adhocOrUndefined';
***REMOVED***;

/**
 * Returns true iff this path is a child of a mixed schema.
 *
 * @param {String***REMOVED*** path
 * @return {Boolean***REMOVED***
 * @api private
 */

Schema.prototype.hasMixedParent = function(path) {
  const subpaths = path.split(/\./g);
  path = '';
  for (let i = 0; i < subpaths.length; ++i) {
    path = i > 0 ? path + '.' + subpaths[i] : subpaths[i];
    if (path in this.paths &&
        this.paths[path] instanceof MongooseTypes.Mixed) {
      return true;
    ***REMOVED***
  ***REMOVED***

  return false;
***REMOVED***;

/**
 * Setup updatedAt and createdAt timestamps to documents if enabled
 *
 * @param {Boolean|Object***REMOVED*** timestamps timestamps options
 * @api private
 */
Schema.prototype.setupTimestamp = function(timestamps) {
  const childHasTimestamp = this.childSchemas.find(withTimestamp);

  function withTimestamp(s) {
    const ts = s.schema.options.timestamps;
    return ts == true || ts == null;
  ***REMOVED***

  if (!timestamps && !childHasTimestamp) {
    return;
  ***REMOVED***

  const createdAt = handleTimestampOption(timestamps, 'createdAt');
  const updatedAt = handleTimestampOption(timestamps, 'updatedAt');
  const schemaAdditions = {***REMOVED***;

  this.$timestamps = { createdAt: createdAt, updatedAt: updatedAt ***REMOVED***;

  if (updatedAt && !this.paths[updatedAt]) {
    schemaAdditions[updatedAt] = Date;
  ***REMOVED***

  if (createdAt && !this.paths[createdAt]) {
    schemaAdditions[createdAt] = Date;
  ***REMOVED***

  this.add(schemaAdditions);

  this.pre('save', function(next) {
    const defaultTimestamp = (this.ownerDocument ? this.ownerDocument() : this).
      constructor.base.now();
    const auto_id = this._id && this._id.auto;

    if (createdAt && !this.get(createdAt) && this.isSelected(createdAt)) {
      this.set(createdAt, auto_id ? this._id.getTimestamp() : defaultTimestamp);
    ***REMOVED***

    if (updatedAt && (this.isNew || this.isModified())) {
      let ts = defaultTimestamp;
      if (this.isNew) {
        if (createdAt != null) {
          ts = this.get(createdAt);
        ***REMOVED*** else if (auto_id) {
          ts = this._id.getTimestamp();
        ***REMOVED***
      ***REMOVED***
      this.set(updatedAt, ts);
    ***REMOVED***

    next();
  ***REMOVED***);

  this.methods.initializeTimestamps = function() {
    if (createdAt && !this.get(createdAt)) {
      this.set(createdAt, new Date());
    ***REMOVED***
    if (updatedAt && !this.get(updatedAt)) {
      this.set(updatedAt, new Date());
    ***REMOVED***
    return this;
  ***REMOVED***;

  this.pre('findOneAndUpdate', _setTimestampsOnUpdate);
  this.pre('replaceOne', _setTimestampsOnUpdate);
  this.pre('update', _setTimestampsOnUpdate);
  this.pre('updateOne', _setTimestampsOnUpdate);
  this.pre('updateMany', _setTimestampsOnUpdate);

  function _setTimestampsOnUpdate(next) {
    const now = this.model.base.now();
    applyTimestampsToUpdate(now, createdAt, updatedAt, this.getUpdate(),
      this.options, this.schema);
    applyTimestampsToChildren(now, this.getUpdate(), this.model.schema);
    next();
  ***REMOVED***
***REMOVED***;

/*!
 * ignore
 */

function getPositionalPathType(self, path) {
  const subpaths = path.split(/\.(\d+)\.|\.(\d+)$/).filter(Boolean);
  if (subpaths.length < 2) {
    return self.paths[subpaths[0]];
  ***REMOVED***

  let val = self.path(subpaths[0]);
  let isNested = false;
  if (!val) {
    return val;
  ***REMOVED***

  const last = subpaths.length - 1;
  let subpath;
  let i = 1;

  for (; i < subpaths.length; ++i) {
    isNested = false;
    subpath = subpaths[i];

    if (i === last && val && !/\D/.test(subpath)) {
      if (val.$isMongooseDocumentArray) {
        const oldVal = val;
        val = new SchemaType(subpath, {
          required: get(val, 'schemaOptions.required', false)
        ***REMOVED***);
        val.cast = function(value, doc, init) {
          return oldVal.cast(value, doc, init)[0];
        ***REMOVED***;
        val.$isMongooseDocumentArrayElement = true;
        val.caster = oldVal.caster;
        val.schema = oldVal.schema;
      ***REMOVED*** else if (val instanceof MongooseTypes.Array) {
        // StringSchema, NumberSchema, etc
        val = val.caster;
      ***REMOVED***
        val = undefined;
      ***REMOVED***
      break;
    ***REMOVED***

    // ignore if its just a position segment: path.0.subpath
    if (!/\D/.test(subpath)) {
      continue;
    ***REMOVED***

    if (!(val && val.schema)) {
      val = undefined;
      break;
    ***REMOVED***

    const type = val.schema.pathType(subpath);
    isNested = (type === 'nested');
    val = val.schema.path(subpath);
  ***REMOVED***

  self.subpaths[path] = val;
  if (val) {
    return 'real';
  ***REMOVED***
  if (isNested) {
    return 'nested';
  ***REMOVED***
  return 'adhocOrUndefined';
***REMOVED***


/*!
 * ignore
 */

function getPositionalPath(self, path) {
  getPositionalPathType(self, path);
  return self.subpaths[path];
***REMOVED***

/**
 * Adds a method call to the queue.
 *
 * @param {String***REMOVED*** name name of the document method to call later
 * @param {Array***REMOVED*** args arguments to pass to the method
 * @api public
 */

Schema.prototype.queue = function(name, args) {
  this.callQueue.push([name, args]);
  return this;
***REMOVED***;

/**
 * Defines a pre hook for the document.
 *
 * ####Example
 *
 *     var toySchema = new Schema({ name: String, created: Date ***REMOVED***);
 *
 *     toySchema.pre('save', function(next) {
 *       if (!this.created) this.created = new Date;
 *       next();
 *     ***REMOVED***);
 *
 *     toySchema.pre('validate', function(next) {
 *       if (this.name !== 'Woody') this.name = 'Woody';
 *       next();
 *     ***REMOVED***);
 *
 *     // Equivalent to calling `pre()` on `find`, `findOne`, `findOneAndUpdate`.
 *     toySchema.pre(/^find/, function(next) {
 *       console.log(this.getQuery());
 *     ***REMOVED***);
 *
 * @param {String|RegExp***REMOVED*** method or regular expression to match method name
 * @param {Object***REMOVED*** [options]
 * @param {Boolean***REMOVED*** [options.document] If `name` is a hook for both document and query middleware, set to `true` to run on document middleware.
 * @param {Boolean***REMOVED*** [options.query] If `name` is a hook for both document and query middleware, set to `true` to run on query middleware.
 * @param {Function***REMOVED*** callback
 * @see hooks.js https://github.com/bnoguchi/hooks-js/tree/31ec571cef0332e21121ee7157e0cf9728572cc3
 * @api public
 */

Schema.prototype.pre = function(name) {
  if (name instanceof RegExp) {
    const remainingArgs = Array.prototype.slice.call(arguments, 1);
    for (const fn of allMiddleware) {
      if (name.test(fn)) {
        this.pre.apply(this, [fn].concat(remainingArgs));
      ***REMOVED***
    ***REMOVED***
    return this;
  ***REMOVED***
  this.s.hooks.pre.apply(this.s.hooks, arguments);
  return this;
***REMOVED***;

/**
 * Defines a post hook for the document
 *
 *     var schema = new Schema(..);
 *     schema.post('save', function (doc) {
 *       console.log('this fired after a document was saved');
 *     ***REMOVED***);
 *
 *     schema.post('find', function(docs) {
 *       console.log('this fired after you ran a find query');
 *     ***REMOVED***);
 *
 *     schema.post(/Many$/, function(res) {
 *       console.log('this fired after you ran `updateMany()` or `deleteMany()`);
 *     ***REMOVED***);
 *
 *     var Model = mongoose.model('Model', schema);
 *
 *     var m = new Model(..);
 *     m.save(function(err) {
 *       console.log('this fires after the `post` hook');
 *     ***REMOVED***);
 *
 *     m.find(function(err, docs) {
 *       console.log('this fires after the post find hook');
 *     ***REMOVED***);
 *
 * @param {String|RegExp***REMOVED*** method or regular expression to match method name
 * @param {Object***REMOVED*** [options]
 * @param {Boolean***REMOVED*** [options.document] If `name` is a hook for both document and query middleware, set to `true` to run on document middleware.
 * @param {Boolean***REMOVED*** [options.query] If `name` is a hook for both document and query middleware, set to `true` to run on query middleware.
 * @param {Function***REMOVED*** fn callback
 * @see middleware http://mongoosejs.com/docs/middleware.html
 * @see kareem http://npmjs.org/package/kareem
 * @api public
 */

Schema.prototype.post = function(name) {
  if (name instanceof RegExp) {
    const remainingArgs = Array.prototype.slice.call(arguments, 1);
    for (const fn of allMiddleware) {
      if (name.test(fn)) {
        this.post.apply(this, [fn].concat(remainingArgs));
      ***REMOVED***
    ***REMOVED***
    return this;
  ***REMOVED***
  this.s.hooks.post.apply(this.s.hooks, arguments);
  return this;
***REMOVED***;

/**
 * Registers a plugin for this schema.
 *
 * @param {Function***REMOVED*** plugin callback
 * @param {Object***REMOVED*** [opts]
 * @see plugins
 * @api public
 */

Schema.prototype.plugin = function(fn, opts) {
  if (typeof fn !== 'function') {
    throw new Error('First param to `schema.plugin()` must be a function, ' +
      'got "' + (typeof fn) + '"');
  ***REMOVED***

  if (opts &&
      opts.deduplicate) {
    for (let i = 0; i < this.plugins.length; ++i) {
      if (this.plugins[i].fn === fn) {
        return this;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  this.plugins.push({ fn: fn, opts: opts ***REMOVED***);

  fn(this, opts);
  return this;
***REMOVED***;

/**
 * Adds an instance method to documents constructed from Models compiled from this schema.
 *
 * ####Example
 *
 *     var schema = kittySchema = new Schema(..);
 *
 *     schema.method('meow', function () {
 *       console.log('meeeeeoooooooooooow');
 *     ***REMOVED***)
 *
 *     var Kitty = mongoose.model('Kitty', schema);
 *
 *     var fizz = new Kitty;
 *     fizz.meow(); // meeeeeooooooooooooow
 *
 * If a hash of name/fn pairs is passed as the only argument, each name/fn pair will be added as methods.
 *
 *     schema.method({
 *         purr: function () {***REMOVED***
 *       , scratch: function () {***REMOVED***
 *     ***REMOVED***);
 *
 *     // later
 *     fizz.purr();
 *     fizz.scratch();
 *
 * NOTE: `Schema.method()` adds instance methods to the `Schema.methods` object. You can also add instance methods directly to the `Schema.methods` object as seen in the [guide](./guide.html#methods)
 *
 * @param {String|Object***REMOVED*** method name
 * @param {Function***REMOVED*** [fn]
 * @api public
 */

Schema.prototype.method = function(name, fn, options) {
  if (typeof name !== 'string') {
    for (const i in name) {
      this.methods[i] = name[i];
      this.methodOptions[i] = utils.clone(options);
    ***REMOVED***
  ***REMOVED***
    this.methods[name] = fn;
    this.methodOptions[name] = utils.clone(options);
  ***REMOVED***
  return this;
***REMOVED***;

/**
 * Adds static "class" methods to Models compiled from this schema.
 *
 * ####Example
 *
 *     var schema = new Schema(..);
 *     schema.static('findByName', function (name, callback) {
 *       return this.find({ name: name ***REMOVED***, callback);
 *     ***REMOVED***);
 *
 *     var Drink = mongoose.model('Drink', schema);
 *     Drink.findByName('sanpellegrino', function (err, drinks) {
 *       //
 *     ***REMOVED***);
 *
 * If a hash of name/fn pairs is passed as the only argument, each name/fn pair will be added as statics.
 *
 * @param {String|Object***REMOVED*** name
 * @param {Function***REMOVED*** [fn]
 * @api public
 */

Schema.prototype.static = function(name, fn) {
  if (typeof name !== 'string') {
    for (const i in name) {
      this.statics[i] = name[i];
    ***REMOVED***
  ***REMOVED***
    this.statics[name] = fn;
  ***REMOVED***
  return this;
***REMOVED***;

/**
 * Defines an index (most likely compound) for this schema.
 *
 * ####Example
 *
 *     schema.index({ first: 1, last: -1 ***REMOVED***)
 *
 * @param {Object***REMOVED*** fields
 * @param {Object***REMOVED*** [options] Options to pass to [MongoDB driver's `createIndex()` function](http://mongodb.github.io/node-mongodb-native/2.0/api/Collection.html#createIndex)
 * @param {String***REMOVED*** [options.expires=null] Mongoose-specific syntactic sugar, uses [ms](https://www.npmjs.com/package/ms) to convert `expires` option into seconds for the `expireAfterSeconds` in the above link.
 * @api public
 */

Schema.prototype.index = function(fields, options) {
  fields || (fields = {***REMOVED***);
  options || (options = {***REMOVED***);

  if (options.expires) {
    utils.expires(options);
  ***REMOVED***

  this._indexes.push([fields, options]);
  return this;
***REMOVED***;

/**
 * Sets/gets a schema option.
 *
 * ####Example
 *
 *     schema.set('strict'); // 'true' by default
 *     schema.set('strict', false); // Sets 'strict' to false
 *     schema.set('strict'); // 'false'
 *
 * @param {String***REMOVED*** key option name
 * @param {Object***REMOVED*** [value] if not passed, the current option value is returned
 * @see Schema ./
 * @api public
 */

Schema.prototype.set = function(key, value, _tags) {
  if (arguments.length === 1) {
    return this.options[key];
  ***REMOVED***

  switch (key) {
    case 'read':
      this.options[key] = readPref(value, _tags);
      this._userProvidedOptions[key] = this.options[key];
      break;
    case 'safe':
      setSafe(this.options, value);
      this._userProvidedOptions[key] = this.options[key];
      break;
    case 'timestamps':
      this.setupTimestamp(value);
      this.options[key] = value;
      this._userProvidedOptions[key] = this.options[key];
      break;
    default:
      this.options[key] = value;
      this._userProvidedOptions[key] = this.options[key];
      break;
  ***REMOVED***

  return this;
***REMOVED***;

/*!
 * ignore
 */

const safeDeprecationWarning = 'Mongoose: The `safe` option for schemas is ' +
  'deprecated. Use the `writeConcern` option instead: ' +
  'http://bit.ly/mongoose-write-concern';

const setSafe = util.deprecate(function setSafe(options, value) {
  options.safe = value === false ?
    {w: 0***REMOVED*** :
    value;
***REMOVED***, safeDeprecationWarning);

/**
 * Gets a schema option.
 *
 * @param {String***REMOVED*** key option name
 * @api public
 */

Schema.prototype.get = function(key) {
  return this.options[key];
***REMOVED***;

/**
 * The allowed index types
 *
 * @receiver Schema
 * @static indexTypes
 * @api public
 */

const indexTypes = '2d 2dsphere hashed text'.split(' ');

Object.defineProperty(Schema, 'indexTypes', {
  get: function() {
    return indexTypes;
  ***REMOVED***,
  set: function() {
    throw new Error('Cannot overwrite Schema.indexTypes');
  ***REMOVED***
***REMOVED***);

/**
 * Returns a list of indexes that this schema declares, via `schema.index()`
 * or by `index: true` in a path's options.
 *
 * @api public
 */

Schema.prototype.indexes = function() {
  return getIndexes(this);
***REMOVED***;

/**
 * Creates a virtual type with the given name.
 *
 * @param {String***REMOVED*** name
 * @param {Object***REMOVED*** [options]
 * @return {VirtualType***REMOVED***
 */

Schema.prototype.virtual = function(name, options) {
  if (options && options.ref) {
    if (!options.localField) {
      throw new Error('Reference virtuals require `localField` option');
    ***REMOVED***

    if (!options.foreignField) {
      throw new Error('Reference virtuals require `foreignField` option');
    ***REMOVED***

    this.pre('init', function(obj) {
      if (mpath.has(name, obj)) {
        const _v = mpath.get(name, obj);
        if (!this.$$populatedVirtuals) {
          this.$$populatedVirtuals = {***REMOVED***;
        ***REMOVED***

        if (options.justOne) {
          this.$$populatedVirtuals[name] = Array.isArray(_v) ?
            _v[0] :
            _v;
        ***REMOVED***
          this.$$populatedVirtuals[name] = Array.isArray(_v) ?
            _v :
            _v == null ? [] : [_v];
        ***REMOVED***

        mpath.unset(name, obj);
      ***REMOVED***
    ***REMOVED***);

    const virtual = this.virtual(name);
    virtual.options = options;
    return virtual.
      get(function() {
        if (!this.$$populatedVirtuals) {
          this.$$populatedVirtuals = {***REMOVED***;
        ***REMOVED***
        if (name in this.$$populatedVirtuals) {
          return this.$$populatedVirtuals[name];
        ***REMOVED***
        return null;
      ***REMOVED***).
      set(function(_v) {
        if (!this.$$populatedVirtuals) {
          this.$$populatedVirtuals = {***REMOVED***;
        ***REMOVED***

        if (options.justOne) {
          this.$$populatedVirtuals[name] = Array.isArray(_v) ?
            _v[0] :
            _v;

          if (typeof this.$$populatedVirtuals[name] !== 'object') {
            this.$$populatedVirtuals[name] = null;
          ***REMOVED***
        ***REMOVED***
          this.$$populatedVirtuals[name] = Array.isArray(_v) ?
            _v :
            _v == null ? [] : [_v];

          this.$$populatedVirtuals[name] = this.$$populatedVirtuals[name].filter(function(doc) {
            return doc && typeof doc === 'object';
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***);
  ***REMOVED***

  const virtuals = this.virtuals;
  const parts = name.split('.');

  if (this.pathType(name) === 'real') {
    throw new Error('Virtual path "' + name + '"' +
      ' conflicts with a real path in the schema');
  ***REMOVED***

  virtuals[name] = parts.reduce(function(mem, part, i) {
    mem[part] || (mem[part] = (i === parts.length - 1)
      ? new VirtualType(options, name)
      : {***REMOVED***);
    return mem[part];
  ***REMOVED***, this.tree);

  return virtuals[name];
***REMOVED***;

/**
 * Returns the virtual type with the given `name`.
 *
 * @param {String***REMOVED*** name
 * @return {VirtualType***REMOVED***
 */

Schema.prototype.virtualpath = function(name) {
  return this.virtuals[name];
***REMOVED***;

/**
 * Removes the given `path` (or [`paths`]).
 *
 * @param {String|Array***REMOVED*** path
 * @return {Schema***REMOVED*** the Schema instance
 * @api public
 */
Schema.prototype.remove = function(path) {
  if (typeof path === 'string') {
    path = [path];
  ***REMOVED***
  if (Array.isArray(path)) {
    path.forEach(function(name) {
      if (this.path(name)) {
        delete this.paths[name];

        const pieces = name.split('.');
        const last = pieces.pop();
        let branch = this.tree;
        for (let i = 0; i < pieces.length; ++i) {
          branch = branch[pieces[i]];
        ***REMOVED***
        delete branch[last];
      ***REMOVED***
    ***REMOVED***, this);
  ***REMOVED***
  return this;
***REMOVED***;

/**
 * Loads an ES6 class into a schema. Maps [setters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set) + [getters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get), [static methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static),
 * and [instance methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#Class_body_and_method_definitions)
 * to schema [virtuals](http://mongoosejs.com/docs/guide.html#virtuals),
 * [statics](http://mongoosejs.com/docs/guide.html#statics), and
 * [methods](http://mongoosejs.com/docs/guide.html#methods).
 *
 * ####Example:
 *
 * ```javascript
 * const md5 = require('md5');
 * const userSchema = new Schema({ email: String ***REMOVED***);
 * class UserClass {
 *   // `gravatarImage` becomes a virtual
 *   get gravatarImage() {
 *     const hash = md5(this.email.toLowerCase());
 *     return `https://www.gravatar.com/avatar/${hash***REMOVED***`;
 *   ***REMOVED***
 *
 *   // `getProfileUrl()` becomes a document method
 *   getProfileUrl() {
 *     return `https://mysite.com/${this.email***REMOVED***`;
 *   ***REMOVED***
 *
 *   // `findByEmail()` becomes a static
 *   static findByEmail(email) {
 *     return this.findOne({ email ***REMOVED***);
 *   ***REMOVED***
 * ***REMOVED***
 *
 * // `schema` will now have a `gravatarImage` virtual, a `getProfileUrl()` method,
 * // and a `findByEmail()` static
 * userSchema.loadClass(UserClass);
 * ```
 *
 * @param {Function***REMOVED*** model
 * @param {Boolean***REMOVED*** [virtualsOnly] if truthy, only pulls virtuals from the class, not methods or statics
 */
Schema.prototype.loadClass = function(model, virtualsOnly) {
  if (model === Object.prototype ||
      model === Function.prototype ||
      model.prototype.hasOwnProperty('$isMongooseModelPrototype')) {
    return this;
  ***REMOVED***

  this.loadClass(Object.getPrototypeOf(model));

  // Add static methods
  if (!virtualsOnly) {
    Object.getOwnPropertyNames(model).forEach(function(name) {
      if (name.match(/^(length|name|prototype)$/)) {
        return;
      ***REMOVED***
      const method = Object.getOwnPropertyDescriptor(model, name);
      if (typeof method.value === 'function') {
        this.static(name, method.value);
      ***REMOVED***
    ***REMOVED***, this);
  ***REMOVED***

  // Add methods and virtuals
  Object.getOwnPropertyNames(model.prototype).forEach(function(name) {
    if (name.match(/^(constructor)$/)) {
      return;
    ***REMOVED***
    const method = Object.getOwnPropertyDescriptor(model.prototype, name);
    if (!virtualsOnly) {
      if (typeof method.value === 'function') {
        this.method(name, method.value);
      ***REMOVED***
    ***REMOVED***
    if (typeof method.get === 'function') {
      this.virtual(name).get(method.get);
    ***REMOVED***
    if (typeof method.set === 'function') {
      this.virtual(name).set(method.set);
    ***REMOVED***
  ***REMOVED***, this);

  return this;
***REMOVED***;

/*!
 * ignore
 */

Schema.prototype._getSchema = function(path) {
  const _this = this;
  const pathschema = _this.path(path);
  const resultPath = [];

  if (pathschema) {
    pathschema.$fullPath = path;
    return pathschema;
  ***REMOVED***

  function search(parts, schema) {
    let p = parts.length + 1;
    let foundschema;
    let trypath;

    while (p--) {
      trypath = parts.slice(0, p).join('.');
      foundschema = schema.path(trypath);
      if (foundschema) {
        resultPath.push(trypath);

        if (foundschema.caster) {
          // array of Mixed?
          if (foundschema.caster instanceof MongooseTypes.Mixed) {
            foundschema.caster.$fullPath = resultPath.join('.');
            return foundschema.caster;
          ***REMOVED***

          // Now that we found the array, we need to check if there
          // are remaining document paths to look up for casting.
          // Also we need to handle array.$.path since schema.path
          // doesn't work for that.
          // If there is no foundschema.schema we are dealing with
          // a path like array.$
          if (p !== parts.length && foundschema.schema) {
            let ret;
            if (parts[p] === '$' || isArrayFilter(parts[p])) {
              if (p + 1 === parts.length) {
                // comments.$
                return foundschema;
              ***REMOVED***
              // comments.$.comments.$.title
              ret = search(parts.slice(p + 1), foundschema.schema);
              if (ret) {
                ret.$isUnderneathDocArray = ret.$isUnderneathDocArray ||
                  !foundschema.schema.$isSingleNested;
              ***REMOVED***
              return ret;
            ***REMOVED***
            // this is the last path of the selector
            ret = search(parts.slice(p), foundschema.schema);
            if (ret) {
              ret.$isUnderneathDocArray = ret.$isUnderneathDocArray ||
                !foundschema.schema.$isSingleNested;
            ***REMOVED***
            return ret;
          ***REMOVED***
        ***REMOVED***

        foundschema.$fullPath = resultPath.join('.');

        return foundschema;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  // look for arrays
  const parts = path.split('.');
  for (let i = 0; i < parts.length; ++i) {
    if (parts[i] === '$' || isArrayFilter(parts[i])) {
      // Re: gh-5628, because `schema.path()` doesn't take $ into account.
      parts[i] = '0';
    ***REMOVED***
  ***REMOVED***
  return search(parts, _this);
***REMOVED***;

/*!
 * ignore
 */

Schema.prototype._getPathType = function(path) {
  const _this = this;
  const pathschema = _this.path(path);

  if (pathschema) {
    return 'real';
  ***REMOVED***

  function search(parts, schema) {
    let p = parts.length + 1,
        foundschema,
        trypath;

    while (p--) {
      trypath = parts.slice(0, p).join('.');
      foundschema = schema.path(trypath);
      if (foundschema) {
        if (foundschema.caster) {
          // array of Mixed?
          if (foundschema.caster instanceof MongooseTypes.Mixed) {
            return { schema: foundschema, pathType: 'mixed' ***REMOVED***;
          ***REMOVED***

          // Now that we found the array, we need to check if there
          // are remaining document paths to look up for casting.
          // Also we need to handle array.$.path since schema.path
          // doesn't work for that.
          // If there is no foundschema.schema we are dealing with
          // a path like array.$
          if (p !== parts.length && foundschema.schema) {
            if (parts[p] === '$' || isArrayFilter(parts[p])) {
              if (p === parts.length - 1) {
                return { schema: foundschema, pathType: 'nested' ***REMOVED***;
              ***REMOVED***
              // comments.$.comments.$.title
              return search(parts.slice(p + 1), foundschema.schema);
            ***REMOVED***
            // this is the last path of the selector
            return search(parts.slice(p), foundschema.schema);
          ***REMOVED***
          return {
            schema: foundschema,
            pathType: foundschema.$isSingleNested ? 'nested' : 'array'
          ***REMOVED***;
        ***REMOVED***
        return { schema: foundschema, pathType: 'real' ***REMOVED***;
      ***REMOVED*** else if (p === parts.length && schema.nested[trypath]) {
        return { schema: schema, pathType: 'nested' ***REMOVED***;
      ***REMOVED***
    ***REMOVED***
    return { schema: foundschema || schema, pathType: 'undefined' ***REMOVED***;
  ***REMOVED***

  // look for arrays
  return search(path.split('.'), _this);
***REMOVED***;

/*!
 * ignore
 */

function isArrayFilter(piece) {
  return piece.indexOf('$[') === 0 &&
    piece.lastIndexOf(']') === piece.length - 1;
***REMOVED***

/*!
 * Module exports.
 */

module.exports = exports = Schema;

// require down here because of reference issues

/**
 * The various built-in Mongoose Schema Types.
 *
 * ####Example:
 *
 *     var mongoose = require('mongoose');
 *     var ObjectId = mongoose.Schema.Types.ObjectId;
 *
 * ####Types:
 *
 * - [String](#schema-string-js)
 * - [Number](#schema-number-js)
 * - [Boolean](#schema-boolean-js) | Bool
 * - [Array](#schema-array-js)
 * - [Buffer](#schema-buffer-js)
 * - [Date](#schema-date-js)
 * - [ObjectId](#schema-objectid-js) | Oid
 * - [Mixed](#schema-mixed-js)
 *
 * Using this exposed access to the `Mixed` SchemaType, we can use them in our schema.
 *
 *     var Mixed = mongoose.Schema.Types.Mixed;
 *     new mongoose.Schema({ _user: Mixed ***REMOVED***)
 *
 * @api public
 */

Schema.Types = MongooseTypes = require('./schema/index');

/*!
 * ignore
 */

exports.ObjectId = MongooseTypes.ObjectId;
