// mongoose/lib/utils.js
'use strict';

/*!
 * Module dependencies.
 */

const Decimal = require('./types/decimal128');
const ObjectId = require('./types/objectid');
const PromiseProvider = require('./promise_provider');
const cloneRegExp = require('regexp-clone');
const get = require('lodash.get');
const sliced = require('sliced');
const mpath = require('mpath');
const ms = require('ms');
const Buffer = require('safe-buffer').Buffer;

let MongooseBuffer;
let MongooseArray;
let Document;

/*!
 * Produces a collection name from model `name`. By default, just returns
 * the model name
 *
 * @param {String***REMOVED*** name a model name
 * @param {Function***REMOVED*** pluralize function that pluralizes the collection name
 * @return {String***REMOVED*** a collection name
 * @api private
 */

exports.toCollectionName = function(name, pluralize) {
  if (name === 'system.profile') {
    return name;
  ***REMOVED***
  if (name === 'system.indexes') {
    return name;
  ***REMOVED***
  if (typeof pluralize === 'function') {
    return pluralize(name);
  ***REMOVED***
  return name;
***REMOVED***;

/*!
 * Determines if `a` and `b` are deep equal.
 *
 * Modified from node/lib/assert.js
 *
 * @param {any***REMOVED*** a a value to compare to `b`
 * @param {any***REMOVED*** b a value to compare to `a`
 * @return {Boolean***REMOVED***
 * @api private
 */

exports.deepEqual = function deepEqual(a, b) {
  if (a === b) {
    return true;
  ***REMOVED***

  if (a instanceof Date && b instanceof Date) {
    return a.getTime() === b.getTime();
  ***REMOVED***

  if ((isBsonType(a, 'ObjectID') && isBsonType(b, 'ObjectID')) ||
      (isBsonType(a, 'Decimal128') && isBsonType(b, 'Decimal128'))) {
    return a.toString() === b.toString();
  ***REMOVED***

  if (a instanceof RegExp && b instanceof RegExp) {
    return a.source === b.source &&
        a.ignoreCase === b.ignoreCase &&
        a.multiline === b.multiline &&
        a.global === b.global;
  ***REMOVED***

  if (typeof a !== 'object' && typeof b !== 'object') {
    return a == b;
  ***REMOVED***

  if (a === null || b === null || a === undefined || b === undefined) {
    return false;
  ***REMOVED***

  if (a.prototype !== b.prototype) {
    return false;
  ***REMOVED***

  // Handle MongooseNumbers
  if (a instanceof Number && b instanceof Number) {
    return a.valueOf() === b.valueOf();
  ***REMOVED***

  if (Buffer.isBuffer(a)) {
    return exports.buffer.areEqual(a, b);
  ***REMOVED***

  if (isMongooseObject(a)) {
    a = a.toObject();
  ***REMOVED***
  if (isMongooseObject(b)) {
    b = b.toObject();
  ***REMOVED***

  let ka;
  let kb;
  let key;
  let i;
  try {
    ka = Object.keys(a);
    kb = Object.keys(b);
  ***REMOVED*** catch (e) {
    // happens when one is a string literal and the other isn't
    return false;
  ***REMOVED***

  // having the same number of owned properties (keys incorporates
  // hasOwnProperty)
  if (ka.length !== kb.length) {
    return false;
  ***REMOVED***

  // the same set of keys (although not necessarily the same order),
  ka.sort();
  kb.sort();

  // ~~~cheap key test
  for (i = ka.length - 1; i >= 0; i--) {
    if (ka[i] !== kb[i]) {
      return false;
    ***REMOVED***
  ***REMOVED***

  // equivalent values for every corresponding key, and
  // ~~~possibly expensive deep test
  for (i = ka.length - 1; i >= 0; i--) {
    key = ka[i];
    if (!deepEqual(a[key], b[key])) {
      return false;
    ***REMOVED***
  ***REMOVED***

  return true;
***REMOVED***;

/*!
 * Get the bson type, if it exists
 */

function isBsonType(obj, typename) {
  return get(obj, '_bsontype', void 0) === typename;
***REMOVED***

/*!
 * Get the last element of an array
 */

exports.last = function(arr) {
  if (arr.length > 0) {
    return arr[arr.length - 1];
  ***REMOVED***
  return void 0;
***REMOVED***;

/*!
 * Object clone with Mongoose natives support.
 *
 * If options.minimize is true, creates a minimal data object. Empty objects and undefined values will not be cloned. This makes the data payload sent to MongoDB as small as possible.
 *
 * Functions are never cloned.
 *
 * @param {Object***REMOVED*** obj the object to clone
 * @param {Object***REMOVED*** options
 * @return {Object***REMOVED*** the cloned object
 * @api private
 */

exports.clone = function clone(obj, options) {
  if (obj === undefined || obj === null) {
    return obj;
  ***REMOVED***

  if (Array.isArray(obj)) {
    return cloneArray(obj, options);
  ***REMOVED***

  if (isMongooseObject(obj)) {
    if (options && options.json && typeof obj.toJSON === 'function') {
      return obj.toJSON(options);
    ***REMOVED***
    return obj.toObject(options);
  ***REMOVED***

  if (obj.constructor) {
    switch (exports.getFunctionName(obj.constructor)) {
      case 'Object':
        return cloneObject(obj, options);
      case 'Date':
        return new obj.constructor(+obj);
      case 'RegExp':
        return cloneRegExp(obj);
      default:
        // ignore
        break;
    ***REMOVED***
  ***REMOVED***

  if (obj instanceof ObjectId) {
    return new ObjectId(obj.id);
  ***REMOVED***
  if (isBsonType(obj, 'Decimal128')) {
    if (options && options.flattenDecimals) {
      return obj.toJSON();
    ***REMOVED***
    return Decimal.fromString(obj.toString());
  ***REMOVED***

  if (!obj.constructor && exports.isObject(obj)) {
    // object created with Object.create(null)
    return cloneObject(obj, options);
  ***REMOVED***

  if (obj.valueOf) {
    return obj.valueOf();
  ***REMOVED***
***REMOVED***;
const clone = exports.clone;

/*!
 * ignore
 */

exports.promiseOrCallback = function promiseOrCallback(callback, fn) {
  if (typeof callback === 'function') {
    try {
      return fn(callback);
    ***REMOVED*** catch (error) {
      return process.nextTick(() => {
        throw error;
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  const Promise = PromiseProvider.get();

  return new Promise((resolve, reject) => {
    fn(function(error, res) {
      if (error != null) {
        return reject(error);
      ***REMOVED***
      if (arguments.length > 2) {
        return resolve(Array.prototype.slice.call(arguments, 1));
      ***REMOVED***
      resolve(res);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***;

/*!
 * ignore
 */

function cloneObject(obj, options) {
  const minimize = options && options.minimize;
  const ret = {***REMOVED***;
  let hasKeys;
  let val;
  let k;

  for (k in obj) {
    val = clone(obj[k], options);

    if (!minimize || (typeof val !== 'undefined')) {
      hasKeys || (hasKeys = true);
      ret[k] = val;
    ***REMOVED***
  ***REMOVED***

  return minimize ? hasKeys && ret : ret;
***REMOVED***

function cloneArray(arr, options) {
  const ret = [];
  for (let i = 0, l = arr.length; i < l; i++) {
    ret.push(clone(arr[i], options));
  ***REMOVED***
  return ret;
***REMOVED***

/*!
 * Shallow copies defaults into options.
 *
 * @param {Object***REMOVED*** defaults
 * @param {Object***REMOVED*** options
 * @return {Object***REMOVED*** the merged object
 * @api private
 */

exports.options = function(defaults, options) {
  const keys = Object.keys(defaults);
  let i = keys.length;
  let k;

  options = options || {***REMOVED***;

  while (i--) {
    k = keys[i];
    if (!(k in options)) {
      options[k] = defaults[k];
    ***REMOVED***
  ***REMOVED***

  return options;
***REMOVED***;

/*!
 * Generates a random string
 *
 * @api private
 */

exports.random = function() {
  return Math.random().toString().substr(3);
***REMOVED***;

/*!
 * Merges `from` into `to` without overwriting existing properties.
 *
 * @param {Object***REMOVED*** to
 * @param {Object***REMOVED*** from
 * @api private
 */

exports.merge = function merge(to, from, options, path) {
  options = options || {***REMOVED***;

  const keys = Object.keys(from);
  let i = 0;
  const len = keys.length;
  let key;

  path = path || '';
  const omitNested = options.omitNested || {***REMOVED***;

  while (i < len) {
    key = keys[i++];
    if (options.omit && options.omit[key]) {
      continue;
    ***REMOVED***
    if (omitNested[path]) {
      continue;
    ***REMOVED***
    if (to[key] == null) {
      to[key] = from[key];
    ***REMOVED*** else if (exports.isObject(from[key])) {
      if (!exports.isObject(to[key])) {
        to[key] = {***REMOVED***;
      ***REMOVED***
      if (from[key] != null && from[key].instanceOfSchema) {
        to[key] = from[key].clone();
        continue;
      ***REMOVED***
      merge(to[key], from[key], options, path ? path + '.' + key : key);
    ***REMOVED*** else if (options.overwrite) {
      to[key] = from[key];
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/*!
 * Applies toObject recursively.
 *
 * @param {Document|Array|Object***REMOVED*** obj
 * @return {Object***REMOVED***
 * @api private
 */

exports.toObject = function toObject(obj) {
  Document || (Document = require('./document'));
  let ret;

  if (obj == null) {
    return obj;
  ***REMOVED***

  if (obj instanceof Document) {
    return obj.toObject();
  ***REMOVED***

  if (Array.isArray(obj)) {
    ret = [];

    for (let i = 0, len = obj.length; i < len; ++i) {
      ret.push(toObject(obj[i]));
    ***REMOVED***

    return ret;
  ***REMOVED***

  if ((obj.constructor && exports.getFunctionName(obj.constructor) === 'Object') ||
      (!obj.constructor && exports.isObject(obj))) {
    ret = {***REMOVED***;

    for (const k in obj) {
      ret[k] = toObject(obj[k]);
    ***REMOVED***

    return ret;
  ***REMOVED***

  return obj;
***REMOVED***;

/*!
 * Determines if `arg` is an object.
 *
 * @param {Object|Array|String|Function|RegExp|any***REMOVED*** arg
 * @api private
 * @return {Boolean***REMOVED***
 */

exports.isObject = function(arg) {
  if (Buffer.isBuffer(arg)) {
    return true;
  ***REMOVED***
  return Object.prototype.toString.call(arg) === '[object Object]';
***REMOVED***;

/*!
 * Determines if `arg` is a plain object.
 *
 * @param {Object|Array|String|Function|RegExp|any***REMOVED*** arg
 * @api private
 * @return {Boolean***REMOVED***
 */

exports.isPOJO = function(arg) {
  return get(arg, 'constructor.name') === 'Object';
***REMOVED***;

/*!
 * A faster Array.prototype.slice.call(arguments) alternative
 * @api private
 */

exports.args = sliced;

/*!
 * process.nextTick helper.
 *
 * Wraps `callback` in a try/catch + nextTick.
 *
 * node-mongodb-native has a habit of state corruption when an error is immediately thrown from within a collection callback.
 *
 * @param {Function***REMOVED*** callback
 * @api private
 */

exports.tick = function tick(callback) {
  if (typeof callback !== 'function') {
    return;
  ***REMOVED***
  return function() {
    try {
      callback.apply(this, arguments);
    ***REMOVED*** catch (err) {
      // only nextTick on err to get out of
      // the event loop and avoid state corruption.
      process.nextTick(function() {
        throw err;
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***;
***REMOVED***;

/*!
 * Returns if `v` is a mongoose object that has a `toObject()` method we can use.
 *
 * This is for compatibility with libs like Date.js which do foolish things to Natives.
 *
 * @param {any***REMOVED*** v
 * @api private
 */

exports.isMongooseObject = function(v) {
  Document || (Document = require('./document'));
  MongooseArray || (MongooseArray = require('./types').Array);
  MongooseBuffer || (MongooseBuffer = require('./types').Buffer);

  if (v == null) {
    return false;
  ***REMOVED***

  return v.$__ != null || // Document
    v.isMongooseArray || // Array or Document Array
    v.isMongooseBuffer || // Buffer
    v.$isMongooseMap; // Map
***REMOVED***;

const isMongooseObject = exports.isMongooseObject;

/*!
 * Converts `expires` options of index objects to `expiresAfterSeconds` options for MongoDB.
 *
 * @param {Object***REMOVED*** object
 * @api private
 */

exports.expires = function expires(object) {
  if (!(object && object.constructor.name === 'Object')) {
    return;
  ***REMOVED***
  if (!('expires' in object)) {
    return;
  ***REMOVED***

  let when;
  if (typeof object.expires !== 'string') {
    when = object.expires;
  ***REMOVED***
    when = Math.round(ms(object.expires) / 1000);
  ***REMOVED***
  object.expireAfterSeconds = when;
  delete object.expires;
***REMOVED***;

/*!
 * Populate options constructor
 */

function PopulateOptions(path, select, match, options, model, subPopulate, justOne) {
  this.path = path;
  this.match = match;
  this.select = select;
  this.options = options;
  this.model = model;
  if (typeof subPopulate === 'object') {
    this.populate = subPopulate;
  ***REMOVED***
  if (justOne != null) {
    this.justOne = justOne;
  ***REMOVED***
  this._docs = {***REMOVED***;
***REMOVED***

// make it compatible with utils.clone
PopulateOptions.prototype.constructor = Object;

// expose
exports.PopulateOptions = PopulateOptions;

/*!
 * populate helper
 */

exports.populate = function populate(path, select, model, match, options, subPopulate, justOne) {
  // The order of select/conditions args is opposite Model.find but
  // necessary to keep backward compatibility (select could be
  // an array, string, or object literal).
  function makeSingles(arr) {
    const ret = [];
    arr.forEach(function(obj) {
      if (/[\s]/.test(obj.path)) {
        const paths = obj.path.split(' ');
        paths.forEach(function(p) {
          const copy = Object.assign({***REMOVED***, obj);
          copy.path = p;
          ret.push(copy);
        ***REMOVED***);
      ***REMOVED***
        ret.push(obj);
      ***REMOVED***
    ***REMOVED***);

    return ret;
  ***REMOVED***

  // might have passed an object specifying all arguments
  if (arguments.length === 1) {
    if (path instanceof PopulateOptions) {
      return [path];
    ***REMOVED***

    if (Array.isArray(path)) {
      const singles = makeSingles(path);
      return singles.map(function(o) {
        if (o.populate && !(o.match || o.options)) {
          return exports.populate(
            o.path,
            o.select,
            o.model,
            o.match,
            o.options,
            o.populate,
            o.justOne)[0];
        ***REMOVED***
          return exports.populate(o)[0];
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***

    if (exports.isObject(path)) {
      match = path.match;
      options = path.options;
      select = path.select;
      model = path.model;
      subPopulate = path.populate;
      justOne = path.justOne;
      path = path.path;
    ***REMOVED***
  ***REMOVED*** else if (typeof model === 'object') {
    options = match;
    match = model;
    model = undefined;
  ***REMOVED***

  if (typeof path !== 'string') {
    throw new TypeError('utils.populate: invalid path. Expected string. Got typeof `' + typeof path + '`');
  ***REMOVED***

  if (Array.isArray(subPopulate)) {
    const ret = [];
    subPopulate.forEach(function(obj) {
      if (/[\s]/.test(obj.path)) {
        const copy = Object.assign({***REMOVED***, obj);
        const paths = copy.path.split(' ');
        paths.forEach(function(p) {
          copy.path = p;
          ret.push(exports.populate(copy)[0]);
        ***REMOVED***);
      ***REMOVED***
        ret.push(exports.populate(obj)[0]);
      ***REMOVED***
    ***REMOVED***);
    subPopulate = exports.populate(ret);
  ***REMOVED*** else if (typeof subPopulate === 'object') {
    subPopulate = exports.populate(subPopulate);
  ***REMOVED***

  const ret = [];
  const paths = path.split(' ');
  options = exports.clone(options);
  for (let i = 0; i < paths.length; ++i) {
    ret.push(new PopulateOptions(paths[i], select, match, options, model, subPopulate, justOne));
  ***REMOVED***

  return ret;
***REMOVED***;

/*!
 * Return the value of `obj` at the given `path`.
 *
 * @param {String***REMOVED*** path
 * @param {Object***REMOVED*** obj
 */

exports.getValue = function(path, obj, map) {
  return mpath.get(path, obj, '_doc', map);
***REMOVED***;

/*!
 * Sets the value of `obj` at the given `path`.
 *
 * @param {String***REMOVED*** path
 * @param {Anything***REMOVED*** val
 * @param {Object***REMOVED*** obj
 */

exports.setValue = function(path, val, obj, map, _copying) {
  mpath.set(path, val, obj, '_doc', map, _copying);
***REMOVED***;

/*!
 * Returns an array of values from object `o`.
 *
 * @param {Object***REMOVED*** o
 * @return {Array***REMOVED***
 * @private
 */

exports.object = {***REMOVED***;
exports.object.vals = function vals(o) {
  const keys = Object.keys(o);
  let i = keys.length;
  const ret = [];

  while (i--) {
    ret.push(o[keys[i]]);
  ***REMOVED***

  return ret;
***REMOVED***;

/*!
 * @see exports.options
 */

exports.object.shallowCopy = exports.options;

/*!
 * Safer helper for hasOwnProperty checks
 *
 * @param {Object***REMOVED*** obj
 * @param {String***REMOVED*** prop
 */

const hop = Object.prototype.hasOwnProperty;
exports.object.hasOwnProperty = function(obj, prop) {
  return hop.call(obj, prop);
***REMOVED***;

/*!
 * Determine if `val` is null or undefined
 *
 * @return {Boolean***REMOVED***
 */

exports.isNullOrUndefined = function(val) {
  return val === null || val === undefined;
***REMOVED***;

/*!
 * ignore
 */

exports.array = {***REMOVED***;

/*!
 * Flattens an array.
 *
 * [ 1, [ 2, 3, [4] ]] -> [1,2,3,4]
 *
 * @param {Array***REMOVED*** arr
 * @param {Function***REMOVED*** [filter] If passed, will be invoked with each item in the array. If `filter` returns a falsey value, the item will not be included in the results.
 * @return {Array***REMOVED***
 * @private
 */

exports.array.flatten = function flatten(arr, filter, ret) {
  ret || (ret = []);

  arr.forEach(function(item) {
    if (Array.isArray(item)) {
      flatten(item, filter, ret);
    ***REMOVED***
      if (!filter || filter(item)) {
        ret.push(item);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);

  return ret;
***REMOVED***;

/*!
 * Removes duplicate values from an array
 *
 * [1, 2, 3, 3, 5] => [1, 2, 3, 5]
 * [ ObjectId("550988ba0c19d57f697dc45e"), ObjectId("550988ba0c19d57f697dc45e") ]
 *    => [ObjectId("550988ba0c19d57f697dc45e")]
 *
 * @param {Array***REMOVED*** arr
 * @return {Array***REMOVED***
 * @private
 */

exports.array.unique = function(arr) {
  const primitives = {***REMOVED***;
  const ids = {***REMOVED***;
  const ret = [];
  const length = arr.length;
  for (let i = 0; i < length; ++i) {
    if (typeof arr[i] === 'number' || typeof arr[i] === 'string' || arr[i] == null) {
      if (primitives[arr[i]]) {
        continue;
      ***REMOVED***
      ret.push(arr[i]);
      primitives[arr[i]] = true;
    ***REMOVED*** else if (arr[i] instanceof ObjectId) {
      if (ids[arr[i].toString()]) {
        continue;
      ***REMOVED***
      ret.push(arr[i]);
      ids[arr[i].toString()] = true;
    ***REMOVED***
      ret.push(arr[i]);
    ***REMOVED***
  ***REMOVED***

  return ret;
***REMOVED***;

/*!
 * Determines if two buffers are equal.
 *
 * @param {Buffer***REMOVED*** a
 * @param {Object***REMOVED*** b
 */

exports.buffer = {***REMOVED***;
exports.buffer.areEqual = function(a, b) {
  if (!Buffer.isBuffer(a)) {
    return false;
  ***REMOVED***
  if (!Buffer.isBuffer(b)) {
    return false;
  ***REMOVED***
  if (a.length !== b.length) {
    return false;
  ***REMOVED***
  for (let i = 0, len = a.length; i < len; ++i) {
    if (a[i] !== b[i]) {
      return false;
    ***REMOVED***
  ***REMOVED***
  return true;
***REMOVED***;

exports.getFunctionName = function(fn) {
  if (fn.name) {
    return fn.name;
  ***REMOVED***
  return (fn.toString().trim().match(/^function\s*([^\s(]+)/) || [])[1];
***REMOVED***;

exports.decorate = function(destination, source) {
  for (const key in source) {
    destination[key] = source[key];
  ***REMOVED***
***REMOVED***;

/**
 * merges to with a copy of from
 *
 * @param {Object***REMOVED*** to
 * @param {Object***REMOVED*** fromObj
 * @api private
 */

exports.mergeClone = function(to, fromObj) {
  if (isMongooseObject(fromObj)) {
    fromObj = fromObj.toObject({
      transform: false,
      virtuals: false,
      depopulate: true,
      getters: false,
      flattenDecimals: false
    ***REMOVED***);
  ***REMOVED***
  const keys = Object.keys(fromObj);
  const len = keys.length;
  let i = 0;
  let key;

  while (i < len) {
    key = keys[i++];
    if (typeof to[key] === 'undefined') {
      to[key] = exports.clone(fromObj[key], {
        transform: false,
        virtuals: false,
        depopulate: true,
        getters: false,
        flattenDecimals: false
      ***REMOVED***);
    ***REMOVED***
      let val = fromObj[key];
      if (val != null && val.valueOf && !(val instanceof Date)) {
        val = val.valueOf();
      ***REMOVED***
      if (exports.isObject(val)) {
        let obj = val;
        if (isMongooseObject(val) && !val.isMongooseBuffer) {
          obj = obj.toObject({
            transform: false,
            virtuals: false,
            depopulate: true,
            getters: false,
            flattenDecimals: false
          ***REMOVED***);
        ***REMOVED***
        if (val.isMongooseBuffer) {
          obj = Buffer.from(obj);
        ***REMOVED***
        exports.mergeClone(to[key], obj);
      ***REMOVED***
        to[key] = exports.clone(val, {
          flattenDecimals: false
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/**
 * Executes a function on each element of an array (like _.each)
 *
 * @param {Array***REMOVED*** arr
 * @param {Function***REMOVED*** fn
 * @api private
 */

exports.each = function(arr, fn) {
  for (let i = 0; i < arr.length; ++i) {
    fn(arr[i]);
  ***REMOVED***
***REMOVED***;

/*!
 * ignore
 */

exports.noop = function() {***REMOVED***;
