// mongoose/lib/virtualtype.js
'use strict';

/**
 * VirtualType constructor
 *
 * This is what mongoose uses to define virtual attributes via `Schema.prototype.virtual`.
 *
 * ####Example:
 *
 *     const fullname = schema.virtual('fullname');
 *     fullname instanceof mongoose.VirtualType // true
 *
 * @param {Object***REMOVED*** options
 * @param {string|function***REMOVED*** [options.ref] if `ref` is not nullish, this becomes a [populated virtual](/docs/populate.html#populate-virtuals)
 * @param {string|function***REMOVED*** [options.localField] the local field to populate on if this is a populated virtual.
 * @param {string|function***REMOVED*** [options.foreignField] the foreign field to populate on if this is a populated virtual.
 * @param {boolean***REMOVED*** [options.justOne=false] by default, a populated virtual is an array. If you set `justOne`, the populated virtual will be a single doc or `null`.
 * @param {boolean***REMOVED*** [options.getters=false] if you set this to `true`, Mongoose will call any custom getters you defined on this virtual
 * @api public
 */

function VirtualType(options, name) {
  this.path = name;
  this.getters = [];
  this.setters = [];
  this.options = Object.assign({***REMOVED***, options);
***REMOVED***

/**
 * If no getters/getters, add a default
 *
 * @param {Function***REMOVED*** fn
 * @return {VirtualType***REMOVED*** this
 * @api public
 */

VirtualType.prototype._applyDefaultGetters = function() {
  if (this.getters.length > 0 || this.setters.length > 0) {
    return;
  ***REMOVED***

  const path = this.path;
  const internalProperty = '$' + path;
  this.getters.push(function() {
    return this[internalProperty];
  ***REMOVED***);
  this.setters.push(function(v) {
    this[internalProperty] = v;
  ***REMOVED***);
***REMOVED***;

/*!
 * ignore
 */

VirtualType.prototype.clone = function() {
  const clone = new VirtualType(this.name, this.options);
  clone.getters = [].concat(this.getters);
  clone.setters = [].concat(this.setters);
  return clone;
***REMOVED***;

/**
 * Defines a getter.
 *
 * ####Example:
 *
 *     var virtual = schema.virtual('fullname');
 *     virtual.get(function () {
 *       return this.name.first + ' ' + this.name.last;
 *     ***REMOVED***);
 *
 * @param {Function***REMOVED*** fn
 * @return {VirtualType***REMOVED*** this
 * @api public
 */

VirtualType.prototype.get = function(fn) {
  this.getters.push(fn);
  return this;
***REMOVED***;

/**
 * Defines a setter.
 *
 * ####Example:
 *
 *     var virtual = schema.virtual('fullname');
 *     virtual.set(function (v) {
 *       var parts = v.split(' ');
 *       this.name.first = parts[0];
 *       this.name.last = parts[1];
 *     ***REMOVED***);
 *
 * @param {Function***REMOVED*** fn
 * @return {VirtualType***REMOVED*** this
 * @api public
 */

VirtualType.prototype.set = function(fn) {
  this.setters.push(fn);
  return this;
***REMOVED***;

/**
 * Applies getters to `value` using optional `scope`.
 *
 * @param {Object***REMOVED*** value
 * @param {Object***REMOVED*** scope
 * @return {any***REMOVED*** the value after applying all getters
 * @api public
 */

VirtualType.prototype.applyGetters = function(value, scope) {
  let v = value;
  for (let l = this.getters.length - 1; l >= 0; l--) {
    v = this.getters[l].call(scope, v, this);
  ***REMOVED***
  return v;
***REMOVED***;

/**
 * Applies setters to `value` using optional `scope`.
 *
 * @param {Object***REMOVED*** value
 * @param {Object***REMOVED*** scope
 * @return {any***REMOVED*** the value after applying all setters
 * @api public
 */

VirtualType.prototype.applySetters = function(value, scope) {
  let v = value;
  for (let l = this.setters.length - 1; l >= 0; l--) {
    v = this.setters[l].call(scope, v, this);
  ***REMOVED***
  return v;
***REMOVED***;

/*!
 * exports
 */

module.exports = VirtualType;
