// mongodb/lib/bulk/common.js
'use strict';

const Long = require('mongodb-core').BSON.Long;
const MongoError = require('mongodb-core').MongoError;
const toError = require('../utils').toError;
const handleCallback = require('../utils').handleCallback;
const applyRetryableWrites = require('../utils').applyRetryableWrites;
const applyWriteConcern = require('../utils').applyWriteConcern;
const ObjectID = require('mongodb-core').BSON.ObjectID;
const BSON = require('mongodb-core').BSON;

// Error codes
const UNKNOWN_ERROR = 8;
const INVALID_BSON_ERROR = 22;
const WRITE_CONCERN_ERROR = 64;
const MULTIPLE_ERROR = 65;

// Insert types
const INSERT = 1;
const UPDATE = 2;
const REMOVE = 3;

const bson = new BSON([
  BSON.Binary,
  BSON.Code,
  BSON.DBRef,
  BSON.Decimal128,
  BSON.Double,
  BSON.Int32,
  BSON.Long,
  BSON.Map,
  BSON.MaxKey,
  BSON.MinKey,
  BSON.ObjectId,
  BSON.BSONRegExp,
  BSON.Symbol,
  BSON.Timestamp
]);

/**
 * Keeps the state of a unordered batch so we can rewrite the results
 * correctly after command execution
 * @ignore
 */
class Batch {
  constructor(batchType, originalZeroIndex) {
    this.originalZeroIndex = originalZeroIndex;
    this.currentIndex = 0;
    this.originalIndexes = [];
    this.batchType = batchType;
    this.operations = [];
    this.size = 0;
    this.sizeBytes = 0;
  ***REMOVED***
***REMOVED***

/**
 * Wraps a legacy operation so we can correctly rewrite it's error
 * @ignore
 */
class LegacyOp {
  constructor(batchType, operation, index) {
    this.batchType = batchType;
    this.index = index;
    this.operation = operation;
  ***REMOVED***
***REMOVED***

/**
 * Create a new BulkWriteResult instance (INTERNAL TYPE, do not instantiate directly)
 *
 * @class
 * @return {BulkWriteResult***REMOVED*** a BulkWriteResult instance
 */
class BulkWriteResult {
  constructor(bulkResult) {
    this.result = bulkResult;
  ***REMOVED***

  /**
   * @return {boolean***REMOVED*** ok Did bulk operation correctly execute
   */
  get ok() {
    return this.result.ok;
  ***REMOVED***

  /**
   * @return {number***REMOVED*** nInserted number of inserted documents
   */
  get nInserted() {
    return this.result.nInserted;
  ***REMOVED***

  /**
   * @return {number***REMOVED*** nUpserted Number of upserted documents
   */
  get nUpserted() {
    return this.result.nUpserted;
  ***REMOVED***

  /**
   * @return {number***REMOVED*** nMatched Number of matched documents
   */
  get nMatched() {
    return this.result.nMatched;
  ***REMOVED***

  /**
   * @return {number***REMOVED*** nModified Number of documents updated physically on disk
   */
  get nModified() {
    return this.result.nModified;
  ***REMOVED***

  /**
   * @return {number***REMOVED*** nRemoved Number of removed documents
   */
  get nRemoved() {
    return this.result.nRemoved;
  ***REMOVED***

  /**
   * Return an array of inserted ids
   *
   * @return {object[]***REMOVED***
   */
  getInsertedIds() {
    return this.result.insertedIds;
  ***REMOVED***

  /**
   * Return an array of upserted ids
   *
   * @return {object[]***REMOVED***
   */
  getUpsertedIds() {
    return this.result.upserted;
  ***REMOVED***

  /**
   * Return the upserted id at position x
   *
   * @param {number***REMOVED*** index the number of the upserted id to return, returns undefined if no result for passed in index
   * @return {object***REMOVED***
   */
  getUpsertedIdAt(index) {
    return this.result.upserted[index];
  ***REMOVED***

  /**
   * Return raw internal result
   *
   * @return {object***REMOVED***
   */
  getRawResponse() {
    return this.result;
  ***REMOVED***

  /**
   * Returns true if the bulk operation contains a write error
   *
   * @return {boolean***REMOVED***
   */
  hasWriteErrors() {
    return this.result.writeErrors.length > 0;
  ***REMOVED***

  /**
   * Returns the number of write errors off the bulk operation
   *
   * @return {number***REMOVED***
   */
  getWriteErrorCount() {
    return this.result.writeErrors.length;
  ***REMOVED***

  /**
   * Returns a specific write error object
   *
   * @param {number***REMOVED*** index of the write error to return, returns null if there is no result for passed in index
   * @return {WriteError***REMOVED***
   */
  getWriteErrorAt(index) {
    if (index < this.result.writeErrors.length) {
      return this.result.writeErrors[index];
    ***REMOVED***
    return null;
  ***REMOVED***

  /**
   * Retrieve all write errors
   *
   * @return {object[]***REMOVED***
   */
  getWriteErrors() {
    return this.result.writeErrors;
  ***REMOVED***

  /**
   * Retrieve lastOp if available
   *
   * @return {object***REMOVED***
   */
  getLastOp() {
    return this.result.lastOp;
  ***REMOVED***

  /**
   * Retrieve the write concern error if any
   *
   * @return {WriteConcernError***REMOVED***
   */
  getWriteConcernError() {
    if (this.result.writeConcernErrors.length === 0) {
      return null;
    ***REMOVED*** else if (this.result.writeConcernErrors.length === 1) {
      // Return the error
      return this.result.writeConcernErrors[0];
    ***REMOVED***
      // Combine the errors
      let errmsg = '';
      for (let i = 0; i < this.result.writeConcernErrors.length; i++) {
        const err = this.result.writeConcernErrors[i];
        errmsg = errmsg + err.errmsg;

        // TODO: Something better
        if (i === 0) errmsg = errmsg + ' and ';
      ***REMOVED***

      return new WriteConcernError({ errmsg: errmsg, code: WRITE_CONCERN_ERROR ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  /**
   * @return {BulkWriteResult***REMOVED*** a BulkWriteResult instance
   */
  toJSON() {
    return this.result;
  ***REMOVED***

  /**
   * @return {string***REMOVED***
   */
  toString() {
    return `BulkWriteResult(${this.toJSON(this.result)***REMOVED***)`;
  ***REMOVED***

  /**
   * @return {boolean***REMOVED***
   */
  isOk() {
    return this.result.ok === 1;
  ***REMOVED***
***REMOVED***

/**
 * Create a new WriteConcernError instance (INTERNAL TYPE, do not instantiate directly)
 *
 * @class
 * @return {WriteConcernError***REMOVED*** a WriteConcernError instance
 */
class WriteConcernError {
  constructor(err) {
    this.err = err;
  ***REMOVED***

  /**
   * @return {number***REMOVED*** code Write concern error code.
   */
  get code() {
    return this.err.code;
  ***REMOVED***

  /**
   * @return {string***REMOVED*** errmsg Write concern error message.
   */
  get errmsg() {
    return this.err.errmsg;
  ***REMOVED***

  /**
   * @return {object***REMOVED***
   */
  toJSON() {
    return { code: this.err.code, errmsg: this.err.errmsg ***REMOVED***;
  ***REMOVED***

  /**
   * @return {string***REMOVED***
   */
  toString() {
    return `WriteConcernError(${this.err.errmsg***REMOVED***)`;
  ***REMOVED***
***REMOVED***

/**
 * Create a new WriteError instance (INTERNAL TYPE, do not instantiate directly)
 *
 * @class
 * @return {WriteConcernError***REMOVED*** a WriteConcernError instance
 */
class WriteError {
  constructor(err) {
    this.err = err;
  ***REMOVED***

  /**
   * @return {number***REMOVED*** code Write concern error code.
   */
  get code() {
    return this.err.code;
  ***REMOVED***

  /**
   * @return {number***REMOVED*** index Write concern error original bulk operation index.
   */
  get index() {
    return this.err.index;
  ***REMOVED***

  /**
   * @return {string***REMOVED*** errmsg Write concern error message.
   */
  get errmsg() {
    return this.err.errmsg;
  ***REMOVED***

  /**
   * Define access methods
   * @return {object***REMOVED***
   */
  getOperation() {
    return this.err.op;
  ***REMOVED***

  /**
   * @return {object***REMOVED***
   */
  toJSON() {
    return { code: this.err.code, index: this.err.index, errmsg: this.err.errmsg, op: this.err.op ***REMOVED***;
  ***REMOVED***

  /**
   * @return {string***REMOVED***
   */
  toString() {
    return `WriteError(${JSON.stringify(this.toJSON())***REMOVED***)`;
  ***REMOVED***
***REMOVED***

/**
 * Merges results into shared data structure
 * @ignore
 */
function mergeBatchResults(ordered, batch, bulkResult, err, result) {
  // If we have an error set the result to be the err object
  if (err) {
    result = err;
  ***REMOVED*** else if (result && result.result) {
    result = result.result;
  ***REMOVED*** else if (result == null) {
    return;
  ***REMOVED***

  // Do we have a top level error stop processing and return
  if (result.ok === 0 && bulkResult.ok === 1) {
    bulkResult.ok = 0;

    const writeError = {
      index: 0,
      code: result.code || 0,
      errmsg: result.message,
      op: batch.operations[0]
    ***REMOVED***;

    bulkResult.writeErrors.push(new WriteError(writeError));
    return;
  ***REMOVED*** else if (result.ok === 0 && bulkResult.ok === 0) {
    return;
  ***REMOVED***

  // Deal with opTime if available
  if (result.opTime || result.lastOp) {
    const opTime = result.lastOp || result.opTime;
    let lastOpTS = null;
    let lastOpT = null;

    // We have a time stamp
    if (opTime && opTime._bsontype === 'Timestamp') {
      if (bulkResult.lastOp == null) {
        bulkResult.lastOp = opTime;
      ***REMOVED*** else if (opTime.greaterThan(bulkResult.lastOp)) {
        bulkResult.lastOp = opTime;
      ***REMOVED***
    ***REMOVED***
      // Existing TS
      if (bulkResult.lastOp) {
        lastOpTS =
          typeof bulkResult.lastOp.ts === 'number'
            ? Long.fromNumber(bulkResult.lastOp.ts)
            : bulkResult.lastOp.ts;
        lastOpT =
          typeof bulkResult.lastOp.t === 'number'
            ? Long.fromNumber(bulkResult.lastOp.t)
            : bulkResult.lastOp.t;
      ***REMOVED***

      // Current OpTime TS
      const opTimeTS = typeof opTime.ts === 'number' ? Long.fromNumber(opTime.ts) : opTime.ts;
      const opTimeT = typeof opTime.t === 'number' ? Long.fromNumber(opTime.t) : opTime.t;

      // Compare the opTime's
      if (bulkResult.lastOp == null) {
        bulkResult.lastOp = opTime;
      ***REMOVED*** else if (opTimeTS.greaterThan(lastOpTS)) {
        bulkResult.lastOp = opTime;
      ***REMOVED*** else if (opTimeTS.equals(lastOpTS)) {
        if (opTimeT.greaterThan(lastOpT)) {
          bulkResult.lastOp = opTime;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  // If we have an insert Batch type
  if (batch.batchType === INSERT && result.n) {
    bulkResult.nInserted = bulkResult.nInserted + result.n;
  ***REMOVED***

  // If we have an insert Batch type
  if (batch.batchType === REMOVE && result.n) {
    bulkResult.nRemoved = bulkResult.nRemoved + result.n;
  ***REMOVED***

  let nUpserted = 0;

  // We have an array of upserted values, we need to rewrite the indexes
  if (Array.isArray(result.upserted)) {
    nUpserted = result.upserted.length;

    for (let i = 0; i < result.upserted.length; i++) {
      bulkResult.upserted.push({
        index: result.upserted[i].index + batch.originalZeroIndex,
        _id: result.upserted[i]._id
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED*** else if (result.upserted) {
    nUpserted = 1;

    bulkResult.upserted.push({
      index: batch.originalZeroIndex,
      _id: result.upserted
    ***REMOVED***);
  ***REMOVED***

  // If we have an update Batch type
  if (batch.batchType === UPDATE && result.n) {
    const nModified = result.nModified;
    bulkResult.nUpserted = bulkResult.nUpserted + nUpserted;
    bulkResult.nMatched = bulkResult.nMatched + (result.n - nUpserted);

    if (typeof nModified === 'number') {
      bulkResult.nModified = bulkResult.nModified + nModified;
    ***REMOVED***
      bulkResult.nModified = null;
    ***REMOVED***
  ***REMOVED***

  if (Array.isArray(result.writeErrors)) {
    for (let i = 0; i < result.writeErrors.length; i++) {
      const writeError = {
        index: batch.originalZeroIndex + result.writeErrors[i].index,
        code: result.writeErrors[i].code,
        errmsg: result.writeErrors[i].errmsg,
        op: batch.operations[result.writeErrors[i].index]
      ***REMOVED***;

      bulkResult.writeErrors.push(new WriteError(writeError));
    ***REMOVED***
  ***REMOVED***

  if (result.writeConcernError) {
    bulkResult.writeConcernErrors.push(new WriteConcernError(result.writeConcernError));
  ***REMOVED***
***REMOVED***

/**
 * handles write concern error
 *
 * @param {object***REMOVED*** batch
 * @param {object***REMOVED*** bulkResult
 * @param {boolean***REMOVED*** ordered
 * @param {WriteConcernError***REMOVED*** err
 * @param {function***REMOVED*** callback
 */
function handleMongoWriteConcernError(batch, bulkResult, ordered, err, callback) {
  mergeBatchResults(ordered, batch, bulkResult, null, err.result);

  const wrappedWriteConcernError = new WriteConcernError({
    errmsg: err.result.writeConcernError.errmsg,
    code: err.result.writeConcernError.result
  ***REMOVED***);
  return handleCallback(
    callback,
    new BulkWriteError(toError(wrappedWriteConcernError), new BulkWriteResult(bulkResult)),
    null
  );
***REMOVED***

/**
 * Creates a new BulkWriteError
 *
 * @class
 * @param {Error|string|object***REMOVED*** message The error message
 * @param {BulkWriteResult***REMOVED*** result The result of the bulk write operation
 * @return {BulkWriteError***REMOVED*** A BulkWriteError instance
 * @extends {MongoError***REMOVED***
 */
class BulkWriteError extends MongoError {
  constructor(error, result) {
    const message = error.err || error.errmsg || error.errMessage || error;
    super(message);

    Object.assign(this, error);

    this.name = 'BulkWriteError';
    this.result = result;
  ***REMOVED***
***REMOVED***

/**
 * Handles the find operators for the bulk operations
 * @class
 */
class FindOperators {
  /**
   * @param {OrderedBulkOperation|UnorderedBulkOperation***REMOVED*** bulkOperation
   */
  constructor(bulkOperation) {
    this.s = bulkOperation.s;
  ***REMOVED***

  /**
   * Add a single update document to the bulk operation
   *
   * @method
   * @param {object***REMOVED*** updateDocument update operations
   * @throws {MongoError***REMOVED***
   * @return {OrderedBulkOperation|UnordedBulkOperation***REMOVED***
   */
  update(updateDocument) {
    // Perform upsert
    const upsert = typeof this.s.currentOp.upsert === 'boolean' ? this.s.currentOp.upsert : false;

    // Establish the update command
    const document = {
      q: this.s.currentOp.selector,
      u: updateDocument,
      multi: true,
      upsert: upsert
    ***REMOVED***;

    // Clear out current Op
    this.s.currentOp = null;
    return this.s.options.addToOperationsList(this, UPDATE, document);
  ***REMOVED***

  /**
   * Add a single update one document to the bulk operation
   *
   * @method
   * @param {object***REMOVED*** updateDocument update operations
   * @throws {MongoError***REMOVED***
   * @return {OrderedBulkOperation|UnordedBulkOperation***REMOVED***
   */
  updateOne(updateDocument) {
    // Perform upsert
    const upsert = typeof this.s.currentOp.upsert === 'boolean' ? this.s.currentOp.upsert : false;

    // Establish the update command
    const document = {
      q: this.s.currentOp.selector,
      u: updateDocument,
      multi: false,
      upsert: upsert
    ***REMOVED***;

    // Clear out current Op
    this.s.currentOp = null;
    return this.s.options.addToOperationsList(this, UPDATE, document);
  ***REMOVED***

  /**
   * Add a replace one operation to the bulk operation
   *
   * @method
   * @param {object***REMOVED*** updateDocument the new document to replace the existing one with
   * @throws {MongoError***REMOVED***
   * @return {OrderedBulkOperation|UnorderedBulkOperation***REMOVED***
   */
  replaceOne(updateDocument) {
    this.updateOne(updateDocument);
  ***REMOVED***

  /**
   * Upsert modifier for update bulk operation
   *
   * @method
   * @throws {MongoError***REMOVED***
   * @return {FindOperators***REMOVED***
   */
  upsert() {
    this.s.currentOp.upsert = true;
    return this;
  ***REMOVED***

  /**
   * Add a delete one operation to the bulk operation
   *
   * @method
   * @throws {MongoError***REMOVED***
   * @return {OrderedBulkOperation|UnordedBulkOperation***REMOVED***
   */
  deleteOne() {
    // Establish the update command
    const document = {
      q: this.s.currentOp.selector,
      limit: 1
    ***REMOVED***;

    // Clear out current Op
    this.s.currentOp = null;
    return this.s.options.addToOperationsList(this, REMOVE, document);
  ***REMOVED***

  /**
   * Add a delete operation to the bulk operation
   *
   * @method
   * @throws {MongoError***REMOVED***
   * @return {OrderedBulkOperation|UnordedBulkOperation***REMOVED***
   */
  delete() {
    // Establish the update command
    const document = {
      q: this.s.currentOp.selector,
      limit: 0
    ***REMOVED***;

    // Clear out current Op
    this.s.currentOp = null;
    return this.s.options.addToOperationsList(this, REMOVE, document);
  ***REMOVED***

  /**
   * backwards compatability for deleteOne
   */
  removeOne() {
    return this.deleteOne();
  ***REMOVED***

  /**
   * backwards compatability for delete
   */
  remove() {
    return this.delete();
  ***REMOVED***
***REMOVED***

/**
 * Parent class to OrderedBulkOperation and UnorderedBulkOperation
 * @class
 */
class BulkOperationBase {
  /**
   * Create a new OrderedBulkOperation or UnorderedBulkOperation instance (INTERNAL TYPE, do not instantiate directly)
   * @class
   * @property {number***REMOVED*** length Get the number of operations in the bulk.
   * @return {OrderedBulkOperation|UnordedBulkOperation***REMOVED***
   */
  constructor(topology, collection, options, isOrdered) {
    // determine whether bulkOperation is ordered or unordered
    this.isOrdered = isOrdered;

    options = options == null ? {***REMOVED*** : options;
    // TODO Bring from driver information in isMaster
    // Get the namespace for the write operations
    const namespace = collection.collectionName;
    // Used to mark operation as executed
    const executed = false;

    // Current item
    const currentOp = null;

    // Handle to the bson serializer, used to calculate running sizes
    const bson = topology.bson;

    // Set max byte size
    const isMaster = topology.lastIsMaster();
    const maxBatchSizeBytes =
      isMaster && isMaster.maxBsonObjectSize ? isMaster.maxBsonObjectSize : 1024 * 1024 * 16;
    const maxWriteBatchSize =
      isMaster && isMaster.maxWriteBatchSize ? isMaster.maxWriteBatchSize : 1000;

    // Final options for retryable writes and write concern
    let finalOptions = Object.assign({***REMOVED***, options);
    finalOptions = applyRetryableWrites(finalOptions, collection.s.db);
    finalOptions = applyWriteConcern(finalOptions, { collection: collection ***REMOVED***, options);
    const writeConcern = finalOptions.writeConcern;

    // Get the promiseLibrary
    const promiseLibrary = options.promiseLibrary || Promise;

    // Final results
    const bulkResult = {
      ok: 1,
      writeErrors: [],
      writeConcernErrors: [],
      insertedIds: [],
      nInserted: 0,
      nUpserted: 0,
      nMatched: 0,
      nModified: 0,
      nRemoved: 0,
      upserted: []
    ***REMOVED***;

    // Internal state
    this.s = {
      // Final result
      bulkResult: bulkResult,
      // Current batch state
      currentBatch: null,
      currentIndex: 0,
      // ordered specific
      currentBatchSize: 0,
      currentBatchSizeBytes: 0,
      // unordered specific
      currentInsertBatch: null,
      currentUpdateBatch: null,
      currentRemoveBatch: null,
      batches: [],
      // Write concern
      writeConcern: writeConcern,
      // Max batch size options
      maxBatchSizeBytes: maxBatchSizeBytes,
      maxWriteBatchSize: maxWriteBatchSize,
      // Namespace
      namespace: namespace,
      // BSON
      bson: bson,
      // Topology
      topology: topology,
      // Options
      options: finalOptions,
      // Current operation
      currentOp: currentOp,
      // Executed
      executed: executed,
      // Collection
      collection: collection,
      // Promise Library
      promiseLibrary: promiseLibrary,
      // Fundamental error
      err: null,
      // check keys
      checkKeys: typeof options.checkKeys === 'boolean' ? options.checkKeys : true
    ***REMOVED***;

    // bypass Validation
    if (options.bypassDocumentValidation === true) {
      this.s.bypassDocumentValidation = true;
    ***REMOVED***
  ***REMOVED***

  /**
   * Add a single insert document to the bulk operation
   *
   * @param {object***REMOVED*** document the document to insert
   * @throws {MongoError***REMOVED***
   * @return {OrderedBulkOperation|UnorderedBulkOperation***REMOVED***
   */
  insert(document) {
    if (this.s.collection.s.db.options.forceServerObjectId !== true && document._id == null)
      document._id = new ObjectID();
    return this.s.options.addToOperationsList(this, INSERT, document);
  ***REMOVED***

  /**
   * Initiate a find operation for an update/updateOne/remove/removeOne/replaceOne
   *
   * @method
   * @param {object***REMOVED*** selector The selector for the bulk operation.
   * @throws {MongoError***REMOVED***
   */
  find(selector) {
    if (!selector) {
      throw toError('Bulk find operation must specify a selector');
    ***REMOVED***

    // Save a current selector
    this.s.currentOp = {
      selector: selector
    ***REMOVED***;

    return new FindOperators(this);
  ***REMOVED***

  /**
   * Raw performs the bulk operation
   *
   * @method
   * @param {object***REMOVED*** op operation
   * @return {OrderedBulkOperation|UnorderedBulkOperation***REMOVED***
   */
  raw(op) {
    const key = Object.keys(op)[0];

    // Set up the force server object id
    const forceServerObjectId =
      typeof this.s.options.forceServerObjectId === 'boolean'
        ? this.s.options.forceServerObjectId
        : this.s.collection.s.db.options.forceServerObjectId;

    // Update operations
    if (
      (op.updateOne && op.updateOne.q) ||
      (op.updateMany && op.updateMany.q) ||
      (op.replaceOne && op.replaceOne.q)
    ) {
      op[key].multi = op.updateOne || op.replaceOne ? false : true;
      return this.s.options.addToOperationsList(this, UPDATE, op[key]);
    ***REMOVED***

    // Crud spec update format
    if (op.updateOne || op.updateMany || op.replaceOne) {
      const multi = op.updateOne || op.replaceOne ? false : true;
      const operation = {
        q: op[key].filter,
        u: op[key].update || op[key].replacement,
        multi: multi
      ***REMOVED***;
      if (this.isOrdered) {
        operation.upsert = op[key].upsert ? true : false;
        if (op.collation) operation.collation = op.collation;
      ***REMOVED***
        if (op[key].upsert) operation.upsert = true;
      ***REMOVED***
      if (op[key].arrayFilters) operation.arrayFilters = op[key].arrayFilters;
      return this.s.options.addToOperationsList(this, UPDATE, operation);
    ***REMOVED***

    // Remove operations
    if (
      op.removeOne ||
      op.removeMany ||
      (op.deleteOne && op.deleteOne.q) ||
      (op.deleteMany && op.deleteMany.q)
    ) {
      op[key].limit = op.removeOne ? 1 : 0;
      return this.s.options.addToOperationsList(this, REMOVE, op[key]);
    ***REMOVED***

    // Crud spec delete operations, less efficient
    if (op.deleteOne || op.deleteMany) {
      const limit = op.deleteOne ? 1 : 0;
      const operation = { q: op[key].filter, limit: limit ***REMOVED***;
      if (this.isOrdered) {
        if (op.collation) operation.collation = op.collation;
      ***REMOVED***
      return this.s.options.addToOperationsList(this, REMOVE, operation);
    ***REMOVED***

    // Insert operations
    if (op.insertOne && op.insertOne.document == null) {
      if (forceServerObjectId !== true && op.insertOne._id == null)
        op.insertOne._id = new ObjectID();
      return this.s.options.addToOperationsList(this, INSERT, op.insertOne);
    ***REMOVED*** else if (op.insertOne && op.insertOne.document) {
      if (forceServerObjectId !== true && op.insertOne.document._id == null)
        op.insertOne.document._id = new ObjectID();
      return this.s.options.addToOperationsList(this, INSERT, op.insertOne.document);
    ***REMOVED***

    if (op.insertMany) {
      for (let i = 0; i < op.insertMany.length; i++) {
        if (forceServerObjectId !== true && op.insertMany[i]._id == null)
          op.insertMany[i]._id = new ObjectID();
        this.s.options.addToOperationsList(this, INSERT, op.insertMany[i]);
      ***REMOVED***

      return;
    ***REMOVED***

    // No valid type of operation
    throw toError(
      'bulkWrite only supports insertOne, insertMany, updateOne, updateMany, removeOne, removeMany, deleteOne, deleteMany'
    );
  ***REMOVED***

  /**
   * Execute next write command in a chain
   *
   * @method
   * @param {class***REMOVED*** bulk either OrderedBulkOperation or UnorderdBulkOperation
   * @param {object***REMOVED*** writeConcern
   * @param {object***REMOVED*** options
   * @param {function***REMOVED*** callback
   */
  bulkExecute(_writeConcern, options, callback) {
    if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
    options = options || {***REMOVED***;

    if (this.s.executed) {
      const executedError = toError('batch cannot be re-executed');
      return typeof callback === 'function'
        ? callback(executedError, null)
        : this.s.promiseLibrary.reject(executedError);
    ***REMOVED***

    if (typeof _writeConcern === 'function') {
      callback = _writeConcern;
    ***REMOVED*** else if (_writeConcern && typeof _writeConcern === 'object') {
      this.s.writeConcern = _writeConcern;
    ***REMOVED***

    // If we have current batch
    if (this.isOrdered) {
      if (this.s.currentBatch) this.s.batches.push(this.s.currentBatch);
    ***REMOVED***
      if (this.s.currentInsertBatch) this.s.batches.push(this.s.currentInsertBatch);
      if (this.s.currentUpdateBatch) this.s.batches.push(this.s.currentUpdateBatch);
      if (this.s.currentRemoveBatch) this.s.batches.push(this.s.currentRemoveBatch);
    ***REMOVED***
    // If we have no operations in the bulk raise an error
    if (this.s.batches.length === 0) {
      const emptyBatchError = toError('Invalid Operation, no operations specified');
      return typeof callback === 'function'
        ? callback(emptyBatchError, null)
        : this.s.promiseLibrary.reject(emptyBatchError);
    ***REMOVED***
    return { options, callback ***REMOVED***;
  ***REMOVED***

  /**
   * Handles final options before executing command
   *
   * @param {object***REMOVED*** config
   * @param {object***REMOVED*** config.options
   * @param {number***REMOVED*** config.batch
   * @param {function***REMOVED*** config.resultHandler
   * @param {function***REMOVED*** callback
   */
  finalOptionsHandler(config, callback) {
    const finalOptions = Object.assign({ ordered: this.isOrdered ***REMOVED***, config.options);
    if (this.s.writeConcern != null) {
      finalOptions.writeConcern = this.s.writeConcern;
    ***REMOVED***

    if (finalOptions.bypassDocumentValidation !== true) {
      delete finalOptions.bypassDocumentValidation;
    ***REMOVED***

    // Set an operationIf if provided
    if (this.operationId) {
      config.resultHandler.operationId = this.operationId;
    ***REMOVED***

    // Serialize functions
    if (this.s.options.serializeFunctions) {
      finalOptions.serializeFunctions = true;
    ***REMOVED***

    // Ignore undefined
    if (this.s.options.ignoreUndefined) {
      finalOptions.ignoreUndefined = true;
    ***REMOVED***

    // Is the bypassDocumentValidation options specific
    if (this.s.bypassDocumentValidation === true) {
      finalOptions.bypassDocumentValidation = true;
    ***REMOVED***

    // Is the checkKeys option disabled
    if (this.s.checkKeys === false) {
      finalOptions.checkKeys = false;
    ***REMOVED***

    if (finalOptions.retryWrites) {
      if (config.batch.batchType === UPDATE) {
        finalOptions.retryWrites =
          finalOptions.retryWrites && !config.batch.operations.some(op => op.multi);
      ***REMOVED***

      if (config.batch.batchType === REMOVE) {
        finalOptions.retryWrites =
          finalOptions.retryWrites && !config.batch.operations.some(op => op.limit === 0);
      ***REMOVED***
    ***REMOVED***

    try {
      if (config.batch.batchType === INSERT) {
        this.s.topology.insert(
          this.s.collection.namespace,
          config.batch.operations,
          finalOptions,
          config.resultHandler
        );
      ***REMOVED*** else if (config.batch.batchType === UPDATE) {
        this.s.topology.update(
          this.s.collection.namespace,
          config.batch.operations,
          finalOptions,
          config.resultHandler
        );
      ***REMOVED*** else if (config.batch.batchType === REMOVE) {
        this.s.topology.remove(
          this.s.collection.namespace,
          config.batch.operations,
          finalOptions,
          config.resultHandler
        );
      ***REMOVED***
    ***REMOVED*** catch (err) {
      // Force top level error
      err.ok = 0;
      // Merge top level error and return
      handleCallback(
        callback,
        null,
        mergeBatchResults(false, config.batch, this.s.bulkResult, err, null)
      );
    ***REMOVED***
  ***REMOVED***

  /**
   * Handles the write error before executing commands
   *
   * @param {function***REMOVED*** callback
   * @param {BulkWriteResult***REMOVED*** writeResult
   * @param {class***REMOVED*** self either OrderedBulkOperation or UnorderdBulkOperation
   */
  handleWriteError(callback, writeResult) {
    if (this.s.bulkResult.writeErrors.length > 0) {
      if (this.s.bulkResult.writeErrors.length === 1) {
        handleCallback(
          callback,
          new BulkWriteError(toError(this.s.bulkResult.writeErrors[0]), writeResult),
          null
        );
        return true;
      ***REMOVED***

      handleCallback(
        callback,
        new BulkWriteError(
          toError({
            message: 'write operation failed',
            code: this.s.bulkResult.writeErrors[0].code,
            writeErrors: this.s.bulkResult.writeErrors
          ***REMOVED***),
          writeResult
        ),
        null
      );
      return true;
    ***REMOVED*** else if (writeResult.getWriteConcernError()) {
      handleCallback(
        callback,
        new BulkWriteError(toError(writeResult.getWriteConcernError()), writeResult),
        null
      );
      return true;
    ***REMOVED***
  ***REMOVED***
***REMOVED***

Object.defineProperty(BulkOperationBase.prototype, 'length', {
  enumerable: true,
  get: function() {
    return this.s.currentIndex;
  ***REMOVED***
***REMOVED***);

// Exports symbols
module.exports = {
  Batch,
  BulkOperationBase,
  BulkWriteError,
  BulkWriteResult,
  bson,
  FindOperators,
  handleMongoWriteConcernError,
  LegacyOp,
  mergeBatchResults,
  INVALID_BSON_ERROR: INVALID_BSON_ERROR,
  MULTIPLE_ERROR: MULTIPLE_ERROR,
  UNKNOWN_ERROR: UNKNOWN_ERROR,
  WRITE_CONCERN_ERROR: WRITE_CONCERN_ERROR,
  INSERT: INSERT,
  UPDATE: UPDATE,
  REMOVE: REMOVE,
  WriteError,
  WriteConcernError
***REMOVED***;
