// mongoose/lib/error/validation.js
/*!
 * Module requirements
 */

'use strict';

const MongooseError = require('./');
const util = require('util');

/**
 * Document Validation Error
 *
 * @api private
 * @param {Document***REMOVED*** instance
 * @inherits MongooseError
 */

function ValidationError(instance) {
  this.errors = {***REMOVED***;
  this._message = '';
  if (instance && instance.constructor.name === 'model') {
    this._message = instance.constructor.modelName + ' validation failed';
    MongooseError.call(this, this._message);
  ***REMOVED***
    this._message = 'Validation failed';
    MongooseError.call(this, this._message);
  ***REMOVED***
  this.name = 'ValidationError';
  if (Error.captureStackTrace) {
    Error.captureStackTrace(this);
  ***REMOVED***
    this.stack = new Error().stack;
  ***REMOVED***
  if (instance) {
    instance.errors = this.errors;
  ***REMOVED***
***REMOVED***

/*!
 * Inherits from MongooseError.
 */

ValidationError.prototype = Object.create(MongooseError.prototype);
ValidationError.prototype.constructor = MongooseError;

/**
 * Console.log helper
 */

ValidationError.prototype.toString = function() {
  return this.name + ': ' + _generateMessage(this);
***REMOVED***;

/*!
 * inspect helper
 */

ValidationError.prototype.inspect = function() {
  return Object.assign(new Error(this.message), this);
***REMOVED***;

if (util.inspect.custom) {
  /*!
  * Avoid Node deprecation warning DEP0079
  */

  ValidationError.prototype[util.inspect.custom] = ValidationError.prototype.inspect;
***REMOVED***

/*!
 * Helper for JSON.stringify
 */

ValidationError.prototype.toJSON = function() {
  return Object.assign({***REMOVED***, this, { message: this.message ***REMOVED***);
***REMOVED***;

/*!
 * add message
 */

ValidationError.prototype.addError = function(path, error) {
  this.errors[path] = error;
  this.message = this._message + ': ' + _generateMessage(this);
***REMOVED***;

/*!
 * ignore
 */

function _generateMessage(err) {
  const keys = Object.keys(err.errors || {***REMOVED***);
  const len = keys.length;
  const msgs = [];
  let key;

  for (let i = 0; i < len; ++i) {
    key = keys[i];
    if (err === err.errors[key]) {
      continue;
    ***REMOVED***
    msgs.push(key + ': ' + err.errors[key].message);
  ***REMOVED***

  return msgs.join(', ');
***REMOVED***

/*!
 * Module exports
 */

module.exports = exports = ValidationError;
