// mongoose/lib/error/validator.js
/*!
 * Module dependencies.
 */

'use strict';

const MongooseError = require('./');

/**
 * Schema validator error
 *
 * @param {Object***REMOVED*** properties
 * @inherits MongooseError
 * @api private
 */

function ValidatorError(properties) {
  let msg = properties.message;
  if (!msg) {
    msg = MongooseError.messages.general.default;
  ***REMOVED***

  const message = this.formatMessage(msg, properties);
  MongooseError.call(this, message);

  properties = Object.assign({***REMOVED***, properties, { message: message ***REMOVED***);
  this.name = 'ValidatorError';
  if (Error.captureStackTrace) {
    Error.captureStackTrace(this);
  ***REMOVED***
    this.stack = new Error().stack;
  ***REMOVED***
  this.properties = properties;
  this.kind = properties.type;
  this.path = properties.path;
  this.value = properties.value;
  this.reason = properties.reason;
***REMOVED***

/*!
 * Inherits from MongooseError
 */

ValidatorError.prototype = Object.create(MongooseError.prototype);
ValidatorError.prototype.constructor = MongooseError;

/*!
 * The object used to define this validator. Not enumerable to hide
 * it from `require('util').inspect()` output re: gh-3925
 */

Object.defineProperty(ValidatorError.prototype, 'properties', {
  enumerable: false,
  writable: true,
  value: null
***REMOVED***);

/*!
 * Formats error messages
 */

ValidatorError.prototype.formatMessage = function(msg, properties) {
  if (typeof msg === 'function') {
    return msg(properties);
  ***REMOVED***
  const propertyNames = Object.keys(properties);
  for (let i = 0; i < propertyNames.length; ++i) {
    const propertyName = propertyNames[i];
    if (propertyName === 'message') {
      continue;
    ***REMOVED***
    msg = msg.replace('{' + propertyName.toUpperCase() + '***REMOVED***', properties[propertyName]);
  ***REMOVED***
  return msg;
***REMOVED***;

/*!
 * toString helper
 */

ValidatorError.prototype.toString = function() {
  return this.message;
***REMOVED***;

/*!
 * exports
 */

module.exports = ValidatorError;
