// proxy-addr/index.js
/*!
 * proxy-addr
 * Copyright(c) 2014-2016 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module exports.
 * @public
 */

module.exports = proxyaddr
module.exports.all = alladdrs
module.exports.compile = compile

/**
 * Module dependencies.
 * @private
 */

var forwarded = require('forwarded')
var ipaddr = require('ipaddr.js')

/**
 * Variables.
 * @private
 */

var DIGIT_REGEXP = /^[0-9]+$/
var isip = ipaddr.isValid
var parseip = ipaddr.parse

/**
 * Pre-defined IP ranges.
 * @private
 */

var IP_RANGES = {
  linklocal: ['169.254.0.0/16', 'fe80::/10'],
  loopback: ['127.0.0.1/8', '::1/128'],
  uniquelocal: ['10.0.0.0/8', '172.16.0.0/12', '192.168.0.0/16', 'fc00::/7']
***REMOVED***

/**
 * Get all addresses in the request, optionally stopping
 * at the first untrusted.
 *
 * @param {Object***REMOVED*** request
 * @param {Function|Array|String***REMOVED*** [trust]
 * @public
 */

function alladdrs (req, trust) {
  // get addresses
  var addrs = forwarded(req)

  if (!trust) {
    // Return all addresses
    return addrs
  ***REMOVED***

  if (typeof trust !== 'function') {
    trust = compile(trust)
  ***REMOVED***

  for (var i = 0; i < addrs.length - 1; i++) {
    if (trust(addrs[i], i)) continue

    addrs.length = i + 1
  ***REMOVED***

  return addrs
***REMOVED***

/**
 * Compile argument into trust function.
 *
 * @param {Array|String***REMOVED*** val
 * @private
 */

function compile (val) {
  if (!val) {
    throw new TypeError('argument is required')
  ***REMOVED***

  var trust

  if (typeof val === 'string') {
    trust = [val]
  ***REMOVED*** else if (Array.isArray(val)) {
    trust = val.slice()
  ***REMOVED***
    throw new TypeError('unsupported trust argument')
  ***REMOVED***

  for (var i = 0; i < trust.length; i++) {
    val = trust[i]

    if (!IP_RANGES.hasOwnProperty(val)) {
      continue
    ***REMOVED***

    // Splice in pre-defined range
    val = IP_RANGES[val]
    trust.splice.apply(trust, [i, 1].concat(val))
    i += val.length - 1
  ***REMOVED***

  return compileTrust(compileRangeSubnets(trust))
***REMOVED***

/**
 * Compile `arr` elements into range subnets.
 *
 * @param {Array***REMOVED*** arr
 * @private
 */

function compileRangeSubnets (arr) {
  var rangeSubnets = new Array(arr.length)

  for (var i = 0; i < arr.length; i++) {
    rangeSubnets[i] = parseipNotation(arr[i])
  ***REMOVED***

  return rangeSubnets
***REMOVED***

/**
 * Compile range subnet array into trust function.
 *
 * @param {Array***REMOVED*** rangeSubnets
 * @private
 */

function compileTrust (rangeSubnets) {
  // Return optimized function based on length
  var len = rangeSubnets.length
  return len === 0
    ? trustNone
    : len === 1
      ? trustSingle(rangeSubnets[0])
      : trustMulti(rangeSubnets)
***REMOVED***

/**
 * Parse IP notation string into range subnet.
 *
 * @param {String***REMOVED*** note
 * @private
 */

function parseipNotation (note) {
  var pos = note.lastIndexOf('/')
  var str = pos !== -1
    ? note.substring(0, pos)
    : note

  if (!isip(str)) {
    throw new TypeError('invalid IP address: ' + str)
  ***REMOVED***

  var ip = parseip(str)

  if (pos === -1 && ip.kind() === 'ipv6' && ip.isIPv4MappedAddress()) {
    // Store as IPv4
    ip = ip.toIPv4Address()
  ***REMOVED***

  var max = ip.kind() === 'ipv6'
    ? 128
    : 32

  var range = pos !== -1
    ? note.substring(pos + 1, note.length)
    : null

  if (range === null) {
    range = max
  ***REMOVED*** else if (DIGIT_REGEXP.test(range)) {
    range = parseInt(range, 10)
  ***REMOVED*** else if (ip.kind() === 'ipv4' && isip(range)) {
    range = parseNetmask(range)
  ***REMOVED***
    range = null
  ***REMOVED***

  if (range <= 0 || range > max) {
    throw new TypeError('invalid range on address: ' + note)
  ***REMOVED***

  return [ip, range]
***REMOVED***

/**
 * Parse netmask string into CIDR range.
 *
 * @param {String***REMOVED*** netmask
 * @private
 */

function parseNetmask (netmask) {
  var ip = parseip(netmask)
  var kind = ip.kind()

  return kind === 'ipv4'
    ? ip.prefixLengthFromSubnetMask()
    : null
***REMOVED***

/**
 * Determine address of proxied request.
 *
 * @param {Object***REMOVED*** request
 * @param {Function|Array|String***REMOVED*** trust
 * @public
 */

function proxyaddr (req, trust) {
  if (!req) {
    throw new TypeError('req argument is required')
  ***REMOVED***

  if (!trust) {
    throw new TypeError('trust argument is required')
  ***REMOVED***

  var addrs = alladdrs(req, trust)
  var addr = addrs[addrs.length - 1]

  return addr
***REMOVED***

/**
 * Static trust function to trust nothing.
 *
 * @private
 */

function trustNone () {
  return false
***REMOVED***

/**
 * Compile trust function for multiple subnets.
 *
 * @param {Array***REMOVED*** subnets
 * @private
 */

function trustMulti (subnets) {
  return function trust (addr) {
    if (!isip(addr)) return false

    var ip = parseip(addr)
    var ipconv
    var kind = ip.kind()

    for (var i = 0; i < subnets.length; i++) {
      var subnet = subnets[i]
      var subnetip = subnet[0]
      var subnetkind = subnetip.kind()
      var subnetrange = subnet[1]
      var trusted = ip

      if (kind !== subnetkind) {
        if (subnetkind === 'ipv4' && !ip.isIPv4MappedAddress()) {
          // Incompatible IP addresses
          continue
        ***REMOVED***

        if (!ipconv) {
          // Convert IP to match subnet IP kind
          ipconv = subnetkind === 'ipv4'
            ? ip.toIPv4Address()
            : ip.toIPv4MappedAddress()
        ***REMOVED***

        trusted = ipconv
      ***REMOVED***

      if (trusted.match(subnetip, subnetrange)) {
        return true
      ***REMOVED***
    ***REMOVED***

    return false
  ***REMOVED***
***REMOVED***

/**
 * Compile trust function for single subnet.
 *
 * @param {Object***REMOVED*** subnet
 * @private
 */

function trustSingle (subnet) {
  var subnetip = subnet[0]
  var subnetkind = subnetip.kind()
  var subnetisipv4 = subnetkind === 'ipv4'
  var subnetrange = subnet[1]

  return function trust (addr) {
    if (!isip(addr)) return false

    var ip = parseip(addr)
    var kind = ip.kind()

    if (kind !== subnetkind) {
      if (subnetisipv4 && !ip.isIPv4MappedAddress()) {
        // Incompatible IP addresses
        return false
      ***REMOVED***

      // Convert IP to match subnet IP kind
      ip = subnetisipv4
        ? ip.toIPv4Address()
        : ip.toIPv4MappedAddress()
    ***REMOVED***

    return ip.match(subnetip, subnetrange)
  ***REMOVED***
***REMOVED***
