// mongodb/lib/cursor.js
'use strict';

const Transform = require('stream').Transform;
const PassThrough = require('stream').PassThrough;
const inherits = require('util').inherits;
const deprecate = require('util').deprecate;
const handleCallback = require('./utils').handleCallback;
const ReadPreference = require('mongodb-core').ReadPreference;
const MongoError = require('mongodb-core').MongoError;
const Readable = require('stream').Readable;
const CoreCursor = require('mongodb-core').Cursor;
const Map = require('mongodb-core').BSON.Map;
const executeOperation = require('./utils').executeOperation;

const count = require('./operations/cursor_ops').count;
const each = require('./operations/cursor_ops').each;
const hasNext = require('./operations/cursor_ops').hasNext;
const next = require('./operations/cursor_ops').next;
const toArray = require('./operations/cursor_ops').toArray;

/**
 * @fileOverview The **Cursor** class is an internal class that embodies a cursor on MongoDB
 * allowing for iteration over the results returned from the underlying query. It supports
 * one by one document iteration, conversion to an array or can be iterated as a Node 4.X
 * or higher stream
 *
 * **CURSORS Cannot directly be instantiated**
 * @example
 * const MongoClient = require('mongodb').MongoClient;
 * const test = require('assert');
 * // Connection url
 * const url = 'mongodb://localhost:27017';
 * // Database Name
 * const dbName = 'test';
 * // Connect using MongoClient
 * MongoClient.connect(url, function(err, client) {
 *   // Create a collection we want to drop later
 *   const col = client.db(dbName).collection('createIndexExample1');
 *   // Insert a bunch of documents
 *   col.insert([{a:1, b:1***REMOVED***
 *     , {a:2, b:2***REMOVED***, {a:3, b:3***REMOVED***
 *     , {a:4, b:4***REMOVED***], {w:1***REMOVED***, function(err, result) {
 *     test.equal(null, err);
 *     // Show that duplicate records got dropped
 *     col.find({***REMOVED***).toArray(function(err, items) {
 *       test.equal(null, err);
 *       test.equal(4, items.length);
 *       client.close();
 *     ***REMOVED***);
 *   ***REMOVED***);
 * ***REMOVED***);
 */

/**
 * Namespace provided by the mongodb-core and node.js
 * @external CoreCursor
 * @external Readable
 */

// Flags allowed for cursor
const flags = ['tailable', 'oplogReplay', 'noCursorTimeout', 'awaitData', 'exhaust', 'partial'];
const fields = ['numberOfRetries', 'tailableRetryInterval'];

/**
 * Creates a new Cursor instance (INTERNAL TYPE, do not instantiate directly)
 * @class Cursor
 * @extends external:CoreCursor
 * @extends external:Readable
 * @property {string***REMOVED*** sortValue Cursor query sort setting.
 * @property {boolean***REMOVED*** timeout Is Cursor able to time out.
 * @property {ReadPreference***REMOVED*** readPreference Get cursor ReadPreference.
 * @fires Cursor#data
 * @fires Cursor#end
 * @fires Cursor#close
 * @fires Cursor#readable
 * @return {Cursor***REMOVED*** a Cursor instance.
 * @example
 * Cursor cursor options.
 *
 * collection.find({***REMOVED***).project({a:1***REMOVED***)                             // Create a projection of field a
 * collection.find({***REMOVED***).skip(1).limit(10)                          // Skip 1 and limit 10
 * collection.find({***REMOVED***).batchSize(5)                               // Set batchSize on cursor to 5
 * collection.find({***REMOVED***).filter({a:1***REMOVED***)                              // Set query on the cursor
 * collection.find({***REMOVED***).comment('add a comment')                   // Add a comment to the query, allowing to correlate queries
 * collection.find({***REMOVED***).addCursorFlag('tailable', true)            // Set cursor as tailable
 * collection.find({***REMOVED***).addCursorFlag('oplogReplay', true)         // Set cursor as oplogReplay
 * collection.find({***REMOVED***).addCursorFlag('noCursorTimeout', true)     // Set cursor as noCursorTimeout
 * collection.find({***REMOVED***).addCursorFlag('awaitData', true)           // Set cursor as awaitData
 * collection.find({***REMOVED***).addCursorFlag('partial', true)             // Set cursor as partial
 * collection.find({***REMOVED***).addQueryModifier('$orderby', {a:1***REMOVED***)        // Set $orderby {a:1***REMOVED***
 * collection.find({***REMOVED***).max(10)                                    // Set the cursor max
 * collection.find({***REMOVED***).maxTimeMS(1000)                            // Set the cursor maxTimeMS
 * collection.find({***REMOVED***).min(100)                                   // Set the cursor min
 * collection.find({***REMOVED***).returnKey(true)                            // Set the cursor returnKey
 * collection.find({***REMOVED***).setReadPreference(ReadPreference.PRIMARY)  // Set the cursor readPreference
 * collection.find({***REMOVED***).showRecordId(true)                         // Set the cursor showRecordId
 * collection.find({***REMOVED***).sort([['a', 1]])                           // Sets the sort order of the cursor query
 * collection.find({***REMOVED***).hint('a_1')                                // Set the cursor hint
 *
 * All options are chainable, so one can do the following.
 *
 * collection.find({***REMOVED***).maxTimeMS(1000).maxScan(100).skip(1).toArray(..)
 */
function Cursor(bson, ns, cmd, options, topology, topologyOptions) {
  CoreCursor.apply(this, Array.prototype.slice.call(arguments, 0));
  const state = Cursor.INIT;
  const streamOptions = {***REMOVED***;

  // Tailable cursor options
  const numberOfRetries = options.numberOfRetries || 5;
  const tailableRetryInterval = options.tailableRetryInterval || 500;
  const currentNumberOfRetries = numberOfRetries;

  // Get the promiseLibrary
  const promiseLibrary = options.promiseLibrary || Promise;

  // Set up
  Readable.call(this, { objectMode: true ***REMOVED***);

  // Internal cursor state
  this.s = {
    // Tailable cursor options
    numberOfRetries: numberOfRetries,
    tailableRetryInterval: tailableRetryInterval,
    currentNumberOfRetries: currentNumberOfRetries,
    // State
    state: state,
    // Stream options
    streamOptions: streamOptions,
    // BSON
    bson: bson,
    // Namespace
    ns: ns,
    // Command
    cmd: cmd,
    // Options
    options: options,
    // Topology
    topology: topology,
    // Topology options
    topologyOptions: topologyOptions,
    // Promise library
    promiseLibrary: promiseLibrary,
    // Current doc
    currentDoc: null,
    // explicitlyIgnoreSession
    explicitlyIgnoreSession: options.explicitlyIgnoreSession
  ***REMOVED***;

  // Optional ClientSession
  if (!options.explicitlyIgnoreSession && options.session) {
    this.s.session = options.session;
  ***REMOVED***

  // Translate correctly
  if (this.s.options.noCursorTimeout === true) {
    this.addCursorFlag('noCursorTimeout', true);
  ***REMOVED***

  // Set the sort value
  this.sortValue = this.s.cmd.sort;

  // Get the batchSize
  const batchSize =
    cmd.cursor && cmd.cursor.batchSize
      ? cmd.cursor && cmd.cursor.batchSize
      : options.cursor && options.cursor.batchSize
        ? options.cursor.batchSize
        : 1000;

  // Set the batchSize
  this.setCursorBatchSize(batchSize);
***REMOVED***

/**
 * Cursor stream data event, fired for each document in the cursor.
 *
 * @event Cursor#data
 * @type {object***REMOVED***
 */

/**
 * Cursor stream end event
 *
 * @event Cursor#end
 * @type {null***REMOVED***
 */

/**
 * Cursor stream close event
 *
 * @event Cursor#close
 * @type {null***REMOVED***
 */

/**
 * Cursor stream readable event
 *
 * @event Cursor#readable
 * @type {null***REMOVED***
 */

// Inherit from Readable
inherits(Cursor, Readable);

// Map core cursor _next method so we can apply mapping
Cursor.prototype._next = function() {
  if (this._initImplicitSession) {
    this._initImplicitSession();
  ***REMOVED***
  return CoreCursor.prototype.next.apply(this, arguments);
***REMOVED***;

for (let name in CoreCursor.prototype) {
  Cursor.prototype[name] = CoreCursor.prototype[name];
***REMOVED***

Cursor.prototype._initImplicitSession = function() {
  if (!this.s.explicitlyIgnoreSession && !this.s.session && this.s.topology.hasSessionSupport()) {
    this.s.session = this.s.topology.startSession({ owner: this ***REMOVED***);
    this.cursorState.session = this.s.session;
  ***REMOVED***
***REMOVED***;

Cursor.prototype._endSession = function() {
  const didCloseCursor = CoreCursor.prototype._endSession.apply(this, arguments);
  if (didCloseCursor) {
    this.s.session = undefined;
  ***REMOVED***
***REMOVED***;

/**
 * Check if there is any document still available in the cursor
 * @method
 * @param {Cursor~resultCallback***REMOVED*** [callback] The result callback.
 * @throws {MongoError***REMOVED***
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.hasNext = function(callback) {
  return executeOperation(this.s.topology, hasNext, [this, callback], {
    skipSessions: true
  ***REMOVED***);
***REMOVED***;

/**
 * Get the next available document from the cursor, returns null if no more documents are available.
 * @method
 * @param {Cursor~resultCallback***REMOVED*** [callback] The result callback.
 * @throws {MongoError***REMOVED***
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.next = function(callback) {
  return executeOperation(this.s.topology, next, [this, callback], {
    skipSessions: true
  ***REMOVED***);
***REMOVED***;

/**
 * Set the cursor query
 * @method
 * @param {object***REMOVED*** filter The filter object used for the cursor.
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.filter = function(filter) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.query = filter;
  return this;
***REMOVED***;

/**
 * Set the cursor maxScan
 * @method
 * @param {object***REMOVED*** maxScan Constrains the query to only scan the specified number of documents when fulfilling the query
 * @deprecated as of MongoDB 4.0
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.maxScan = deprecate(function(maxScan) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.maxScan = maxScan;
  return this;
***REMOVED***, 'Cursor.maxScan is deprecated, and will be removed in a later version');

/**
 * Set the cursor hint
 * @method
 * @param {object***REMOVED*** hint If specified, then the query system will only consider plans using the hinted index.
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.hint = function(hint) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.hint = hint;
  return this;
***REMOVED***;

/**
 * Set the cursor min
 * @method
 * @param {object***REMOVED*** min Specify a $min value to specify the inclusive lower bound for a specific index in order to constrain the results of find(). The $min specifies the lower bound for all keys of a specific index in order.
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.min = function(min) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead())
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  this.s.cmd.min = min;
  return this;
***REMOVED***;

/**
 * Set the cursor max
 * @method
 * @param {object***REMOVED*** max Specify a $max value to specify the exclusive upper bound for a specific index in order to constrain the results of find(). The $max specifies the upper bound for all keys of a specific index in order.
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.max = function(max) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.max = max;
  return this;
***REMOVED***;

/**
 * Set the cursor returnKey. If set to true, modifies the cursor to only return the index field or fields for the results of the query, rather than documents. If set to true and the query does not use an index to perform the read operation, the returned documents will not contain any fields.
 * @method
 * @param {bool***REMOVED*** returnKey the returnKey value.
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.returnKey = function(value) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.returnKey = value;
  return this;
***REMOVED***;

/**
 * Set the cursor showRecordId
 * @method
 * @param {object***REMOVED*** showRecordId The $showDiskLoc option has now been deprecated and replaced with the showRecordId field. $showDiskLoc will still be accepted for OP_QUERY stye find.
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.showRecordId = function(value) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.showDiskLoc = value;
  return this;
***REMOVED***;

/**
 * Set the cursor snapshot
 * @method
 * @param {object***REMOVED*** snapshot The $snapshot operator prevents the cursor from returning a document more than once because an intervening write operation results in a move of the document.
 * @deprecated as of MongoDB 4.0
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.snapshot = deprecate(function(value) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.snapshot = value;
  return this;
***REMOVED***, 'Cursor Snapshot is deprecated, and will be removed in a later version');

/**
 * Set a node.js specific cursor option
 * @method
 * @param {string***REMOVED*** field The cursor option to set ['numberOfRetries', 'tailableRetryInterval'].
 * @param {object***REMOVED*** value The field value.
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.setCursorOption = function(field, value) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  if (fields.indexOf(field) === -1) {
    throw MongoError.create({
      message: `option ${field***REMOVED*** is not a supported option ${fields***REMOVED***`,
      driver: true
    ***REMOVED***);
  ***REMOVED***

  this.s[field] = value;
  if (field === 'numberOfRetries') this.s.currentNumberOfRetries = value;
  return this;
***REMOVED***;

/**
 * Add a cursor flag to the cursor
 * @method
 * @param {string***REMOVED*** flag The flag to set, must be one of following ['tailable', 'oplogReplay', 'noCursorTimeout', 'awaitData', 'partial'].
 * @param {boolean***REMOVED*** value The flag boolean value.
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.addCursorFlag = function(flag, value) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  if (flags.indexOf(flag) === -1) {
    throw MongoError.create({
      message: `flag ${flag***REMOVED*** is not a supported flag ${flags***REMOVED***`,
      driver: true
    ***REMOVED***);
  ***REMOVED***

  if (typeof value !== 'boolean') {
    throw MongoError.create({ message: `flag ${flag***REMOVED*** must be a boolean value`, driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd[flag] = value;
  return this;
***REMOVED***;

/**
 * Add a query modifier to the cursor query
 * @method
 * @param {string***REMOVED*** name The query modifier (must start with $, such as $orderby etc)
 * @param {string|boolean|number***REMOVED*** value The modifier value.
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.addQueryModifier = function(name, value) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  if (name[0] !== '$') {
    throw MongoError.create({ message: `${name***REMOVED*** is not a valid query modifier`, driver: true ***REMOVED***);
  ***REMOVED***

  // Strip of the $
  const field = name.substr(1);
  // Set on the command
  this.s.cmd[field] = value;
  // Deal with the special case for sort
  if (field === 'orderby') this.s.cmd.sort = this.s.cmd[field];
  return this;
***REMOVED***;

/**
 * Add a comment to the cursor query allowing for tracking the comment in the log.
 * @method
 * @param {string***REMOVED*** value The comment attached to this query.
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.comment = function(value) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.comment = value;
  return this;
***REMOVED***;

/**
 * Set a maxAwaitTimeMS on a tailing cursor query to allow to customize the timeout value for the option awaitData (Only supported on MongoDB 3.2 or higher, ignored otherwise)
 * @method
 * @param {number***REMOVED*** value Number of milliseconds to wait before aborting the tailed query.
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.maxAwaitTimeMS = function(value) {
  if (typeof value !== 'number') {
    throw MongoError.create({ message: 'maxAwaitTimeMS must be a number', driver: true ***REMOVED***);
  ***REMOVED***

  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.maxAwaitTimeMS = value;
  return this;
***REMOVED***;

/**
 * Set a maxTimeMS on the cursor query, allowing for hard timeout limits on queries (Only supported on MongoDB 2.6 or higher)
 * @method
 * @param {number***REMOVED*** value Number of milliseconds to wait before aborting the query.
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.maxTimeMS = function(value) {
  if (typeof value !== 'number') {
    throw MongoError.create({ message: 'maxTimeMS must be a number', driver: true ***REMOVED***);
  ***REMOVED***

  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.maxTimeMS = value;
  return this;
***REMOVED***;

Cursor.prototype.maxTimeMs = Cursor.prototype.maxTimeMS;

/**
 * Sets a field projection for the query.
 * @method
 * @param {object***REMOVED*** value The field projection object.
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.project = function(value) {
  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.fields = value;
  return this;
***REMOVED***;

/**
 * Sets the sort order of the cursor query.
 * @method
 * @param {(string|array|object)***REMOVED*** keyOrList The key or keys set for the sort.
 * @param {number***REMOVED*** [direction] The direction of the sorting (1 or -1).
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.sort = function(keyOrList, direction) {
  if (this.s.options.tailable) {
    throw MongoError.create({ message: "Tailable cursor doesn't support sorting", driver: true ***REMOVED***);
  ***REMOVED***

  if (this.s.state === Cursor.CLOSED || this.s.state === Cursor.OPEN || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  let order = keyOrList;

  // We have an array of arrays, we need to preserve the order of the sort
  // so we will us a Map
  if (Array.isArray(order) && Array.isArray(order[0])) {
    order = new Map(
      order.map(x => {
        const value = [x[0], null];
        if (x[1] === 'asc') {
          value[1] = 1;
        ***REMOVED*** else if (x[1] === 'desc') {
          value[1] = -1;
        ***REMOVED*** else if (x[1] === 1 || x[1] === -1 || x[1].$meta) {
          value[1] = x[1];
        ***REMOVED***
          throw new MongoError(
            "Illegal sort clause, must be of the form [['field1', '(ascending|descending)'], ['field2', '(ascending|descending)']]"
          );
        ***REMOVED***

        return value;
      ***REMOVED***)
    );
  ***REMOVED***

  if (direction != null) {
    order = [[keyOrList, direction]];
  ***REMOVED***

  this.s.cmd.sort = order;
  this.sortValue = order;
  return this;
***REMOVED***;

/**
 * Set the batch size for the cursor.
 * @method
 * @param {number***REMOVED*** value The batchSize for the cursor.
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.batchSize = function(value) {
  if (this.s.options.tailable) {
    throw MongoError.create({ message: "Tailable cursor doesn't support batchSize", driver: true ***REMOVED***);
  ***REMOVED***

  if (this.s.state === Cursor.CLOSED || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  if (typeof value !== 'number') {
    throw MongoError.create({ message: 'batchSize requires an integer', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.batchSize = value;
  this.setCursorBatchSize(value);
  return this;
***REMOVED***;

/**
 * Set the collation options for the cursor.
 * @method
 * @param {object***REMOVED*** value The cursor collation options (MongoDB 3.4 or higher) settings for update operation (see 3.4 documentation for available fields).
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.collation = function(value) {
  this.s.cmd.collation = value;
  return this;
***REMOVED***;

/**
 * Set the limit for the cursor.
 * @method
 * @param {number***REMOVED*** value The limit for the cursor query.
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.limit = function(value) {
  if (this.s.options.tailable) {
    throw MongoError.create({ message: "Tailable cursor doesn't support limit", driver: true ***REMOVED***);
  ***REMOVED***

  if (this.s.state === Cursor.OPEN || this.s.state === Cursor.CLOSED || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  if (typeof value !== 'number') {
    throw MongoError.create({ message: 'limit requires an integer', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.limit = value;
  // this.cursorLimit = value;
  this.setCursorLimit(value);
  return this;
***REMOVED***;

/**
 * Set the skip for the cursor.
 * @method
 * @param {number***REMOVED*** value The skip for the cursor query.
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.skip = function(value) {
  if (this.s.options.tailable) {
    throw MongoError.create({ message: "Tailable cursor doesn't support skip", driver: true ***REMOVED***);
  ***REMOVED***

  if (this.s.state === Cursor.OPEN || this.s.state === Cursor.CLOSED || this.isDead()) {
    throw MongoError.create({ message: 'Cursor is closed', driver: true ***REMOVED***);
  ***REMOVED***

  if (typeof value !== 'number') {
    throw MongoError.create({ message: 'skip requires an integer', driver: true ***REMOVED***);
  ***REMOVED***

  this.s.cmd.skip = value;
  this.setCursorSkip(value);
  return this;
***REMOVED***;

/**
 * The callback format for results
 * @callback Cursor~resultCallback
 * @param {MongoError***REMOVED*** error An error instance representing the error during the execution.
 * @param {(object|null|boolean)***REMOVED*** result The result object if the command was executed successfully.
 */

/**
 * Clone the cursor
 * @function external:CoreCursor#clone
 * @return {Cursor***REMOVED***
 */

/**
 * Resets the cursor
 * @function external:CoreCursor#rewind
 * @return {null***REMOVED***
 */

/**
 * Iterates over all the documents for this cursor. As with **{cursor.toArray***REMOVED*****,
 * not all of the elements will be iterated if this cursor had been previouly accessed.
 * In that case, **{cursor.rewind***REMOVED***** can be used to reset the cursor. However, unlike
 * **{cursor.toArray***REMOVED*****, the cursor will only hold a maximum of batch size elements
 * at any given time if batch size is specified. Otherwise, the caller is responsible
 * for making sure that the entire result can fit the memory.
 * @method
 * @deprecated
 * @param {Cursor~resultCallback***REMOVED*** callback The result callback.
 * @throws {MongoError***REMOVED***
 * @return {null***REMOVED***
 */
Cursor.prototype.each = deprecate(function(callback) {
  // Rewind cursor state
  this.rewind();
  // Set current cursor to INIT
  this.s.state = Cursor.INIT;
  // Run the query
  each(this, callback);
***REMOVED***, 'Cursor.each is deprecated. Use Cursor.forEach instead.');

/**
 * The callback format for the forEach iterator method
 * @callback Cursor~iteratorCallback
 * @param {Object***REMOVED*** doc An emitted document for the iterator
 */

/**
 * The callback error format for the forEach iterator method
 * @callback Cursor~endCallback
 * @param {MongoError***REMOVED*** error An error instance representing the error during the execution.
 */

/**
 * Iterates over all the documents for this cursor using the iterator, callback pattern.
 * @method
 * @param {Cursor~iteratorCallback***REMOVED*** iterator The iteration callback.
 * @param {Cursor~endCallback***REMOVED*** callback The end callback.
 * @throws {MongoError***REMOVED***
 * @return {Promise***REMOVED*** if no callback supplied
 */
Cursor.prototype.forEach = function(iterator, callback) {
  // Rewind cursor state
  this.rewind();

  // Set current cursor to INIT
  this.s.state = Cursor.INIT;

  if (typeof callback === 'function') {
    each(this, (err, doc) => {
      if (err) {
        callback(err);
        return false;
      ***REMOVED***
      if (doc != null) {
        iterator(doc);
        return true;
      ***REMOVED***
      if (doc == null && callback) {
        const internalCallback = callback;
        callback = null;
        internalCallback(null);
        return false;
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***
    return new this.s.promiseLibrary((fulfill, reject) => {
      each(this, (err, doc) => {
        if (err) {
          reject(err);
          return false;
        ***REMOVED*** else if (doc == null) {
          fulfill(null);
          return false;
        ***REMOVED***
          iterator(doc);
          return true;
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;

/**
 * Set the ReadPreference for the cursor.
 * @method
 * @param {(string|ReadPreference)***REMOVED*** readPreference The new read preference for the cursor.
 * @throws {MongoError***REMOVED***
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.setReadPreference = function(readPreference) {
  if (this.s.state !== Cursor.INIT) {
    throw MongoError.create({
      message: 'cannot change cursor readPreference after cursor has been accessed',
      driver: true
    ***REMOVED***);
  ***REMOVED***

  if (readPreference instanceof ReadPreference) {
    this.s.options.readPreference = readPreference;
  ***REMOVED*** else if (typeof readPreference === 'string') {
    this.s.options.readPreference = new ReadPreference(readPreference);
  ***REMOVED***
    throw new TypeError('Invalid read preference: ' + readPreference);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * The callback format for results
 * @callback Cursor~toArrayResultCallback
 * @param {MongoError***REMOVED*** error An error instance representing the error during the execution.
 * @param {object[]***REMOVED*** documents All the documents the satisfy the cursor.
 */

/**
 * Returns an array of documents. The caller is responsible for making sure that there
 * is enough memory to store the results. Note that the array only contains partial
 * results when this cursor had been previouly accessed. In that case,
 * cursor.rewind() can be used to reset the cursor.
 * @method
 * @param {Cursor~toArrayResultCallback***REMOVED*** [callback] The result callback.
 * @throws {MongoError***REMOVED***
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.toArray = function(callback) {
  if (this.s.options.tailable) {
    throw MongoError.create({
      message: 'Tailable cursor cannot be converted to array',
      driver: true
    ***REMOVED***);
  ***REMOVED***

  return executeOperation(this.s.topology, toArray, [this, callback], {
    skipSessions: true
  ***REMOVED***);
***REMOVED***;

/**
 * The callback format for results
 * @callback Cursor~countResultCallback
 * @param {MongoError***REMOVED*** error An error instance representing the error during the execution.
 * @param {number***REMOVED*** count The count of documents.
 */

/**
 * Get the count of documents for this cursor
 * @method
 * @param {boolean***REMOVED*** [applySkipLimit=true] Should the count command apply limit and skip settings on the cursor or in the passed in options.
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {number***REMOVED*** [options.skip] The number of documents to skip.
 * @param {number***REMOVED*** [options.limit] The maximum amounts to count before aborting.
 * @param {number***REMOVED*** [options.maxTimeMS] Number of miliseconds to wait before aborting the query.
 * @param {string***REMOVED*** [options.hint] An index name hint for the query.
 * @param {(ReadPreference|string)***REMOVED*** [options.readPreference] The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST).
 * @param {Cursor~countResultCallback***REMOVED*** [callback] The result callback.
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.count = function(applySkipLimit, opts, callback) {
  if (this.s.cmd.query == null)
    throw MongoError.create({ message: 'count can only be used with find command', driver: true ***REMOVED***);
  if (typeof opts === 'function') (callback = opts), (opts = {***REMOVED***);
  opts = opts || {***REMOVED***;

  if (typeof applySkipLimit === 'function') {
    callback = applySkipLimit;
    applySkipLimit = true;
  ***REMOVED***

  if (this.s.session) {
    opts = Object.assign({***REMOVED***, opts, { session: this.s.session ***REMOVED***);
  ***REMOVED***

  return executeOperation(this.s.topology, count, [this, applySkipLimit, opts, callback], {
    skipSessions: !!this.s.session
  ***REMOVED***);
***REMOVED***;

/**
 * Close the cursor, sending a KillCursor command and emitting close.
 * @method
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {boolean***REMOVED*** [options.skipKillCursors] Bypass calling killCursors when closing the cursor.
 * @param {Cursor~resultCallback***REMOVED*** [callback] The result callback.
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.close = function(options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = Object.assign({***REMOVED***, { skipKillCursors: false ***REMOVED***, options);

  this.s.state = Cursor.CLOSED;
  if (!options.skipKillCursors) {
    // Kill the cursor
    this.kill();
  ***REMOVED***

  const completeClose = () => {
    // Emit the close event for the cursor
    this.emit('close');

    // Callback if provided
    if (typeof callback === 'function') {
      return handleCallback(callback, null, this);
    ***REMOVED***

    // Return a Promise
    return new this.s.promiseLibrary(resolve => {
      resolve();
    ***REMOVED***);
  ***REMOVED***;

  if (this.s.session) {
    if (typeof callback === 'function') {
      return this._endSession(() => completeClose());
    ***REMOVED***

    return new this.s.promiseLibrary(resolve => {
      this._endSession(() => completeClose().then(resolve));
    ***REMOVED***);
  ***REMOVED***

  return completeClose();
***REMOVED***;

/**
 * Map all documents using the provided function
 * @method
 * @param {function***REMOVED*** [transform] The mapping transformation method.
 * @return {Cursor***REMOVED***
 */
Cursor.prototype.map = function(transform) {
  if (this.cursorState.transforms && this.cursorState.transforms.doc) {
    const oldTransform = this.cursorState.transforms.doc;
    this.cursorState.transforms.doc = doc => {
      return transform(oldTransform(doc));
    ***REMOVED***;
  ***REMOVED***
    this.cursorState.transforms = { doc: transform ***REMOVED***;
  ***REMOVED***
  return this;
***REMOVED***;

/**
 * Is the cursor closed
 * @method
 * @return {boolean***REMOVED***
 */
Cursor.prototype.isClosed = function() {
  return this.isDead();
***REMOVED***;

Cursor.prototype.destroy = function(err) {
  if (err) this.emit('error', err);
  this.pause();
  this.close();
***REMOVED***;

/**
 * Return a modified Readable stream including a possible transform method.
 * @method
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {function***REMOVED*** [options.transform] A transformation method applied to each document emitted by the stream.
 * @return {Cursor***REMOVED***
 * TODO: replace this method with transformStream in next major release
 */
Cursor.prototype.stream = function(options) {
  this.s.streamOptions = options || {***REMOVED***;
  return this;
***REMOVED***;

/**
 * Return a modified Readable stream that applies a given transform function, if supplied. If none supplied,
 * returns a stream of unmodified docs.
 * @method
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {function***REMOVED*** [options.transform] A transformation method applied to each document emitted by the stream.
 * @return {stream***REMOVED***
 */
Cursor.prototype.transformStream = function(options) {
  const streamOptions = options || {***REMOVED***;
  if (typeof streamOptions.transform === 'function') {
    const stream = new Transform({
      objectMode: true,
      transform: function(chunk, encoding, callback) {
        this.push(streamOptions.transform(chunk));
        callback();
      ***REMOVED***
    ***REMOVED***);

    return this.pipe(stream);
  ***REMOVED***
  return this.pipe(new PassThrough({ objectMode: true ***REMOVED***));
***REMOVED***;

/**
 * Execute the explain for the cursor
 * @method
 * @param {Cursor~resultCallback***REMOVED*** [callback] The result callback.
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.explain = function(callback) {
  this.s.cmd.explain = true;

  // Do we have a readConcern
  if (this.s.cmd.readConcern) {
    delete this.s.cmd['readConcern'];
  ***REMOVED***

  return executeOperation(this.s.topology, this._next.bind(this), [callback], {
    skipSessions: true
  ***REMOVED***);
***REMOVED***;

Cursor.prototype._read = function() {
  if (this.s.state === Cursor.CLOSED || this.isDead()) {
    return this.push(null);
  ***REMOVED***

  // Get the next item
  this.next((err, result) => {
    if (err) {
      if (this.listeners('error') && this.listeners('error').length > 0) {
        this.emit('error', err);
      ***REMOVED***
      if (!this.isDead()) this.close();

      // Emit end event
      this.emit('end');
      return this.emit('finish');
    ***REMOVED***

    // If we provided a transformation method
    if (typeof this.s.streamOptions.transform === 'function' && result != null) {
      return this.push(this.s.streamOptions.transform(result));
    ***REMOVED***

    // If we provided a map function
    if (
      this.cursorState.transforms &&
      typeof this.cursorState.transforms.doc === 'function' &&
      result != null
    ) {
      return this.push(this.cursorState.transforms.doc(result));
    ***REMOVED***

    // Return the result
    this.push(result);

    if (result === null && this.isDead()) {
      this.once('end', () => {
        this.close();
        this.emit('finish');
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***;

/**
 * Return the cursor logger
 * @method
 * @return {Logger***REMOVED*** return the cursor logger
 * @ignore
 */
Cursor.prototype.getLogger = function() {
  return this.logger;
***REMOVED***;

Object.defineProperty(Cursor.prototype, 'readPreference', {
  enumerable: true,
  get: function() {
    if (!this || !this.s) {
      return null;
    ***REMOVED***

    return this.s.options.readPreference;
  ***REMOVED***
***REMOVED***);

Object.defineProperty(Cursor.prototype, 'namespace', {
  enumerable: true,
  get: function() {
    if (!this || !this.s) {
      return null;
    ***REMOVED***

    // TODO: refactor this logic into core
    const ns = this.s.ns || '';
    const firstDot = ns.indexOf('.');
    if (firstDot < 0) {
      return {
        database: this.s.ns,
        collection: ''
      ***REMOVED***;
    ***REMOVED***
    return {
      database: ns.substr(0, firstDot),
      collection: ns.substr(firstDot + 1)
    ***REMOVED***;
  ***REMOVED***
***REMOVED***);

/**
 * The read() method pulls some data out of the internal buffer and returns it. If there is no data available, then it will return null.
 * @function external:Readable#read
 * @param {number***REMOVED*** size Optional argument to specify how much data to read.
 * @return {(String | Buffer | null)***REMOVED***
 */

/**
 * Call this function to cause the stream to return strings of the specified encoding instead of Buffer objects.
 * @function external:Readable#setEncoding
 * @param {string***REMOVED*** encoding The encoding to use.
 * @return {null***REMOVED***
 */

/**
 * This method will cause the readable stream to resume emitting data events.
 * @function external:Readable#resume
 * @return {null***REMOVED***
 */

/**
 * This method will cause a stream in flowing-mode to stop emitting data events. Any data that becomes available will remain in the internal buffer.
 * @function external:Readable#pause
 * @return {null***REMOVED***
 */

/**
 * This method pulls all the data out of a readable stream, and writes it to the supplied destination, automatically managing the flow so that the destination is not overwhelmed by a fast readable stream.
 * @function external:Readable#pipe
 * @param {Writable***REMOVED*** destination The destination for writing data
 * @param {object***REMOVED*** [options] Pipe options
 * @return {null***REMOVED***
 */

/**
 * This method will remove the hooks set up for a previous pipe() call.
 * @function external:Readable#unpipe
 * @param {Writable***REMOVED*** [destination] The destination for writing data
 * @return {null***REMOVED***
 */

/**
 * This is useful in certain cases where a stream is being consumed by a parser, which needs to "un-consume" some data that it has optimistically pulled out of the source, so that the stream can be passed on to some other party.
 * @function external:Readable#unshift
 * @param {(Buffer|string)***REMOVED*** chunk Chunk of data to unshift onto the read queue.
 * @return {null***REMOVED***
 */

/**
 * Versions of Node prior to v0.10 had streams that did not implement the entire Streams API as it is today. (See "Compatibility" below for more information.)
 * @function external:Readable#wrap
 * @param {Stream***REMOVED*** stream An "old style" readable stream.
 * @return {null***REMOVED***
 */

Cursor.INIT = 0;
Cursor.OPEN = 1;
Cursor.CLOSED = 2;
Cursor.GET_MORE = 3;

module.exports = Cursor;
