// mongodb/lib/db.js
'use strict';

const EventEmitter = require('events').EventEmitter;
const inherits = require('util').inherits;
const getSingleProperty = require('./utils').getSingleProperty;
const CommandCursor = require('./command_cursor');
const handleCallback = require('./utils').handleCallback;
const filterOptions = require('./utils').filterOptions;
const toError = require('./utils').toError;
const ReadPreference = require('mongodb-core').ReadPreference;
const MongoError = require('mongodb-core').MongoError;
const ObjectID = require('mongodb-core').ObjectID;
const Logger = require('mongodb-core').Logger;
const Collection = require('./collection');
const mergeOptionsAndWriteConcern = require('./utils').mergeOptionsAndWriteConcern;
const executeOperation = require('./utils').executeOperation;
const applyWriteConcern = require('./utils').applyWriteConcern;
const resolveReadPreference = require('./utils').resolveReadPreference;
const ChangeStream = require('./change_stream');
const deprecate = require('util').deprecate;
const deprecateOptions = require('./utils').deprecateOptions;

// Operations
const addUser = require('./operations/db_ops').addUser;
const collections = require('./operations/db_ops').collections;
const createCollection = require('./operations/db_ops').createCollection;
const createIndex = require('./operations/db_ops').createIndex;
const createListener = require('./operations/db_ops').createListener;
const dropCollection = require('./operations/db_ops').dropCollection;
const dropDatabase = require('./operations/db_ops').dropDatabase;
const ensureIndex = require('./operations/db_ops').ensureIndex;
const evaluate = require('./operations/db_ops').evaluate;
const executeCommand = require('./operations/db_ops').executeCommand;
const executeDbAdminCommand = require('./operations/db_ops').executeDbAdminCommand;
const indexInformation = require('./operations/db_ops').indexInformation;
const listCollectionsTransforms = require('./operations/db_ops').listCollectionsTransforms;
const profilingInfo = require('./operations/db_ops').profilingInfo;
const profilingLevel = require('./operations/db_ops').profilingLevel;
const removeUser = require('./operations/db_ops').removeUser;
const setProfilingLevel = require('./operations/db_ops').setProfilingLevel;
const validateDatabaseName = require('./operations/db_ops').validateDatabaseName;

/**
 * @fileOverview The **Db** class is a class that represents a MongoDB Database.
 *
 * @example
 * const MongoClient = require('mongodb').MongoClient;
 * // Connection url
 * const url = 'mongodb://localhost:27017';
 * // Database Name
 * const dbName = 'test';
 * // Connect using MongoClient
 * MongoClient.connect(url, function(err, client) {
 *   // Select the database by name
 *   const testDb = client.db(dbName);
 *   client.close();
 * ***REMOVED***);
 */

// Allowed parameters
const legalOptionNames = [
  'w',
  'wtimeout',
  'fsync',
  'j',
  'readPreference',
  'readPreferenceTags',
  'native_parser',
  'forceServerObjectId',
  'pkFactory',
  'serializeFunctions',
  'raw',
  'bufferMaxEntries',
  'authSource',
  'ignoreUndefined',
  'promoteLongs',
  'promiseLibrary',
  'readConcern',
  'retryMiliSeconds',
  'numberOfRetries',
  'parentDb',
  'noListener',
  'loggerLevel',
  'logger',
  'promoteBuffers',
  'promoteLongs',
  'promoteValues',
  'compression',
  'retryWrites'
];

/**
 * Creates a new Db instance
 * @class
 * @param {string***REMOVED*** databaseName The name of the database this instance represents.
 * @param {(Server|ReplSet|Mongos)***REMOVED*** topology The server topology for the database.
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {string***REMOVED*** [options.authSource] If the database authentication is dependent on another databaseName.
 * @param {(number|string)***REMOVED*** [options.w] The write concern.
 * @param {number***REMOVED*** [options.wtimeout] The write concern timeout.
 * @param {boolean***REMOVED*** [options.j=false] Specify a journal write concern.
 * @param {boolean***REMOVED*** [options.forceServerObjectId=false] Force server to assign _id values instead of driver.
 * @param {boolean***REMOVED*** [options.serializeFunctions=false] Serialize functions on any object.
 * @param {Boolean***REMOVED*** [options.ignoreUndefined=false] Specify if the BSON serializer should ignore undefined fields.
 * @param {boolean***REMOVED*** [options.raw=false] Return document results as raw BSON buffers.
 * @param {boolean***REMOVED*** [options.promoteLongs=true] Promotes Long values to number if they fit inside the 53 bits resolution.
 * @param {boolean***REMOVED*** [options.promoteBuffers=false] Promotes Binary BSON values to native Node Buffers.
 * @param {boolean***REMOVED*** [options.promoteValues=true] Promotes BSON values to native types where possible, set to false to only receive wrapper types.
 * @param {number***REMOVED*** [options.bufferMaxEntries=-1] Sets a cap on how many operations the driver will buffer up before giving up on getting a working connection, default is -1 which is unlimited.
 * @param {(ReadPreference|string)***REMOVED*** [options.readPreference] The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST).
 * @param {object***REMOVED*** [options.pkFactory] A primary key factory object for generation of custom _id keys.
 * @param {object***REMOVED*** [options.promiseLibrary] A Promise library class the application wishes to use such as Bluebird, must be ES6 compatible
 * @param {object***REMOVED*** [options.readConcern] Specify a read concern for the collection. (only MongoDB 3.2 or higher supported)
 * @param {object***REMOVED*** [options.readConcern.level='local'] Specify a read concern level for the collection operations, one of [local|majority]. (only MongoDB 3.2 or higher supported)
 * @property {(Server|ReplSet|Mongos)***REMOVED*** serverConfig Get the current db topology.
 * @property {number***REMOVED*** bufferMaxEntries Current bufferMaxEntries value for the database
 * @property {string***REMOVED*** databaseName The name of the database this instance represents.
 * @property {object***REMOVED*** options The options associated with the db instance.
 * @property {boolean***REMOVED*** native_parser The current value of the parameter native_parser.
 * @property {boolean***REMOVED*** slaveOk The current slaveOk value for the db instance.
 * @property {object***REMOVED*** writeConcern The current write concern values.
 * @property {object***REMOVED*** topology Access the topology object (single server, replicaset or mongos).
 * @fires Db#close
 * @fires Db#reconnect
 * @fires Db#error
 * @fires Db#timeout
 * @fires Db#parseError
 * @fires Db#fullsetup
 * @return {Db***REMOVED*** a Db instance.
 */
function Db(databaseName, topology, options) {
  options = options || {***REMOVED***;
  if (!(this instanceof Db)) return new Db(databaseName, topology, options);
  EventEmitter.call(this);

  // Get the promiseLibrary
  const promiseLibrary = options.promiseLibrary || Promise;

  // Filter the options
  options = filterOptions(options, legalOptionNames);

  // Ensure we put the promiseLib in the options
  options.promiseLibrary = promiseLibrary;

  // Internal state of the db object
  this.s = {
    // Database name
    databaseName: databaseName,
    // DbCache
    dbCache: {***REMOVED***,
    // Children db's
    children: [],
    // Topology
    topology: topology,
    // Options
    options: options,
    // Logger instance
    logger: Logger('Db', options),
    // Get the bson parser
    bson: topology ? topology.bson : null,
    // Unpack read preference
    readPreference: options.readPreference,
    // Set buffermaxEntries
    bufferMaxEntries: typeof options.bufferMaxEntries === 'number' ? options.bufferMaxEntries : -1,
    // Parent db (if chained)
    parentDb: options.parentDb || null,
    // Set up the primary key factory or fallback to ObjectID
    pkFactory: options.pkFactory || ObjectID,
    // Get native parser
    nativeParser: options.nativeParser || options.native_parser,
    // Promise library
    promiseLibrary: promiseLibrary,
    // No listener
    noListener: typeof options.noListener === 'boolean' ? options.noListener : false,
    // ReadConcern
    readConcern: options.readConcern
  ***REMOVED***;

  // Ensure we have a valid db name
  validateDatabaseName(this.s.databaseName);

  // Add a read Only property
  getSingleProperty(this, 'serverConfig', this.s.topology);
  getSingleProperty(this, 'bufferMaxEntries', this.s.bufferMaxEntries);
  getSingleProperty(this, 'databaseName', this.s.databaseName);

  // This is a child db, do not register any listeners
  if (options.parentDb) return;
  if (this.s.noListener) return;

  // Add listeners
  topology.on('error', createListener(this, 'error', this));
  topology.on('timeout', createListener(this, 'timeout', this));
  topology.on('close', createListener(this, 'close', this));
  topology.on('parseError', createListener(this, 'parseError', this));
  topology.once('open', createListener(this, 'open', this));
  topology.once('fullsetup', createListener(this, 'fullsetup', this));
  topology.once('all', createListener(this, 'all', this));
  topology.on('reconnect', createListener(this, 'reconnect', this));
***REMOVED***

inherits(Db, EventEmitter);

// Topology
Object.defineProperty(Db.prototype, 'topology', {
  enumerable: true,
  get: function() {
    return this.s.topology;
  ***REMOVED***
***REMOVED***);

// Options
Object.defineProperty(Db.prototype, 'options', {
  enumerable: true,
  get: function() {
    return this.s.options;
  ***REMOVED***
***REMOVED***);

// slaveOk specified
Object.defineProperty(Db.prototype, 'slaveOk', {
  enumerable: true,
  get: function() {
    if (
      this.s.options.readPreference != null &&
      (this.s.options.readPreference !== 'primary' ||
        this.s.options.readPreference.mode !== 'primary')
    ) {
      return true;
    ***REMOVED***
    return false;
  ***REMOVED***
***REMOVED***);

// get the write Concern
Object.defineProperty(Db.prototype, 'writeConcern', {
  enumerable: true,
  get: function() {
    const ops = {***REMOVED***;
    if (this.s.options.w != null) ops.w = this.s.options.w;
    if (this.s.options.j != null) ops.j = this.s.options.j;
    if (this.s.options.fsync != null) ops.fsync = this.s.options.fsync;
    if (this.s.options.wtimeout != null) ops.wtimeout = this.s.options.wtimeout;
    return ops;
  ***REMOVED***
***REMOVED***);

/**
 * Execute a command
 * @method
 * @param {object***REMOVED*** command The command hash
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {(ReadPreference|string)***REMOVED*** [options.readPreference] The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST).
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The command result callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.command = function(command, options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = Object.assign({***REMOVED***, options);

  return executeOperation(this.s.topology, executeCommand, [this, command, options, callback]);
***REMOVED***;

/**
 * Return the Admin db instance
 * @method
 * @return {Admin***REMOVED*** return the new Admin db instance
 */
Db.prototype.admin = function() {
  const Admin = require('./admin');

  return new Admin(this, this.s.topology, this.s.promiseLibrary);
***REMOVED***;

/**
 * The callback format for the collection method, must be used if strict is specified
 * @callback Db~collectionResultCallback
 * @param {MongoError***REMOVED*** error An error instance representing the error during the execution.
 * @param {Collection***REMOVED*** collection The collection instance.
 */

const collectionKeys = [
  'pkFactory',
  'readPreference',
  'serializeFunctions',
  'strict',
  'readConcern',
  'ignoreUndefined',
  'promoteValues',
  'promoteBuffers',
  'promoteLongs'
];

/**
 * Fetch a specific collection (containing the actual collection information). If the application does not use strict mode you
 * can use it without a callback in the following way: `const collection = db.collection('mycollection');`
 *
 * @method
 * @param {string***REMOVED*** name the collection name we wish to access.
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {(number|string)***REMOVED*** [options.w] The write concern.
 * @param {number***REMOVED*** [options.wtimeout] The write concern timeout.
 * @param {boolean***REMOVED*** [options.j=false] Specify a journal write concern.
 * @param {boolean***REMOVED*** [options.raw=false] Return document results as raw BSON buffers.
 * @param {object***REMOVED*** [options.pkFactory] A primary key factory object for generation of custom _id keys.
 * @param {(ReadPreference|string)***REMOVED*** [options.readPreference] The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST).
 * @param {boolean***REMOVED*** [options.serializeFunctions=false] Serialize functions on any object.
 * @param {boolean***REMOVED*** [options.strict=false] Returns an error if the collection does not exist
 * @param {object***REMOVED*** [options.readConcern] Specify a read concern for the collection. (only MongoDB 3.2 or higher supported)
 * @param {object***REMOVED*** [options.readConcern.level='local'] Specify a read concern level for the collection operations, one of [local|majority]. (only MongoDB 3.2 or higher supported)
 * @param {Db~collectionResultCallback***REMOVED*** [callback] The collection result callback
 * @return {Collection***REMOVED*** return the new Collection instance if not in strict mode
 */
Db.prototype.collection = function(name, options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;
  options = Object.assign({***REMOVED***, options);

  // Set the promise library
  options.promiseLibrary = this.s.promiseLibrary;

  // If we have not set a collection level readConcern set the db level one
  options.readConcern = options.readConcern || this.s.readConcern;

  // Do we have ignoreUndefined set
  if (this.s.options.ignoreUndefined) {
    options.ignoreUndefined = this.s.options.ignoreUndefined;
  ***REMOVED***

  // Merge in all needed options and ensure correct writeConcern merging from db level
  options = mergeOptionsAndWriteConcern(options, this.s.options, collectionKeys, true);

  // Execute
  if (options == null || !options.strict) {
    try {
      const collection = new Collection(
        this,
        this.s.topology,
        this.s.databaseName,
        name,
        this.s.pkFactory,
        options
      );
      if (callback) callback(null, collection);
      return collection;
    ***REMOVED*** catch (err) {
      if (err instanceof MongoError && callback) return callback(err);
      throw err;
    ***REMOVED***
  ***REMOVED***

  // Strict mode
  if (typeof callback !== 'function') {
    throw toError(`A callback is required in strict mode. While getting collection ${name***REMOVED***`);
  ***REMOVED***

  // Did the user destroy the topology
  if (this.serverConfig && this.serverConfig.isDestroyed()) {
    return callback(new MongoError('topology was destroyed'));
  ***REMOVED***

  const listCollectionOptions = Object.assign({***REMOVED***, options, { nameOnly: true ***REMOVED***);

  // Strict mode
  this.listCollections({ name: name ***REMOVED***, listCollectionOptions).toArray((err, collections) => {
    if (err != null) return handleCallback(callback, err, null);
    if (collections.length === 0)
      return handleCallback(
        callback,
        toError(`Collection ${name***REMOVED*** does not exist. Currently in strict mode.`),
        null
      );

    try {
      return handleCallback(
        callback,
        null,
        new Collection(this, this.s.topology, this.s.databaseName, name, this.s.pkFactory, options)
      );
    ***REMOVED*** catch (err) {
      return handleCallback(callback, err, null);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***;

/**
 * Create a new collection on a server with the specified options. Use this to create capped collections.
 * More information about command options available at https://docs.mongodb.com/manual/reference/command/create/
 *
 * @method
 * @param {string***REMOVED*** name the collection name we wish to access.
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {(number|string)***REMOVED*** [options.w] The write concern.
 * @param {number***REMOVED*** [options.wtimeout] The write concern timeout.
 * @param {boolean***REMOVED*** [options.j=false] Specify a journal write concern.
 * @param {boolean***REMOVED*** [options.raw=false] Return document results as raw BSON buffers.
 * @param {object***REMOVED*** [options.pkFactory] A primary key factory object for generation of custom _id keys.
 * @param {(ReadPreference|string)***REMOVED*** [options.readPreference] The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST).
 * @param {boolean***REMOVED*** [options.serializeFunctions=false] Serialize functions on any object.
 * @param {boolean***REMOVED*** [options.strict=false] Returns an error if the collection does not exist
 * @param {boolean***REMOVED*** [options.capped=false] Create a capped collection.
 * @param {boolean***REMOVED*** [options.autoIndexId=true] DEPRECATED: Create an index on the _id field of the document, True by default on MongoDB 2.6 - 3.0
 * @param {number***REMOVED*** [options.size] The size of the capped collection in bytes.
 * @param {number***REMOVED*** [options.max] The maximum number of documents in the capped collection.
 * @param {number***REMOVED*** [options.flags] Optional. Available for the MMAPv1 storage engine only to set the usePowerOf2Sizes and the noPadding flag.
 * @param {object***REMOVED*** [options.storageEngine] Allows users to specify configuration to the storage engine on a per-collection basis when creating a collection on MongoDB 3.0 or higher.
 * @param {object***REMOVED*** [options.validator] Allows users to specify validation rules or expressions for the collection. For more information, see Document Validation on MongoDB 3.2 or higher.
 * @param {string***REMOVED*** [options.validationLevel] Determines how strictly MongoDB applies the validation rules to existing documents during an update on MongoDB 3.2 or higher.
 * @param {string***REMOVED*** [options.validationAction] Determines whether to error on invalid documents or just warn about the violations but allow invalid documents to be inserted on MongoDB 3.2 or higher.
 * @param {object***REMOVED*** [options.indexOptionDefaults] Allows users to specify a default configuration for indexes when creating a collection on MongoDB 3.2 or higher.
 * @param {string***REMOVED*** [options.viewOn] The name of the source collection or view from which to create the view. The name is not the full namespace of the collection or view; i.e. does not include the database name and implies the same database as the view to create on MongoDB 3.4 or higher.
 * @param {array***REMOVED*** [options.pipeline] An array that consists of the aggregation pipeline stage. create creates the view by applying the specified pipeline to the viewOn collection or view on MongoDB 3.4 or higher.
 * @param {object***REMOVED*** [options.collation] Specify collation (MongoDB 3.4 or higher) settings for update operation (see 3.4 documentation for available fields).
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~collectionResultCallback***REMOVED*** [callback] The results callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.createCollection = deprecateOptions(
  {
    name: 'Db.createCollection',
    deprecatedOptions: ['autoIndexId'],
    optionsIndex: 1
  ***REMOVED***,
  function(name, options, callback) {
    if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
    options = options || {***REMOVED***;
    options.promiseLibrary = options.promiseLibrary || this.s.promiseLibrary;

    return executeOperation(this.s.topology, createCollection, [this, name, options, callback]);
  ***REMOVED***
);

/**
 * Get all the db statistics.
 *
 * @method
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {number***REMOVED*** [options.scale] Divide the returned sizes by scale value.
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The collection result callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.stats = function(options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;
  // Build command object
  const commandObject = { dbStats: true ***REMOVED***;
  // Check if we have the scale value
  if (options['scale'] != null) commandObject['scale'] = options['scale'];

  // If we have a readPreference set
  if (options.readPreference == null && this.s.readPreference) {
    options.readPreference = this.s.readPreference;
  ***REMOVED***

  // Execute the command
  return this.command(commandObject, options, callback);
***REMOVED***;

/**
 * Get the list of all collection information for the specified db.
 *
 * @method
 * @param {object***REMOVED*** [filter={***REMOVED***] Query to filter collections by
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {boolean***REMOVED*** [options.nameOnly=false] Since 4.0: If true, will only return the collection name in the response, and will omit additional info
 * @param {number***REMOVED*** [options.batchSize] The batchSize for the returned command cursor or if pre 2.8 the systems batch collection
 * @param {(ReadPreference|string)***REMOVED*** [options.readPreference] The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST).
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @return {CommandCursor***REMOVED***
 */
Db.prototype.listCollections = function(filter, options) {
  filter = filter || {***REMOVED***;
  options = options || {***REMOVED***;

  // Shallow clone the object
  options = Object.assign({***REMOVED***, options);
  // Set the promise library
  options.promiseLibrary = this.s.promiseLibrary;

  // Ensure valid readPreference
  options.readPreference = resolveReadPreference(options, {
    db: this,
    default: ReadPreference.primary
  ***REMOVED***);

  // Cursor options
  let cursor = options.batchSize ? { batchSize: options.batchSize ***REMOVED*** : {***REMOVED***;

  // We have a list collections command
  if (this.serverConfig.capabilities().hasListCollectionsCommand) {
    const nameOnly = typeof options.nameOnly === 'boolean' ? options.nameOnly : false;
    // Build the command
    const command = { listCollections: true, filter, cursor, nameOnly ***REMOVED***;
    // Set the AggregationCursor constructor
    options.cursorFactory = CommandCursor;
    // Create the cursor
    cursor = this.s.topology.cursor(`${this.s.databaseName***REMOVED***.$cmd`, command, options);
    // Do we have a readPreference, apply it
    if (options.readPreference) {
      cursor.setReadPreference(options.readPreference);
    ***REMOVED***
    // Return the cursor
    return cursor;
  ***REMOVED***

  // We cannot use the listCollectionsCommand
  if (!this.serverConfig.capabilities().hasListCollectionsCommand) {
    // If we have legacy mode and have not provided a full db name filter it
    if (
      typeof filter.name === 'string' &&
      !new RegExp('^' + this.databaseName + '\\.').test(filter.name)
    ) {
      filter = Object.assign({***REMOVED***, filter);
      filter.name = `${this.s.databaseName***REMOVED***.${filter.name***REMOVED***`;
    ***REMOVED***
  ***REMOVED***

  // No filter, filter by current database
  if (filter == null) {
    filter.name = `/${this.s.databaseName***REMOVED***/`;
  ***REMOVED***

  // Rewrite the filter to use $and to filter out indexes
  if (filter.name) {
    filter = { $and: [{ name: filter.name ***REMOVED***, { name: /^((?!\$).)*$/ ***REMOVED***] ***REMOVED***;
  ***REMOVED***
    filter = { name: /^((?!\$).)*$/ ***REMOVED***;
  ***REMOVED***

  // Return options
  const _options = { transforms: listCollectionsTransforms(this.s.databaseName) ***REMOVED***;
  // Get the cursor
  cursor = this.collection(Db.SYSTEM_NAMESPACE_COLLECTION).find(filter, _options);
  // Do we have a readPreference, apply it
  if (options.readPreference) cursor.setReadPreference(options.readPreference);
  // Set the passed in batch size if one was provided
  if (options.batchSize) cursor = cursor.batchSize(options.batchSize);
  // We have a fallback mode using legacy systems collections
  return cursor;
***REMOVED***;

/**
 * Evaluate JavaScript on the server
 *
 * @method
 * @param {Code***REMOVED*** code JavaScript to execute on server.
 * @param {(object|array)***REMOVED*** parameters The parameters for the call.
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {boolean***REMOVED*** [options.nolock=false] Tell MongoDB not to block on the evaulation of the javascript.
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The results callback
 * @deprecated Eval is deprecated on MongoDB 3.2 and forward
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.eval = deprecate(function(code, parameters, options, callback) {
  const args = Array.prototype.slice.call(arguments, 1);
  callback = typeof args[args.length - 1] === 'function' ? args.pop() : undefined;
  parameters = args.length ? args.shift() : parameters;
  options = args.length ? args.shift() || {***REMOVED*** : {***REMOVED***;

  return executeOperation(this.s.topology, evaluate, [this, code, parameters, options, callback]);
***REMOVED***, 'Db.eval is deprecated as of MongoDB version 3.2');

/**
 * Rename a collection.
 *
 * @method
 * @param {string***REMOVED*** fromCollection Name of current collection to rename.
 * @param {string***REMOVED*** toCollection New name of of the collection.
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {boolean***REMOVED*** [options.dropTarget=false] Drop the target name collection if it previously exists.
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~collectionResultCallback***REMOVED*** [callback] The results callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.renameCollection = function(fromCollection, toCollection, options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;
  // Add return new collection
  options.new_collection = true;

  const collection = this.collection(fromCollection);
  return executeOperation(this.s.topology, collection.rename.bind(collection), [
    toCollection,
    options,
    callback
  ]);
***REMOVED***;

/**
 * Drop a collection from the database, removing it permanently. New accesses will create a new collection.
 *
 * @method
 * @param {string***REMOVED*** name Name of collection to drop
 * @param {Object***REMOVED*** [options] Optional settings
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The results callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.dropCollection = function(name, options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;

  // Command to execute
  const cmd = { drop: name ***REMOVED***;

  // Decorate with write concern
  applyWriteConcern(cmd, { db: this ***REMOVED***, options);

  // options
  const opts = Object.assign({***REMOVED***, this.s.options, { readPreference: ReadPreference.PRIMARY ***REMOVED***);
  if (options.session) opts.session = options.session;

  return executeOperation(this.s.topology, dropCollection, [this, cmd, opts, callback]);
***REMOVED***;

/**
 * Drop a database, removing it permanently from the server.
 *
 * @method
 * @param {Object***REMOVED*** [options] Optional settings
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The results callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.dropDatabase = function(options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;
  // Drop database command
  const cmd = { dropDatabase: 1 ***REMOVED***;

  // Decorate with write concern
  applyWriteConcern(cmd, { db: this ***REMOVED***, options);

  // Ensure primary only
  const finalOptions = Object.assign({***REMOVED***, this.s.options, {
    readPreference: ReadPreference.PRIMARY
  ***REMOVED***);

  if (options.session) {
    finalOptions.session = options.session;
  ***REMOVED***

  return executeOperation(this.s.topology, dropDatabase, [this, cmd, finalOptions, callback]);
***REMOVED***;

/**
 * Fetch all collections for the current db.
 *
 * @method
 * @param {Object***REMOVED*** [options] Optional settings
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~collectionsResultCallback***REMOVED*** [callback] The results callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.collections = function(options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;

  return executeOperation(this.s.topology, collections, [this, options, callback]);
***REMOVED***;

/**
 * Runs a command on the database as admin.
 * @method
 * @param {object***REMOVED*** command The command hash
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {(ReadPreference|string)***REMOVED*** [options.readPreference] The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST).
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The command result callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.executeDbAdminCommand = function(selector, options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;
  options.readPreference = resolveReadPreference(options);

  return executeOperation(this.s.topology, executeDbAdminCommand, [
    this,
    selector,
    options,
    callback
  ]);
***REMOVED***;

/**
 * Creates an index on the db and collection.
 * @method
 * @param {string***REMOVED*** name Name of the collection to create the index on.
 * @param {(string|object)***REMOVED*** fieldOrSpec Defines the index.
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {(number|string)***REMOVED*** [options.w] The write concern.
 * @param {number***REMOVED*** [options.wtimeout] The write concern timeout.
 * @param {boolean***REMOVED*** [options.j=false] Specify a journal write concern.
 * @param {boolean***REMOVED*** [options.unique=false] Creates an unique index.
 * @param {boolean***REMOVED*** [options.sparse=false] Creates a sparse index.
 * @param {boolean***REMOVED*** [options.background=false] Creates the index in the background, yielding whenever possible.
 * @param {boolean***REMOVED*** [options.dropDups=false] A unique index cannot be created on a key that has pre-existing duplicate values. If you would like to create the index anyway, keeping the first document the database indexes and deleting all subsequent documents that have duplicate value
 * @param {number***REMOVED*** [options.min] For geospatial indexes set the lower bound for the co-ordinates.
 * @param {number***REMOVED*** [options.max] For geospatial indexes set the high bound for the co-ordinates.
 * @param {number***REMOVED*** [options.v] Specify the format version of the indexes.
 * @param {number***REMOVED*** [options.expireAfterSeconds] Allows you to expire data on indexes applied to a data (MongoDB 2.2 or higher)
 * @param {number***REMOVED*** [options.name] Override the autogenerated index name (useful if the resulting name is larger than 128 bytes)
 * @param {object***REMOVED*** [options.partialFilterExpression] Creates a partial index based on the given filter object (MongoDB 3.2 or higher)
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The command result callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.createIndex = function(name, fieldOrSpec, options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options ? Object.assign({***REMOVED***, options) : {***REMOVED***;

  return executeOperation(this.s.topology, createIndex, [
    this,
    name,
    fieldOrSpec,
    options,
    callback
  ]);
***REMOVED***;

/**
 * Ensures that an index exists, if it does not it creates it
 * @method
 * @deprecated since version 2.0
 * @param {string***REMOVED*** name The index name
 * @param {(string|object)***REMOVED*** fieldOrSpec Defines the index.
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {(number|string)***REMOVED*** [options.w] The write concern.
 * @param {number***REMOVED*** [options.wtimeout] The write concern timeout.
 * @param {boolean***REMOVED*** [options.j=false] Specify a journal write concern.
 * @param {boolean***REMOVED*** [options.unique=false] Creates an unique index.
 * @param {boolean***REMOVED*** [options.sparse=false] Creates a sparse index.
 * @param {boolean***REMOVED*** [options.background=false] Creates the index in the background, yielding whenever possible.
 * @param {boolean***REMOVED*** [options.dropDups=false] A unique index cannot be created on a key that has pre-existing duplicate values. If you would like to create the index anyway, keeping the first document the database indexes and deleting all subsequent documents that have duplicate value
 * @param {number***REMOVED*** [options.min] For geospatial indexes set the lower bound for the co-ordinates.
 * @param {number***REMOVED*** [options.max] For geospatial indexes set the high bound for the co-ordinates.
 * @param {number***REMOVED*** [options.v] Specify the format version of the indexes.
 * @param {number***REMOVED*** [options.expireAfterSeconds] Allows you to expire data on indexes applied to a data (MongoDB 2.2 or higher)
 * @param {number***REMOVED*** [options.name] Override the autogenerated index name (useful if the resulting name is larger than 128 bytes)
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The command result callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.ensureIndex = deprecate(function(name, fieldOrSpec, options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;

  return executeOperation(this.s.topology, ensureIndex, [
    this,
    name,
    fieldOrSpec,
    options,
    callback
  ]);
***REMOVED***, 'Db.ensureIndex is deprecated as of MongoDB version 3.0 / driver version 2.0');

Db.prototype.addChild = function(db) {
  if (this.s.parentDb) return this.s.parentDb.addChild(db);
  this.s.children.push(db);
***REMOVED***;

/**
 * Add a user to the database.
 * @method
 * @param {string***REMOVED*** username The username.
 * @param {string***REMOVED*** password The password.
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {(number|string)***REMOVED*** [options.w] The write concern.
 * @param {number***REMOVED*** [options.wtimeout] The write concern timeout.
 * @param {boolean***REMOVED*** [options.j=false] Specify a journal write concern.
 * @param {object***REMOVED*** [options.customData] Custom data associated with the user (only Mongodb 2.6 or higher)
 * @param {object[]***REMOVED*** [options.roles] Roles associated with the created user (only Mongodb 2.6 or higher)
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The command result callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.addUser = function(username, password, options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;

  return executeOperation(this.s.topology, addUser, [this, username, password, options, callback]);
***REMOVED***;

/**
 * Remove a user from a database
 * @method
 * @param {string***REMOVED*** username The username.
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {(number|string)***REMOVED*** [options.w] The write concern.
 * @param {number***REMOVED*** [options.wtimeout] The write concern timeout.
 * @param {boolean***REMOVED*** [options.j=false] Specify a journal write concern.
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The command result callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.removeUser = function(username, options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;

  return executeOperation(this.s.topology, removeUser, [this, username, options, callback]);
***REMOVED***;

/**
 * Set the current profiling level of MongoDB
 *
 * @param {string***REMOVED*** level The new profiling level (off, slow_only, all).
 * @param {Object***REMOVED*** [options] Optional settings
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The command result callback.
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.setProfilingLevel = function(level, options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;

  return executeOperation(this.s.topology, setProfilingLevel, [this, level, options, callback]);
***REMOVED***;

/**
 * Retrive the current profiling information for MongoDB
 *
 * @param {Object***REMOVED*** [options] Optional settings
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The command result callback.
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 * @deprecated Query the system.profile collection directly.
 */
Db.prototype.profilingInfo = deprecate(function(options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;

  return executeOperation(this.s.topology, profilingInfo, [this, options, callback]);
***REMOVED***, 'Db.profilingInfo is deprecated. Query the system.profile collection directly.');

/**
 * Retrieve the current profiling Level for MongoDB
 *
 * @param {Object***REMOVED*** [options] Optional settings
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The command result callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.profilingLevel = function(options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;

  return executeOperation(this.s.topology, profilingLevel, [this, options, callback]);
***REMOVED***;

/**
 * Retrieves this collections index info.
 * @method
 * @param {string***REMOVED*** name The name of the collection.
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {boolean***REMOVED*** [options.full=false] Returns the full raw index information.
 * @param {(ReadPreference|string)***REMOVED*** [options.readPreference] The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST).
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param {Db~resultCallback***REMOVED*** [callback] The command result callback
 * @return {Promise***REMOVED*** returns Promise if no callback passed
 */
Db.prototype.indexInformation = function(name, options, callback) {
  if (typeof options === 'function') (callback = options), (options = {***REMOVED***);
  options = options || {***REMOVED***;

  return executeOperation(this.s.topology, indexInformation, [this, name, options, callback]);
***REMOVED***;

/**
 * Unref all sockets
 * @method
 */
Db.prototype.unref = function() {
  this.s.topology.unref();
***REMOVED***;

/**
 * Create a new Change Stream, watching for new changes (insertions, updates, replacements, deletions, and invalidations) in this database. Will ignore all changes to system collections.
 * @method
 * @since 3.1.0
 * @param {Array***REMOVED*** [pipeline] An array of {@link https://docs.mongodb.com/manual/reference/operator/aggregation-pipeline/|aggregation pipeline stages***REMOVED*** through which to pass change stream documents. This allows for filtering (using $match) and manipulating the change stream documents.
 * @param {object***REMOVED*** [options] Optional settings
 * @param {string***REMOVED*** [options.fullDocument='default'] Allowed values: ‘default’, ‘updateLookup’. When set to ‘updateLookup’, the change stream will include both a delta describing the changes to the document, as well as a copy of the entire document that was changed from some time after the change occurred.
 * @param {object***REMOVED*** [options.resumeAfter] Specifies the logical starting point for the new change stream. This should be the _id field from a previously returned change stream document.
 * @param {number***REMOVED*** [options.maxAwaitTimeMS] The maximum amount of time for the server to wait on new documents to satisfy a change stream query
 * @param {number***REMOVED*** [options.batchSize] The number of documents to return per batch. See {@link https://docs.mongodb.com/manual/reference/command/aggregate|aggregation documentation***REMOVED***.
 * @param {object***REMOVED*** [options.collation] Specify collation settings for operation. See {@link https://docs.mongodb.com/manual/reference/command/aggregate|aggregation documentation***REMOVED***.
 * @param {ReadPreference***REMOVED*** [options.readPreference] The read preference. Defaults to the read preference of the database. See {@link https://docs.mongodb.com/manual/reference/read-preference|read preference documentation***REMOVED***.
 * @param {Timestamp***REMOVED*** [options.startAtClusterTime] receive change events that occur after the specified timestamp
 * @param {ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @return {ChangeStream***REMOVED*** a ChangeStream instance.
 */
Db.prototype.watch = function(pipeline, options) {
  pipeline = pipeline || [];
  options = options || {***REMOVED***;

  // Allow optionally not specifying a pipeline
  if (!Array.isArray(pipeline)) {
    options = pipeline;
    pipeline = [];
  ***REMOVED***

  return new ChangeStream(this, pipeline, options);
***REMOVED***;

/**
 * Return the db logger
 * @method
 * @return {Logger***REMOVED*** return the db logger
 * @ignore
 */
Db.prototype.getLogger = function() {
  return this.s.logger;
***REMOVED***;

/**
 * Db close event
 *
 * Emitted after a socket closed against a single server or mongos proxy.
 *
 * @event Db#close
 * @type {MongoError***REMOVED***
 */

/**
 * Db reconnect event
 *
 *  * Server: Emitted when the driver has reconnected and re-authenticated.
 *  * ReplicaSet: N/A
 *  * Mongos: Emitted when the driver reconnects and re-authenticates successfully against a Mongos.
 *
 * @event Db#reconnect
 * @type {object***REMOVED***
 */

/**
 * Db error event
 *
 * Emitted after an error occurred against a single server or mongos proxy.
 *
 * @event Db#error
 * @type {MongoError***REMOVED***
 */

/**
 * Db timeout event
 *
 * Emitted after a socket timeout occurred against a single server or mongos proxy.
 *
 * @event Db#timeout
 * @type {MongoError***REMOVED***
 */

/**
 * Db parseError event
 *
 * The parseError event is emitted if the driver detects illegal or corrupt BSON being received from the server.
 *
 * @event Db#parseError
 * @type {MongoError***REMOVED***
 */

/**
 * Db fullsetup event, emitted when all servers in the topology have been connected to at start up time.
 *
 * * Server: Emitted when the driver has connected to the single server and has authenticated.
 * * ReplSet: Emitted after the driver has attempted to connect to all replicaset members.
 * * Mongos: Emitted after the driver has attempted to connect to all mongos proxies.
 *
 * @event Db#fullsetup
 * @type {Db***REMOVED***
 */

// Constants
Db.SYSTEM_NAMESPACE_COLLECTION = 'system.namespaces';
Db.SYSTEM_INDEX_COLLECTION = 'system.indexes';
Db.SYSTEM_PROFILE_COLLECTION = 'system.profile';
Db.SYSTEM_USER_COLLECTION = 'system.users';
Db.SYSTEM_COMMAND_COLLECTION = '$cmd';
Db.SYSTEM_JS_COLLECTION = 'system.js';

module.exports = Db;
