// mongodb-core/lib/connection/logger.js
'use strict';

var f = require('util').format,
  MongoError = require('../error').MongoError;

// Filters for classes
var classFilters = {***REMOVED***;
var filteredClasses = {***REMOVED***;
var level = null;
// Save the process id
var pid = process.pid;
// current logger
var currentLogger = null;

/**
 * Creates a new Logger instance
 * @class
 * @param {string***REMOVED*** className The Class name associated with the logging instance
 * @param {object***REMOVED*** [options=null] Optional settings.
 * @param {Function***REMOVED*** [options.logger=null] Custom logger function;
 * @param {string***REMOVED*** [options.loggerLevel=error] Override default global log level.
 * @return {Logger***REMOVED*** a Logger instance.
 */
var Logger = function(className, options) {
  if (!(this instanceof Logger)) return new Logger(className, options);
  options = options || {***REMOVED***;

  // Current reference
  this.className = className;

  // Current logger
  if (options.logger) {
    currentLogger = options.logger;
  ***REMOVED*** else if (currentLogger == null) {
    currentLogger = console.log;
  ***REMOVED***

  // Set level of logging, default is error
  if (options.loggerLevel) {
    level = options.loggerLevel || 'error';
  ***REMOVED***

  // Add all class names
  if (filteredClasses[this.className] == null) classFilters[this.className] = true;
***REMOVED***;

/**
 * Log a message at the debug level
 * @method
 * @param {string***REMOVED*** message The message to log
 * @param {object***REMOVED*** object additional meta data to log
 * @return {null***REMOVED***
 */
Logger.prototype.debug = function(message, object) {
  if (
    this.isDebug() &&
    ((Object.keys(filteredClasses).length > 0 && filteredClasses[this.className]) ||
      (Object.keys(filteredClasses).length === 0 && classFilters[this.className]))
  ) {
    var dateTime = new Date().getTime();
    var msg = f('[%s-%s:%s] %s %s', 'DEBUG', this.className, pid, dateTime, message);
    var state = {
      type: 'debug',
      message: message,
      className: this.className,
      pid: pid,
      date: dateTime
    ***REMOVED***;
    if (object) state.meta = object;
    currentLogger(msg, state);
  ***REMOVED***
***REMOVED***;

/**
 * Log a message at the warn level
 * @method
 * @param {string***REMOVED*** message The message to log
 * @param {object***REMOVED*** object additional meta data to log
 * @return {null***REMOVED***
 */
(Logger.prototype.warn = function(message, object) {
  if (
    this.isWarn() &&
    ((Object.keys(filteredClasses).length > 0 && filteredClasses[this.className]) ||
      (Object.keys(filteredClasses).length === 0 && classFilters[this.className]))
  ) {
    var dateTime = new Date().getTime();
    var msg = f('[%s-%s:%s] %s %s', 'WARN', this.className, pid, dateTime, message);
    var state = {
      type: 'warn',
      message: message,
      className: this.className,
      pid: pid,
      date: dateTime
    ***REMOVED***;
    if (object) state.meta = object;
    currentLogger(msg, state);
  ***REMOVED***
***REMOVED***),
  /**
   * Log a message at the info level
   * @method
   * @param {string***REMOVED*** message The message to log
   * @param {object***REMOVED*** object additional meta data to log
   * @return {null***REMOVED***
   */
  (Logger.prototype.info = function(message, object) {
    if (
      this.isInfo() &&
      ((Object.keys(filteredClasses).length > 0 && filteredClasses[this.className]) ||
        (Object.keys(filteredClasses).length === 0 && classFilters[this.className]))
    ) {
      var dateTime = new Date().getTime();
      var msg = f('[%s-%s:%s] %s %s', 'INFO', this.className, pid, dateTime, message);
      var state = {
        type: 'info',
        message: message,
        className: this.className,
        pid: pid,
        date: dateTime
      ***REMOVED***;
      if (object) state.meta = object;
      currentLogger(msg, state);
    ***REMOVED***
  ***REMOVED***),
  /**
   * Log a message at the error level
   * @method
   * @param {string***REMOVED*** message The message to log
   * @param {object***REMOVED*** object additional meta data to log
   * @return {null***REMOVED***
   */
  (Logger.prototype.error = function(message, object) {
    if (
      this.isError() &&
      ((Object.keys(filteredClasses).length > 0 && filteredClasses[this.className]) ||
        (Object.keys(filteredClasses).length === 0 && classFilters[this.className]))
    ) {
      var dateTime = new Date().getTime();
      var msg = f('[%s-%s:%s] %s %s', 'ERROR', this.className, pid, dateTime, message);
      var state = {
        type: 'error',
        message: message,
        className: this.className,
        pid: pid,
        date: dateTime
      ***REMOVED***;
      if (object) state.meta = object;
      currentLogger(msg, state);
    ***REMOVED***
  ***REMOVED***),
  /**
   * Is the logger set at info level
   * @method
   * @return {boolean***REMOVED***
   */
  (Logger.prototype.isInfo = function() {
    return level === 'info' || level === 'debug';
  ***REMOVED***),
  /**
   * Is the logger set at error level
   * @method
   * @return {boolean***REMOVED***
   */
  (Logger.prototype.isError = function() {
    return level === 'error' || level === 'info' || level === 'debug';
  ***REMOVED***),
  /**
   * Is the logger set at error level
   * @method
   * @return {boolean***REMOVED***
   */
  (Logger.prototype.isWarn = function() {
    return level === 'error' || level === 'warn' || level === 'info' || level === 'debug';
  ***REMOVED***),
  /**
   * Is the logger set at debug level
   * @method
   * @return {boolean***REMOVED***
   */
  (Logger.prototype.isDebug = function() {
    return level === 'debug';
  ***REMOVED***);

/**
 * Resets the logger to default settings, error and no filtered classes
 * @method
 * @return {null***REMOVED***
 */
Logger.reset = function() {
  level = 'error';
  filteredClasses = {***REMOVED***;
***REMOVED***;

/**
 * Get the current logger function
 * @method
 * @return {function***REMOVED***
 */
Logger.currentLogger = function() {
  return currentLogger;
***REMOVED***;

/**
 * Set the current logger function
 * @method
 * @param {function***REMOVED*** logger Logger function.
 * @return {null***REMOVED***
 */
Logger.setCurrentLogger = function(logger) {
  if (typeof logger !== 'function') throw new MongoError('current logger must be a function');
  currentLogger = logger;
***REMOVED***;

/**
 * Set what classes to log.
 * @method
 * @param {string***REMOVED*** type The type of filter (currently only class)
 * @param {string[]***REMOVED*** values The filters to apply
 * @return {null***REMOVED***
 */
Logger.filter = function(type, values) {
  if (type === 'class' && Array.isArray(values)) {
    filteredClasses = {***REMOVED***;

    values.forEach(function(x) {
      filteredClasses[x] = true;
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;

/**
 * Set the current log level
 * @method
 * @param {string***REMOVED*** level Set current log level (debug, info, error)
 * @return {null***REMOVED***
 */
Logger.setLevel = function(_level) {
  if (_level !== 'info' && _level !== 'error' && _level !== 'debug' && _level !== 'warn') {
    throw new Error(f('%s is an illegal logging level', _level));
  ***REMOVED***

  level = _level;
***REMOVED***;

module.exports = Logger;
