// mongodb/lib/gridfs-stream/index.js
'use strict';

var Emitter = require('events').EventEmitter;
var GridFSBucketReadStream = require('./download');
var GridFSBucketWriteStream = require('./upload');
var shallowClone = require('../utils').shallowClone;
var toError = require('../utils').toError;
var util = require('util');
var executeOperation = require('../utils').executeOperation;

var DEFAULT_GRIDFS_BUCKET_OPTIONS = {
  bucketName: 'fs',
  chunkSizeBytes: 255 * 1024
***REMOVED***;

module.exports = GridFSBucket;

/**
 * Constructor for a streaming GridFS interface
 * @class
 * @param {Db***REMOVED*** db A db handle
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {string***REMOVED*** [options.bucketName="fs"] The 'files' and 'chunks' collections will be prefixed with the bucket name followed by a dot.
 * @param {number***REMOVED*** [options.chunkSizeBytes=255 * 1024] Number of bytes stored in each chunk. Defaults to 255KB
 * @param {object***REMOVED*** [options.writeConcern] Optional write concern to be passed to write operations, for instance `{ w: 1 ***REMOVED***`
 * @param {object***REMOVED*** [options.readPreference] Optional read preference to be passed to read operations
 * @fires GridFSBucketWriteStream#index
 * @return {GridFSBucket***REMOVED***
 */

function GridFSBucket(db, options) {
  Emitter.apply(this);
  this.setMaxListeners(0);

  if (options && typeof options === 'object') {
    options = shallowClone(options);
    var keys = Object.keys(DEFAULT_GRIDFS_BUCKET_OPTIONS);
    for (var i = 0; i < keys.length; ++i) {
      if (!options[keys[i]]) {
        options[keys[i]] = DEFAULT_GRIDFS_BUCKET_OPTIONS[keys[i]];
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
    options = DEFAULT_GRIDFS_BUCKET_OPTIONS;
  ***REMOVED***

  this.s = {
    db: db,
    options: options,
    _chunksCollection: db.collection(options.bucketName + '.chunks'),
    _filesCollection: db.collection(options.bucketName + '.files'),
    checkedIndexes: false,
    calledOpenUploadStream: false,
    promiseLibrary: db.s.promiseLibrary || Promise
  ***REMOVED***;
***REMOVED***

util.inherits(GridFSBucket, Emitter);

/**
 * When the first call to openUploadStream is made, the upload stream will
 * check to see if it needs to create the proper indexes on the chunks and
 * files collections. This event is fired either when 1) it determines that
 * no index creation is necessary, 2) when it successfully creates the
 * necessary indexes.
 *
 * @event GridFSBucket#index
 * @type {Error***REMOVED***
 */

/**
 * Returns a writable stream (GridFSBucketWriteStream) for writing
 * buffers to GridFS. The stream's 'id' property contains the resulting
 * file's id.
 * @method
 * @param {string***REMOVED*** filename The value of the 'filename' key in the files doc
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {number***REMOVED*** [options.chunkSizeBytes] Optional overwrite this bucket's chunkSizeBytes for this file
 * @param {object***REMOVED*** [options.metadata] Optional object to store in the file document's `metadata` field
 * @param {string***REMOVED*** [options.contentType] Optional string to store in the file document's `contentType` field
 * @param {array***REMOVED*** [options.aliases] Optional array of strings to store in the file document's `aliases` field
 * @param {boolean***REMOVED*** [options.disableMD5=false] If true, disables adding an md5 field to file data
 * @return {GridFSBucketWriteStream***REMOVED***
 */

GridFSBucket.prototype.openUploadStream = function(filename, options) {
  if (options) {
    options = shallowClone(options);
  ***REMOVED***
    options = {***REMOVED***;
  ***REMOVED***
  if (!options.chunkSizeBytes) {
    options.chunkSizeBytes = this.s.options.chunkSizeBytes;
  ***REMOVED***
  return new GridFSBucketWriteStream(this, filename, options);
***REMOVED***;

/**
 * Returns a writable stream (GridFSBucketWriteStream) for writing
 * buffers to GridFS for a custom file id. The stream's 'id' property contains the resulting
 * file's id.
 * @method
 * @param {string|number|object***REMOVED*** id A custom id used to identify the file
 * @param {string***REMOVED*** filename The value of the 'filename' key in the files doc
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {number***REMOVED*** [options.chunkSizeBytes] Optional overwrite this bucket's chunkSizeBytes for this file
 * @param {object***REMOVED*** [options.metadata] Optional object to store in the file document's `metadata` field
 * @param {string***REMOVED*** [options.contentType] Optional string to store in the file document's `contentType` field
 * @param {array***REMOVED*** [options.aliases] Optional array of strings to store in the file document's `aliases` field
 * @param {boolean***REMOVED*** [options.disableMD5=false] If true, disables adding an md5 field to file data
 * @return {GridFSBucketWriteStream***REMOVED***
 */

GridFSBucket.prototype.openUploadStreamWithId = function(id, filename, options) {
  if (options) {
    options = shallowClone(options);
  ***REMOVED***
    options = {***REMOVED***;
  ***REMOVED***

  if (!options.chunkSizeBytes) {
    options.chunkSizeBytes = this.s.options.chunkSizeBytes;
  ***REMOVED***

  options.id = id;

  return new GridFSBucketWriteStream(this, filename, options);
***REMOVED***;

/**
 * Returns a readable stream (GridFSBucketReadStream) for streaming file
 * data from GridFS.
 * @method
 * @param {ObjectId***REMOVED*** id The id of the file doc
 * @param {Object***REMOVED*** [options] Optional settings.
 * @param {Number***REMOVED*** [options.start] Optional 0-based offset in bytes to start streaming from
 * @param {Number***REMOVED*** [options.end] Optional 0-based offset in bytes to stop streaming before
 * @return {GridFSBucketReadStream***REMOVED***
 */

GridFSBucket.prototype.openDownloadStream = function(id, options) {
  var filter = { _id: id ***REMOVED***;
  options = {
    start: options && options.start,
    end: options && options.end
  ***REMOVED***;

  return new GridFSBucketReadStream(
    this.s._chunksCollection,
    this.s._filesCollection,
    this.s.options.readPreference,
    filter,
    options
  );
***REMOVED***;

/**
 * Deletes a file with the given id
 * @method
 * @param {ObjectId***REMOVED*** id The id of the file doc
 * @param {GridFSBucket~errorCallback***REMOVED*** [callback]
 */

GridFSBucket.prototype.delete = function(id, callback) {
  return executeOperation(this.s.db.s.topology, _delete, [this, id, callback], {
    skipSessions: true
  ***REMOVED***);
***REMOVED***;

/**
 * @ignore
 */

function _delete(_this, id, callback) {
  _this.s._filesCollection.deleteOne({ _id: id ***REMOVED***, function(error, res) {
    if (error) {
      return callback(error);
    ***REMOVED***

    _this.s._chunksCollection.deleteMany({ files_id: id ***REMOVED***, function(error) {
      if (error) {
        return callback(error);
      ***REMOVED***

      // Delete orphaned chunks before returning FileNotFound
      if (!res.result.n) {
        var errmsg = 'FileNotFound: no file with id ' + id + ' found';
        return callback(new Error(errmsg));
      ***REMOVED***

      callback();
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

/**
 * Convenience wrapper around find on the files collection
 * @method
 * @param {Object***REMOVED*** filter
 * @param {Object***REMOVED*** [options] Optional settings for cursor
 * @param {number***REMOVED*** [options.batchSize] Optional batch size for cursor
 * @param {number***REMOVED*** [options.limit] Optional limit for cursor
 * @param {number***REMOVED*** [options.maxTimeMS] Optional maxTimeMS for cursor
 * @param {boolean***REMOVED*** [options.noCursorTimeout] Optionally set cursor's `noCursorTimeout` flag
 * @param {number***REMOVED*** [options.skip] Optional skip for cursor
 * @param {object***REMOVED*** [options.sort] Optional sort for cursor
 * @return {Cursor***REMOVED***
 */

GridFSBucket.prototype.find = function(filter, options) {
  filter = filter || {***REMOVED***;
  options = options || {***REMOVED***;

  var cursor = this.s._filesCollection.find(filter);

  if (options.batchSize != null) {
    cursor.batchSize(options.batchSize);
  ***REMOVED***
  if (options.limit != null) {
    cursor.limit(options.limit);
  ***REMOVED***
  if (options.maxTimeMS != null) {
    cursor.maxTimeMS(options.maxTimeMS);
  ***REMOVED***
  if (options.noCursorTimeout != null) {
    cursor.addCursorFlag('noCursorTimeout', options.noCursorTimeout);
  ***REMOVED***
  if (options.skip != null) {
    cursor.skip(options.skip);
  ***REMOVED***
  if (options.sort != null) {
    cursor.sort(options.sort);
  ***REMOVED***

  return cursor;
***REMOVED***;

/**
 * Returns a readable stream (GridFSBucketReadStream) for streaming the
 * file with the given name from GridFS. If there are multiple files with
 * the same name, this will stream the most recent file with the given name
 * (as determined by the `uploadDate` field). You can set the `revision`
 * option to change this behavior.
 * @method
 * @param {String***REMOVED*** filename The name of the file to stream
 * @param {Object***REMOVED*** [options] Optional settings
 * @param {number***REMOVED*** [options.revision=-1] The revision number relative to the oldest file with the given filename. 0 gets you the oldest file, 1 gets you the 2nd oldest, -1 gets you the newest.
 * @param {Number***REMOVED*** [options.start] Optional 0-based offset in bytes to start streaming from
 * @param {Number***REMOVED*** [options.end] Optional 0-based offset in bytes to stop streaming before
 * @return {GridFSBucketReadStream***REMOVED***
 */

GridFSBucket.prototype.openDownloadStreamByName = function(filename, options) {
  var sort = { uploadDate: -1 ***REMOVED***;
  var skip = null;
  if (options && options.revision != null) {
    if (options.revision >= 0) {
      sort = { uploadDate: 1 ***REMOVED***;
      skip = options.revision;
    ***REMOVED***
      skip = -options.revision - 1;
    ***REMOVED***
  ***REMOVED***

  var filter = { filename: filename ***REMOVED***;
  options = {
    sort: sort,
    skip: skip,
    start: options && options.start,
    end: options && options.end
  ***REMOVED***;
  return new GridFSBucketReadStream(
    this.s._chunksCollection,
    this.s._filesCollection,
    this.s.options.readPreference,
    filter,
    options
  );
***REMOVED***;

/**
 * Renames the file with the given _id to the given string
 * @method
 * @param {ObjectId***REMOVED*** id the id of the file to rename
 * @param {String***REMOVED*** filename new name for the file
 * @param {GridFSBucket~errorCallback***REMOVED*** [callback]
 */

GridFSBucket.prototype.rename = function(id, filename, callback) {
  return executeOperation(this.s.db.s.topology, _rename, [this, id, filename, callback], {
    skipSessions: true
  ***REMOVED***);
***REMOVED***;

/**
 * @ignore
 */

function _rename(_this, id, filename, callback) {
  var filter = { _id: id ***REMOVED***;
  var update = { $set: { filename: filename ***REMOVED*** ***REMOVED***;
  _this.s._filesCollection.updateOne(filter, update, function(error, res) {
    if (error) {
      return callback(error);
    ***REMOVED***
    if (!res.result.n) {
      return callback(toError('File with id ' + id + ' not found'));
    ***REMOVED***
    callback();
  ***REMOVED***);
***REMOVED***

/**
 * Removes this bucket's files collection, followed by its chunks collection.
 * @method
 * @param {GridFSBucket~errorCallback***REMOVED*** [callback]
 */

GridFSBucket.prototype.drop = function(callback) {
  return executeOperation(this.s.db.s.topology, _drop, [this, callback], {
    skipSessions: true
  ***REMOVED***);
***REMOVED***;

/**
 * Return the db logger
 * @method
 * @return {Logger***REMOVED*** return the db logger
 * @ignore
 */
GridFSBucket.prototype.getLogger = function() {
  return this.s.db.s.logger;
***REMOVED***;

/**
 * @ignore
 */

function _drop(_this, callback) {
  _this.s._filesCollection.drop(function(error) {
    if (error) {
      return callback(error);
    ***REMOVED***
    _this.s._chunksCollection.drop(function(error) {
      if (error) {
        return callback(error);
      ***REMOVED***

      return callback();
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

/**
 * Callback format for all GridFSBucket methods that can accept a callback.
 * @callback GridFSBucket~errorCallback
 * @param {MongoError***REMOVED*** error An error instance representing any errors that occurred
 */
