// mongodb/lib/gridfs-stream/upload.js
'use strict';

var core = require('mongodb-core');
var crypto = require('crypto');
var stream = require('stream');
var util = require('util');
var Buffer = require('safe-buffer').Buffer;

var ERROR_NAMESPACE_NOT_FOUND = 26;

module.exports = GridFSBucketWriteStream;

/**
 * A writable stream that enables you to write buffers to GridFS.
 *
 * Do not instantiate this class directly. Use `openUploadStream()` instead.
 *
 * @class
 * @param {GridFSBucket***REMOVED*** bucket Handle for this stream's corresponding bucket
 * @param {string***REMOVED*** filename The value of the 'filename' key in the files doc
 * @param {object***REMOVED*** [options] Optional settings.
 * @param {string|number|object***REMOVED*** [options.id] Custom file id for the GridFS file.
 * @param {number***REMOVED*** [options.chunkSizeBytes] The chunk size to use, in bytes
 * @param {number***REMOVED*** [options.w] The write concern
 * @param {number***REMOVED*** [options.wtimeout] The write concern timeout
 * @param {number***REMOVED*** [options.j] The journal write concern
 * @param {boolean***REMOVED*** [options.disableMD5=false] If true, disables adding an md5 field to file data
 * @fires GridFSBucketWriteStream#error
 * @fires GridFSBucketWriteStream#finish
 * @return {GridFSBucketWriteStream***REMOVED*** a GridFSBucketWriteStream instance.
 */

function GridFSBucketWriteStream(bucket, filename, options) {
  options = options || {***REMOVED***;
  this.bucket = bucket;
  this.chunks = bucket.s._chunksCollection;
  this.filename = filename;
  this.files = bucket.s._filesCollection;
  this.options = options;
  // Signals the write is all done
  this.done = false;

  this.id = options.id ? options.id : core.BSON.ObjectId();
  this.chunkSizeBytes = this.options.chunkSizeBytes;
  this.bufToStore = Buffer.alloc(this.chunkSizeBytes);
  this.length = 0;
  this.md5 = !options.disableMD5 && crypto.createHash('md5');
  this.n = 0;
  this.pos = 0;
  this.state = {
    streamEnd: false,
    outstandingRequests: 0,
    errored: false,
    aborted: false,
    promiseLibrary: this.bucket.s.promiseLibrary
  ***REMOVED***;

  if (!this.bucket.s.calledOpenUploadStream) {
    this.bucket.s.calledOpenUploadStream = true;

    var _this = this;
    checkIndexes(this, function() {
      _this.bucket.s.checkedIndexes = true;
      _this.bucket.emit('index');
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

util.inherits(GridFSBucketWriteStream, stream.Writable);

/**
 * An error occurred
 *
 * @event GridFSBucketWriteStream#error
 * @type {Error***REMOVED***
 */

/**
 * `end()` was called and the write stream successfully wrote the file
 * metadata and all the chunks to MongoDB.
 *
 * @event GridFSBucketWriteStream#finish
 * @type {object***REMOVED***
 */

/**
 * Write a buffer to the stream.
 *
 * @method
 * @param {Buffer***REMOVED*** chunk Buffer to write
 * @param {String***REMOVED*** encoding Optional encoding for the buffer
 * @param {Function***REMOVED*** callback Function to call when the chunk was added to the buffer, or if the entire chunk was persisted to MongoDB if this chunk caused a flush.
 * @return {Boolean***REMOVED*** False if this write required flushing a chunk to MongoDB. True otherwise.
 */

GridFSBucketWriteStream.prototype.write = function(chunk, encoding, callback) {
  var _this = this;
  return waitForIndexes(this, function() {
    return doWrite(_this, chunk, encoding, callback);
  ***REMOVED***);
***REMOVED***;

/**
 * Places this write stream into an aborted state (all future writes fail)
 * and deletes all chunks that have already been written.
 *
 * @method
 * @param {GridFSBucket~errorCallback***REMOVED*** callback called when chunks are successfully removed or error occurred
 * @return {Promise***REMOVED*** if no callback specified
 */

GridFSBucketWriteStream.prototype.abort = function(callback) {
  if (this.state.streamEnd) {
    var error = new Error('Cannot abort a stream that has already completed');
    if (typeof callback === 'function') {
      return callback(error);
    ***REMOVED***
    return this.state.promiseLibrary.reject(error);
  ***REMOVED***
  if (this.state.aborted) {
    error = new Error('Cannot call abort() on a stream twice');
    if (typeof callback === 'function') {
      return callback(error);
    ***REMOVED***
    return this.state.promiseLibrary.reject(error);
  ***REMOVED***
  this.state.aborted = true;
  this.chunks.deleteMany({ files_id: this.id ***REMOVED***, function(error) {
    if (typeof callback === 'function') callback(error);
  ***REMOVED***);
***REMOVED***;

/**
 * Tells the stream that no more data will be coming in. The stream will
 * persist the remaining data to MongoDB, write the files document, and
 * then emit a 'finish' event.
 *
 * @method
 * @param {Buffer***REMOVED*** chunk Buffer to write
 * @param {String***REMOVED*** encoding Optional encoding for the buffer
 * @param {Function***REMOVED*** callback Function to call when all files and chunks have been persisted to MongoDB
 */

GridFSBucketWriteStream.prototype.end = function(chunk, encoding, callback) {
  var _this = this;
  if (typeof chunk === 'function') {
    (callback = chunk), (chunk = null), (encoding = null);
  ***REMOVED*** else if (typeof encoding === 'function') {
    (callback = encoding), (encoding = null);
  ***REMOVED***

  if (checkAborted(this, callback)) {
    return;
  ***REMOVED***
  this.state.streamEnd = true;

  if (callback) {
    this.once('finish', function(result) {
      callback(null, result);
    ***REMOVED***);
  ***REMOVED***

  if (!chunk) {
    waitForIndexes(this, function() {
      writeRemnant(_this);
    ***REMOVED***);
    return;
  ***REMOVED***

  this.write(chunk, encoding, function() {
    writeRemnant(_this);
  ***REMOVED***);
***REMOVED***;

/**
 * @ignore
 */

function __handleError(_this, error, callback) {
  if (_this.state.errored) {
    return;
  ***REMOVED***
  _this.state.errored = true;
  if (callback) {
    return callback(error);
  ***REMOVED***
  _this.emit('error', error);
***REMOVED***

/**
 * @ignore
 */

function createChunkDoc(filesId, n, data) {
  return {
    _id: core.BSON.ObjectId(),
    files_id: filesId,
    n: n,
    data: data
  ***REMOVED***;
***REMOVED***

/**
 * @ignore
 */

function checkChunksIndex(_this, callback) {
  _this.chunks.listIndexes().toArray(function(error, indexes) {
    if (error) {
      // Collection doesn't exist so create index
      if (error.code === ERROR_NAMESPACE_NOT_FOUND) {
        var index = { files_id: 1, n: 1 ***REMOVED***;
        _this.chunks.createIndex(index, { background: false, unique: true ***REMOVED***, function(error) {
          if (error) {
            return callback(error);
          ***REMOVED***

          callback();
        ***REMOVED***);
        return;
      ***REMOVED***
      return callback(error);
    ***REMOVED***

    var hasChunksIndex = false;
    indexes.forEach(function(index) {
      if (index.key) {
        var keys = Object.keys(index.key);
        if (keys.length === 2 && index.key.files_id === 1 && index.key.n === 1) {
          hasChunksIndex = true;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***);

    if (hasChunksIndex) {
      callback();
    ***REMOVED***
      index = { files_id: 1, n: 1 ***REMOVED***;
      var indexOptions = getWriteOptions(_this);

      indexOptions.background = false;
      indexOptions.unique = true;

      _this.chunks.createIndex(index, indexOptions, function(error) {
        if (error) {
          return callback(error);
        ***REMOVED***

        callback();
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

/**
 * @ignore
 */

function checkDone(_this, callback) {
  if (_this.done) return true;
  if (_this.state.streamEnd && _this.state.outstandingRequests === 0 && !_this.state.errored) {
    // Set done so we dont' trigger duplicate createFilesDoc
    _this.done = true;
    // Create a new files doc
    var filesDoc = createFilesDoc(
      _this.id,
      _this.length,
      _this.chunkSizeBytes,
      _this.md5 && _this.md5.digest('hex'),
      _this.filename,
      _this.options.contentType,
      _this.options.aliases,
      _this.options.metadata
    );

    if (checkAborted(_this, callback)) {
      return false;
    ***REMOVED***

    _this.files.insertOne(filesDoc, getWriteOptions(_this), function(error) {
      if (error) {
        return __handleError(_this, error, callback);
      ***REMOVED***
      _this.emit('finish', filesDoc);
    ***REMOVED***);

    return true;
  ***REMOVED***

  return false;
***REMOVED***

/**
 * @ignore
 */

function checkIndexes(_this, callback) {
  _this.files.findOne({***REMOVED***, { _id: 1 ***REMOVED***, function(error, doc) {
    if (error) {
      return callback(error);
    ***REMOVED***
    if (doc) {
      return callback();
    ***REMOVED***

    _this.files.listIndexes().toArray(function(error, indexes) {
      if (error) {
        // Collection doesn't exist so create index
        if (error.code === ERROR_NAMESPACE_NOT_FOUND) {
          var index = { filename: 1, uploadDate: 1 ***REMOVED***;
          _this.files.createIndex(index, { background: false ***REMOVED***, function(error) {
            if (error) {
              return callback(error);
            ***REMOVED***

            checkChunksIndex(_this, callback);
          ***REMOVED***);
          return;
        ***REMOVED***
        return callback(error);
      ***REMOVED***

      var hasFileIndex = false;
      indexes.forEach(function(index) {
        var keys = Object.keys(index.key);
        if (keys.length === 2 && index.key.filename === 1 && index.key.uploadDate === 1) {
          hasFileIndex = true;
        ***REMOVED***
      ***REMOVED***);

      if (hasFileIndex) {
        checkChunksIndex(_this, callback);
      ***REMOVED***
        index = { filename: 1, uploadDate: 1 ***REMOVED***;

        var indexOptions = getWriteOptions(_this);

        indexOptions.background = false;

        _this.files.createIndex(index, indexOptions, function(error) {
          if (error) {
            return callback(error);
          ***REMOVED***

          checkChunksIndex(_this, callback);
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

/**
 * @ignore
 */

function createFilesDoc(_id, length, chunkSize, md5, filename, contentType, aliases, metadata) {
  var ret = {
    _id: _id,
    length: length,
    chunkSize: chunkSize,
    uploadDate: new Date(),
    filename: filename
  ***REMOVED***;

  if (md5) {
    ret.md5 = md5;
  ***REMOVED***

  if (contentType) {
    ret.contentType = contentType;
  ***REMOVED***

  if (aliases) {
    ret.aliases = aliases;
  ***REMOVED***

  if (metadata) {
    ret.metadata = metadata;
  ***REMOVED***

  return ret;
***REMOVED***

/**
 * @ignore
 */

function doWrite(_this, chunk, encoding, callback) {
  if (checkAborted(_this, callback)) {
    return false;
  ***REMOVED***

  var inputBuf = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk, encoding);

  _this.length += inputBuf.length;

  // Input is small enough to fit in our buffer
  if (_this.pos + inputBuf.length < _this.chunkSizeBytes) {
    inputBuf.copy(_this.bufToStore, _this.pos);
    _this.pos += inputBuf.length;

    callback && callback();

    // Note that we reverse the typical semantics of write's return value
    // to be compatible with node's `.pipe()` function.
    // True means client can keep writing.
    return true;
  ***REMOVED***

  // Otherwise, buffer is too big for current chunk, so we need to flush
  // to MongoDB.
  var inputBufRemaining = inputBuf.length;
  var spaceRemaining = _this.chunkSizeBytes - _this.pos;
  var numToCopy = Math.min(spaceRemaining, inputBuf.length);
  var outstandingRequests = 0;
  while (inputBufRemaining > 0) {
    var inputBufPos = inputBuf.length - inputBufRemaining;
    inputBuf.copy(_this.bufToStore, _this.pos, inputBufPos, inputBufPos + numToCopy);
    _this.pos += numToCopy;
    spaceRemaining -= numToCopy;
    if (spaceRemaining === 0) {
      if (_this.md5) {
        _this.md5.update(_this.bufToStore);
      ***REMOVED***
      var doc = createChunkDoc(_this.id, _this.n, _this.bufToStore);
      ++_this.state.outstandingRequests;
      ++outstandingRequests;

      if (checkAborted(_this, callback)) {
        return false;
      ***REMOVED***

      _this.chunks.insertOne(doc, getWriteOptions(_this), function(error) {
        if (error) {
          return __handleError(_this, error);
        ***REMOVED***
        --_this.state.outstandingRequests;
        --outstandingRequests;

        if (!outstandingRequests) {
          _this.emit('drain', doc);
          callback && callback();
          checkDone(_this);
        ***REMOVED***
      ***REMOVED***);

      spaceRemaining = _this.chunkSizeBytes;
      _this.pos = 0;
      ++_this.n;
    ***REMOVED***
    inputBufRemaining -= numToCopy;
    numToCopy = Math.min(spaceRemaining, inputBufRemaining);
  ***REMOVED***

  // Note that we reverse the typical semantics of write's return value
  // to be compatible with node's `.pipe()` function.
  // False means the client should wait for the 'drain' event.
  return false;
***REMOVED***

/**
 * @ignore
 */

function getWriteOptions(_this) {
  var obj = {***REMOVED***;
  if (_this.options.writeConcern) {
    obj.w = _this.options.writeConcern.w;
    obj.wtimeout = _this.options.writeConcern.wtimeout;
    obj.j = _this.options.writeConcern.j;
  ***REMOVED***
  return obj;
***REMOVED***

/**
 * @ignore
 */

function waitForIndexes(_this, callback) {
  if (_this.bucket.s.checkedIndexes) {
    return callback(false);
  ***REMOVED***

  _this.bucket.once('index', function() {
    callback(true);
  ***REMOVED***);

  return true;
***REMOVED***

/**
 * @ignore
 */

function writeRemnant(_this, callback) {
  // Buffer is empty, so don't bother to insert
  if (_this.pos === 0) {
    return checkDone(_this, callback);
  ***REMOVED***

  ++_this.state.outstandingRequests;

  // Create a new buffer to make sure the buffer isn't bigger than it needs
  // to be.
  var remnant = Buffer.alloc(_this.pos);
  _this.bufToStore.copy(remnant, 0, 0, _this.pos);
  if (_this.md5) {
    _this.md5.update(remnant);
  ***REMOVED***
  var doc = createChunkDoc(_this.id, _this.n, remnant);

  // If the stream was aborted, do not write remnant
  if (checkAborted(_this, callback)) {
    return false;
  ***REMOVED***

  _this.chunks.insertOne(doc, getWriteOptions(_this), function(error) {
    if (error) {
      return __handleError(_this, error);
    ***REMOVED***
    --_this.state.outstandingRequests;
    checkDone(_this);
  ***REMOVED***);
***REMOVED***

/**
 * @ignore
 */

function checkAborted(_this, callback) {
  if (_this.state.aborted) {
    if (typeof callback === 'function') {
      callback(new Error('this stream has been aborted'));
    ***REMOVED***
    return true;
  ***REMOVED***
  return false;
***REMOVED***
