// bluebird/js/release/direct_resolve.js
"use strict";
module.exports = function(Promise) {
function returner() {
    return this.value;
***REMOVED***
function thrower() {
    throw this.reason;
***REMOVED***

Promise.prototype["return"] =
Promise.prototype.thenReturn = function (value) {
    if (value instanceof Promise) value.suppressUnhandledRejections();
    return this._then(
        returner, undefined, undefined, {value: value***REMOVED***, undefined);
***REMOVED***;

Promise.prototype["throw"] =
Promise.prototype.thenThrow = function (reason) {
    return this._then(
        thrower, undefined, undefined, {reason: reason***REMOVED***, undefined);
***REMOVED***;

Promise.prototype.catchThrow = function (reason) {
    if (arguments.length <= 1) {
        return this._then(
            undefined, thrower, undefined, {reason: reason***REMOVED***, undefined);
    ***REMOVED***
        var _reason = arguments[1];
        var handler = function() {throw _reason;***REMOVED***;
        return this.caught(reason, handler);
    ***REMOVED***
***REMOVED***;

Promise.prototype.catchReturn = function (value) {
    if (arguments.length <= 1) {
        if (value instanceof Promise) value.suppressUnhandledRejections();
        return this._then(
            undefined, returner, undefined, {value: value***REMOVED***, undefined);
    ***REMOVED***
        var _value = arguments[1];
        if (_value instanceof Promise) _value.suppressUnhandledRejections();
        var handler = function() {return _value;***REMOVED***;
        return this.caught(value, handler);
    ***REMOVED***
***REMOVED***;
***REMOVED***;
