// bluebird/js/release/es5.js
var isES5 = (function(){
    "use strict";
    return this === undefined;
***REMOVED***)();

if (isES5) {
    module.exports = {
        freeze: Object.freeze,
        defineProperty: Object.defineProperty,
        getDescriptor: Object.getOwnPropertyDescriptor,
        keys: Object.keys,
        names: Object.getOwnPropertyNames,
        getPrototypeOf: Object.getPrototypeOf,
        isArray: Array.isArray,
        isES5: isES5,
        propertyIsWritable: function(obj, prop) {
            var descriptor = Object.getOwnPropertyDescriptor(obj, prop);
            return !!(!descriptor || descriptor.writable || descriptor.set);
        ***REMOVED***
    ***REMOVED***;
***REMOVED***
    var has = {***REMOVED***.hasOwnProperty;
    var str = {***REMOVED***.toString;
    var proto = {***REMOVED***.constructor.prototype;

    var ObjectKeys = function (o) {
        var ret = [];
        for (var key in o) {
            if (has.call(o, key)) {
                ret.push(key);
            ***REMOVED***
        ***REMOVED***
        return ret;
    ***REMOVED***;

    var ObjectGetDescriptor = function(o, key) {
        return {value: o[key]***REMOVED***;
    ***REMOVED***;

    var ObjectDefineProperty = function (o, key, desc) {
        o[key] = desc.value;
        return o;
    ***REMOVED***;

    var ObjectFreeze = function (obj) {
        return obj;
    ***REMOVED***;

    var ObjectGetPrototypeOf = function (obj) {
        try {
            return Object(obj).constructor.prototype;
        ***REMOVED***
        catch (e) {
            return proto;
        ***REMOVED***
    ***REMOVED***;

    var ArrayIsArray = function (obj) {
        try {
            return str.call(obj) === "[object Array]";
        ***REMOVED***
        catch(e) {
            return false;
        ***REMOVED***
    ***REMOVED***;

    module.exports = {
        isArray: ArrayIsArray,
        keys: ObjectKeys,
        names: ObjectKeys,
        defineProperty: ObjectDefineProperty,
        getDescriptor: ObjectGetDescriptor,
        freeze: ObjectFreeze,
        getPrototypeOf: ObjectGetPrototypeOf,
        isES5: isES5,
        propertyIsWritable: function() {
            return true;
        ***REMOVED***
    ***REMOVED***;
***REMOVED***
