// bluebird/js/release/some.js
"use strict";
module.exports =
function(Promise, PromiseArray, apiRejection) {
var util = require("./util");
var RangeError = require("./errors").RangeError;
var AggregateError = require("./errors").AggregateError;
var isArray = util.isArray;
var CANCELLATION = {***REMOVED***;


function SomePromiseArray(values) {
    this.constructor$(values);
    this._howMany = 0;
    this._unwrap = false;
    this._initialized = false;
***REMOVED***
util.inherits(SomePromiseArray, PromiseArray);

SomePromiseArray.prototype._init = function () {
    if (!this._initialized) {
        return;
    ***REMOVED***
    if (this._howMany === 0) {
        this._resolve([]);
        return;
    ***REMOVED***
    this._init$(undefined, -5);
    var isArrayResolved = isArray(this._values);
    if (!this._isResolved() &&
        isArrayResolved &&
        this._howMany > this._canPossiblyFulfill()) {
        this._reject(this._getRangeError(this.length()));
    ***REMOVED***
***REMOVED***;

SomePromiseArray.prototype.init = function () {
    this._initialized = true;
    this._init();
***REMOVED***;

SomePromiseArray.prototype.setUnwrap = function () {
    this._unwrap = true;
***REMOVED***;

SomePromiseArray.prototype.howMany = function () {
    return this._howMany;
***REMOVED***;

SomePromiseArray.prototype.setHowMany = function (count) {
    this._howMany = count;
***REMOVED***;

SomePromiseArray.prototype._promiseFulfilled = function (value) {
    this._addFulfilled(value);
    if (this._fulfilled() === this.howMany()) {
        this._values.length = this.howMany();
        if (this.howMany() === 1 && this._unwrap) {
            this._resolve(this._values[0]);
        ***REMOVED***
            this._resolve(this._values);
        ***REMOVED***
        return true;
    ***REMOVED***
    return false;

***REMOVED***;
SomePromiseArray.prototype._promiseRejected = function (reason) {
    this._addRejected(reason);
    return this._checkOutcome();
***REMOVED***;

SomePromiseArray.prototype._promiseCancelled = function () {
    if (this._values instanceof Promise || this._values == null) {
        return this._cancel();
    ***REMOVED***
    this._addRejected(CANCELLATION);
    return this._checkOutcome();
***REMOVED***;

SomePromiseArray.prototype._checkOutcome = function() {
    if (this.howMany() > this._canPossiblyFulfill()) {
        var e = new AggregateError();
        for (var i = this.length(); i < this._values.length; ++i) {
            if (this._values[i] !== CANCELLATION) {
                e.push(this._values[i]);
            ***REMOVED***
        ***REMOVED***
        if (e.length > 0) {
            this._reject(e);
        ***REMOVED***
            this._cancel();
        ***REMOVED***
        return true;
    ***REMOVED***
    return false;
***REMOVED***;

SomePromiseArray.prototype._fulfilled = function () {
    return this._totalResolved;
***REMOVED***;

SomePromiseArray.prototype._rejected = function () {
    return this._values.length - this.length();
***REMOVED***;

SomePromiseArray.prototype._addRejected = function (reason) {
    this._values.push(reason);
***REMOVED***;

SomePromiseArray.prototype._addFulfilled = function (value) {
    this._values[this._totalResolved++] = value;
***REMOVED***;

SomePromiseArray.prototype._canPossiblyFulfill = function () {
    return this.length() - this._rejected();
***REMOVED***;

SomePromiseArray.prototype._getRangeError = function (count) {
    var message = "Input array must contain at least " +
            this._howMany + " items but contains only " + count + " items";
    return new RangeError(message);
***REMOVED***;

SomePromiseArray.prototype._resolveEmptyArray = function () {
    this._reject(this._getRangeError(0));
***REMOVED***;

function some(promises, howMany) {
    if ((howMany | 0) !== howMany || howMany < 0) {
        return apiRejection("expecting a positive integer\u000a\u000a    See http://goo.gl/MqrFmX\u000a");
    ***REMOVED***
    var ret = new SomePromiseArray(promises);
    var promise = ret.promise();
    ret.setHowMany(howMany);
    ret.init();
    return promise;
***REMOVED***

Promise.some = function (promises, howMany) {
    return some(promises, howMany);
***REMOVED***;

Promise.prototype.some = function (howMany) {
    return some(this, howMany);
***REMOVED***;

Promise._SomePromiseArray = SomePromiseArray;
***REMOVED***;
