// mongoose/lib/helpers/model/castBulkWrite.js
'use strict';

const applyTimestampsToChildren = require('../update/applyTimestampsToChildren');
const applyTimestampsToUpdate = require('../update/applyTimestampsToUpdate');
const cast = require('../../cast');
const castUpdate = require('../query/castUpdate');
const setDefaultsOnInsert = require('../setDefaultsOnInsert');

/*!
 * Given a model and a bulkWrite op, return a thunk that handles casting and
 * validating the individual op.
 */

module.exports = function castBulkWrite(model, op) {
  const now = model.base.now();

  if (op['insertOne']) {
    return (callback) => {
      const doc = new model(op['insertOne']['document']);
      if (model.schema.options.timestamps != null) {
        doc.initializeTimestamps();
      ***REMOVED***

      op['insertOne']['document'] = doc;
      op['insertOne']['document'].validate({ __noPromise: true ***REMOVED***, function(error) {
        if (error) {
          return callback(error, null);
        ***REMOVED***
        callback(null);
      ***REMOVED***);
    ***REMOVED***;
  ***REMOVED*** else if (op['updateOne']) {
    op = op['updateOne'];
    return (callback) => {
      try {
        op['filter'] = cast(model.schema, op['filter']);
        op['update'] = castUpdate(model.schema, op['update'], {
          strict: model.schema.options.strict,
          overwrite: false
        ***REMOVED***);
        if (op.setDefaultsOnInsert) {
          setDefaultsOnInsert(op['filter'], model.schema, op['update'], {
            setDefaultsOnInsert: true,
            upsert: op.upsert
          ***REMOVED***);
        ***REMOVED***
        if (model.schema.$timestamps != null) {
          const createdAt = model.schema.$timestamps.createdAt;
          const updatedAt = model.schema.$timestamps.updatedAt;
          applyTimestampsToUpdate(now, createdAt, updatedAt, op['update'], {***REMOVED***);
        ***REMOVED***
        applyTimestampsToChildren(now, op['update'], model.schema);
      ***REMOVED*** catch (error) {
        return callback(error, null);
      ***REMOVED***

      callback(null);
    ***REMOVED***;
  ***REMOVED*** else if (op['updateMany']) {
    op = op['updateMany'];
    return (callback) => {
      try {
        op['filter'] = cast(model.schema, op['filter']);
        op['update'] = castUpdate(model.schema, op['update'], {
          strict: model.schema.options.strict,
          overwrite: false
        ***REMOVED***);
        if (op.setDefaultsOnInsert) {
          setDefaultsOnInsert(op['filter'], model.schema, op['update'], {
            setDefaultsOnInsert: true,
            upsert: op.upsert
          ***REMOVED***);
        ***REMOVED***
        if (model.schema.$timestamps != null) {
          const createdAt = model.schema.$timestamps.createdAt;
          const updatedAt = model.schema.$timestamps.updatedAt;
          applyTimestampsToUpdate(now, createdAt, updatedAt, op['update'], {***REMOVED***);
        ***REMOVED***
        applyTimestampsToChildren(now, op['update'], model.schema);
      ***REMOVED*** catch (error) {
        return callback(error, null);
      ***REMOVED***

      callback(null);
    ***REMOVED***;
  ***REMOVED*** else if (op['replaceOne']) {
    return (callback) => {
      try {
        op['replaceOne']['filter'] = cast(model.schema,
          op['replaceOne']['filter']);
      ***REMOVED*** catch (error) {
        return callback(error, null);
      ***REMOVED***

      // set `skipId`, otherwise we get "_id field cannot be changed"
      const doc = new model(op['replaceOne']['replacement'], null, true);
      if (model.schema.options.timestamps != null) {
        doc.initializeTimestamps();
      ***REMOVED***
      op['replaceOne']['replacement'] = doc;

      op['replaceOne']['replacement'].validate({ __noPromise: true ***REMOVED***, function(error) {
        if (error) {
          return callback(error, null);
        ***REMOVED***
        callback(null);
      ***REMOVED***);
    ***REMOVED***;
  ***REMOVED*** else if (op['deleteOne']) {
    return (callback) => {
      try {
        op['deleteOne']['filter'] = cast(model.schema,
          op['deleteOne']['filter']);
      ***REMOVED*** catch (error) {
        return callback(error, null);
      ***REMOVED***

      callback(null);
    ***REMOVED***;
  ***REMOVED*** else if (op['deleteMany']) {
    return (callback) => {
      try {
        op['deleteMany']['filter'] = cast(model.schema,
          op['deleteMany']['filter']);
      ***REMOVED*** catch (error) {
        return callback(error, null);
      ***REMOVED***

      callback(null);
    ***REMOVED***;
  ***REMOVED***
    return (callback) => {
      callback(new Error('Invalid op passed to `bulkWrite()`'), null);
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
