// mongoose/lib/schema/buffer.js
/*!
 * Module dependencies.
 */

'use strict';

const handleBitwiseOperator = require('./operators/bitwise');
const utils = require('../utils');

const MongooseBuffer = require('../types/buffer');
const SchemaType = require('../schematype');

const Binary = MongooseBuffer.Binary;
const CastError = SchemaType.CastError;
let Document;

/**
 * Buffer SchemaType constructor
 *
 * @param {String***REMOVED*** key
 * @param {Object***REMOVED*** options
 * @inherits SchemaType
 * @api public
 */

function SchemaBuffer(key, options) {
  SchemaType.call(this, key, options, 'Buffer');
***REMOVED***

/**
 * This schema type's name, to defend against minifiers that mangle
 * function names.
 *
 * @api public
 */
SchemaBuffer.schemaName = 'Buffer';

/*!
 * Inherits from SchemaType.
 */
SchemaBuffer.prototype = Object.create(SchemaType.prototype);
SchemaBuffer.prototype.constructor = SchemaBuffer;

/**
 * Check if the given value satisfies a required validator. To satisfy a
 * required validator, a buffer must not be null or undefined and have
 * non-zero length.
 *
 * @param {Any***REMOVED*** value
 * @param {Document***REMOVED*** doc
 * @return {Boolean***REMOVED***
 * @api public
 */

SchemaBuffer.prototype.checkRequired = function(value, doc) {
  if (SchemaType._isRef(this, value, doc, true)) {
    return !!value;
  ***REMOVED***
  return !!(value && value.length);
***REMOVED***;

/**
 * Casts contents
 *
 * @param {Object***REMOVED*** value
 * @param {Document***REMOVED*** doc document that triggers the casting
 * @param {Boolean***REMOVED*** init
 * @api private
 */

SchemaBuffer.prototype.cast = function(value, doc, init) {
  let ret;
  if (SchemaType._isRef(this, value, doc, init)) {
    // wait! we may need to cast this to a document

    if (value === null || value === undefined) {
      return value;
    ***REMOVED***

    // lazy load
    Document || (Document = require('./../document'));

    if (value instanceof Document) {
      value.$__.wasPopulated = true;
      return value;
    ***REMOVED***

    // setting a populated path
    if (Buffer.isBuffer(value)) {
      return value;
    ***REMOVED*** else if (!utils.isObject(value)) {
      throw new CastError('buffer', value, this.path);
    ***REMOVED***

    // Handle the case where user directly sets a populated
    // path to a plain object; cast to the Model used in
    // the population query.
    const path = doc.$__fullPath(this.path);
    const owner = doc.ownerDocument ? doc.ownerDocument() : doc;
    const pop = owner.populated(path, true);
    ret = new pop.options.model(value);
    ret.$__.wasPopulated = true;
    return ret;
  ***REMOVED***

  // documents
  if (value && value._id) {
    value = value._id;
  ***REMOVED***

  if (value && value.isMongooseBuffer) {
    return value;
  ***REMOVED***

  if (Buffer.isBuffer(value)) {
    if (!value || !value.isMongooseBuffer) {
      value = new MongooseBuffer(value, [this.path, doc]);
      if (this.options.subtype != null) {
        value._subtype = this.options.subtype;
      ***REMOVED***
    ***REMOVED***
    return value;
  ***REMOVED***

  if (value instanceof Binary) {
    ret = new MongooseBuffer(value.value(true), [this.path, doc]);
    if (typeof value.sub_type !== 'number') {
      throw new CastError('buffer', value, this.path);
    ***REMOVED***
    ret._subtype = value.sub_type;
    return ret;
  ***REMOVED***

  if (value === null) {
    return value;
  ***REMOVED***


  const type = typeof value;
  if (
    type === 'string' || type === 'number' || Array.isArray(value) ||
    (type === 'object' && value.type === 'Buffer' && Array.isArray(value.data)) // gh-6863
  ) {
    if (type === 'number') {
      value = [value];
    ***REMOVED***
    ret = new MongooseBuffer(value, [this.path, doc]);
    if (this.options.subtype != null) {
      ret._subtype = this.options.subtype;
    ***REMOVED***
    return ret;
  ***REMOVED***

  throw new CastError('buffer', value, this.path);
***REMOVED***;

/**
 * Sets the default [subtype](https://studio3t.com/whats-new/best-practices-uuid-mongodb/)
 * for this buffer. You can find a [list of allowed subtypes here](http://api.mongodb.com/python/current/api/bson/binary.html).
 *
 * ####Example:
 *
 *     var s = new Schema({ uuid: { type: Buffer, subtype: 4 ***REMOVED***);
 *     var M = db.model('M', s);
 *     var m = new M({ uuid: 'test string' ***REMOVED***);
 *     m.uuid._subtype; // 4
 *
 * @param {Number***REMOVED*** subtype the default subtype
 * @return {SchemaType***REMOVED*** this
 * @api public
 */

SchemaBuffer.prototype.subtype = function(subtype) {
  this.options.subtype = subtype;
  return this;
***REMOVED***;

/*!
 * ignore
 */
function handleSingle(val) {
  return this.castForQuery(val);
***REMOVED***

SchemaBuffer.prototype.$conditionalHandlers =
    utils.options(SchemaType.prototype.$conditionalHandlers, {
      $bitsAllClear: handleBitwiseOperator,
      $bitsAnyClear: handleBitwiseOperator,
      $bitsAllSet: handleBitwiseOperator,
      $bitsAnySet: handleBitwiseOperator,
      $gt: handleSingle,
      $gte: handleSingle,
      $lt: handleSingle,
      $lte: handleSingle
    ***REMOVED***);

/**
 * Casts contents for queries.
 *
 * @param {String***REMOVED*** $conditional
 * @param {any***REMOVED*** [value]
 * @api private
 */

SchemaBuffer.prototype.castForQuery = function($conditional, val) {
  let handler;
  if (arguments.length === 2) {
    handler = this.$conditionalHandlers[$conditional];
    if (!handler) {
      throw new Error('Can\'t use ' + $conditional + ' with Buffer.');
    ***REMOVED***
    return handler.call(this, val);
  ***REMOVED***
  val = $conditional;
  const casted = this._castForQuery(val);
  return casted ? casted.toObject({ transform: false, virtuals: false ***REMOVED***) : casted;
***REMOVED***;

/*!
 * Module exports.
 */

module.exports = SchemaBuffer;
