// mongoose/lib/schema/decimal128.js
/*!
 * Module dependencies.
 */

'use strict';

const SchemaType = require('../schematype');
const CastError = SchemaType.CastError;
const Decimal128Type = require('../types/decimal128');
const utils = require('../utils');
let Document;

/**
 * Decimal128 SchemaType constructor.
 *
 * @param {String***REMOVED*** key
 * @param {Object***REMOVED*** options
 * @inherits SchemaType
 * @api public
 */

function Decimal128(key, options) {
  SchemaType.call(this, key, options, 'Decimal128');
***REMOVED***

/**
 * This schema type's name, to defend against minifiers that mangle
 * function names.
 *
 * @api public
 */
Decimal128.schemaName = 'Decimal128';

/*!
 * Inherits from SchemaType.
 */
Decimal128.prototype = Object.create(SchemaType.prototype);
Decimal128.prototype.constructor = Decimal128;

/**
 * Check if the given value satisfies a required validator.
 *
 * @param {Any***REMOVED*** value
 * @param {Document***REMOVED*** doc
 * @return {Boolean***REMOVED***
 * @api public
 */

Decimal128.prototype.checkRequired = function checkRequired(value, doc) {
  if (SchemaType._isRef(this, value, doc, true)) {
    return !!value;
  ***REMOVED***
  return value instanceof Decimal128Type;
***REMOVED***;

/**
 * Casts to Decimal128
 *
 * @param {Object***REMOVED*** value
 * @param {Object***REMOVED*** doc
 * @param {Boolean***REMOVED*** init whether this is an initialization cast
 * @api private
 */

Decimal128.prototype.cast = function(value, doc, init) {
  if (SchemaType._isRef(this, value, doc, init)) {
    // wait! we may need to cast this to a document

    if (value === null || value === undefined) {
      return value;
    ***REMOVED***

    // lazy load
    Document || (Document = require('./../document'));

    if (value instanceof Document) {
      value.$__.wasPopulated = true;
      return value;
    ***REMOVED***

    // setting a populated path
    if (value instanceof Decimal128Type) {
      return value;
    ***REMOVED*** else if (Buffer.isBuffer(value) || !utils.isObject(value)) {
      throw new CastError('Decimal128', value, this.path);
    ***REMOVED***

    // Handle the case where user directly sets a populated
    // path to a plain object; cast to the Model used in
    // the population query.
    const path = doc.$__fullPath(this.path);
    const owner = doc.ownerDocument ? doc.ownerDocument() : doc;
    const pop = owner.populated(path, true);
    let ret = value;
    if (!doc.$__.populated ||
        !doc.$__.populated[path] ||
        !doc.$__.populated[path].options ||
        !doc.$__.populated[path].options.options ||
        !doc.$__.populated[path].options.options.lean) {
      ret = new pop.options.model(value);
      ret.$__.wasPopulated = true;
    ***REMOVED***

    return ret;
  ***REMOVED***

  if (value == null) {
    return value;
  ***REMOVED***

  if (typeof value === 'object' && typeof value.$numberDecimal === 'string') {
    return Decimal128Type.fromString(value.$numberDecimal);
  ***REMOVED***

  if (value instanceof Decimal128Type) {
    return value;
  ***REMOVED***

  if (typeof value === 'string') {
    return Decimal128Type.fromString(value);
  ***REMOVED***

  if (Buffer.isBuffer(value)) {
    return new Decimal128Type(value);
  ***REMOVED***

  if (typeof value === 'number') {
    return Decimal128Type.fromString(String(value));
  ***REMOVED***

  if (typeof value.valueOf === 'function' && typeof value.valueOf() === 'string') {
    return Decimal128Type.fromString(value.valueOf());
  ***REMOVED***

  throw new CastError('Decimal128', value, this.path);
***REMOVED***;

/*!
 * ignore
 */

function handleSingle(val) {
  return this.cast(val);
***REMOVED***

Decimal128.prototype.$conditionalHandlers =
    utils.options(SchemaType.prototype.$conditionalHandlers, {
      $gt: handleSingle,
      $gte: handleSingle,
      $lt: handleSingle,
      $lte: handleSingle
    ***REMOVED***);

/*!
 * Module exports.
 */

module.exports = Decimal128;
