// mongoose/lib/schema/number.js
'use strict';

/*!
 * Module requirements.
 */

const MongooseError = require('../error');
const SchemaType = require('../schematype');
const castNumber = require('../cast/number');
const handleBitwiseOperator = require('./operators/bitwise');
const utils = require('../utils');

const CastError = SchemaType.CastError;
let Document;

/**
 * Number SchemaType constructor.
 *
 * @param {String***REMOVED*** key
 * @param {Object***REMOVED*** options
 * @inherits SchemaType
 * @api public
 */

function SchemaNumber(key, options) {
  SchemaType.call(this, key, options, 'Number');
***REMOVED***

/**
 * This schema type's name, to defend against minifiers that mangle
 * function names.
 *
 * @api public
 */
SchemaNumber.schemaName = 'Number';

/*!
 * Inherits from SchemaType.
 */
SchemaNumber.prototype = Object.create(SchemaType.prototype);
SchemaNumber.prototype.constructor = SchemaNumber;

/**
 * Check if the given value satisfies a required validator.
 *
 * @param {Any***REMOVED*** value
 * @param {Document***REMOVED*** doc
 * @return {Boolean***REMOVED***
 * @api public
 */

SchemaNumber.prototype.checkRequired = function checkRequired(value, doc) {
  if (SchemaType._isRef(this, value, doc, true)) {
    return !!value;
  ***REMOVED***
  return typeof value === 'number' || value instanceof Number;
***REMOVED***;

/**
 * Sets a minimum number validator.
 *
 * ####Example:
 *
 *     var s = new Schema({ n: { type: Number, min: 10 ***REMOVED***)
 *     var M = db.model('M', s)
 *     var m = new M({ n: 9 ***REMOVED***)
 *     m.save(function (err) {
 *       console.error(err) // validator error
 *       m.n = 10;
 *       m.save() // success
 *     ***REMOVED***)
 *
 *     // custom error messages
 *     // We can also use the special {MIN***REMOVED*** token which will be replaced with the invalid value
 *     var min = [10, 'The value of path `{PATH***REMOVED***` ({VALUE***REMOVED***) is beneath the limit ({MIN***REMOVED***).'];
 *     var schema = new Schema({ n: { type: Number, min: min ***REMOVED***)
 *     var M = mongoose.model('Measurement', schema);
 *     var s= new M({ n: 4 ***REMOVED***);
 *     s.validate(function (err) {
 *       console.log(String(err)) // ValidationError: The value of path `n` (4) is beneath the limit (10).
 *     ***REMOVED***)
 *
 * @param {Number***REMOVED*** value minimum number
 * @param {String***REMOVED*** [message] optional custom error message
 * @return {SchemaType***REMOVED*** this
 * @see Customized Error Messages #error_messages_MongooseError-messages
 * @api public
 */

SchemaNumber.prototype.min = function(value, message) {
  if (this.minValidator) {
    this.validators = this.validators.filter(function(v) {
      return v.validator !== this.minValidator;
    ***REMOVED***, this);
  ***REMOVED***

  if (value !== null && value !== undefined) {
    let msg = message || MongooseError.messages.Number.min;
    msg = msg.replace(/{MIN***REMOVED***/, value);
    this.validators.push({
      validator: this.minValidator = function(v) {
        return v == null || v >= value;
      ***REMOVED***,
      message: msg,
      type: 'min',
      min: value
    ***REMOVED***);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Sets a maximum number validator.
 *
 * ####Example:
 *
 *     var s = new Schema({ n: { type: Number, max: 10 ***REMOVED***)
 *     var M = db.model('M', s)
 *     var m = new M({ n: 11 ***REMOVED***)
 *     m.save(function (err) {
 *       console.error(err) // validator error
 *       m.n = 10;
 *       m.save() // success
 *     ***REMOVED***)
 *
 *     // custom error messages
 *     // We can also use the special {MAX***REMOVED*** token which will be replaced with the invalid value
 *     var max = [10, 'The value of path `{PATH***REMOVED***` ({VALUE***REMOVED***) exceeds the limit ({MAX***REMOVED***).'];
 *     var schema = new Schema({ n: { type: Number, max: max ***REMOVED***)
 *     var M = mongoose.model('Measurement', schema);
 *     var s= new M({ n: 4 ***REMOVED***);
 *     s.validate(function (err) {
 *       console.log(String(err)) // ValidationError: The value of path `n` (4) exceeds the limit (10).
 *     ***REMOVED***)
 *
 * @param {Number***REMOVED*** maximum number
 * @param {String***REMOVED*** [message] optional custom error message
 * @return {SchemaType***REMOVED*** this
 * @see Customized Error Messages #error_messages_MongooseError-messages
 * @api public
 */

SchemaNumber.prototype.max = function(value, message) {
  if (this.maxValidator) {
    this.validators = this.validators.filter(function(v) {
      return v.validator !== this.maxValidator;
    ***REMOVED***, this);
  ***REMOVED***

  if (value !== null && value !== undefined) {
    let msg = message || MongooseError.messages.Number.max;
    msg = msg.replace(/{MAX***REMOVED***/, value);
    this.validators.push({
      validator: this.maxValidator = function(v) {
        return v == null || v <= value;
      ***REMOVED***,
      message: msg,
      type: 'max',
      max: value
    ***REMOVED***);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Casts to number
 *
 * @param {Object***REMOVED*** value value to cast
 * @param {Document***REMOVED*** doc document that triggers the casting
 * @param {Boolean***REMOVED*** init
 * @api private
 */

SchemaNumber.prototype.cast = function(value, doc, init) {
  if (SchemaType._isRef(this, value, doc, init)) {
    // wait! we may need to cast this to a document

    if (value === null || value === undefined) {
      return value;
    ***REMOVED***

    // lazy load
    Document || (Document = require('./../document'));

    if (value instanceof Document) {
      value.$__.wasPopulated = true;
      return value;
    ***REMOVED***

    // setting a populated path
    if (typeof value === 'number') {
      return value;
    ***REMOVED*** else if (Buffer.isBuffer(value) || !utils.isObject(value)) {
      throw new CastError('number', value, this.path);
    ***REMOVED***

    // Handle the case where user directly sets a populated
    // path to a plain object; cast to the Model used in
    // the population query.
    const path = doc.$__fullPath(this.path);
    const owner = doc.ownerDocument ? doc.ownerDocument() : doc;
    const pop = owner.populated(path, true);
    const ret = new pop.options.model(value);
    ret.$__.wasPopulated = true;
    return ret;
  ***REMOVED***

  const val = value && typeof value._id !== 'undefined' ?
    value._id : // documents
    value;

  return castNumber(val, this.path);
***REMOVED***;

/*!
 * ignore
 */

function handleSingle(val) {
  return this.cast(val);
***REMOVED***

function handleArray(val) {
  const _this = this;
  if (!Array.isArray(val)) {
    return [this.cast(val)];
  ***REMOVED***
  return val.map(function(m) {
    return _this.cast(m);
  ***REMOVED***);
***REMOVED***

SchemaNumber.prototype.$conditionalHandlers =
    utils.options(SchemaType.prototype.$conditionalHandlers, {
      $bitsAllClear: handleBitwiseOperator,
      $bitsAnyClear: handleBitwiseOperator,
      $bitsAllSet: handleBitwiseOperator,
      $bitsAnySet: handleBitwiseOperator,
      $gt: handleSingle,
      $gte: handleSingle,
      $lt: handleSingle,
      $lte: handleSingle,
      $mod: handleArray
    ***REMOVED***);

/**
 * Casts contents for queries.
 *
 * @param {String***REMOVED*** $conditional
 * @param {any***REMOVED*** [value]
 * @api private
 */

SchemaNumber.prototype.castForQuery = function($conditional, val) {
  let handler;
  if (arguments.length === 2) {
    handler = this.$conditionalHandlers[$conditional];
    if (!handler) {
      throw new CastError('Can\'t use ' + $conditional + ' with Number.');
    ***REMOVED***
    return handler.call(this, val);
  ***REMOVED***
  val = this._castForQuery($conditional);
  return val;
***REMOVED***;

/*!
 * Module exports.
 */

module.exports = SchemaNumber;
