// mongoose/lib/schema/objectid.js
/*!
 * Module dependencies.
 */

'use strict';

const SchemaType = require('../schematype');
const oid = require('../types/objectid');
const utils = require('../utils');

const CastError = SchemaType.CastError;
let Document;

/**
 * ObjectId SchemaType constructor.
 *
 * @param {String***REMOVED*** key
 * @param {Object***REMOVED*** options
 * @inherits SchemaType
 * @api public
 */

function ObjectId(key, options) {
  const isKeyHexStr = typeof key === 'string' && key.length === 24 && /^[a-f0-9]+$/i.test(key);
  const suppressWarning = options && options.suppressWarning;
  if ((isKeyHexStr || typeof key === 'undefined') && !suppressWarning) {
    console.warn('mongoose: To create a new ObjectId please try ' +
      '`Mongoose.Types.ObjectId` instead of using ' +
      '`Mongoose.Schema.ObjectId`. Set the `suppressWarning` option if ' +
      'you\'re trying to create a hex char path in your schema.');
    console.trace();
  ***REMOVED***
  SchemaType.call(this, key, options, 'ObjectID');
***REMOVED***

/**
 * This schema type's name, to defend against minifiers that mangle
 * function names.
 *
 * @api public
 */
ObjectId.schemaName = 'ObjectId';

/*!
 * Inherits from SchemaType.
 */
ObjectId.prototype = Object.create(SchemaType.prototype);
ObjectId.prototype.constructor = ObjectId;

/**
 * Adds an auto-generated ObjectId default if turnOn is true.
 * @param {Boolean***REMOVED*** turnOn auto generated ObjectId defaults
 * @api public
 * @return {SchemaType***REMOVED*** this
 */

ObjectId.prototype.auto = function(turnOn) {
  if (turnOn) {
    this.default(defaultId);
    this.set(resetId);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Check if the given value satisfies a required validator.
 *
 * @param {Any***REMOVED*** value
 * @param {Document***REMOVED*** doc
 * @return {Boolean***REMOVED***
 * @api public
 */

ObjectId.prototype.checkRequired = function checkRequired(value, doc) {
  if (SchemaType._isRef(this, value, doc, true)) {
    return !!value;
  ***REMOVED***
  return value instanceof oid;
***REMOVED***;

/**
 * Casts to ObjectId
 *
 * @param {Object***REMOVED*** value
 * @param {Object***REMOVED*** doc
 * @param {Boolean***REMOVED*** init whether this is an initialization cast
 * @api private
 */

ObjectId.prototype.cast = function(value, doc, init) {
  if (SchemaType._isRef(this, value, doc, init)) {
    // wait! we may need to cast this to a document

    if (value === null || value === undefined) {
      return value;
    ***REMOVED***

    // lazy load
    Document || (Document = require('./../document'));

    if (value instanceof Document) {
      value.$__.wasPopulated = true;
      return value;
    ***REMOVED***

    // setting a populated path
    if (value instanceof oid) {
      return value;
    ***REMOVED*** else if ((value.constructor.name || '').toLowerCase() === 'objectid') {
      return new oid(value.toHexString());
    ***REMOVED*** else if (Buffer.isBuffer(value) || !utils.isObject(value)) {
      throw new CastError('ObjectId', value, this.path);
    ***REMOVED***

    // Handle the case where user directly sets a populated
    // path to a plain object; cast to the Model used in
    // the population query.
    const path = doc.$__fullPath(this.path);
    const owner = doc.ownerDocument ? doc.ownerDocument() : doc;
    const pop = owner.populated(path, true);
    let ret = value;
    if (!doc.$__.populated ||
        !doc.$__.populated[path] ||
        !doc.$__.populated[path].options ||
        !doc.$__.populated[path].options.options ||
        !doc.$__.populated[path].options.options.lean) {
      ret = new pop.options.model(value);
      ret.$__.wasPopulated = true;
    ***REMOVED***

    return ret;
  ***REMOVED***

  if (value === null || value === undefined) {
    return value;
  ***REMOVED***

  if (value instanceof oid) {
    return value;
  ***REMOVED***

  if (value._id) {
    if (value._id instanceof oid) {
      return value._id;
    ***REMOVED***
    if (value._id.toString instanceof Function) {
      return new oid(value._id.toString());
    ***REMOVED***
  ***REMOVED***

  if (value.toString instanceof Function) {
    try {
      return new oid(value.toString());
    ***REMOVED*** catch (err) {
      throw new CastError('ObjectId', value, this.path);
    ***REMOVED***
  ***REMOVED***

  throw new CastError('ObjectId', value, this.path);
***REMOVED***;

/*!
 * ignore
 */

function handleSingle(val) {
  return this.cast(val);
***REMOVED***

ObjectId.prototype.$conditionalHandlers =
    utils.options(SchemaType.prototype.$conditionalHandlers, {
      $gt: handleSingle,
      $gte: handleSingle,
      $lt: handleSingle,
      $lte: handleSingle
    ***REMOVED***);

/*!
 * ignore
 */

function defaultId() {
  return new oid();
***REMOVED***

defaultId.$runBeforeSetters = true;

function resetId(v) {
  Document || (Document = require('./../document'));

  if (this instanceof Document) {
    if (v === void 0) {
      const _v = new oid;
      this.$__._id = _v;
      return _v;
    ***REMOVED***

    this.$__._id = v;
  ***REMOVED***

  return v;
***REMOVED***

/*!
 * Module exports.
 */

module.exports = ObjectId;
