// mongoose/lib/schema/string.js
'use strict';

/*!
 * Module dependencies.
 */

const SchemaType = require('../schematype');
const CastError = SchemaType.CastError;
const MongooseError = require('../error');
const castString = require('../cast/string');
const utils = require('../utils');

let Document;

/**
 * String SchemaType constructor.
 *
 * @param {String***REMOVED*** key
 * @param {Object***REMOVED*** options
 * @inherits SchemaType
 * @api public
 */

function SchemaString(key, options) {
  this.enumValues = [];
  this.regExp = null;
  SchemaType.call(this, key, options, 'String');
***REMOVED***

/**
 * This schema type's name, to defend against minifiers that mangle
 * function names.
 *
 * @api public
 */
SchemaString.schemaName = 'String';

/*!
 * Inherits from SchemaType.
 */
SchemaString.prototype = Object.create(SchemaType.prototype);
SchemaString.prototype.constructor = SchemaString;

/**
 * Adds an enum validator
 *
 * ####Example:
 *
 *     var states = ['opening', 'open', 'closing', 'closed']
 *     var s = new Schema({ state: { type: String, enum: states ***REMOVED******REMOVED***)
 *     var M = db.model('M', s)
 *     var m = new M({ state: 'invalid' ***REMOVED***)
 *     m.save(function (err) {
 *       console.error(String(err)) // ValidationError: `invalid` is not a valid enum value for path `state`.
 *       m.state = 'open'
 *       m.save(callback) // success
 *     ***REMOVED***)
 *
 *     // or with custom error messages
 *     var enum = {
 *       values: ['opening', 'open', 'closing', 'closed'],
 *       message: 'enum validator failed for path `{PATH***REMOVED***` with value `{VALUE***REMOVED***`'
 *     ***REMOVED***
 *     var s = new Schema({ state: { type: String, enum: enum ***REMOVED***)
 *     var M = db.model('M', s)
 *     var m = new M({ state: 'invalid' ***REMOVED***)
 *     m.save(function (err) {
 *       console.error(String(err)) // ValidationError: enum validator failed for path `state` with value `invalid`
 *       m.state = 'open'
 *       m.save(callback) // success
 *     ***REMOVED***)
 *
 * @param {String|Object***REMOVED*** [args...] enumeration values
 * @return {SchemaType***REMOVED*** this
 * @see Customized Error Messages #error_messages_MongooseError-messages
 * @api public
 */

SchemaString.prototype.enum = function() {
  if (this.enumValidator) {
    this.validators = this.validators.filter(function(v) {
      return v.validator !== this.enumValidator;
    ***REMOVED***, this);
    this.enumValidator = false;
  ***REMOVED***

  if (arguments[0] === void 0 || arguments[0] === false) {
    return this;
  ***REMOVED***

  let values;
  let errorMessage;

  if (utils.isObject(arguments[0])) {
    values = arguments[0].values;
    errorMessage = arguments[0].message;
  ***REMOVED***
    values = arguments;
    errorMessage = MongooseError.messages.String.enum;
  ***REMOVED***

  for (let i = 0; i < values.length; i++) {
    if (undefined !== values[i]) {
      this.enumValues.push(this.cast(values[i]));
    ***REMOVED***
  ***REMOVED***

  const vals = this.enumValues;
  this.enumValidator = function(v) {
    return undefined === v || ~vals.indexOf(v);
  ***REMOVED***;
  this.validators.push({
    validator: this.enumValidator,
    message: errorMessage,
    type: 'enum',
    enumValues: vals
  ***REMOVED***);

  return this;
***REMOVED***;

/**
 * Adds a lowercase [setter](http://mongoosejs.com/docs/api.html#schematype_SchemaType-set).
 *
 * ####Example:
 *
 *     var s = new Schema({ email: { type: String, lowercase: true ***REMOVED******REMOVED***)
 *     var M = db.model('M', s);
 *     var m = new M({ email: 'SomeEmail@example.COM' ***REMOVED***);
 *     console.log(m.email) // someemail@example.com
 *     M.find({ email: 'SomeEmail@example.com' ***REMOVED***); // Queries by 'someemail@example.com'
 *
 * @api public
 * @return {SchemaType***REMOVED*** this
 */

SchemaString.prototype.lowercase = function(shouldApply) {
  if (arguments.length > 0 && !shouldApply) {
    return this;
  ***REMOVED***
  return this.set(function(v, self) {
    if (typeof v !== 'string') {
      v = self.cast(v);
    ***REMOVED***
    if (v) {
      return v.toLowerCase();
    ***REMOVED***
    return v;
  ***REMOVED***);
***REMOVED***;

/**
 * Adds an uppercase [setter](http://mongoosejs.com/docs/api.html#schematype_SchemaType-set).
 *
 * ####Example:
 *
 *     var s = new Schema({ caps: { type: String, uppercase: true ***REMOVED******REMOVED***)
 *     var M = db.model('M', s);
 *     var m = new M({ caps: 'an example' ***REMOVED***);
 *     console.log(m.caps) // AN EXAMPLE
 *     M.find({ caps: 'an example' ***REMOVED***) // Matches documents where caps = 'AN EXAMPLE'
 *
 * @api public
 * @return {SchemaType***REMOVED*** this
 */

SchemaString.prototype.uppercase = function(shouldApply) {
  if (arguments.length > 0 && !shouldApply) {
    return this;
  ***REMOVED***
  return this.set(function(v, self) {
    if (typeof v !== 'string') {
      v = self.cast(v);
    ***REMOVED***
    if (v) {
      return v.toUpperCase();
    ***REMOVED***
    return v;
  ***REMOVED***);
***REMOVED***;

/**
 * Adds a trim [setter](http://mongoosejs.com/docs/api.html#schematype_SchemaType-set).
 *
 * The string value will be trimmed when set.
 *
 * ####Example:
 *
 *     var s = new Schema({ name: { type: String, trim: true ***REMOVED******REMOVED***)
 *     var M = db.model('M', s)
 *     var string = ' some name '
 *     console.log(string.length) // 11
 *     var m = new M({ name: string ***REMOVED***)
 *     console.log(m.name.length) // 9
 *
 * @api public
 * @return {SchemaType***REMOVED*** this
 */

SchemaString.prototype.trim = function(shouldTrim) {
  if (arguments.length > 0 && !shouldTrim) {
    return this;
  ***REMOVED***
  return this.set(function(v, self) {
    if (typeof v !== 'string') {
      v = self.cast(v);
    ***REMOVED***
    if (v) {
      return v.trim();
    ***REMOVED***
    return v;
  ***REMOVED***);
***REMOVED***;

/**
 * Sets a minimum length validator.
 *
 * ####Example:
 *
 *     var schema = new Schema({ postalCode: { type: String, minlength: 5 ***REMOVED***)
 *     var Address = db.model('Address', schema)
 *     var address = new Address({ postalCode: '9512' ***REMOVED***)
 *     address.save(function (err) {
 *       console.error(err) // validator error
 *       address.postalCode = '95125';
 *       address.save() // success
 *     ***REMOVED***)
 *
 *     // custom error messages
 *     // We can also use the special {MINLENGTH***REMOVED*** token which will be replaced with the minimum allowed length
 *     var minlength = [5, 'The value of path `{PATH***REMOVED***` (`{VALUE***REMOVED***`) is shorter than the minimum allowed length ({MINLENGTH***REMOVED***).'];
 *     var schema = new Schema({ postalCode: { type: String, minlength: minlength ***REMOVED***)
 *     var Address = mongoose.model('Address', schema);
 *     var address = new Address({ postalCode: '9512' ***REMOVED***);
 *     address.validate(function (err) {
 *       console.log(String(err)) // ValidationError: The value of path `postalCode` (`9512`) is shorter than the minimum length (5).
 *     ***REMOVED***)
 *
 * @param {Number***REMOVED*** value minimum string length
 * @param {String***REMOVED*** [message] optional custom error message
 * @return {SchemaType***REMOVED*** this
 * @see Customized Error Messages #error_messages_MongooseError-messages
 * @api public
 */

SchemaString.prototype.minlength = function(value, message) {
  if (this.minlengthValidator) {
    this.validators = this.validators.filter(function(v) {
      return v.validator !== this.minlengthValidator;
    ***REMOVED***, this);
  ***REMOVED***

  if (value !== null && value !== undefined) {
    let msg = message || MongooseError.messages.String.minlength;
    msg = msg.replace(/{MINLENGTH***REMOVED***/, value);
    this.validators.push({
      validator: this.minlengthValidator = function(v) {
        return v === null || v.length >= value;
      ***REMOVED***,
      message: msg,
      type: 'minlength',
      minlength: value
    ***REMOVED***);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Sets a maximum length validator.
 *
 * ####Example:
 *
 *     var schema = new Schema({ postalCode: { type: String, maxlength: 9 ***REMOVED***)
 *     var Address = db.model('Address', schema)
 *     var address = new Address({ postalCode: '9512512345' ***REMOVED***)
 *     address.save(function (err) {
 *       console.error(err) // validator error
 *       address.postalCode = '95125';
 *       address.save() // success
 *     ***REMOVED***)
 *
 *     // custom error messages
 *     // We can also use the special {MAXLENGTH***REMOVED*** token which will be replaced with the maximum allowed length
 *     var maxlength = [9, 'The value of path `{PATH***REMOVED***` (`{VALUE***REMOVED***`) exceeds the maximum allowed length ({MAXLENGTH***REMOVED***).'];
 *     var schema = new Schema({ postalCode: { type: String, maxlength: maxlength ***REMOVED***)
 *     var Address = mongoose.model('Address', schema);
 *     var address = new Address({ postalCode: '9512512345' ***REMOVED***);
 *     address.validate(function (err) {
 *       console.log(String(err)) // ValidationError: The value of path `postalCode` (`9512512345`) exceeds the maximum allowed length (9).
 *     ***REMOVED***)
 *
 * @param {Number***REMOVED*** value maximum string length
 * @param {String***REMOVED*** [message] optional custom error message
 * @return {SchemaType***REMOVED*** this
 * @see Customized Error Messages #error_messages_MongooseError-messages
 * @api public
 */

SchemaString.prototype.maxlength = function(value, message) {
  if (this.maxlengthValidator) {
    this.validators = this.validators.filter(function(v) {
      return v.validator !== this.maxlengthValidator;
    ***REMOVED***, this);
  ***REMOVED***

  if (value !== null && value !== undefined) {
    let msg = message || MongooseError.messages.String.maxlength;
    msg = msg.replace(/{MAXLENGTH***REMOVED***/, value);
    this.validators.push({
      validator: this.maxlengthValidator = function(v) {
        return v === null || v.length <= value;
      ***REMOVED***,
      message: msg,
      type: 'maxlength',
      maxlength: value
    ***REMOVED***);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Sets a regexp validator.
 *
 * Any value that does not pass `regExp`.test(val) will fail validation.
 *
 * ####Example:
 *
 *     var s = new Schema({ name: { type: String, match: /^a/ ***REMOVED******REMOVED***)
 *     var M = db.model('M', s)
 *     var m = new M({ name: 'I am invalid' ***REMOVED***)
 *     m.validate(function (err) {
 *       console.error(String(err)) // "ValidationError: Path `name` is invalid (I am invalid)."
 *       m.name = 'apples'
 *       m.validate(function (err) {
 *         assert.ok(err) // success
 *       ***REMOVED***)
 *     ***REMOVED***)
 *
 *     // using a custom error message
 *     var match = [ /\.html$/, "That file doesn't end in .html ({VALUE***REMOVED***)" ];
 *     var s = new Schema({ file: { type: String, match: match ***REMOVED******REMOVED***)
 *     var M = db.model('M', s);
 *     var m = new M({ file: 'invalid' ***REMOVED***);
 *     m.validate(function (err) {
 *       console.log(String(err)) // "ValidationError: That file doesn't end in .html (invalid)"
 *     ***REMOVED***)
 *
 * Empty strings, `undefined`, and `null` values always pass the match validator. If you require these values, enable the `required` validator also.
 *
 *     var s = new Schema({ name: { type: String, match: /^a/, required: true ***REMOVED******REMOVED***)
 *
 * @param {RegExp***REMOVED*** regExp regular expression to test against
 * @param {String***REMOVED*** [message] optional custom error message
 * @return {SchemaType***REMOVED*** this
 * @see Customized Error Messages #error_messages_MongooseError-messages
 * @api public
 */

SchemaString.prototype.match = function match(regExp, message) {
  // yes, we allow multiple match validators

  const msg = message || MongooseError.messages.String.match;

  const matchValidator = function(v) {
    if (!regExp) {
      return false;
    ***REMOVED***

    const ret = ((v != null && v !== '')
      ? regExp.test(v)
      : true);
    return ret;
  ***REMOVED***;

  this.validators.push({
    validator: matchValidator,
    message: msg,
    type: 'regexp',
    regexp: regExp
  ***REMOVED***);
  return this;
***REMOVED***;

/**
 * Check if the given value satisfies the `required` validator. The value is
 * considered valid if it is a string (that is, not `null` or `undefined`) and
 * has positive length. The `required` validator **will** fail for empty
 * strings.
 *
 * @param {Any***REMOVED*** value
 * @param {Document***REMOVED*** doc
 * @return {Boolean***REMOVED***
 * @api public
 */

SchemaString.prototype.checkRequired = function checkRequired(value, doc) {
  if (SchemaType._isRef(this, value, doc, true)) {
    return !!value;
  ***REMOVED***
  return (value instanceof String || typeof value === 'string') && value.length;
***REMOVED***;

/**
 * Casts to String
 *
 * @api private
 */

SchemaString.prototype.cast = function(value, doc, init) {
  if (SchemaType._isRef(this, value, doc, init)) {
    // wait! we may need to cast this to a document

    if (value === null || value === undefined) {
      return value;
    ***REMOVED***

    // lazy load
    Document || (Document = require('./../document'));

    if (value instanceof Document) {
      value.$__.wasPopulated = true;
      return value;
    ***REMOVED***

    // setting a populated path
    if (typeof value === 'string') {
      return value;
    ***REMOVED*** else if (Buffer.isBuffer(value) || !utils.isObject(value)) {
      throw new CastError('string', value, this.path);
    ***REMOVED***

    // Handle the case where user directly sets a populated
    // path to a plain object; cast to the Model used in
    // the population query.
    const path = doc.$__fullPath(this.path);
    const owner = doc.ownerDocument ? doc.ownerDocument() : doc;
    const pop = owner.populated(path, true);
    const ret = new pop.options.model(value);
    ret.$__.wasPopulated = true;
    return ret;
  ***REMOVED***

  return castString(value, this.path);
***REMOVED***;

/*!
 * ignore
 */

function handleSingle(val) {
  return this.castForQuery(val);
***REMOVED***

function handleArray(val) {
  const _this = this;
  if (!Array.isArray(val)) {
    return [this.castForQuery(val)];
  ***REMOVED***
  return val.map(function(m) {
    return _this.castForQuery(m);
  ***REMOVED***);
***REMOVED***

SchemaString.prototype.$conditionalHandlers =
    utils.options(SchemaType.prototype.$conditionalHandlers, {
      $all: handleArray,
      $gt: handleSingle,
      $gte: handleSingle,
      $lt: handleSingle,
      $lte: handleSingle,
      $options: handleSingle,
      $regex: handleSingle,
      $not: handleSingle
    ***REMOVED***);

/**
 * Casts contents for queries.
 *
 * @param {String***REMOVED*** $conditional
 * @param {any***REMOVED*** [val]
 * @api private
 */

SchemaString.prototype.castForQuery = function($conditional, val) {
  let handler;
  if (arguments.length === 2) {
    handler = this.$conditionalHandlers[$conditional];
    if (!handler) {
      throw new Error('Can\'t use ' + $conditional + ' with String.');
    ***REMOVED***
    return handler.call(this, val);
  ***REMOVED***
  val = $conditional;
  if (Object.prototype.toString.call(val) === '[object RegExp]') {
    return val;
  ***REMOVED***

  return this._castForQuery(val);
***REMOVED***;

/*!
 * Module exports.
 */

module.exports = SchemaString;
