// mongoose/lib/types/embedded.js
/* eslint no-func-assign: 1 */

/*!
 * Module dependencies.
 */

'use strict';

const Document = require('../document_provider')();
const EventEmitter = require('events').EventEmitter;
const immediate = require('../helpers/immediate');
const internalToObjectOptions = require('../options').internalToObjectOptions;
const get = require('lodash.get');
const utils = require('../utils');
const util = require('util');
const validatorErrorSymbol = require('../helpers/symbols').validatorErrorSymbol;

/**
 * EmbeddedDocument constructor.
 *
 * @param {Object***REMOVED*** obj js object returned from the db
 * @param {MongooseDocumentArray***REMOVED*** parentArr the parent array of this document
 * @param {Boolean***REMOVED*** skipId
 * @inherits Document
 * @api private
 */

function EmbeddedDocument(obj, parentArr, skipId, fields, index) {
  if (parentArr) {
    this.__parentArray = parentArr;
    this.__parent = parentArr._parent;
  ***REMOVED***
    this.__parentArray = undefined;
    this.__parent = undefined;
  ***REMOVED***
  this.$setIndex(index);
  this.$isDocumentArrayElement = true;

  Document.call(this, obj, fields, skipId);

  const _this = this;
  this.on('isNew', function(val) {
    _this.isNew = val;
  ***REMOVED***);

  _this.on('save', function() {
    _this.constructor.emit('save', _this);
  ***REMOVED***);
***REMOVED***

/*!
 * Inherit from Document
 */
EmbeddedDocument.prototype = Object.create(Document.prototype);
EmbeddedDocument.prototype.constructor = EmbeddedDocument;

for (const i in EventEmitter.prototype) {
  EmbeddedDocument[i] = EventEmitter.prototype[i];
***REMOVED***

EmbeddedDocument.prototype.toBSON = function() {
  return this.toObject(internalToObjectOptions);
***REMOVED***;

/*!
 * ignore
 */

EmbeddedDocument.prototype.$setIndex = function(index) {
  this.__index = index;

  if (get(this, '$__.validationError', null) != null) {
    const keys = Object.keys(this.$__.validationError.errors);
    for (const key of keys) {
      this.invalidate(key, this.$__.validationError.errors[key]);
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/**
 * Marks the embedded doc modified.
 *
 * ####Example:
 *
 *     var doc = blogpost.comments.id(hexstring);
 *     doc.mixed.type = 'changed';
 *     doc.markModified('mixed.type');
 *
 * @param {String***REMOVED*** path the path which changed
 * @api public
 * @receiver EmbeddedDocument
 */

EmbeddedDocument.prototype.markModified = function(path) {
  this.$__.activePaths.modify(path);
  if (!this.__parentArray) {
    return;
  ***REMOVED***

  if (this.isNew) {
    // Mark the WHOLE parent array as modified
    // if this is a new document (i.e., we are initializing
    // a document),
    this.__parentArray._markModified();
  ***REMOVED***
    this.__parentArray._markModified(this, path);
  ***REMOVED***
***REMOVED***;

/*!
 * ignore
 */

EmbeddedDocument.prototype.populate = function() {
  throw new Error('Mongoose does not support calling populate() on nested ' +
    'docs. Instead of `doc.arr[0].populate("path")`, use ' +
    '`doc.populate("arr.0.path")`');
***REMOVED***;

/**
 * Used as a stub for [hooks.js](https://github.com/bnoguchi/hooks-js/tree/31ec571cef0332e21121ee7157e0cf9728572cc3)
 *
 * ####NOTE:
 *
 * _This is a no-op. Does not actually save the doc to the db._
 *
 * @param {Function***REMOVED*** [fn]
 * @return {Promise***REMOVED*** resolved Promise
 * @api private
 */

EmbeddedDocument.prototype.save = function(options, fn) {
  if (typeof options === 'function') {
    fn = options;
    options = {***REMOVED***;
  ***REMOVED***
  options = options || {***REMOVED***;

  if (!options.suppressWarning) {
    console.warn('mongoose: calling `save()` on a subdoc does **not** save ' +
      'the document to MongoDB, it only runs save middleware. ' +
      'Use `subdoc.save({ suppressWarning: true ***REMOVED***)` to hide this warning ' +
      'if you\'re sure this behavior is right for your app.');
  ***REMOVED***

  return utils.promiseOrCallback(fn, cb => {
    this.$__save(cb);
  ***REMOVED***);
***REMOVED***;

/**
 * Used as a stub for middleware
 *
 * ####NOTE:
 *
 * _This is a no-op. Does not actually save the doc to the db._
 *
 * @param {Function***REMOVED*** [fn]
 * @method $__save
 * @api private
 */

EmbeddedDocument.prototype.$__save = function(fn) {
  return immediate(() => fn(null, this));
***REMOVED***;

/*!
 * Registers remove event listeners for triggering
 * on subdocuments.
 *
 * @param {EmbeddedDocument***REMOVED*** sub
 * @api private
 */

function registerRemoveListener(sub) {
  let owner = sub.ownerDocument();

  function emitRemove() {
    owner.removeListener('save', emitRemove);
    owner.removeListener('remove', emitRemove);
    sub.emit('remove', sub);
    sub.constructor.emit('remove', sub);
    owner = sub = null;
  ***REMOVED***

  owner.on('save', emitRemove);
  owner.on('remove', emitRemove);
***REMOVED***

/*!
 * no-op for hooks
 */

EmbeddedDocument.prototype.$__remove = function(cb) {
  return cb(null, this);
***REMOVED***;

/**
 * Removes the subdocument from its parent array.
 *
 * @param {Object***REMOVED*** [options]
 * @param {Function***REMOVED*** [fn]
 * @api public
 */

EmbeddedDocument.prototype.remove = function(options, fn) {
  if ( typeof options === 'function' && !fn ) {
    fn = options;
    options = undefined;
  ***REMOVED***
  if (!this.__parentArray || (options && options.noop)) {
    fn && fn(null);
    return this;
  ***REMOVED***

  let _id;
  if (!this.willRemove) {
    _id = this._doc._id;
    if (!_id) {
      throw new Error('For your own good, Mongoose does not know ' +
          'how to remove an EmbeddedDocument that has no _id');
    ***REMOVED***
    this.__parentArray.pull({_id: _id***REMOVED***);
    this.willRemove = true;
    registerRemoveListener(this);
  ***REMOVED***

  if (fn) {
    fn(null);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Override #update method of parent documents.
 * @api private
 */

EmbeddedDocument.prototype.update = function() {
  throw new Error('The #update method is not available on EmbeddedDocuments');
***REMOVED***;

/**
 * Helper for console.log
 *
 * @api public
 */

EmbeddedDocument.prototype.inspect = function() {
  return this.toObject({
    transform: false,
    virtuals: false,
    flattenDecimals: false
  ***REMOVED***);
***REMOVED***;

if (util.inspect.custom) {
  /*!
  * Avoid Node deprecation warning DEP0079
  */

  EmbeddedDocument.prototype[util.inspect.custom] = EmbeddedDocument.prototype.inspect;
***REMOVED***

/**
 * Marks a path as invalid, causing validation to fail.
 *
 * @param {String***REMOVED*** path the field to invalidate
 * @param {String|Error***REMOVED*** err error which states the reason `path` was invalid
 * @return {Boolean***REMOVED***
 * @api public
 */

EmbeddedDocument.prototype.invalidate = function(path, err, val) {
  Document.prototype.invalidate.call(this, path, err, val);

  if (!this.__parent) {
    if (err[validatorErrorSymbol] || err.name === 'ValidationError') {
      return true;
    ***REMOVED***
    throw err;
  ***REMOVED***

  const index = this.__index;
  if (typeof index !== 'undefined') {
    const parentPath = this.__parentArray._path;
    const fullPath = [parentPath, index, path].join('.');
    this.__parent.invalidate(fullPath, err, val);
  ***REMOVED***

  return true;
***REMOVED***;

/**
 * Marks a path as valid, removing existing validation errors.
 *
 * @param {String***REMOVED*** path the field to mark as valid
 * @api private
 * @method $markValid
 * @receiver EmbeddedDocument
 */

EmbeddedDocument.prototype.$markValid = function(path) {
  if (!this.__parent) {
    return;
  ***REMOVED***

  const index = this.__index;
  if (typeof index !== 'undefined') {
    const parentPath = this.__parentArray._path;
    const fullPath = [parentPath, index, path].join('.');
    this.__parent.$markValid(fullPath);
  ***REMOVED***
***REMOVED***;

/*!
 * ignore
 */

EmbeddedDocument.prototype.$ignore = function(path) {
  Document.prototype.$ignore.call(this, path);

  if (!this.__parent) {
    return;
  ***REMOVED***

  const index = this.__index;
  if (typeof index !== 'undefined') {
    const parentPath = this.__parentArray._path;
    const fullPath = [parentPath, index, path].join('.');
    this.__parent.$ignore(fullPath);
  ***REMOVED***
***REMOVED***;

/**
 * Checks if a path is invalid
 *
 * @param {String***REMOVED*** path the field to check
 * @api private
 * @method $isValid
 * @receiver EmbeddedDocument
 */

EmbeddedDocument.prototype.$isValid = function(path) {
  const index = this.__index;
  if (typeof index !== 'undefined' && this.__parent) {
    return !this.__parent.$__.validationError ||
      !this.__parent.$__.validationError.errors[this.$__fullPath(path)];
  ***REMOVED***

  return true;
***REMOVED***;

/**
 * Returns the top level document of this sub-document.
 *
 * @return {Document***REMOVED***
 */

EmbeddedDocument.prototype.ownerDocument = function() {
  if (this.$__.ownerDocument) {
    return this.$__.ownerDocument;
  ***REMOVED***

  let parent = this.__parent;
  if (!parent) {
    return this;
  ***REMOVED***

  while (parent.__parent || parent.$parent) {
    parent = parent.__parent || parent.$parent;
  ***REMOVED***

  this.$__.ownerDocument = parent;
  return this.$__.ownerDocument;
***REMOVED***;

/**
 * Returns the full path to this document. If optional `path` is passed, it is appended to the full path.
 *
 * @param {String***REMOVED*** [path]
 * @return {String***REMOVED***
 * @api private
 * @method $__fullPath
 * @memberOf EmbeddedDocument
 * @instance
 */

EmbeddedDocument.prototype.$__fullPath = function(path) {
  if (!this.$__.fullPath) {
    let parent = this; // eslint-disable-line consistent-this
    if (!parent.__parent) {
      return path;
    ***REMOVED***

    const paths = [];
    while (parent.__parent || parent.$parent) {
      if (parent.__parent) {
        paths.unshift(parent.__parentArray._path);
      ***REMOVED***
        paths.unshift(parent.$basePath);
      ***REMOVED***
      parent = parent.__parent || parent.$parent;
    ***REMOVED***

    this.$__.fullPath = paths.join('.');

    if (!this.$__.ownerDocument) {
      // optimization
      this.$__.ownerDocument = parent;
    ***REMOVED***
  ***REMOVED***

  return path
    ? this.$__.fullPath + '.' + path
    : this.$__.fullPath;
***REMOVED***;

/**
 * Returns this sub-documents parent document.
 *
 * @api public
 */

EmbeddedDocument.prototype.parent = function() {
  return this.__parent;
***REMOVED***;

/**
 * Returns this sub-documents parent array.
 *
 * @api public
 */

EmbeddedDocument.prototype.parentArray = function() {
  return this.__parentArray;
***REMOVED***;

/*!
 * Module exports.
 */

module.exports = EmbeddedDocument;
